<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="generator" content="pdoc 14.7.0"/>
    <title>dcatoolkit API documentation</title>

    <style>/*! * Bootstrap Reboot v5.0.0 (https://getbootstrap.com/) * Copyright 2011-2021 The Bootstrap Authors * Copyright 2011-2021 Twitter, Inc. * Licensed under MIT (https://github.com/twbs/bootstrap/blob/main/LICENSE) * Forked from Normalize.css, licensed MIT (https://github.com/necolas/normalize.css/blob/master/LICENSE.md) */*,::after,::before{box-sizing:border-box}@media (prefers-reduced-motion:no-preference){:root{scroll-behavior:smooth}}body{margin:0;font-family:system-ui,-apple-system,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans","Liberation Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";font-size:1rem;font-weight:400;line-height:1.5;color:#212529;background-color:#fff;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:transparent}hr{margin:1rem 0;color:inherit;background-color:currentColor;border:0;opacity:.25}hr:not([size]){height:1px}h1,h2,h3,h4,h5,h6{margin-top:0;margin-bottom:.5rem;font-weight:500;line-height:1.2}h1{font-size:calc(1.375rem + 1.5vw)}@media (min-width:1200px){h1{font-size:2.5rem}}h2{font-size:calc(1.325rem + .9vw)}@media (min-width:1200px){h2{font-size:2rem}}h3{font-size:calc(1.3rem + .6vw)}@media (min-width:1200px){h3{font-size:1.75rem}}h4{font-size:calc(1.275rem + .3vw)}@media (min-width:1200px){h4{font-size:1.5rem}}h5{font-size:1.25rem}h6{font-size:1rem}p{margin-top:0;margin-bottom:1rem}abbr[data-bs-original-title],abbr[title]{-webkit-text-decoration:underline dotted;text-decoration:underline dotted;cursor:help;-webkit-text-decoration-skip-ink:none;text-decoration-skip-ink:none}address{margin-bottom:1rem;font-style:normal;line-height:inherit}ol,ul{padding-left:2rem}dl,ol,ul{margin-top:0;margin-bottom:1rem}ol ol,ol ul,ul ol,ul ul{margin-bottom:0}dt{font-weight:700}dd{margin-bottom:.5rem;margin-left:0}blockquote{margin:0 0 1rem}b,strong{font-weight:bolder}small{font-size:.875em}mark{padding:.2em;background-color:#fcf8e3}sub,sup{position:relative;font-size:.75em;line-height:0;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}a{color:#0d6efd;text-decoration:underline}a:hover{color:#0a58ca}a:not([href]):not([class]),a:not([href]):not([class]):hover{color:inherit;text-decoration:none}code,kbd,pre,samp{font-family:SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace;font-size:1em;direction:ltr;unicode-bidi:bidi-override}pre{display:block;margin-top:0;margin-bottom:1rem;overflow:auto;font-size:.875em}pre code{font-size:inherit;color:inherit;word-break:normal}code{font-size:.875em;color:#d63384;word-wrap:break-word}a>code{color:inherit}kbd{padding:.2rem .4rem;font-size:.875em;color:#fff;background-color:#212529;border-radius:.2rem}kbd kbd{padding:0;font-size:1em;font-weight:700}figure{margin:0 0 1rem}img,svg{vertical-align:middle}table{caption-side:bottom;border-collapse:collapse}caption{padding-top:.5rem;padding-bottom:.5rem;color:#6c757d;text-align:left}th{text-align:inherit;text-align:-webkit-match-parent}tbody,td,tfoot,th,thead,tr{border-color:inherit;border-style:solid;border-width:0}label{display:inline-block}button{border-radius:0}button:focus:not(:focus-visible){outline:0}button,input,optgroup,select,textarea{margin:0;font-family:inherit;font-size:inherit;line-height:inherit}button,select{text-transform:none}[role=button]{cursor:pointer}select{word-wrap:normal}select:disabled{opacity:1}[list]::-webkit-calendar-picker-indicator{display:none}[type=button],[type=reset],[type=submit],button{-webkit-appearance:button}[type=button]:not(:disabled),[type=reset]:not(:disabled),[type=submit]:not(:disabled),button:not(:disabled){cursor:pointer}::-moz-focus-inner{padding:0;border-style:none}textarea{resize:vertical}fieldset{min-width:0;padding:0;margin:0;border:0}legend{float:left;width:100%;padding:0;margin-bottom:.5rem;font-size:calc(1.275rem + .3vw);line-height:inherit}@media (min-width:1200px){legend{font-size:1.5rem}}legend+*{clear:left}::-webkit-datetime-edit-day-field,::-webkit-datetime-edit-fields-wrapper,::-webkit-datetime-edit-hour-field,::-webkit-datetime-edit-minute,::-webkit-datetime-edit-month-field,::-webkit-datetime-edit-text,::-webkit-datetime-edit-year-field{padding:0}::-webkit-inner-spin-button{height:auto}[type=search]{outline-offset:-2px;-webkit-appearance:textfield}::-webkit-search-decoration{-webkit-appearance:none}::-webkit-color-swatch-wrapper{padding:0}::file-selector-button{font:inherit}::-webkit-file-upload-button{font:inherit;-webkit-appearance:button}output{display:inline-block}iframe{border:0}summary{display:list-item;cursor:pointer}progress{vertical-align:baseline}[hidden]{display:none!important}</style>
    <style>/*! syntax-highlighting.css */pre{line-height:125%;}span.linenos{color:inherit; background-color:transparent; padding-left:5px; padding-right:20px;}.pdoc-code .hll{background-color:#ffffcc}.pdoc-code{background:#f8f8f8;}.pdoc-code .c{color:#3D7B7B; font-style:italic}.pdoc-code .err{border:1px solid #FF0000}.pdoc-code .k{color:#008000; font-weight:bold}.pdoc-code .o{color:#666666}.pdoc-code .ch{color:#3D7B7B; font-style:italic}.pdoc-code .cm{color:#3D7B7B; font-style:italic}.pdoc-code .cp{color:#9C6500}.pdoc-code .cpf{color:#3D7B7B; font-style:italic}.pdoc-code .c1{color:#3D7B7B; font-style:italic}.pdoc-code .cs{color:#3D7B7B; font-style:italic}.pdoc-code .gd{color:#A00000}.pdoc-code .ge{font-style:italic}.pdoc-code .gr{color:#E40000}.pdoc-code .gh{color:#000080; font-weight:bold}.pdoc-code .gi{color:#008400}.pdoc-code .go{color:#717171}.pdoc-code .gp{color:#000080; font-weight:bold}.pdoc-code .gs{font-weight:bold}.pdoc-code .gu{color:#800080; font-weight:bold}.pdoc-code .gt{color:#0044DD}.pdoc-code .kc{color:#008000; font-weight:bold}.pdoc-code .kd{color:#008000; font-weight:bold}.pdoc-code .kn{color:#008000; font-weight:bold}.pdoc-code .kp{color:#008000}.pdoc-code .kr{color:#008000; font-weight:bold}.pdoc-code .kt{color:#B00040}.pdoc-code .m{color:#666666}.pdoc-code .s{color:#BA2121}.pdoc-code .na{color:#687822}.pdoc-code .nb{color:#008000}.pdoc-code .nc{color:#0000FF; font-weight:bold}.pdoc-code .no{color:#880000}.pdoc-code .nd{color:#AA22FF}.pdoc-code .ni{color:#717171; font-weight:bold}.pdoc-code .ne{color:#CB3F38; font-weight:bold}.pdoc-code .nf{color:#0000FF}.pdoc-code .nl{color:#767600}.pdoc-code .nn{color:#0000FF; font-weight:bold}.pdoc-code .nt{color:#008000; font-weight:bold}.pdoc-code .nv{color:#19177C}.pdoc-code .ow{color:#AA22FF; font-weight:bold}.pdoc-code .w{color:#bbbbbb}.pdoc-code .mb{color:#666666}.pdoc-code .mf{color:#666666}.pdoc-code .mh{color:#666666}.pdoc-code .mi{color:#666666}.pdoc-code .mo{color:#666666}.pdoc-code .sa{color:#BA2121}.pdoc-code .sb{color:#BA2121}.pdoc-code .sc{color:#BA2121}.pdoc-code .dl{color:#BA2121}.pdoc-code .sd{color:#BA2121; font-style:italic}.pdoc-code .s2{color:#BA2121}.pdoc-code .se{color:#AA5D1F; font-weight:bold}.pdoc-code .sh{color:#BA2121}.pdoc-code .si{color:#A45A77; font-weight:bold}.pdoc-code .sx{color:#008000}.pdoc-code .sr{color:#A45A77}.pdoc-code .s1{color:#BA2121}.pdoc-code .ss{color:#19177C}.pdoc-code .bp{color:#008000}.pdoc-code .fm{color:#0000FF}.pdoc-code .vc{color:#19177C}.pdoc-code .vg{color:#19177C}.pdoc-code .vi{color:#19177C}.pdoc-code .vm{color:#19177C}.pdoc-code .il{color:#666666}</style>
    <style>/*! theme.css */:root{--pdoc-background:#fff;}.pdoc{--text:#212529;--muted:#6c757d;--link:#3660a5;--link-hover:#1659c5;--code:#f8f8f8;--active:#fff598;--accent:#eee;--accent2:#c1c1c1;--nav-hover:rgba(255, 255, 255, 0.5);--name:#0066BB;--def:#008800;--annotation:#007020;}</style>
    <style>/*! layout.css */html, body{width:100%;height:100%;}html, main{scroll-behavior:smooth;}body{background-color:var(--pdoc-background);}@media (max-width:769px){#navtoggle{cursor:pointer;position:absolute;width:50px;height:40px;top:1rem;right:1rem;border-color:var(--text);color:var(--text);display:flex;opacity:0.8;z-index:999;}#navtoggle:hover{opacity:1;}#togglestate + div{display:none;}#togglestate:checked + div{display:inherit;}main, header{padding:2rem 3vw;}header + main{margin-top:-3rem;}.git-button{display:none !important;}nav input[type="search"]{max-width:77%;}nav input[type="search"]:first-child{margin-top:-6px;}nav input[type="search"]:valid ~ *{display:none !important;}}@media (min-width:770px){:root{--sidebar-width:clamp(12.5rem, 28vw, 22rem);}nav{position:fixed;overflow:auto;height:100vh;width:var(--sidebar-width);}main, header{padding:3rem 2rem 3rem calc(var(--sidebar-width) + 3rem);width:calc(54rem + var(--sidebar-width));max-width:100%;}header + main{margin-top:-4rem;}#navtoggle{display:none;}}#togglestate{position:absolute;height:0;opacity:0;}nav.pdoc{--pad:clamp(0.5rem, 2vw, 1.75rem);--indent:1.5rem;background-color:var(--accent);border-right:1px solid var(--accent2);box-shadow:0 0 20px rgba(50, 50, 50, .2) inset;padding:0 0 0 var(--pad);overflow-wrap:anywhere;scrollbar-width:thin; scrollbar-color:var(--accent2) transparent; z-index:1}nav.pdoc::-webkit-scrollbar{width:.4rem; }nav.pdoc::-webkit-scrollbar-thumb{background-color:var(--accent2); }nav.pdoc > div{padding:var(--pad) 0;}nav.pdoc .module-list-button{display:inline-flex;align-items:center;color:var(--text);border-color:var(--muted);margin-bottom:1rem;}nav.pdoc .module-list-button:hover{border-color:var(--text);}nav.pdoc input[type=search]{display:block;outline-offset:0;width:calc(100% - var(--pad));}nav.pdoc .logo{max-width:calc(100% - var(--pad));max-height:35vh;display:block;margin:0 auto 1rem;transform:translate(calc(-.5 * var(--pad)), 0);}nav.pdoc ul{list-style:none;padding-left:0;}nav.pdoc > div > ul{margin-left:calc(0px - var(--pad));}nav.pdoc li a{padding:.2rem 0 .2rem calc(var(--pad) + var(--indent));}nav.pdoc > div > ul > li > a{padding-left:var(--pad);}nav.pdoc li{transition:all 100ms;}nav.pdoc li:hover{background-color:var(--nav-hover);}nav.pdoc a, nav.pdoc a:hover{color:var(--text);}nav.pdoc a{display:block;}nav.pdoc > h2:first-of-type{margin-top:1.5rem;}nav.pdoc .class:before{content:"class ";color:var(--muted);}nav.pdoc .function:after{content:"()";color:var(--muted);}nav.pdoc footer:before{content:"";display:block;width:calc(100% - var(--pad));border-top:solid var(--accent2) 1px;margin-top:1.5rem;padding-top:.5rem;}nav.pdoc footer{font-size:small;}</style>
    <style>/*! content.css */.pdoc{color:var(--text);box-sizing:border-box;line-height:1.5;background:none;}.pdoc .pdoc-button{cursor:pointer;display:inline-block;border:solid black 1px;border-radius:2px;font-size:.75rem;padding:calc(0.5em - 1px) 1em;transition:100ms all;}.pdoc .alert{padding:1rem 1rem 1rem calc(1.5rem + 24px);border:1px solid transparent;border-radius:.25rem;background-repeat:no-repeat;background-position:.75rem center;margin-bottom:1rem;}.pdoc .alert > em{display:none;}.pdoc .alert > *:last-child{margin-bottom:0;}.pdoc .alert.note {color:#084298;background-color:#cfe2ff;border-color:#b6d4fe;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23084298%22%20viewBox%3D%220%200%2016%2016%22%3E%3Cpath%20d%3D%22M8%2016A8%208%200%201%200%208%200a8%208%200%200%200%200%2016zm.93-9.412-1%204.705c-.07.34.029.533.304.533.194%200%20.487-.07.686-.246l-.088.416c-.287.346-.92.598-1.465.598-.703%200-1.002-.422-.808-1.319l.738-3.468c.064-.293.006-.399-.287-.47l-.451-.081.082-.381%202.29-.287zM8%205.5a1%201%200%201%201%200-2%201%201%200%200%201%200%202z%22/%3E%3C/svg%3E");}.pdoc .alert.warning{color:#664d03;background-color:#fff3cd;border-color:#ffecb5;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23664d03%22%20viewBox%3D%220%200%2016%2016%22%3E%3Cpath%20d%3D%22M8.982%201.566a1.13%201.13%200%200%200-1.96%200L.165%2013.233c-.457.778.091%201.767.98%201.767h13.713c.889%200%201.438-.99.98-1.767L8.982%201.566zM8%205c.535%200%20.954.462.9.995l-.35%203.507a.552.552%200%200%201-1.1%200L7.1%205.995A.905.905%200%200%201%208%205zm.002%206a1%201%200%201%201%200%202%201%201%200%200%201%200-2z%22/%3E%3C/svg%3E");}.pdoc .alert.danger{color:#842029;background-color:#f8d7da;border-color:#f5c2c7;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23842029%22%20viewBox%3D%220%200%2016%2016%22%3E%3Cpath%20d%3D%22M5.52.359A.5.5%200%200%201%206%200h4a.5.5%200%200%201%20.474.658L8.694%206H12.5a.5.5%200%200%201%20.395.807l-7%209a.5.5%200%200%201-.873-.454L6.823%209.5H3.5a.5.5%200%200%201-.48-.641l2.5-8.5z%22/%3E%3C/svg%3E");}.pdoc .visually-hidden{position:absolute !important;width:1px !important;height:1px !important;padding:0 !important;margin:-1px !important;overflow:hidden !important;clip:rect(0, 0, 0, 0) !important;white-space:nowrap !important;border:0 !important;}.pdoc h1, .pdoc h2, .pdoc h3{font-weight:300;margin:.3em 0;padding:.2em 0;}.pdoc > section:not(.module-info) h1{font-size:1.5rem;font-weight:500;}.pdoc > section:not(.module-info) h2{font-size:1.4rem;font-weight:500;}.pdoc > section:not(.module-info) h3{font-size:1.3rem;font-weight:500;}.pdoc > section:not(.module-info) h4{font-size:1.2rem;}.pdoc > section:not(.module-info) h5{font-size:1.1rem;}.pdoc a{text-decoration:none;color:var(--link);}.pdoc a:hover{color:var(--link-hover);}.pdoc blockquote{margin-left:2rem;}.pdoc pre{border-top:1px solid var(--accent2);border-bottom:1px solid var(--accent2);margin-top:0;margin-bottom:1em;padding:.5rem 0 .5rem .5rem;overflow-x:auto;background-color:var(--code);}.pdoc code{color:var(--text);padding:.2em .4em;margin:0;font-size:85%;background-color:var(--accent);border-radius:6px;}.pdoc a > code{color:inherit;}.pdoc pre > code{display:inline-block;font-size:inherit;background:none;border:none;padding:0;}.pdoc > section:not(.module-info){margin-bottom:1.5rem;}.pdoc .modulename{margin-top:0;font-weight:bold;}.pdoc .modulename a{color:var(--link);transition:100ms all;}.pdoc .git-button{float:right;border:solid var(--link) 1px;}.pdoc .git-button:hover{background-color:var(--link);color:var(--pdoc-background);}.view-source-toggle-state,.view-source-toggle-state ~ .pdoc-code{display:none;}.view-source-toggle-state:checked ~ .pdoc-code{display:block;}.view-source-button{display:inline-block;float:right;font-size:.75rem;line-height:1.5rem;color:var(--muted);padding:0 .4rem 0 1.3rem;cursor:pointer;text-indent:-2px;}.view-source-button > span{visibility:hidden;}.module-info .view-source-button{float:none;display:flex;justify-content:flex-end;margin:-1.2rem .4rem -.2rem 0;}.view-source-button::before{position:absolute;content:"View Source";display:list-item;list-style-type:disclosure-closed;}.view-source-toggle-state:checked ~ .attr .view-source-button::before,.view-source-toggle-state:checked ~ .view-source-button::before{list-style-type:disclosure-open;}.pdoc .docstring{margin-bottom:1.5rem;}.pdoc section:not(.module-info) .docstring{margin-left:clamp(0rem, 5vw - 2rem, 1rem);}.pdoc .docstring .pdoc-code{margin-left:1em;margin-right:1em;}.pdoc h1:target,.pdoc h2:target,.pdoc h3:target,.pdoc h4:target,.pdoc h5:target,.pdoc h6:target,.pdoc .pdoc-code > pre > span:target{background-color:var(--active);box-shadow:-1rem 0 0 0 var(--active);}.pdoc .pdoc-code > pre > span:target{display:block;}.pdoc div:target > .attr,.pdoc section:target > .attr,.pdoc dd:target > a{background-color:var(--active);}.pdoc *{scroll-margin:2rem;}.pdoc .pdoc-code .linenos{user-select:none;}.pdoc .attr:hover{filter:contrast(0.95);}.pdoc section, .pdoc .classattr{position:relative;}.pdoc .headerlink{--width:clamp(1rem, 3vw, 2rem);position:absolute;top:0;left:calc(0rem - var(--width));transition:all 100ms ease-in-out;opacity:0;}.pdoc .headerlink::before{content:"#";display:block;text-align:center;width:var(--width);height:2.3rem;line-height:2.3rem;font-size:1.5rem;}.pdoc .attr:hover ~ .headerlink,.pdoc *:target > .headerlink,.pdoc .headerlink:hover{opacity:1;}.pdoc .attr{display:block;margin:.5rem 0 .5rem;padding:.4rem .4rem .4rem 1rem;background-color:var(--accent);overflow-x:auto;}.pdoc .classattr{margin-left:2rem;}.pdoc .name{color:var(--name);font-weight:bold;}.pdoc .def{color:var(--def);font-weight:bold;}.pdoc .signature{background-color:transparent;}.pdoc .param, .pdoc .return-annotation{white-space:pre;}.pdoc .signature.multiline .param{display:block;}.pdoc .signature.condensed .param{display:inline-block;}.pdoc .annotation{color:var(--annotation);}.pdoc .view-value-toggle-state,.pdoc .view-value-toggle-state ~ .default_value{display:none;}.pdoc .view-value-toggle-state:checked ~ .default_value{display:inherit;}.pdoc .view-value-button{font-size:.5rem;vertical-align:middle;border-style:dashed;margin-top:-0.1rem;}.pdoc .view-value-button:hover{background:white;}.pdoc .view-value-button::before{content:"show";text-align:center;width:2.2em;display:inline-block;}.pdoc .view-value-toggle-state:checked ~ .view-value-button::before{content:"hide";}.pdoc .inherited{margin-left:2rem;}.pdoc .inherited dt{font-weight:700;}.pdoc .inherited dt, .pdoc .inherited dd{display:inline;margin-left:0;margin-bottom:.5rem;}.pdoc .inherited dd:not(:last-child):after{content:", ";}.pdoc .inherited .class:before{content:"class ";}.pdoc .inherited .function a:after{content:"()";}.pdoc .search-result .docstring{overflow:auto;max-height:25vh;}.pdoc .search-result.focused > .attr{background-color:var(--active);}.pdoc .attribution{margin-top:2rem;display:block;opacity:0.5;transition:all 200ms;filter:grayscale(100%);}.pdoc .attribution:hover{opacity:1;filter:grayscale(0%);}.pdoc .attribution img{margin-left:5px;height:35px;vertical-align:middle;width:70px;transition:all 200ms;}.pdoc table{display:block;width:max-content;max-width:100%;overflow:auto;margin-bottom:1rem;}.pdoc table th{font-weight:600;}.pdoc table th, .pdoc table td{padding:6px 13px;border:1px solid var(--accent2);}</style>
    <style>/*! custom.css */</style></head>
<body>
    <nav class="pdoc">
        <label id="navtoggle" for="togglestate" class="pdoc-button"><svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 30 30'><path stroke-linecap='round' stroke="currentColor" stroke-miterlimit='10' stroke-width='2' d='M4 7h22M4 15h22M4 23h22'/></svg></label>
        <input id="togglestate" type="checkbox" aria-hidden="true" tabindex="-1">
        <div>




            <h2>API Documentation</h2>
                <ul class="memberlist">
            <li>
                    <a class="class" href="#Pairs">Pairs</a>
                            <ul class="memberlist">
                        <li>
                                <a class="function" href="#Pairs.__init__">Pairs</a>
                        </li>
                        <li>
                                <a class="function" href="#Pairs.load_from_file">load_from_file</a>
                        </li>
                        <li>
                                <a class="function" href="#Pairs.load_from_ndarray">load_from_ndarray</a>
                        </li>
                        <li>
                                <a class="function" href="#Pairs.mirror_diagonal">mirror_diagonal</a>
                        </li>
                        <li>
                                <a class="function" href="#Pairs.subset_pairs">subset_pairs</a>
                        </li>
                        <li>
                                <a class="function" href="#Pairs.mirror_pairs">mirror_pairs</a>
                        </li>
                        <li>
                                <a class="function" href="#Pairs.get_pairs">get_pairs</a>
                        </li>
                </ul>

            </li>
            <li>
                    <a class="class" href="#DirectInformationData">DirectInformationData</a>
                            <ul class="memberlist">
                        <li>
                                <a class="function" href="#DirectInformationData.__init__">DirectInformationData</a>
                        </li>
                        <li>
                                <a class="variable" href="#DirectInformationData.DI_data">DI_data</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.load_from_dca_output">load_from_dca_output</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.load_from_DI_file">load_from_DI_file</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.load_as_ndarray">load_as_ndarray</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.get_ranked_mapped_pairs">get_ranked_mapped_pairs</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.map_DIs">map_DIs</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.rank_pairs">rank_pairs</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.nonlocal_pairs">nonlocal_pairs</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.find_DI_with_residues">find_DI_with_residues</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.get_dist_commands">get_dist_commands</a>
                        </li>
                        <li>
                                <a class="function" href="#DirectInformationData.write_DI_data">write_DI_data</a>
                        </li>
                </ul>

            </li>
            <li>
                    <a class="class" href="#StructureInformation">StructureInformation</a>
                            <ul class="memberlist">
                        <li>
                                <a class="function" href="#StructureInformation.fetch_pdb">fetch_pdb</a>
                        </li>
                        <li>
                                <a class="function" href="#StructureInformation.read_mmCIF_file">read_mmCIF_file</a>
                        </li>
                        <li>
                                <a class="function" href="#StructureInformation.read_pdb_file">read_pdb_file</a>
                        </li>
                        <li>
                                <a class="function" href="#StructureInformation.write_contacts_set">write_contacts_set</a>
                        </li>
                </ul>

            </li>
            <li>
                    <a class="class" href="#ResidueAlignment">ResidueAlignment</a>
                            <ul class="memberlist">
                        <li>
                                <a class="function" href="#ResidueAlignment.__init__">ResidueAlignment</a>
                        </li>
                        <li>
                                <a class="variable" href="#ResidueAlignment.domain_name">domain_name</a>
                        </li>
                        <li>
                                <a class="variable" href="#ResidueAlignment.protein_name">protein_name</a>
                        </li>
                        <li>
                                <a class="function" href="#ResidueAlignment.set_reference_mapping">set_reference_mapping</a>
                        </li>
                        <li>
                                <a class="function" href="#ResidueAlignment.load_from_align_file">load_from_align_file</a>
                        </li>
                        <li>
                                <a class="function" href="#ResidueAlignment.read_align_file">read_align_file</a>
                        </li>
                </ul>

            </li>
            <li>
                    <a class="class" href="#MMCIFInformation">MMCIFInformation</a>
                            <ul class="memberlist">
                        <li>
                                <a class="function" href="#MMCIFInformation.__init__">MMCIFInformation</a>
                        </li>
                        <li>
                                <a class="variable" href="#MMCIFInformation.structure">structure</a>
                        </li>
                        <li>
                                <a class="variable" href="#MMCIFInformation.pdbx_file">pdbx_file</a>
                        </li>
                        <li>
                                <a class="variable" href="#MMCIFInformation.model_num">model_num</a>
                        </li>
                        <li>
                                <a class="variable" href="#MMCIFInformation.full_sequences">full_sequences</a>
                        </li>
                        <li>
                                <a class="variable" href="#MMCIFInformation.non_missing_sequences">non_missing_sequences</a>
                        </li>
                        <li>
                                <a class="function" href="#MMCIFInformation.generate_auth_info">generate_auth_info</a>
                        </li>
                        <li>
                                <a class="function" href="#MMCIFInformation.get_full_sequence">get_full_sequence</a>
                        </li>
                        <li>
                                <a class="function" href="#MMCIFInformation.get_non_missing_sequence">get_non_missing_sequence</a>
                        </li>
                        <li>
                                <a class="function" href="#MMCIFInformation.get_chain_specific_structure">get_chain_specific_structure</a>
                        </li>
                        <li>
                                <a class="function" href="#MMCIFInformation.generate_dist_matrix">generate_dist_matrix</a>
                        </li>
                        <li>
                                <a class="function" href="#MMCIFInformation.get_shift_values">get_shift_values</a>
                        </li>
                        <li>
                                <a class="function" href="#MMCIFInformation.get_min_dist_atom_info">get_min_dist_atom_info</a>
                        </li>
                        <li>
                                <a class="function" href="#MMCIFInformation.get_contacts">get_contacts</a>
                        </li>
                </ul>

            </li>
            <li>
                    <a class="class" href="#PDBInformation">PDBInformation</a>
                            <ul class="memberlist">
                        <li>
                                <a class="function" href="#PDBInformation.__init__">PDBInformation</a>
                        </li>
                        <li>
                                <a class="variable" href="#PDBInformation.structure">structure</a>
                        </li>
                        <li>
                                <a class="variable" href="#PDBInformation.pdb_file">pdb_file</a>
                        </li>
                        <li>
                                <a class="variable" href="#PDBInformation.model_num">model_num</a>
                        </li>
                        <li>
                                <a class="variable" href="#PDBInformation.non_missing_sequences">non_missing_sequences</a>
                        </li>
                        <li>
                                <a class="variable" href="#PDBInformation.unique_chains">unique_chains</a>
                        </li>
                        <li>
                                <a class="function" href="#PDBInformation.get_non_missing_sequence">get_non_missing_sequence</a>
                        </li>
                        <li>
                                <a class="function" href="#PDBInformation.get_chain_specific_structure">get_chain_specific_structure</a>
                        </li>
                        <li>
                                <a class="function" href="#PDBInformation.generate_dist_matrix">generate_dist_matrix</a>
                        </li>
                        <li>
                                <a class="function" href="#PDBInformation.get_shift_values">get_shift_values</a>
                        </li>
                        <li>
                                <a class="function" href="#PDBInformation.get_min_dist_atom_info">get_min_dist_atom_info</a>
                        </li>
                        <li>
                                <a class="function" href="#PDBInformation.get_contacts">get_contacts</a>
                        </li>
                </ul>

            </li>
            <li>
                    <a class="class" href="#MSATools">MSATools</a>
                            <ul class="memberlist">
                        <li>
                                <a class="function" href="#MSATools.__init__">MSATools</a>
                        </li>
                        <li>
                                <a class="variable" href="#MSATools.MSA">MSA</a>
                        </li>
                        <li>
                                <a class="function" href="#MSATools.load_from_file">load_from_file</a>
                        </li>
                        <li>
                                <a class="function" href="#MSATools.get_sequence_max_cont_gaps">get_sequence_max_cont_gaps</a>
                        </li>
                        <li>
                                <a class="function" href="#MSATools.gap_frequency">gap_frequency</a>
                        </li>
                        <li>
                                <a class="function" href="#MSATools.filter_by_continuous_gaps">filter_by_continuous_gaps</a>
                        </li>
                        <li>
                                <a class="function" href="#MSATools.write">write</a>
                        </li>
                </ul>

            </li>
    </ul>



        <a class="attribution" title="pdoc: Python API documentation generator" href="https://pdoc.dev" target="_blank">
            built with <span class="visually-hidden">pdoc</span><img
                alt="pdoc logo"
                src="data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20role%3D%22img%22%20aria-label%3D%22pdoc%20logo%22%20width%3D%22300%22%20height%3D%22150%22%20viewBox%3D%22-1%200%2060%2030%22%3E%3Ctitle%3Epdoc%3C/title%3E%3Cpath%20d%3D%22M29.621%2021.293c-.011-.273-.214-.475-.511-.481a.5.5%200%200%200-.489.503l-.044%201.393c-.097.551-.695%201.215-1.566%201.704-.577.428-1.306.486-2.193.182-1.426-.617-2.467-1.654-3.304-2.487l-.173-.172a3.43%203.43%200%200%200-.365-.306.49.49%200%200%200-.286-.196c-1.718-1.06-4.931-1.47-7.353.191l-.219.15c-1.707%201.187-3.413%202.131-4.328%201.03-.02-.027-.49-.685-.141-1.763.233-.721.546-2.408.772-4.076.042-.09.067-.187.046-.288.166-1.347.277-2.625.241-3.351%201.378-1.008%202.271-2.586%202.271-4.362%200-.976-.272-1.935-.788-2.774-.057-.094-.122-.18-.184-.268.033-.167.052-.339.052-.516%200-1.477-1.202-2.679-2.679-2.679-.791%200-1.496.352-1.987.9a6.3%206.3%200%200%200-1.001.029c-.492-.564-1.207-.929-2.012-.929-1.477%200-2.679%201.202-2.679%202.679A2.65%202.65%200%200%200%20.97%206.554c-.383.747-.595%201.572-.595%202.41%200%202.311%201.507%204.29%203.635%205.107-.037.699-.147%202.27-.423%203.294l-.137.461c-.622%202.042-2.515%208.257%201.727%2010.643%201.614.908%203.06%201.248%204.317%201.248%202.665%200%204.492-1.524%205.322-2.401%201.476-1.559%202.886-1.854%206.491.82%201.877%201.393%203.514%201.753%204.861%201.068%202.223-1.713%202.811-3.867%203.399-6.374.077-.846.056-1.469.054-1.537zm-4.835%204.313c-.054.305-.156.586-.242.629-.034-.007-.131-.022-.307-.157-.145-.111-.314-.478-.456-.908.221.121.432.25.675.355.115.039.219.051.33.081zm-2.251-1.238c-.05.33-.158.648-.252.694-.022.001-.125-.018-.307-.157-.217-.166-.488-.906-.639-1.573.358.344.754.693%201.198%201.036zm-3.887-2.337c-.006-.116-.018-.231-.041-.342.635.145%201.189.368%201.599.625.097.231.166.481.174.642-.03.049-.055.101-.067.158-.046.013-.128.026-.298.004-.278-.037-.901-.57-1.367-1.087zm-1.127-.497c.116.306.176.625.12.71-.019.014-.117.045-.345.016-.206-.027-.604-.332-.986-.695.41-.051.816-.056%201.211-.031zm-4.535%201.535c.209.22.379.47.358.598-.006.041-.088.138-.351.234-.144.055-.539-.063-.979-.259a11.66%2011.66%200%200%200%20.972-.573zm.983-.664c.359-.237.738-.418%201.126-.554.25.237.479.548.457.694-.006.042-.087.138-.351.235-.174.064-.694-.105-1.232-.375zm-3.381%201.794c-.022.145-.061.29-.149.401-.133.166-.358.248-.69.251h-.002c-.133%200-.306-.26-.45-.621.417.091.854.07%201.291-.031zm-2.066-8.077a4.78%204.78%200%200%201-.775-.584c.172-.115.505-.254.88-.378l-.105.962zm-.331%202.302a10.32%2010.32%200%200%201-.828-.502c.202-.143.576-.328.984-.49l-.156.992zm-.45%202.157l-.701-.403c.214-.115.536-.249.891-.376a11.57%2011.57%200%200%201-.19.779zm-.181%201.716c.064.398.194.702.298.893-.194-.051-.435-.162-.736-.398.061-.119.224-.3.438-.495zM8.87%204.141c0%20.152-.123.276-.276.276s-.275-.124-.275-.276.123-.276.276-.276.275.124.275.276zm-.735-.389a1.15%201.15%200%200%200-.314.783%201.16%201.16%200%200%200%201.162%201.162c.457%200%20.842-.27%201.032-.653.026.117.042.238.042.362a1.68%201.68%200%200%201-1.679%201.679%201.68%201.68%200%200%201-1.679-1.679c0-.843.626-1.535%201.436-1.654zM5.059%205.406A1.68%201.68%200%200%201%203.38%207.085a1.68%201.68%200%200%201-1.679-1.679c0-.037.009-.072.011-.109.21.3.541.508.935.508a1.16%201.16%200%200%200%201.162-1.162%201.14%201.14%200%200%200-.474-.912c.015%200%20.03-.005.045-.005.926.001%201.679.754%201.679%201.68zM3.198%204.141c0%20.152-.123.276-.276.276s-.275-.124-.275-.276.123-.276.276-.276.275.124.275.276zM1.375%208.964c0-.52.103-1.035.288-1.52.466.394%201.06.64%201.717.64%201.144%200%202.116-.725%202.499-1.738.383%201.012%201.355%201.738%202.499%201.738.867%200%201.631-.421%202.121-1.062.307.605.478%201.267.478%201.942%200%202.486-2.153%204.51-4.801%204.51s-4.801-2.023-4.801-4.51zm24.342%2019.349c-.985.498-2.267.168-3.813-.979-3.073-2.281-5.453-3.199-7.813-.705-1.315%201.391-4.163%203.365-8.423.97-3.174-1.786-2.239-6.266-1.261-9.479l.146-.492c.276-1.02.395-2.457.444-3.268a6.11%206.11%200%200%200%201.18.115%206.01%206.01%200%200%200%202.536-.562l-.006.175c-.802.215-1.848.612-2.021%201.25-.079.295.021.601.274.837.219.203.415.364.598.501-.667.304-1.243.698-1.311%201.179-.02.144-.022.507.393.787.213.144.395.26.564.365-1.285.521-1.361.96-1.381%201.126-.018.142-.011.496.427.746l.854.489c-.473.389-.971.914-.999%201.429-.018.278.095.532.316.713.675.556%201.231.721%201.653.721.059%200%20.104-.014.158-.02.207.707.641%201.64%201.513%201.64h.013c.8-.008%201.236-.345%201.462-.626.173-.216.268-.457.325-.692.424.195.93.374%201.372.374.151%200%20.294-.021.423-.068.732-.27.944-.704.993-1.021.009-.061.003-.119.002-.179.266.086.538.147.789.147.15%200%20.294-.021.423-.069.542-.2.797-.489.914-.754.237.147.478.258.704.288.106.014.205.021.296.021.356%200%20.595-.101.767-.229.438.435%201.094.992%201.656%201.067.106.014.205.021.296.021a1.56%201.56%200%200%200%20.323-.035c.17.575.453%201.289.866%201.605.358.273.665.362.914.362a.99.99%200%200%200%20.421-.093%201.03%201.03%200%200%200%20.245-.164c.168.428.39.846.68%201.068.358.273.665.362.913.362a.99.99%200%200%200%20.421-.093c.317-.148.512-.448.639-.762.251.157.495.257.726.257.127%200%20.25-.024.37-.071.427-.17.706-.617.841-1.314.022-.015.047-.022.068-.038.067-.051.133-.104.196-.159-.443%201.486-1.107%202.761-2.086%203.257zM8.66%209.925a.5.5%200%201%200-1%200c0%20.653-.818%201.205-1.787%201.205s-1.787-.552-1.787-1.205a.5.5%200%201%200-1%200c0%201.216%201.25%202.205%202.787%202.205s2.787-.989%202.787-2.205zm4.4%2015.965l-.208.097c-2.661%201.258-4.708%201.436-6.086.527-1.542-1.017-1.88-3.19-1.844-4.198a.4.4%200%200%200-.385-.414c-.242-.029-.406.164-.414.385-.046%201.249.367%203.686%202.202%204.896.708.467%201.547.7%202.51.7%201.248%200%202.706-.392%204.362-1.174l.185-.086a.4.4%200%200%200%20.205-.527c-.089-.204-.326-.291-.527-.206zM9.547%202.292c.093.077.205.114.317.114a.5.5%200%200%200%20.318-.886L8.817.397a.5.5%200%200%200-.703.068.5.5%200%200%200%20.069.703l1.364%201.124zm-7.661-.065c.086%200%20.173-.022.253-.068l1.523-.893a.5.5%200%200%200-.506-.863l-1.523.892a.5.5%200%200%200-.179.685c.094.158.261.247.432.247z%22%20transform%3D%22matrix%28-1%200%200%201%2058%200%29%22%20fill%3D%22%233bb300%22/%3E%3Cpath%20d%3D%22M.3%2021.86V10.18q0-.46.02-.68.04-.22.18-.5.28-.54%201.34-.54%201.06%200%201.42.28.38.26.44.78.76-1.04%202.38-1.04%201.64%200%203.1%201.54%201.46%201.54%201.46%203.58%200%202.04-1.46%203.58-1.44%201.54-3.08%201.54-1.64%200-2.38-.92v4.04q0%20.46-.04.68-.02.22-.18.5-.14.3-.5.42-.36.12-.98.12-.62%200-1-.12-.36-.12-.52-.4-.14-.28-.18-.5-.02-.22-.02-.68zm3.96-9.42q-.46.54-.46%201.18%200%20.64.46%201.18.48.52%201.2.52.74%200%201.24-.52.52-.52.52-1.18%200-.66-.48-1.18-.48-.54-1.26-.54-.76%200-1.22.54zm14.741-8.36q.16-.3.54-.42.38-.12%201-.12.64%200%201.02.12.38.12.52.42.16.3.18.54.04.22.04.68v11.94q0%20.46-.04.7-.02.22-.18.5-.3.54-1.7.54-1.38%200-1.54-.98-.84.96-2.34.96-1.8%200-3.28-1.56-1.48-1.58-1.48-3.66%200-2.1%201.48-3.68%201.5-1.58%203.28-1.58%201.48%200%202.3%201v-4.2q0-.46.02-.68.04-.24.18-.52zm-3.24%2010.86q.52.54%201.26.54.74%200%201.22-.54.5-.54.5-1.18%200-.66-.48-1.22-.46-.56-1.26-.56-.8%200-1.28.56-.48.54-.48%201.2%200%20.66.52%201.2zm7.833-1.2q0-2.4%201.68-3.96%201.68-1.56%203.84-1.56%202.16%200%203.82%201.56%201.66%201.54%201.66%203.94%200%201.66-.86%202.96-.86%201.28-2.1%201.9-1.22.6-2.54.6-1.32%200-2.56-.64-1.24-.66-2.1-1.92-.84-1.28-.84-2.88zm4.18%201.44q.64.48%201.3.48.66%200%201.32-.5.66-.5.66-1.48%200-.98-.62-1.46-.62-.48-1.34-.48-.72%200-1.34.5-.62.5-.62%201.48%200%20.96.64%201.46zm11.412-1.44q0%20.84.56%201.32.56.46%201.18.46.64%200%201.18-.36.56-.38.9-.38.6%200%201.46%201.06.46.58.46%201.04%200%20.76-1.1%201.42-1.14.8-2.8.8-1.86%200-3.58-1.34-.82-.64-1.34-1.7-.52-1.08-.52-2.36%200-1.3.52-2.34.52-1.06%201.34-1.7%201.66-1.32%203.54-1.32.76%200%201.48.22.72.2%201.06.4l.32.2q.36.24.56.38.52.4.52.92%200%20.5-.42%201.14-.72%201.1-1.38%201.1-.38%200-1.08-.44-.36-.34-1.04-.34-.66%200-1.24.48-.58.48-.58%201.34z%22%20fill%3D%22green%22/%3E%3C/svg%3E"/>
        </a>
</div>
    </nav>
    <main class="pdoc">
            <section class="module-info">
                    <h1 class="modulename">
dcatoolkit    </h1>

                
                        <input id="mod-dcatoolkit-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">

                        <label class="view-source-button" for="mod-dcatoolkit-view-source"><span>View Source</span></label>

                        <div class="pdoc-code codehilite"><pre><span></span><span id="L-1"><a href="#L-1"><span class="linenos">1</span></a><span class="n">__version__</span> <span class="o">=</span> <span class="s2">&quot;0.1.9&quot;</span>
</span><span id="L-2"><a href="#L-2"><span class="linenos">2</span></a><span class="kn">from</span> <span class="nn">.representation</span> <span class="kn">import</span> <span class="n">Pairs</span><span class="p">,</span> <span class="n">DirectInformationData</span><span class="p">,</span> <span class="n">StructureInformation</span><span class="p">,</span> <span class="n">ResidueAlignment</span><span class="p">,</span> <span class="n">MMCIFInformation</span><span class="p">,</span> <span class="n">PDBInformation</span>
</span><span id="L-3"><a href="#L-3"><span class="linenos">3</span></a><span class="kn">from</span> <span class="nn">.analytics</span> <span class="kn">import</span> <span class="n">MSATools</span>
</span><span id="L-4"><a href="#L-4"><span class="linenos">4</span></a>
</span><span id="L-5"><a href="#L-5"><span class="linenos">5</span></a><span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;DirectInformationData&#39;</span><span class="p">,</span> <span class="s1">&#39;StructureInformation&#39;</span><span class="p">,</span> <span class="s1">&#39;ResidueAlignment&#39;</span><span class="p">,</span> <span class="s1">&#39;MMCIFInformation&#39;</span><span class="p">,</span> <span class="s1">&#39;PDBInformation&#39;</span><span class="p">,</span> <span class="s1">&#39;MSATools&#39;</span><span class="p">]</span>
</span></pre></div>


            </section>
                <section id="Pairs">
                            <input id="Pairs-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr class">
            
    <span class="def">class</span>
    <span class="name">Pairs</span>:

                <label class="view-source-button" for="Pairs-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#Pairs"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="Pairs-16"><a href="#Pairs-16"><span class="linenos"> 16</span></a><span class="k">class</span> <span class="nc">Pairs</span><span class="p">:</span>
</span><span id="Pairs-17"><a href="#Pairs-17"><span class="linenos"> 17</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs-18"><a href="#Pairs-18"><span class="linenos"> 18</span></a><span class="sd">    Object that contains a representation (as an ndarray) of pairs of entities that are related. This may extend to Direct Information Pairs or Structural contacts, where each residue is one component of the pair.</span>
</span><span id="Pairs-19"><a href="#Pairs-19"><span class="linenos"> 19</span></a><span class="sd">    </span>
</span><span id="Pairs-20"><a href="#Pairs-20"><span class="linenos"> 20</span></a><span class="sd">    Note</span>
</span><span id="Pairs-21"><a href="#Pairs-21"><span class="linenos"> 21</span></a><span class="sd">    ----</span>
</span><span id="Pairs-22"><a href="#Pairs-22"><span class="linenos"> 22</span></a><span class="sd">    Either a filepath or a ndarr has to be specified in order to produce a Pairs representation.</span>
</span><span id="Pairs-23"><a href="#Pairs-23"><span class="linenos"> 23</span></a>
</span><span id="Pairs-24"><a href="#Pairs-24"><span class="linenos"> 24</span></a><span class="sd">    Parameters</span>
</span><span id="Pairs-25"><a href="#Pairs-25"><span class="linenos"> 25</span></a><span class="sd">    ----------</span>
</span><span id="Pairs-26"><a href="#Pairs-26"><span class="linenos"> 26</span></a><span class="sd">    filepath : str, optional</span>
</span><span id="Pairs-27"><a href="#Pairs-27"><span class="linenos"> 27</span></a><span class="sd">        Filepath of the pairs in tabular representation, separated by whitespace between the pair components and newlines between each pair.</span>
</span><span id="Pairs-28"><a href="#Pairs-28"><span class="linenos"> 28</span></a><span class="sd">    ndarr : numpy.ndarray, optional</span>
</span><span id="Pairs-29"><a href="#Pairs-29"><span class="linenos"> 29</span></a><span class="sd">        Populated ndarray that contains pair information.</span>
</span><span id="Pairs-30"><a href="#Pairs-30"><span class="linenos"> 30</span></a><span class="sd">    delimiter : str, optional</span>
</span><span id="Pairs-31"><a href="#Pairs-31"><span class="linenos"> 31</span></a><span class="sd">        String used to specify separator between two pairs. See numpy.loadtxt() for details.</span>
</span><span id="Pairs-32"><a href="#Pairs-32"><span class="linenos"> 32</span></a><span class="sd">    </span>
</span><span id="Pairs-33"><a href="#Pairs-33"><span class="linenos"> 33</span></a><span class="sd">    Attributes</span>
</span><span id="Pairs-34"><a href="#Pairs-34"><span class="linenos"> 34</span></a><span class="sd">    ----------</span>
</span><span id="Pairs-35"><a href="#Pairs-35"><span class="linenos"> 35</span></a><span class="sd">    pairs : numpy.ndarray</span>
</span><span id="Pairs-36"><a href="#Pairs-36"><span class="linenos"> 36</span></a><span class="sd">        Ndarray representation of pairs supplied by the user. This is produced via the np.loadtxt() function.</span>
</span><span id="Pairs-37"><a href="#Pairs-37"><span class="linenos"> 37</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="Pairs-38"><a href="#Pairs-38"><span class="linenos"> 38</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ndarr</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delimiter</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs-39"><a href="#Pairs-39"><span class="linenos"> 39</span></a>        <span class="k">if</span> <span class="p">(</span><span class="n">filepath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">ndarr</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">filepath</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">ndarr</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">):</span>
</span><span id="Pairs-40"><a href="#Pairs-40"><span class="linenos"> 40</span></a>            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Please specify either a filepath or a NumPy array to populate your pairs.&quot;</span><span class="p">)</span>
</span><span id="Pairs-41"><a href="#Pairs-41"><span class="linenos"> 41</span></a>        <span class="k">elif</span> <span class="n">filepath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs-42"><a href="#Pairs-42"><span class="linenos"> 42</span></a>            <span class="k">if</span> <span class="n">delimiter</span><span class="p">:</span>
</span><span id="Pairs-43"><a href="#Pairs-43"><span class="linenos"> 43</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="n">delimiter</span><span class="p">)</span>
</span><span id="Pairs-44"><a href="#Pairs-44"><span class="linenos"> 44</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs-45"><a href="#Pairs-45"><span class="linenos"> 45</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
</span><span id="Pairs-46"><a href="#Pairs-46"><span class="linenos"> 46</span></a>        <span class="k">elif</span> <span class="n">ndarr</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs-47"><a href="#Pairs-47"><span class="linenos"> 47</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">pairs</span> <span class="o">=</span> <span class="n">ndarr</span>
</span><span id="Pairs-48"><a href="#Pairs-48"><span class="linenos"> 48</span></a>    
</span><span id="Pairs-49"><a href="#Pairs-49"><span class="linenos"> 49</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs-50"><a href="#Pairs-50"><span class="linenos"> 50</span></a>    <span class="k">def</span> <span class="nf">load_from_file</span><span class="p">(</span><span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
</span><span id="Pairs-51"><a href="#Pairs-51"><span class="linenos"> 51</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs-52"><a href="#Pairs-52"><span class="linenos"> 52</span></a><span class="sd">        Loads file containing whitespace-delimited data in columns of residues being column 1 and column 2.</span>
</span><span id="Pairs-53"><a href="#Pairs-53"><span class="linenos"> 53</span></a>
</span><span id="Pairs-54"><a href="#Pairs-54"><span class="linenos"> 54</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs-55"><a href="#Pairs-55"><span class="linenos"> 55</span></a><span class="sd">        ----------</span>
</span><span id="Pairs-56"><a href="#Pairs-56"><span class="linenos"> 56</span></a><span class="sd">        filepath : str</span>
</span><span id="Pairs-57"><a href="#Pairs-57"><span class="linenos"> 57</span></a><span class="sd">            Filepath with residue columns corresponding to the indices of first and second components (proteins, chains, etc.) constituting a pair.</span>
</span><span id="Pairs-58"><a href="#Pairs-58"><span class="linenos"> 58</span></a>
</span><span id="Pairs-59"><a href="#Pairs-59"><span class="linenos"> 59</span></a><span class="sd">        Returns</span>
</span><span id="Pairs-60"><a href="#Pairs-60"><span class="linenos"> 60</span></a><span class="sd">        -------</span>
</span><span id="Pairs-61"><a href="#Pairs-61"><span class="linenos"> 61</span></a><span class="sd">        Pairs</span>
</span><span id="Pairs-62"><a href="#Pairs-62"><span class="linenos"> 62</span></a><span class="sd">            Pairs object with a loaded, structured ndarray with dtype=[(&#39;residue1&#39;, int), (&#39;residue2&#39;, int)]</span>
</span><span id="Pairs-63"><a href="#Pairs-63"><span class="linenos"> 63</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs-64"><a href="#Pairs-64"><span class="linenos"> 64</span></a>        <span class="k">return</span> <span class="n">Pairs</span><span class="p">(</span><span class="n">ndarr</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)]))</span>
</span><span id="Pairs-65"><a href="#Pairs-65"><span class="linenos"> 65</span></a>
</span><span id="Pairs-66"><a href="#Pairs-66"><span class="linenos"> 66</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs-67"><a href="#Pairs-67"><span class="linenos"> 67</span></a>    <span class="k">def</span> <span class="nf">load_from_ndarray</span><span class="p">(</span><span class="n">ndarray</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Iterable</span><span class="p">]]):</span>
</span><span id="Pairs-68"><a href="#Pairs-68"><span class="linenos"> 68</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs-69"><a href="#Pairs-69"><span class="linenos"> 69</span></a><span class="sd">        Loads 2d ndarray of residue pairs in columnar format into Pairs object.</span>
</span><span id="Pairs-70"><a href="#Pairs-70"><span class="linenos"> 70</span></a>
</span><span id="Pairs-71"><a href="#Pairs-71"><span class="linenos"> 71</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs-72"><a href="#Pairs-72"><span class="linenos"> 72</span></a><span class="sd">        ----------</span>
</span><span id="Pairs-73"><a href="#Pairs-73"><span class="linenos"> 73</span></a><span class="sd">        ndarray : numpy.ndarray or Iterable of Iterable (excluding dict)</span>
</span><span id="Pairs-74"><a href="#Pairs-74"><span class="linenos"> 74</span></a><span class="sd">            Unstructured ndarray or iterable of iterables with pairs of residue indices with residue1 and residue 2 in separate columns or as two separate elements.</span>
</span><span id="Pairs-75"><a href="#Pairs-75"><span class="linenos"> 75</span></a>
</span><span id="Pairs-76"><a href="#Pairs-76"><span class="linenos"> 76</span></a><span class="sd">        Returns</span>
</span><span id="Pairs-77"><a href="#Pairs-77"><span class="linenos"> 77</span></a><span class="sd">        -------</span>
</span><span id="Pairs-78"><a href="#Pairs-78"><span class="linenos"> 78</span></a><span class="sd">        Pairs</span>
</span><span id="Pairs-79"><a href="#Pairs-79"><span class="linenos"> 79</span></a><span class="sd">            Pairs object with a loaded, structured ndarray with dtype=[(&#39;residue1&#39;, int), (&#39;residue2&#39;, int)]</span>
</span><span id="Pairs-80"><a href="#Pairs-80"><span class="linenos"> 80</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs-81"><a href="#Pairs-81"><span class="linenos"> 81</span></a>        <span class="k">return</span> <span class="n">Pairs</span><span class="p">(</span><span class="n">ndarr</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ndarray</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">)}))</span>
</span><span id="Pairs-82"><a href="#Pairs-82"><span class="linenos"> 82</span></a>    
</span><span id="Pairs-83"><a href="#Pairs-83"><span class="linenos"> 83</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs-84"><a href="#Pairs-84"><span class="linenos"> 84</span></a>    <span class="k">def</span> <span class="nf">mirror_diagonal</span><span class="p">(</span><span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="Pairs-85"><a href="#Pairs-85"><span class="linenos"> 85</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot; </span>
</span><span id="Pairs-86"><a href="#Pairs-86"><span class="linenos"> 86</span></a><span class="sd">        Flip 2D ndarray with 2 columns columnwise. Flips pair positions for diagonal-mirrored representation.</span>
</span><span id="Pairs-87"><a href="#Pairs-87"><span class="linenos"> 87</span></a>
</span><span id="Pairs-88"><a href="#Pairs-88"><span class="linenos"> 88</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs-89"><a href="#Pairs-89"><span class="linenos"> 89</span></a><span class="sd">        ----------</span>
</span><span id="Pairs-90"><a href="#Pairs-90"><span class="linenos"> 90</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs-91"><a href="#Pairs-91"><span class="linenos"> 91</span></a><span class="sd">            2d array with (n, 2) shape.</span>
</span><span id="Pairs-92"><a href="#Pairs-92"><span class="linenos"> 92</span></a>
</span><span id="Pairs-93"><a href="#Pairs-93"><span class="linenos"> 93</span></a><span class="sd">        Returns</span>
</span><span id="Pairs-94"><a href="#Pairs-94"><span class="linenos"> 94</span></a><span class="sd">        -------</span>
</span><span id="Pairs-95"><a href="#Pairs-95"><span class="linenos"> 95</span></a><span class="sd">        numpy.ndarray </span>
</span><span id="Pairs-96"><a href="#Pairs-96"><span class="linenos"> 96</span></a><span class="sd">            Values flipped along the column axis.</span>
</span><span id="Pairs-97"><a href="#Pairs-97"><span class="linenos"> 97</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs-98"><a href="#Pairs-98"><span class="linenos"> 98</span></a>        <span class="k">if</span> <span class="n">pairs</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs-99"><a href="#Pairs-99"><span class="linenos"> 99</span></a>            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">)})</span>
</span><span id="Pairs-100"><a href="#Pairs-100"><span class="linenos">100</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs-101"><a href="#Pairs-101"><span class="linenos">101</span></a>            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">pairs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</span><span id="Pairs-102"><a href="#Pairs-102"><span class="linenos">102</span></a>    
</span><span id="Pairs-103"><a href="#Pairs-103"><span class="linenos">103</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs-104"><a href="#Pairs-104"><span class="linenos">104</span></a>    <span class="k">def</span> <span class="nf">subset_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">number</span> <span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="Pairs-105"><a href="#Pairs-105"><span class="linenos">105</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs-106"><a href="#Pairs-106"><span class="linenos">106</span></a><span class="sd">        Picks out a subset of &#39;number&#39; pairs if a number is supplied. Otherwise, returns all pairs.</span>
</span><span id="Pairs-107"><a href="#Pairs-107"><span class="linenos">107</span></a>
</span><span id="Pairs-108"><a href="#Pairs-108"><span class="linenos">108</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs-109"><a href="#Pairs-109"><span class="linenos">109</span></a><span class="sd">        ----------</span>
</span><span id="Pairs-110"><a href="#Pairs-110"><span class="linenos">110</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs-111"><a href="#Pairs-111"><span class="linenos">111</span></a><span class="sd">            Ndarray to select number of rows from.</span>
</span><span id="Pairs-112"><a href="#Pairs-112"><span class="linenos">112</span></a><span class="sd">        number :  int, optional</span>
</span><span id="Pairs-113"><a href="#Pairs-113"><span class="linenos">113</span></a><span class="sd">            Specific number of rows of pairs to subset.</span>
</span><span id="Pairs-114"><a href="#Pairs-114"><span class="linenos">114</span></a>
</span><span id="Pairs-115"><a href="#Pairs-115"><span class="linenos">115</span></a><span class="sd">        Returns</span>
</span><span id="Pairs-116"><a href="#Pairs-116"><span class="linenos">116</span></a><span class="sd">        -------</span>
</span><span id="Pairs-117"><a href="#Pairs-117"><span class="linenos">117</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="Pairs-118"><a href="#Pairs-118"><span class="linenos">118</span></a><span class="sd">            Subset of pairs from rows 0 to number.</span>
</span><span id="Pairs-119"><a href="#Pairs-119"><span class="linenos">119</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs-120"><a href="#Pairs-120"><span class="linenos">120</span></a><span class="sd">            All pairs specified from the parameters section.</span>
</span><span id="Pairs-121"><a href="#Pairs-121"><span class="linenos">121</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs-122"><a href="#Pairs-122"><span class="linenos">122</span></a>        <span class="k">if</span> <span class="n">number</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs-123"><a href="#Pairs-123"><span class="linenos">123</span></a>            <span class="k">return</span> <span class="n">pairs</span><span class="p">[:</span><span class="n">number</span><span class="p">,</span> <span class="p">]</span>
</span><span id="Pairs-124"><a href="#Pairs-124"><span class="linenos">124</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs-125"><a href="#Pairs-125"><span class="linenos">125</span></a>            <span class="k">return</span> <span class="n">pairs</span>
</span><span id="Pairs-126"><a href="#Pairs-126"><span class="linenos">126</span></a>    
</span><span id="Pairs-127"><a href="#Pairs-127"><span class="linenos">127</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs-128"><a href="#Pairs-128"><span class="linenos">128</span></a>    <span class="k">def</span> <span class="nf">mirror_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">mirror</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="Pairs-129"><a href="#Pairs-129"><span class="linenos">129</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs-130"><a href="#Pairs-130"><span class="linenos">130</span></a><span class="sd">        Produces combined array of pairs and potentially their mirrored representation.</span>
</span><span id="Pairs-131"><a href="#Pairs-131"><span class="linenos">131</span></a>
</span><span id="Pairs-132"><a href="#Pairs-132"><span class="linenos">132</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs-133"><a href="#Pairs-133"><span class="linenos">133</span></a><span class="sd">        ----------</span>
</span><span id="Pairs-134"><a href="#Pairs-134"><span class="linenos">134</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs-135"><a href="#Pairs-135"><span class="linenos">135</span></a><span class="sd">            Ndarray to mirror and vertically append if mirror is set to True.</span>
</span><span id="Pairs-136"><a href="#Pairs-136"><span class="linenos">136</span></a><span class="sd">        mirror : bool</span>
</span><span id="Pairs-137"><a href="#Pairs-137"><span class="linenos">137</span></a><span class="sd">            Whether or not to append mirrored representation of pairs to the original pairs ndarray.</span>
</span><span id="Pairs-138"><a href="#Pairs-138"><span class="linenos">138</span></a>
</span><span id="Pairs-139"><a href="#Pairs-139"><span class="linenos">139</span></a><span class="sd">        Returns</span>
</span><span id="Pairs-140"><a href="#Pairs-140"><span class="linenos">140</span></a><span class="sd">        -------</span>
</span><span id="Pairs-141"><a href="#Pairs-141"><span class="linenos">141</span></a><span class="sd">        mirrored_ndarray : numpy.ndarray</span>
</span><span id="Pairs-142"><a href="#Pairs-142"><span class="linenos">142</span></a><span class="sd">            combined ndarray of pairs and mirrored pairs.</span>
</span><span id="Pairs-143"><a href="#Pairs-143"><span class="linenos">143</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs-144"><a href="#Pairs-144"><span class="linenos">144</span></a><span class="sd">            The original pairs specified from the parameters section.</span>
</span><span id="Pairs-145"><a href="#Pairs-145"><span class="linenos">145</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs-146"><a href="#Pairs-146"><span class="linenos">146</span></a>        <span class="k">if</span> <span class="n">mirror</span><span class="p">:</span>
</span><span id="Pairs-147"><a href="#Pairs-147"><span class="linenos">147</span></a>            <span class="k">if</span> <span class="n">pairs</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs-148"><a href="#Pairs-148"><span class="linenos">148</span></a>                <span class="n">unstruc_pairs</span> <span class="o">=</span> <span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">]</span>
</span><span id="Pairs-149"><a href="#Pairs-149"><span class="linenos">149</span></a>                <span class="n">unstruc_pairs_mirrored</span> <span class="o">=</span> <span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">]</span>
</span><span id="Pairs-150"><a href="#Pairs-150"><span class="linenos">150</span></a>                <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">unstruc_pairs</span> <span class="o">+</span> <span class="n">unstruc_pairs_mirrored</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)])</span>     
</span><span id="Pairs-151"><a href="#Pairs-151"><span class="linenos">151</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs-152"><a href="#Pairs-152"><span class="linenos">152</span></a>                <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">pairs</span><span class="p">,</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">mirror_diagonal</span><span class="p">(</span><span class="n">pairs</span><span class="p">)))</span>
</span><span id="Pairs-153"><a href="#Pairs-153"><span class="linenos">153</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs-154"><a href="#Pairs-154"><span class="linenos">154</span></a>            <span class="k">return</span> <span class="n">pairs</span>
</span><span id="Pairs-155"><a href="#Pairs-155"><span class="linenos">155</span></a>    
</span><span id="Pairs-156"><a href="#Pairs-156"><span class="linenos">156</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs-157"><a href="#Pairs-157"><span class="linenos">157</span></a>    <span class="k">def</span> <span class="nf">get_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">mirror</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">number</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="Pairs-158"><a href="#Pairs-158"><span class="linenos">158</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs-159"><a href="#Pairs-159"><span class="linenos">159</span></a><span class="sd">        Returns pairs based on user specification, offering options to produce mirrored representation of pairs and to select a specific number of pairs.</span>
</span><span id="Pairs-160"><a href="#Pairs-160"><span class="linenos">160</span></a>
</span><span id="Pairs-161"><a href="#Pairs-161"><span class="linenos">161</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs-162"><a href="#Pairs-162"><span class="linenos">162</span></a><span class="sd">        ----------</span>
</span><span id="Pairs-163"><a href="#Pairs-163"><span class="linenos">163</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs-164"><a href="#Pairs-164"><span class="linenos">164</span></a><span class="sd">            ndarray of pairs to select from or to mirror.</span>
</span><span id="Pairs-165"><a href="#Pairs-165"><span class="linenos">165</span></a><span class="sd">        mirror : bool</span>
</span><span id="Pairs-166"><a href="#Pairs-166"><span class="linenos">166</span></a><span class="sd">            Whether or not to append mirrored representation of pairs to the original pairs ndarray.</span>
</span><span id="Pairs-167"><a href="#Pairs-167"><span class="linenos">167</span></a><span class="sd">        number : int</span>
</span><span id="Pairs-168"><a href="#Pairs-168"><span class="linenos">168</span></a><span class="sd">            Specific number of rows of pairs to subset.</span>
</span><span id="Pairs-169"><a href="#Pairs-169"><span class="linenos">169</span></a>
</span><span id="Pairs-170"><a href="#Pairs-170"><span class="linenos">170</span></a><span class="sd">        Returns</span>
</span><span id="Pairs-171"><a href="#Pairs-171"><span class="linenos">171</span></a><span class="sd">        -------</span>
</span><span id="Pairs-172"><a href="#Pairs-172"><span class="linenos">172</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="Pairs-173"><a href="#Pairs-173"><span class="linenos">173</span></a><span class="sd">            mirrored, subset version of pairs produced via subset_pairs() and mirror_pairs() on pairs.</span>
</span><span id="Pairs-174"><a href="#Pairs-174"><span class="linenos">174</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs-175"><a href="#Pairs-175"><span class="linenos">175</span></a>        <span class="c1"># Check to see if user requested mirrored pairs, if so, add in pairs that are mirrored across diagonal</span>
</span><span id="Pairs-176"><a href="#Pairs-176"><span class="linenos">176</span></a>        <span class="n">pairs</span> <span class="o">=</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">subset_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">,</span> <span class="n">number</span><span class="p">)</span>
</span><span id="Pairs-177"><a href="#Pairs-177"><span class="linenos">177</span></a>        <span class="k">if</span> <span class="n">mirror</span><span class="p">:</span>
</span><span id="Pairs-178"><a href="#Pairs-178"><span class="linenos">178</span></a>            <span class="n">pairs</span> <span class="o">=</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">mirror_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">,</span> <span class="n">mirror</span><span class="p">)</span>
</span><span id="Pairs-179"><a href="#Pairs-179"><span class="linenos">179</span></a>        <span class="k">return</span> <span class="n">pairs</span>
</span></pre></div>


            <div class="docstring"><p>Object that contains a representation (as an ndarray) of pairs of entities that are related. This may extend to Direct Information Pairs or Structural contacts, where each residue is one component of the pair.</p>

<h6 id="note">Note</h6>

<p>Either a filepath or a ndarr has to be specified in order to produce a Pairs representation.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>filepath</strong> (str, optional):
Filepath of the pairs in tabular representation, separated by whitespace between the pair components and newlines between each pair.</li>
<li><strong>ndarr</strong> (numpy.ndarray, optional):
Populated ndarray that contains pair information.</li>
<li><strong>delimiter</strong> (str, optional):
String used to specify separator between two pairs. See numpy.loadtxt() for details.</li>
</ul>

<h6 id="attributes">Attributes</h6>

<ul>
<li><strong>pairs</strong> (numpy.ndarray):
Ndarray representation of pairs supplied by the user. This is produced via the np.loadtxt() function.</li>
</ul>
</div>


                            <div id="Pairs.__init__" class="classattr">
                                        <input id="Pairs.__init__-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="name">Pairs</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">filepath</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>,</span><span class="param">	<span class="n">ndarr</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span>,</span><span class="param">	<span class="n">delimiter</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span></span>)</span>

                <label class="view-source-button" for="Pairs.__init__-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#Pairs.__init__"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="Pairs.__init__-38"><a href="#Pairs.__init__-38"><span class="linenos">38</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ndarr</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delimiter</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs.__init__-39"><a href="#Pairs.__init__-39"><span class="linenos">39</span></a>        <span class="k">if</span> <span class="p">(</span><span class="n">filepath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">ndarr</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">filepath</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">ndarr</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">):</span>
</span><span id="Pairs.__init__-40"><a href="#Pairs.__init__-40"><span class="linenos">40</span></a>            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Please specify either a filepath or a NumPy array to populate your pairs.&quot;</span><span class="p">)</span>
</span><span id="Pairs.__init__-41"><a href="#Pairs.__init__-41"><span class="linenos">41</span></a>        <span class="k">elif</span> <span class="n">filepath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs.__init__-42"><a href="#Pairs.__init__-42"><span class="linenos">42</span></a>            <span class="k">if</span> <span class="n">delimiter</span><span class="p">:</span>
</span><span id="Pairs.__init__-43"><a href="#Pairs.__init__-43"><span class="linenos">43</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="n">delimiter</span><span class="p">)</span>
</span><span id="Pairs.__init__-44"><a href="#Pairs.__init__-44"><span class="linenos">44</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs.__init__-45"><a href="#Pairs.__init__-45"><span class="linenos">45</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">pairs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
</span><span id="Pairs.__init__-46"><a href="#Pairs.__init__-46"><span class="linenos">46</span></a>        <span class="k">elif</span> <span class="n">ndarr</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs.__init__-47"><a href="#Pairs.__init__-47"><span class="linenos">47</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">pairs</span> <span class="o">=</span> <span class="n">ndarr</span>
</span></pre></div>


    

                            </div>
                            <div id="Pairs.load_from_file" class="classattr">
                                        <input id="Pairs.load_from_file-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">load_from_file</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">):</span></span>

                <label class="view-source-button" for="Pairs.load_from_file-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#Pairs.load_from_file"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="Pairs.load_from_file-49"><a href="#Pairs.load_from_file-49"><span class="linenos">49</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs.load_from_file-50"><a href="#Pairs.load_from_file-50"><span class="linenos">50</span></a>    <span class="k">def</span> <span class="nf">load_from_file</span><span class="p">(</span><span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
</span><span id="Pairs.load_from_file-51"><a href="#Pairs.load_from_file-51"><span class="linenos">51</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs.load_from_file-52"><a href="#Pairs.load_from_file-52"><span class="linenos">52</span></a><span class="sd">        Loads file containing whitespace-delimited data in columns of residues being column 1 and column 2.</span>
</span><span id="Pairs.load_from_file-53"><a href="#Pairs.load_from_file-53"><span class="linenos">53</span></a>
</span><span id="Pairs.load_from_file-54"><a href="#Pairs.load_from_file-54"><span class="linenos">54</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs.load_from_file-55"><a href="#Pairs.load_from_file-55"><span class="linenos">55</span></a><span class="sd">        ----------</span>
</span><span id="Pairs.load_from_file-56"><a href="#Pairs.load_from_file-56"><span class="linenos">56</span></a><span class="sd">        filepath : str</span>
</span><span id="Pairs.load_from_file-57"><a href="#Pairs.load_from_file-57"><span class="linenos">57</span></a><span class="sd">            Filepath with residue columns corresponding to the indices of first and second components (proteins, chains, etc.) constituting a pair.</span>
</span><span id="Pairs.load_from_file-58"><a href="#Pairs.load_from_file-58"><span class="linenos">58</span></a>
</span><span id="Pairs.load_from_file-59"><a href="#Pairs.load_from_file-59"><span class="linenos">59</span></a><span class="sd">        Returns</span>
</span><span id="Pairs.load_from_file-60"><a href="#Pairs.load_from_file-60"><span class="linenos">60</span></a><span class="sd">        -------</span>
</span><span id="Pairs.load_from_file-61"><a href="#Pairs.load_from_file-61"><span class="linenos">61</span></a><span class="sd">        Pairs</span>
</span><span id="Pairs.load_from_file-62"><a href="#Pairs.load_from_file-62"><span class="linenos">62</span></a><span class="sd">            Pairs object with a loaded, structured ndarray with dtype=[(&#39;residue1&#39;, int), (&#39;residue2&#39;, int)]</span>
</span><span id="Pairs.load_from_file-63"><a href="#Pairs.load_from_file-63"><span class="linenos">63</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs.load_from_file-64"><a href="#Pairs.load_from_file-64"><span class="linenos">64</span></a>        <span class="k">return</span> <span class="n">Pairs</span><span class="p">(</span><span class="n">ndarr</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)]))</span>
</span></pre></div>


            <div class="docstring"><p>Loads file containing whitespace-delimited data in columns of residues being column 1 and column 2.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>filepath</strong> (str):
Filepath with residue columns corresponding to the indices of first and second components (proteins, chains, etc.) constituting a pair.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>Pairs</strong>: Pairs object with a loaded, structured ndarray with dtype=[('residue1', int), ('residue2', int)]</li>
</ul>
</div>


                            </div>
                            <div id="Pairs.load_from_ndarray" class="classattr">
                                        <input id="Pairs.load_from_ndarray-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">load_from_ndarray</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">ndarray</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]],</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Iterable</span><span class="p">]]</span></span><span class="return-annotation">):</span></span>

                <label class="view-source-button" for="Pairs.load_from_ndarray-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#Pairs.load_from_ndarray"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="Pairs.load_from_ndarray-66"><a href="#Pairs.load_from_ndarray-66"><span class="linenos">66</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs.load_from_ndarray-67"><a href="#Pairs.load_from_ndarray-67"><span class="linenos">67</span></a>    <span class="k">def</span> <span class="nf">load_from_ndarray</span><span class="p">(</span><span class="n">ndarray</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Iterable</span><span class="p">]]):</span>
</span><span id="Pairs.load_from_ndarray-68"><a href="#Pairs.load_from_ndarray-68"><span class="linenos">68</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs.load_from_ndarray-69"><a href="#Pairs.load_from_ndarray-69"><span class="linenos">69</span></a><span class="sd">        Loads 2d ndarray of residue pairs in columnar format into Pairs object.</span>
</span><span id="Pairs.load_from_ndarray-70"><a href="#Pairs.load_from_ndarray-70"><span class="linenos">70</span></a>
</span><span id="Pairs.load_from_ndarray-71"><a href="#Pairs.load_from_ndarray-71"><span class="linenos">71</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs.load_from_ndarray-72"><a href="#Pairs.load_from_ndarray-72"><span class="linenos">72</span></a><span class="sd">        ----------</span>
</span><span id="Pairs.load_from_ndarray-73"><a href="#Pairs.load_from_ndarray-73"><span class="linenos">73</span></a><span class="sd">        ndarray : numpy.ndarray or Iterable of Iterable (excluding dict)</span>
</span><span id="Pairs.load_from_ndarray-74"><a href="#Pairs.load_from_ndarray-74"><span class="linenos">74</span></a><span class="sd">            Unstructured ndarray or iterable of iterables with pairs of residue indices with residue1 and residue 2 in separate columns or as two separate elements.</span>
</span><span id="Pairs.load_from_ndarray-75"><a href="#Pairs.load_from_ndarray-75"><span class="linenos">75</span></a>
</span><span id="Pairs.load_from_ndarray-76"><a href="#Pairs.load_from_ndarray-76"><span class="linenos">76</span></a><span class="sd">        Returns</span>
</span><span id="Pairs.load_from_ndarray-77"><a href="#Pairs.load_from_ndarray-77"><span class="linenos">77</span></a><span class="sd">        -------</span>
</span><span id="Pairs.load_from_ndarray-78"><a href="#Pairs.load_from_ndarray-78"><span class="linenos">78</span></a><span class="sd">        Pairs</span>
</span><span id="Pairs.load_from_ndarray-79"><a href="#Pairs.load_from_ndarray-79"><span class="linenos">79</span></a><span class="sd">            Pairs object with a loaded, structured ndarray with dtype=[(&#39;residue1&#39;, int), (&#39;residue2&#39;, int)]</span>
</span><span id="Pairs.load_from_ndarray-80"><a href="#Pairs.load_from_ndarray-80"><span class="linenos">80</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs.load_from_ndarray-81"><a href="#Pairs.load_from_ndarray-81"><span class="linenos">81</span></a>        <span class="k">return</span> <span class="n">Pairs</span><span class="p">(</span><span class="n">ndarr</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ndarray</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">)}))</span>
</span></pre></div>


            <div class="docstring"><p>Loads 2d ndarray of residue pairs in columnar format into Pairs object.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>ndarray</strong> (numpy.ndarray or Iterable of Iterable (excluding dict)):
Unstructured ndarray or iterable of iterables with pairs of residue indices with residue1 and residue 2 in separate columns or as two separate elements.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>Pairs</strong>: Pairs object with a loaded, structured ndarray with dtype=[('residue1', int), ('residue2', int)]</li>
</ul>
</div>


                            </div>
                            <div id="Pairs.mirror_diagonal" class="classattr">
                                        <input id="Pairs.mirror_diagonal-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">mirror_diagonal</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">pairs</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="Pairs.mirror_diagonal-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#Pairs.mirror_diagonal"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="Pairs.mirror_diagonal-83"><a href="#Pairs.mirror_diagonal-83"><span class="linenos"> 83</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs.mirror_diagonal-84"><a href="#Pairs.mirror_diagonal-84"><span class="linenos"> 84</span></a>    <span class="k">def</span> <span class="nf">mirror_diagonal</span><span class="p">(</span><span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="Pairs.mirror_diagonal-85"><a href="#Pairs.mirror_diagonal-85"><span class="linenos"> 85</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot; </span>
</span><span id="Pairs.mirror_diagonal-86"><a href="#Pairs.mirror_diagonal-86"><span class="linenos"> 86</span></a><span class="sd">        Flip 2D ndarray with 2 columns columnwise. Flips pair positions for diagonal-mirrored representation.</span>
</span><span id="Pairs.mirror_diagonal-87"><a href="#Pairs.mirror_diagonal-87"><span class="linenos"> 87</span></a>
</span><span id="Pairs.mirror_diagonal-88"><a href="#Pairs.mirror_diagonal-88"><span class="linenos"> 88</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs.mirror_diagonal-89"><a href="#Pairs.mirror_diagonal-89"><span class="linenos"> 89</span></a><span class="sd">        ----------</span>
</span><span id="Pairs.mirror_diagonal-90"><a href="#Pairs.mirror_diagonal-90"><span class="linenos"> 90</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs.mirror_diagonal-91"><a href="#Pairs.mirror_diagonal-91"><span class="linenos"> 91</span></a><span class="sd">            2d array with (n, 2) shape.</span>
</span><span id="Pairs.mirror_diagonal-92"><a href="#Pairs.mirror_diagonal-92"><span class="linenos"> 92</span></a>
</span><span id="Pairs.mirror_diagonal-93"><a href="#Pairs.mirror_diagonal-93"><span class="linenos"> 93</span></a><span class="sd">        Returns</span>
</span><span id="Pairs.mirror_diagonal-94"><a href="#Pairs.mirror_diagonal-94"><span class="linenos"> 94</span></a><span class="sd">        -------</span>
</span><span id="Pairs.mirror_diagonal-95"><a href="#Pairs.mirror_diagonal-95"><span class="linenos"> 95</span></a><span class="sd">        numpy.ndarray </span>
</span><span id="Pairs.mirror_diagonal-96"><a href="#Pairs.mirror_diagonal-96"><span class="linenos"> 96</span></a><span class="sd">            Values flipped along the column axis.</span>
</span><span id="Pairs.mirror_diagonal-97"><a href="#Pairs.mirror_diagonal-97"><span class="linenos"> 97</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs.mirror_diagonal-98"><a href="#Pairs.mirror_diagonal-98"><span class="linenos"> 98</span></a>        <span class="k">if</span> <span class="n">pairs</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs.mirror_diagonal-99"><a href="#Pairs.mirror_diagonal-99"><span class="linenos"> 99</span></a>            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">)})</span>
</span><span id="Pairs.mirror_diagonal-100"><a href="#Pairs.mirror_diagonal-100"><span class="linenos">100</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs.mirror_diagonal-101"><a href="#Pairs.mirror_diagonal-101"><span class="linenos">101</span></a>            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">pairs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Flip 2D ndarray with 2 columns columnwise. Flips pair positions for diagonal-mirrored representation.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>pairs</strong> (numpy.ndarray):
2d array with (n, 2) shape.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>numpy.ndarray</strong>: Values flipped along the column axis.</li>
</ul>
</div>


                            </div>
                            <div id="Pairs.subset_pairs" class="classattr">
                                        <input id="Pairs.subset_pairs-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">subset_pairs</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">pairs</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>,</span><span class="param">	<span class="n">number</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="Pairs.subset_pairs-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#Pairs.subset_pairs"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="Pairs.subset_pairs-103"><a href="#Pairs.subset_pairs-103"><span class="linenos">103</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs.subset_pairs-104"><a href="#Pairs.subset_pairs-104"><span class="linenos">104</span></a>    <span class="k">def</span> <span class="nf">subset_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">number</span> <span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="Pairs.subset_pairs-105"><a href="#Pairs.subset_pairs-105"><span class="linenos">105</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs.subset_pairs-106"><a href="#Pairs.subset_pairs-106"><span class="linenos">106</span></a><span class="sd">        Picks out a subset of &#39;number&#39; pairs if a number is supplied. Otherwise, returns all pairs.</span>
</span><span id="Pairs.subset_pairs-107"><a href="#Pairs.subset_pairs-107"><span class="linenos">107</span></a>
</span><span id="Pairs.subset_pairs-108"><a href="#Pairs.subset_pairs-108"><span class="linenos">108</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs.subset_pairs-109"><a href="#Pairs.subset_pairs-109"><span class="linenos">109</span></a><span class="sd">        ----------</span>
</span><span id="Pairs.subset_pairs-110"><a href="#Pairs.subset_pairs-110"><span class="linenos">110</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs.subset_pairs-111"><a href="#Pairs.subset_pairs-111"><span class="linenos">111</span></a><span class="sd">            Ndarray to select number of rows from.</span>
</span><span id="Pairs.subset_pairs-112"><a href="#Pairs.subset_pairs-112"><span class="linenos">112</span></a><span class="sd">        number :  int, optional</span>
</span><span id="Pairs.subset_pairs-113"><a href="#Pairs.subset_pairs-113"><span class="linenos">113</span></a><span class="sd">            Specific number of rows of pairs to subset.</span>
</span><span id="Pairs.subset_pairs-114"><a href="#Pairs.subset_pairs-114"><span class="linenos">114</span></a>
</span><span id="Pairs.subset_pairs-115"><a href="#Pairs.subset_pairs-115"><span class="linenos">115</span></a><span class="sd">        Returns</span>
</span><span id="Pairs.subset_pairs-116"><a href="#Pairs.subset_pairs-116"><span class="linenos">116</span></a><span class="sd">        -------</span>
</span><span id="Pairs.subset_pairs-117"><a href="#Pairs.subset_pairs-117"><span class="linenos">117</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="Pairs.subset_pairs-118"><a href="#Pairs.subset_pairs-118"><span class="linenos">118</span></a><span class="sd">            Subset of pairs from rows 0 to number.</span>
</span><span id="Pairs.subset_pairs-119"><a href="#Pairs.subset_pairs-119"><span class="linenos">119</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs.subset_pairs-120"><a href="#Pairs.subset_pairs-120"><span class="linenos">120</span></a><span class="sd">            All pairs specified from the parameters section.</span>
</span><span id="Pairs.subset_pairs-121"><a href="#Pairs.subset_pairs-121"><span class="linenos">121</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs.subset_pairs-122"><a href="#Pairs.subset_pairs-122"><span class="linenos">122</span></a>        <span class="k">if</span> <span class="n">number</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs.subset_pairs-123"><a href="#Pairs.subset_pairs-123"><span class="linenos">123</span></a>            <span class="k">return</span> <span class="n">pairs</span><span class="p">[:</span><span class="n">number</span><span class="p">,</span> <span class="p">]</span>
</span><span id="Pairs.subset_pairs-124"><a href="#Pairs.subset_pairs-124"><span class="linenos">124</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs.subset_pairs-125"><a href="#Pairs.subset_pairs-125"><span class="linenos">125</span></a>            <span class="k">return</span> <span class="n">pairs</span>
</span></pre></div>


            <div class="docstring"><p>Picks out a subset of 'number' pairs if a number is supplied. Otherwise, returns all pairs.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>pairs</strong> (numpy.ndarray):
Ndarray to select number of rows from.</li>
<li><strong>number</strong> (int, optional):
Specific number of rows of pairs to subset.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>numpy.ndarray</strong>: Subset of pairs from rows 0 to number.</li>
<li><strong>pairs</strong> (numpy.ndarray):
All pairs specified from the parameters section.</li>
</ul>
</div>


                            </div>
                            <div id="Pairs.mirror_pairs" class="classattr">
                                        <input id="Pairs.mirror_pairs-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">mirror_pairs</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">pairs</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>,</span><span class="param">	<span class="n">mirror</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="Pairs.mirror_pairs-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#Pairs.mirror_pairs"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="Pairs.mirror_pairs-127"><a href="#Pairs.mirror_pairs-127"><span class="linenos">127</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs.mirror_pairs-128"><a href="#Pairs.mirror_pairs-128"><span class="linenos">128</span></a>    <span class="k">def</span> <span class="nf">mirror_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">mirror</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="Pairs.mirror_pairs-129"><a href="#Pairs.mirror_pairs-129"><span class="linenos">129</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs.mirror_pairs-130"><a href="#Pairs.mirror_pairs-130"><span class="linenos">130</span></a><span class="sd">        Produces combined array of pairs and potentially their mirrored representation.</span>
</span><span id="Pairs.mirror_pairs-131"><a href="#Pairs.mirror_pairs-131"><span class="linenos">131</span></a>
</span><span id="Pairs.mirror_pairs-132"><a href="#Pairs.mirror_pairs-132"><span class="linenos">132</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs.mirror_pairs-133"><a href="#Pairs.mirror_pairs-133"><span class="linenos">133</span></a><span class="sd">        ----------</span>
</span><span id="Pairs.mirror_pairs-134"><a href="#Pairs.mirror_pairs-134"><span class="linenos">134</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs.mirror_pairs-135"><a href="#Pairs.mirror_pairs-135"><span class="linenos">135</span></a><span class="sd">            Ndarray to mirror and vertically append if mirror is set to True.</span>
</span><span id="Pairs.mirror_pairs-136"><a href="#Pairs.mirror_pairs-136"><span class="linenos">136</span></a><span class="sd">        mirror : bool</span>
</span><span id="Pairs.mirror_pairs-137"><a href="#Pairs.mirror_pairs-137"><span class="linenos">137</span></a><span class="sd">            Whether or not to append mirrored representation of pairs to the original pairs ndarray.</span>
</span><span id="Pairs.mirror_pairs-138"><a href="#Pairs.mirror_pairs-138"><span class="linenos">138</span></a>
</span><span id="Pairs.mirror_pairs-139"><a href="#Pairs.mirror_pairs-139"><span class="linenos">139</span></a><span class="sd">        Returns</span>
</span><span id="Pairs.mirror_pairs-140"><a href="#Pairs.mirror_pairs-140"><span class="linenos">140</span></a><span class="sd">        -------</span>
</span><span id="Pairs.mirror_pairs-141"><a href="#Pairs.mirror_pairs-141"><span class="linenos">141</span></a><span class="sd">        mirrored_ndarray : numpy.ndarray</span>
</span><span id="Pairs.mirror_pairs-142"><a href="#Pairs.mirror_pairs-142"><span class="linenos">142</span></a><span class="sd">            combined ndarray of pairs and mirrored pairs.</span>
</span><span id="Pairs.mirror_pairs-143"><a href="#Pairs.mirror_pairs-143"><span class="linenos">143</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs.mirror_pairs-144"><a href="#Pairs.mirror_pairs-144"><span class="linenos">144</span></a><span class="sd">            The original pairs specified from the parameters section.</span>
</span><span id="Pairs.mirror_pairs-145"><a href="#Pairs.mirror_pairs-145"><span class="linenos">145</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs.mirror_pairs-146"><a href="#Pairs.mirror_pairs-146"><span class="linenos">146</span></a>        <span class="k">if</span> <span class="n">mirror</span><span class="p">:</span>
</span><span id="Pairs.mirror_pairs-147"><a href="#Pairs.mirror_pairs-147"><span class="linenos">147</span></a>            <span class="k">if</span> <span class="n">pairs</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="Pairs.mirror_pairs-148"><a href="#Pairs.mirror_pairs-148"><span class="linenos">148</span></a>                <span class="n">unstruc_pairs</span> <span class="o">=</span> <span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">]</span>
</span><span id="Pairs.mirror_pairs-149"><a href="#Pairs.mirror_pairs-149"><span class="linenos">149</span></a>                <span class="n">unstruc_pairs_mirrored</span> <span class="o">=</span> <span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">]</span>
</span><span id="Pairs.mirror_pairs-150"><a href="#Pairs.mirror_pairs-150"><span class="linenos">150</span></a>                <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">unstruc_pairs</span> <span class="o">+</span> <span class="n">unstruc_pairs_mirrored</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)])</span>     
</span><span id="Pairs.mirror_pairs-151"><a href="#Pairs.mirror_pairs-151"><span class="linenos">151</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs.mirror_pairs-152"><a href="#Pairs.mirror_pairs-152"><span class="linenos">152</span></a>                <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">pairs</span><span class="p">,</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">mirror_diagonal</span><span class="p">(</span><span class="n">pairs</span><span class="p">)))</span>
</span><span id="Pairs.mirror_pairs-153"><a href="#Pairs.mirror_pairs-153"><span class="linenos">153</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="Pairs.mirror_pairs-154"><a href="#Pairs.mirror_pairs-154"><span class="linenos">154</span></a>            <span class="k">return</span> <span class="n">pairs</span>
</span></pre></div>


            <div class="docstring"><p>Produces combined array of pairs and potentially their mirrored representation.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>pairs</strong> (numpy.ndarray):
Ndarray to mirror and vertically append if mirror is set to True.</li>
<li><strong>mirror</strong> (bool):
Whether or not to append mirrored representation of pairs to the original pairs ndarray.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>mirrored_ndarray</strong> (numpy.ndarray):
combined ndarray of pairs and mirrored pairs.</li>
<li><strong>pairs</strong> (numpy.ndarray):
The original pairs specified from the parameters section.</li>
</ul>
</div>


                            </div>
                            <div id="Pairs.get_pairs" class="classattr">
                                        <input id="Pairs.get_pairs-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">get_pairs</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">pairs</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>,</span><span class="param">	<span class="n">mirror</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>,</span><span class="param">	<span class="n">number</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="Pairs.get_pairs-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#Pairs.get_pairs"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="Pairs.get_pairs-156"><a href="#Pairs.get_pairs-156"><span class="linenos">156</span></a>    <span class="nd">@staticmethod</span>
</span><span id="Pairs.get_pairs-157"><a href="#Pairs.get_pairs-157"><span class="linenos">157</span></a>    <span class="k">def</span> <span class="nf">get_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">mirror</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">number</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="Pairs.get_pairs-158"><a href="#Pairs.get_pairs-158"><span class="linenos">158</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="Pairs.get_pairs-159"><a href="#Pairs.get_pairs-159"><span class="linenos">159</span></a><span class="sd">        Returns pairs based on user specification, offering options to produce mirrored representation of pairs and to select a specific number of pairs.</span>
</span><span id="Pairs.get_pairs-160"><a href="#Pairs.get_pairs-160"><span class="linenos">160</span></a>
</span><span id="Pairs.get_pairs-161"><a href="#Pairs.get_pairs-161"><span class="linenos">161</span></a><span class="sd">        Parameters</span>
</span><span id="Pairs.get_pairs-162"><a href="#Pairs.get_pairs-162"><span class="linenos">162</span></a><span class="sd">        ----------</span>
</span><span id="Pairs.get_pairs-163"><a href="#Pairs.get_pairs-163"><span class="linenos">163</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="Pairs.get_pairs-164"><a href="#Pairs.get_pairs-164"><span class="linenos">164</span></a><span class="sd">            ndarray of pairs to select from or to mirror.</span>
</span><span id="Pairs.get_pairs-165"><a href="#Pairs.get_pairs-165"><span class="linenos">165</span></a><span class="sd">        mirror : bool</span>
</span><span id="Pairs.get_pairs-166"><a href="#Pairs.get_pairs-166"><span class="linenos">166</span></a><span class="sd">            Whether or not to append mirrored representation of pairs to the original pairs ndarray.</span>
</span><span id="Pairs.get_pairs-167"><a href="#Pairs.get_pairs-167"><span class="linenos">167</span></a><span class="sd">        number : int</span>
</span><span id="Pairs.get_pairs-168"><a href="#Pairs.get_pairs-168"><span class="linenos">168</span></a><span class="sd">            Specific number of rows of pairs to subset.</span>
</span><span id="Pairs.get_pairs-169"><a href="#Pairs.get_pairs-169"><span class="linenos">169</span></a>
</span><span id="Pairs.get_pairs-170"><a href="#Pairs.get_pairs-170"><span class="linenos">170</span></a><span class="sd">        Returns</span>
</span><span id="Pairs.get_pairs-171"><a href="#Pairs.get_pairs-171"><span class="linenos">171</span></a><span class="sd">        -------</span>
</span><span id="Pairs.get_pairs-172"><a href="#Pairs.get_pairs-172"><span class="linenos">172</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="Pairs.get_pairs-173"><a href="#Pairs.get_pairs-173"><span class="linenos">173</span></a><span class="sd">            mirrored, subset version of pairs produced via subset_pairs() and mirror_pairs() on pairs.</span>
</span><span id="Pairs.get_pairs-174"><a href="#Pairs.get_pairs-174"><span class="linenos">174</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="Pairs.get_pairs-175"><a href="#Pairs.get_pairs-175"><span class="linenos">175</span></a>        <span class="c1"># Check to see if user requested mirrored pairs, if so, add in pairs that are mirrored across diagonal</span>
</span><span id="Pairs.get_pairs-176"><a href="#Pairs.get_pairs-176"><span class="linenos">176</span></a>        <span class="n">pairs</span> <span class="o">=</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">subset_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">,</span> <span class="n">number</span><span class="p">)</span>
</span><span id="Pairs.get_pairs-177"><a href="#Pairs.get_pairs-177"><span class="linenos">177</span></a>        <span class="k">if</span> <span class="n">mirror</span><span class="p">:</span>
</span><span id="Pairs.get_pairs-178"><a href="#Pairs.get_pairs-178"><span class="linenos">178</span></a>            <span class="n">pairs</span> <span class="o">=</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">mirror_pairs</span><span class="p">(</span><span class="n">pairs</span><span class="p">,</span> <span class="n">mirror</span><span class="p">)</span>
</span><span id="Pairs.get_pairs-179"><a href="#Pairs.get_pairs-179"><span class="linenos">179</span></a>        <span class="k">return</span> <span class="n">pairs</span>
</span></pre></div>


            <div class="docstring"><p>Returns pairs based on user specification, offering options to produce mirrored representation of pairs and to select a specific number of pairs.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>pairs</strong> (numpy.ndarray):
ndarray of pairs to select from or to mirror.</li>
<li><strong>mirror</strong> (bool):
Whether or not to append mirrored representation of pairs to the original pairs ndarray.</li>
<li><strong>number</strong> (int):
Specific number of rows of pairs to subset.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>numpy.ndarray</strong>: mirrored, subset version of pairs produced via subset_pairs() and mirror_pairs() on pairs.</li>
</ul>
</div>


                            </div>
                </section>
                <section id="DirectInformationData">
                            <input id="DirectInformationData-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr class">
            
    <span class="def">class</span>
    <span class="name">DirectInformationData</span>:

                <label class="view-source-button" for="DirectInformationData-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData-338"><a href="#DirectInformationData-338"><span class="linenos">338</span></a><span class="k">class</span> <span class="nc">DirectInformationData</span><span class="p">:</span>
</span><span id="DirectInformationData-339"><a href="#DirectInformationData-339"><span class="linenos">339</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-340"><a href="#DirectInformationData-340"><span class="linenos">340</span></a><span class="sd">    Representation and interface for Direct Information data including residue indices for a pair and its corresponding DI value represented as a 3-column ndarray.</span>
</span><span id="DirectInformationData-341"><a href="#DirectInformationData-341"><span class="linenos">341</span></a>
</span><span id="DirectInformationData-342"><a href="#DirectInformationData-342"><span class="linenos">342</span></a><span class="sd">    Parameters</span>
</span><span id="DirectInformationData-343"><a href="#DirectInformationData-343"><span class="linenos">343</span></a><span class="sd">    ----------</span>
</span><span id="DirectInformationData-344"><a href="#DirectInformationData-344"><span class="linenos">344</span></a><span class="sd">    structured_ndarray : numpy.ndarray</span>
</span><span id="DirectInformationData-345"><a href="#DirectInformationData-345"><span class="linenos">345</span></a><span class="sd">        Ndarray with the shape (n,3) with dtype={&#39;names&#39;: (&#39;residue1&#39;, &#39;residue2&#39;, &#39;DI&#39;), &#39;formats&#39;: (int, int, float, float)}</span>
</span><span id="DirectInformationData-346"><a href="#DirectInformationData-346"><span class="linenos">346</span></a>
</span><span id="DirectInformationData-347"><a href="#DirectInformationData-347"><span class="linenos">347</span></a><span class="sd">    Attributes</span>
</span><span id="DirectInformationData-348"><a href="#DirectInformationData-348"><span class="linenos">348</span></a><span class="sd">    ----------</span>
</span><span id="DirectInformationData-349"><a href="#DirectInformationData-349"><span class="linenos">349</span></a><span class="sd">    DI_data : numpy.ndarray</span>
</span><span id="DirectInformationData-350"><a href="#DirectInformationData-350"><span class="linenos">350</span></a><span class="sd">        The structured_ndarray in the parameters section where column 1 corresponds to a pair&#39;s first residue, column 2 corresponds to the pair&#39;s second residue, and column 3 corresponds to the Direct Information of the pair.</span>
</span><span id="DirectInformationData-351"><a href="#DirectInformationData-351"><span class="linenos">351</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-352"><a href="#DirectInformationData-352"><span class="linenos">352</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structured_ndarray</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="DirectInformationData-353"><a href="#DirectInformationData-353"><span class="linenos">353</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">DI_data</span> <span class="o">=</span> <span class="n">structured_ndarray</span>
</span><span id="DirectInformationData-354"><a href="#DirectInformationData-354"><span class="linenos">354</span></a>
</span><span id="DirectInformationData-355"><a href="#DirectInformationData-355"><span class="linenos">355</span></a>    <span class="nd">@staticmethod</span>    
</span><span id="DirectInformationData-356"><a href="#DirectInformationData-356"><span class="linenos">356</span></a>    <span class="k">def</span> <span class="nf">load_from_dca_output</span><span class="p">(</span><span class="n">dca_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DirectInformationData&#39;</span><span class="p">:</span>
</span><span id="DirectInformationData-357"><a href="#DirectInformationData-357"><span class="linenos">357</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-358"><a href="#DirectInformationData-358"><span class="linenos">358</span></a><span class="sd">        Function to generate a DirectInformationData object from the direct output of the MATLab dca function.</span>
</span><span id="DirectInformationData-359"><a href="#DirectInformationData-359"><span class="linenos">359</span></a>
</span><span id="DirectInformationData-360"><a href="#DirectInformationData-360"><span class="linenos">360</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-361"><a href="#DirectInformationData-361"><span class="linenos">361</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-362"><a href="#DirectInformationData-362"><span class="linenos">362</span></a><span class="sd">        dca_filepath : str</span>
</span><span id="DirectInformationData-363"><a href="#DirectInformationData-363"><span class="linenos">363</span></a><span class="sd">            Filepath of the DCA output to be read and compiled into a structured ndarray. DCA output is a 4 column text file with the following columns: (residue 1, residue 2, Mutual Information, Direct Information).</span>
</span><span id="DirectInformationData-364"><a href="#DirectInformationData-364"><span class="linenos">364</span></a>
</span><span id="DirectInformationData-365"><a href="#DirectInformationData-365"><span class="linenos">365</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-366"><a href="#DirectInformationData-366"><span class="linenos">366</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-367"><a href="#DirectInformationData-367"><span class="linenos">367</span></a><span class="sd">        DirectInformationData</span>
</span><span id="DirectInformationData-368"><a href="#DirectInformationData-368"><span class="linenos">368</span></a><span class="sd">            DirectInformationData object with named structured array containing residue indices and the DI value of the pair.</span>
</span><span id="DirectInformationData-369"><a href="#DirectInformationData-369"><span class="linenos">369</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-370"><a href="#DirectInformationData-370"><span class="linenos">370</span></a>        <span class="n">file_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">dca_filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="s1">&#39;MI&#39;</span><span class="p">,</span> <span class="s1">&#39;DI&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">)})</span>
</span><span id="DirectInformationData-371"><a href="#DirectInformationData-371"><span class="linenos">371</span></a>        <span class="k">return</span> <span class="n">DirectInformationData</span><span class="p">(</span><span class="n">file_data</span><span class="p">[[</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="s1">&#39;DI&#39;</span><span class="p">]])</span>
</span><span id="DirectInformationData-372"><a href="#DirectInformationData-372"><span class="linenos">372</span></a>
</span><span id="DirectInformationData-373"><a href="#DirectInformationData-373"><span class="linenos">373</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData-374"><a href="#DirectInformationData-374"><span class="linenos">374</span></a>    <span class="k">def</span> <span class="nf">load_from_DI_file</span><span class="p">(</span><span class="n">DI_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DirectInformationData&#39;</span><span class="p">:</span>
</span><span id="DirectInformationData-375"><a href="#DirectInformationData-375"><span class="linenos">375</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-376"><a href="#DirectInformationData-376"><span class="linenos">376</span></a><span class="sd">        Function to generate a DirectInformationData object from the modified DI-only version of the DCA output generated via the MATLab dca function.</span>
</span><span id="DirectInformationData-377"><a href="#DirectInformationData-377"><span class="linenos">377</span></a>
</span><span id="DirectInformationData-378"><a href="#DirectInformationData-378"><span class="linenos">378</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-379"><a href="#DirectInformationData-379"><span class="linenos">379</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-380"><a href="#DirectInformationData-380"><span class="linenos">380</span></a><span class="sd">        DI_filepath : str</span>
</span><span id="DirectInformationData-381"><a href="#DirectInformationData-381"><span class="linenos">381</span></a><span class="sd">            Filepath of the DI file to be read and compile into a structured ndarray. DI file is a 3 column text file with the following columns: (residue 1, residue 2, Direct Information).</span>
</span><span id="DirectInformationData-382"><a href="#DirectInformationData-382"><span class="linenos">382</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-383"><a href="#DirectInformationData-383"><span class="linenos">383</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-384"><a href="#DirectInformationData-384"><span class="linenos">384</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-385"><a href="#DirectInformationData-385"><span class="linenos">385</span></a><span class="sd">        DirectInformationData</span>
</span><span id="DirectInformationData-386"><a href="#DirectInformationData-386"><span class="linenos">386</span></a><span class="sd">            DirectInformationData object with named structured array containing residue indices and the DI value of the pair.</span>
</span><span id="DirectInformationData-387"><a href="#DirectInformationData-387"><span class="linenos">387</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-388"><a href="#DirectInformationData-388"><span class="linenos">388</span></a>        <span class="k">return</span> <span class="n">DirectInformationData</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">DI_filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="s1">&#39;DI&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">)}))</span>
</span><span id="DirectInformationData-389"><a href="#DirectInformationData-389"><span class="linenos">389</span></a>
</span><span id="DirectInformationData-390"><a href="#DirectInformationData-390"><span class="linenos">390</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData-391"><a href="#DirectInformationData-391"><span class="linenos">391</span></a>    <span class="k">def</span> <span class="nf">load_as_ndarray</span><span class="p">(</span><span class="n">ndarray</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Iterable</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="s1">&#39;DirectInformationData&#39;</span><span class="p">:</span>
</span><span id="DirectInformationData-392"><a href="#DirectInformationData-392"><span class="linenos">392</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-393"><a href="#DirectInformationData-393"><span class="linenos">393</span></a><span class="sd">        Function to generate a Direct Information object from a ndarray.</span>
</span><span id="DirectInformationData-394"><a href="#DirectInformationData-394"><span class="linenos">394</span></a>
</span><span id="DirectInformationData-395"><a href="#DirectInformationData-395"><span class="linenos">395</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-396"><a href="#DirectInformationData-396"><span class="linenos">396</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-397"><a href="#DirectInformationData-397"><span class="linenos">397</span></a><span class="sd">        ndarray : numpy.ndarray or Iterable of Iterable (excluding dict)</span>
</span><span id="DirectInformationData-398"><a href="#DirectInformationData-398"><span class="linenos">398</span></a><span class="sd">            An ndarray of shape (n,3) where its columns are (residue 1, residue 2, and Direct Information). Can also be parsed from an iterable of iterable provided that the aforementioned format is followed.</span>
</span><span id="DirectInformationData-399"><a href="#DirectInformationData-399"><span class="linenos">399</span></a>
</span><span id="DirectInformationData-400"><a href="#DirectInformationData-400"><span class="linenos">400</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-401"><a href="#DirectInformationData-401"><span class="linenos">401</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-402"><a href="#DirectInformationData-402"><span class="linenos">402</span></a><span class="sd">        DirectInformationData</span>
</span><span id="DirectInformationData-403"><a href="#DirectInformationData-403"><span class="linenos">403</span></a><span class="sd">            DirectInformationData object with named structured array containing residue indices and the DI value of the pair.</span>
</span><span id="DirectInformationData-404"><a href="#DirectInformationData-404"><span class="linenos">404</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-405"><a href="#DirectInformationData-405"><span class="linenos">405</span></a>        
</span><span id="DirectInformationData-406"><a href="#DirectInformationData-406"><span class="linenos">406</span></a>        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">and</span> <span class="n">ndarray</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">3</span><span class="p">:</span>
</span><span id="DirectInformationData-407"><a href="#DirectInformationData-407"><span class="linenos">407</span></a>            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Dimensions of numpy array supplied are different from what is expected. Please supply residue1, residue2, and DI column in int, int, float format and with shape of (n, 3).&quot;</span><span class="p">)</span>
</span><span id="DirectInformationData-408"><a href="#DirectInformationData-408"><span class="linenos">408</span></a>        <span class="c1"># Structured ndarrays require list of tuples for conversion.</span>
</span><span id="DirectInformationData-409"><a href="#DirectInformationData-409"><span class="linenos">409</span></a>        <span class="n">DI_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ndarray</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="s1">&#39;DI&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">)})</span>
</span><span id="DirectInformationData-410"><a href="#DirectInformationData-410"><span class="linenos">410</span></a>        
</span><span id="DirectInformationData-411"><a href="#DirectInformationData-411"><span class="linenos">411</span></a>        <span class="k">return</span> <span class="n">DirectInformationData</span><span class="p">(</span><span class="n">DI_data</span><span class="p">)</span>
</span><span id="DirectInformationData-412"><a href="#DirectInformationData-412"><span class="linenos">412</span></a>    
</span><span id="DirectInformationData-413"><a href="#DirectInformationData-413"><span class="linenos">413</span></a>    <span class="k">def</span> <span class="nf">get_ranked_mapped_pairs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">RA1</span><span class="p">:</span> <span class="n">ResidueAlignment</span><span class="p">,</span> <span class="n">RA2</span><span class="p">:</span> <span class="n">ResidueAlignment</span><span class="p">,</span> <span class="n">pairs_only</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">mirror</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">number</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="DirectInformationData-414"><a href="#DirectInformationData-414"><span class="linenos">414</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-415"><a href="#DirectInformationData-415"><span class="linenos">415</span></a><span class="sd">        Uses DirectInformationData and Pairs interface methods to obtain ranked, mapped residues. See rank_pairs() function and map_DIs() function for details on rank and mapping. Residue Alignments can be the same for intra-domain / intra-protein mapping.</span>
</span><span id="DirectInformationData-416"><a href="#DirectInformationData-416"><span class="linenos">416</span></a>
</span><span id="DirectInformationData-417"><a href="#DirectInformationData-417"><span class="linenos">417</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-418"><a href="#DirectInformationData-418"><span class="linenos">418</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-419"><a href="#DirectInformationData-419"><span class="linenos">419</span></a><span class="sd">        RA1 : ResidueAlignment</span>
</span><span id="DirectInformationData-420"><a href="#DirectInformationData-420"><span class="linenos">420</span></a><span class="sd">            The ResidueAlignment used for mapping the first column of residues to the appropriate target sequence.</span>
</span><span id="DirectInformationData-421"><a href="#DirectInformationData-421"><span class="linenos">421</span></a><span class="sd">        RA2 : ResidueAlignment</span>
</span><span id="DirectInformationData-422"><a href="#DirectInformationData-422"><span class="linenos">422</span></a><span class="sd">            The ResidueAlignment used for mapping the second column of residues to the appropriate target sequence.</span>
</span><span id="DirectInformationData-423"><a href="#DirectInformationData-423"><span class="linenos">423</span></a><span class="sd">        pairs_only : bool</span>
</span><span id="DirectInformationData-424"><a href="#DirectInformationData-424"><span class="linenos">424</span></a><span class="sd">            True if the final ndarray should contain only columns 1 and 2, corresponding to the residues that constitute the pair. This would drop the DI column.</span>
</span><span id="DirectInformationData-425"><a href="#DirectInformationData-425"><span class="linenos">425</span></a><span class="sd">        mirror : bool</span>
</span><span id="DirectInformationData-426"><a href="#DirectInformationData-426"><span class="linenos">426</span></a><span class="sd">            See Pairs.mirror_pairs() or get_pairs() for details. NOTE: This option is overriden entirely if pairs_only is False. If true, this will produce an ndarray that has the original residue indices and repeated residue indices but with residue 1 and residue 2 switched. This is useful for plotting across the upper diagonal of a contact map.</span>
</span><span id="DirectInformationData-427"><a href="#DirectInformationData-427"><span class="linenos">427</span></a><span class="sd">        number : int, None</span>
</span><span id="DirectInformationData-428"><a href="#DirectInformationData-428"><span class="linenos">428</span></a><span class="sd">            Number of ranked, mapped pairs to return.</span>
</span><span id="DirectInformationData-429"><a href="#DirectInformationData-429"><span class="linenos">429</span></a>
</span><span id="DirectInformationData-430"><a href="#DirectInformationData-430"><span class="linenos">430</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-431"><a href="#DirectInformationData-431"><span class="linenos">431</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-432"><a href="#DirectInformationData-432"><span class="linenos">432</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="DirectInformationData-433"><a href="#DirectInformationData-433"><span class="linenos">433</span></a><span class="sd">            Structured ndarray with columns residue 1, residue 2 and optionally DI. Only has specified number of pairs if `number` is specified and mirrored pairs if `mirror` is True and pairs_only is False.</span>
</span><span id="DirectInformationData-434"><a href="#DirectInformationData-434"><span class="linenos">434</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-435"><a href="#DirectInformationData-435"><span class="linenos">435</span></a><span class="sd">        Notes</span>
</span><span id="DirectInformationData-436"><a href="#DirectInformationData-436"><span class="linenos">436</span></a><span class="sd">        -----</span>
</span><span id="DirectInformationData-437"><a href="#DirectInformationData-437"><span class="linenos">437</span></a><span class="sd">        ResidueAlignments contain dictionaries like domain_to_protein to map residues produced via Direct Coupling Analysis (DCA) on an MSA generated in context to an HMM. The residues are mapped to a protein structure via alignment of the HMM hit / domain to the protein sequence.</span>
</span><span id="DirectInformationData-438"><a href="#DirectInformationData-438"><span class="linenos">438</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-439"><a href="#DirectInformationData-439"><span class="linenos">439</span></a>        <span class="n">ranked_pairs</span> <span class="o">=</span> <span class="n">DirectInformationData</span><span class="o">.</span><span class="n">rank_pairs</span><span class="p">(</span><span class="n">DirectInformationData</span><span class="o">.</span><span class="n">nonlocal_pairs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DI_data</span><span class="p">))</span>
</span><span id="DirectInformationData-440"><a href="#DirectInformationData-440"><span class="linenos">440</span></a>        <span class="n">ranked_mapped_pairs</span> <span class="o">=</span> <span class="n">DirectInformationData</span><span class="o">.</span><span class="n">map_DIs</span><span class="p">(</span><span class="n">ranked_pairs</span><span class="p">,</span> <span class="n">RA1</span><span class="p">,</span> <span class="n">RA2</span><span class="p">)</span>
</span><span id="DirectInformationData-441"><a href="#DirectInformationData-441"><span class="linenos">441</span></a>        <span class="k">if</span> <span class="n">pairs_only</span><span class="p">:</span>
</span><span id="DirectInformationData-442"><a href="#DirectInformationData-442"><span class="linenos">442</span></a>            <span class="k">return</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">get_pairs</span><span class="p">(</span><span class="n">ranked_mapped_pairs</span><span class="p">[[</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">]],</span> <span class="n">mirror</span><span class="o">=</span><span class="n">mirror</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="n">number</span><span class="p">)</span>
</span><span id="DirectInformationData-443"><a href="#DirectInformationData-443"><span class="linenos">443</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData-444"><a href="#DirectInformationData-444"><span class="linenos">444</span></a>            <span class="k">return</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">get_pairs</span><span class="p">(</span><span class="n">ranked_mapped_pairs</span><span class="p">,</span> <span class="n">mirror</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="n">number</span><span class="p">)</span>
</span><span id="DirectInformationData-445"><a href="#DirectInformationData-445"><span class="linenos">445</span></a>    
</span><span id="DirectInformationData-446"><a href="#DirectInformationData-446"><span class="linenos">446</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData-447"><a href="#DirectInformationData-447"><span class="linenos">447</span></a>    <span class="k">def</span> <span class="nf">map_DIs</span><span class="p">(</span><span class="n">DI_data</span> <span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">RA1</span><span class="p">:</span> <span class="n">ResidueAlignment</span><span class="p">,</span> <span class="n">RA2</span><span class="p">:</span> <span class="n">ResidueAlignment</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="DirectInformationData-448"><a href="#DirectInformationData-448"><span class="linenos">448</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-449"><a href="#DirectInformationData-449"><span class="linenos">449</span></a><span class="sd">        Uses domain-to-protein mappings present in the Residue Alignments provided to generate mapped representations of the residues from the DI_data structured ndarray provided.</span>
</span><span id="DirectInformationData-450"><a href="#DirectInformationData-450"><span class="linenos">450</span></a>
</span><span id="DirectInformationData-451"><a href="#DirectInformationData-451"><span class="linenos">451</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-452"><a href="#DirectInformationData-452"><span class="linenos">452</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-453"><a href="#DirectInformationData-453"><span class="linenos">453</span></a><span class="sd">        DI_data : numpy.ndarray</span>
</span><span id="DirectInformationData-454"><a href="#DirectInformationData-454"><span class="linenos">454</span></a><span class="sd">            Structured ndarray that contains columns &quot;residue1&quot; and &quot;residue2&quot;.</span>
</span><span id="DirectInformationData-455"><a href="#DirectInformationData-455"><span class="linenos">455</span></a><span class="sd">        RA1 : ResidueAlignment</span>
</span><span id="DirectInformationData-456"><a href="#DirectInformationData-456"><span class="linenos">456</span></a><span class="sd">            The ResidueAlignment used for mapping the first column of residues to the appropriate target sequence.</span>
</span><span id="DirectInformationData-457"><a href="#DirectInformationData-457"><span class="linenos">457</span></a><span class="sd">        RA2 : ResidueAlignment</span>
</span><span id="DirectInformationData-458"><a href="#DirectInformationData-458"><span class="linenos">458</span></a><span class="sd">            The ResidueAlignment used for mapping the second column of residues to the appropriate target sequence.</span>
</span><span id="DirectInformationData-459"><a href="#DirectInformationData-459"><span class="linenos">459</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-460"><a href="#DirectInformationData-460"><span class="linenos">460</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-461"><a href="#DirectInformationData-461"><span class="linenos">461</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-462"><a href="#DirectInformationData-462"><span class="linenos">462</span></a><span class="sd">        mappable_DI_data : numpy.ndarray</span>
</span><span id="DirectInformationData-463"><a href="#DirectInformationData-463"><span class="linenos">463</span></a><span class="sd">            DI_data that has been mapped to the target sequence specified in the generation of the corresponding ResidueAlignments.</span>
</span><span id="DirectInformationData-464"><a href="#DirectInformationData-464"><span class="linenos">464</span></a>
</span><span id="DirectInformationData-465"><a href="#DirectInformationData-465"><span class="linenos">465</span></a><span class="sd">        Note:</span>
</span><span id="DirectInformationData-466"><a href="#DirectInformationData-466"><span class="linenos">466</span></a><span class="sd">            Residues that do not map to the target sequence of the ResidueAlignment are dropped.</span>
</span><span id="DirectInformationData-467"><a href="#DirectInformationData-467"><span class="linenos">467</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-468"><a href="#DirectInformationData-468"><span class="linenos">468</span></a>        <span class="c1"># Alternative approach is to just use get function instead of [x] and default to np.nan and drop nans row-wise.</span>
</span><span id="DirectInformationData-469"><a href="#DirectInformationData-469"><span class="linenos">469</span></a>        <span class="n">mapping_key_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">DI_data</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">],</span> <span class="nb">list</span><span class="p">(</span><span class="n">RA1</span><span class="o">.</span><span class="n">domain_to_protein</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">DI_data</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">],</span> <span class="nb">list</span><span class="p">(</span><span class="n">RA2</span><span class="o">.</span><span class="n">domain_to_protein</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
</span><span id="DirectInformationData-470"><a href="#DirectInformationData-470"><span class="linenos">470</span></a>        <span class="n">mappable_DI_data</span> <span class="o">=</span> <span class="n">DI_data</span><span class="p">[</span><span class="n">mapping_key_mask</span><span class="p">]</span>
</span><span id="DirectInformationData-471"><a href="#DirectInformationData-471"><span class="linenos">471</span></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mappable_DI_data</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="DirectInformationData-472"><a href="#DirectInformationData-472"><span class="linenos">472</span></a>            <span class="k">return</span> <span class="n">mappable_DI_data</span>
</span><span id="DirectInformationData-473"><a href="#DirectInformationData-473"><span class="linenos">473</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData-474"><a href="#DirectInformationData-474"><span class="linenos">474</span></a>            <span class="n">mappable_DI_data</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">RA1</span><span class="o">.</span><span class="n">domain_to_protein</span><span class="p">[</span><span class="n">x</span><span class="p">])(</span><span class="n">mappable_DI_data</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">])</span>
</span><span id="DirectInformationData-475"><a href="#DirectInformationData-475"><span class="linenos">475</span></a>            <span class="n">mappable_DI_data</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">RA2</span><span class="o">.</span><span class="n">domain_to_protein</span><span class="p">[</span><span class="n">x</span><span class="p">])(</span><span class="n">mappable_DI_data</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">])</span>
</span><span id="DirectInformationData-476"><a href="#DirectInformationData-476"><span class="linenos">476</span></a>            <span class="k">return</span> <span class="n">mappable_DI_data</span>
</span><span id="DirectInformationData-477"><a href="#DirectInformationData-477"><span class="linenos">477</span></a>    
</span><span id="DirectInformationData-478"><a href="#DirectInformationData-478"><span class="linenos">478</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData-479"><a href="#DirectInformationData-479"><span class="linenos">479</span></a>    <span class="k">def</span> <span class="nf">rank_pairs</span><span class="p">(</span><span class="n">DI_data</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="DirectInformationData-480"><a href="#DirectInformationData-480"><span class="linenos">480</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-481"><a href="#DirectInformationData-481"><span class="linenos">481</span></a><span class="sd">        Sorts a structured ndarray of pairs information to order the ndarray by Direct Information (DI) score.</span>
</span><span id="DirectInformationData-482"><a href="#DirectInformationData-482"><span class="linenos">482</span></a>
</span><span id="DirectInformationData-483"><a href="#DirectInformationData-483"><span class="linenos">483</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-484"><a href="#DirectInformationData-484"><span class="linenos">484</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-485"><a href="#DirectInformationData-485"><span class="linenos">485</span></a><span class="sd">        DI_data : numpy.ndarray</span>
</span><span id="DirectInformationData-486"><a href="#DirectInformationData-486"><span class="linenos">486</span></a><span class="sd">            Structured ndarray that contains columns with names &quot;residue1&quot;, &quot;residue2&quot;, and &quot;DI&quot; (Direct Information)</span>
</span><span id="DirectInformationData-487"><a href="#DirectInformationData-487"><span class="linenos">487</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-488"><a href="#DirectInformationData-488"><span class="linenos">488</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-489"><a href="#DirectInformationData-489"><span class="linenos">489</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-490"><a href="#DirectInformationData-490"><span class="linenos">490</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="DirectInformationData-491"><a href="#DirectInformationData-491"><span class="linenos">491</span></a><span class="sd">            Structured ndarray sorted upon column that is named DI in descending order.</span>
</span><span id="DirectInformationData-492"><a href="#DirectInformationData-492"><span class="linenos">492</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-493"><a href="#DirectInformationData-493"><span class="linenos">493</span></a>        <span class="c1"># [::-1] reverses the order from ascending DI Score to descending DI score.</span>
</span><span id="DirectInformationData-494"><a href="#DirectInformationData-494"><span class="linenos">494</span></a>        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">DI_data</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="s1">&#39;DI&#39;</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span><span id="DirectInformationData-495"><a href="#DirectInformationData-495"><span class="linenos">495</span></a>    
</span><span id="DirectInformationData-496"><a href="#DirectInformationData-496"><span class="linenos">496</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData-497"><a href="#DirectInformationData-497"><span class="linenos">497</span></a>    <span class="k">def</span> <span class="nf">nonlocal_pairs</span><span class="p">(</span><span class="n">DI_data</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="DirectInformationData-498"><a href="#DirectInformationData-498"><span class="linenos">498</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-499"><a href="#DirectInformationData-499"><span class="linenos">499</span></a><span class="sd">        Subsets a structured ndarray of pairs information to find nonlocal pairs, where residue interactions are likely not involved in secondary structure formation i.e. helices and sheet interactions. Nonlocal pairs must be at least 4 residues apart.</span>
</span><span id="DirectInformationData-500"><a href="#DirectInformationData-500"><span class="linenos">500</span></a>
</span><span id="DirectInformationData-501"><a href="#DirectInformationData-501"><span class="linenos">501</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-502"><a href="#DirectInformationData-502"><span class="linenos">502</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-503"><a href="#DirectInformationData-503"><span class="linenos">503</span></a><span class="sd">        DI_data : numpy.ndarray</span>
</span><span id="DirectInformationData-504"><a href="#DirectInformationData-504"><span class="linenos">504</span></a><span class="sd">            Structured ndarray that contains (at least) the first and second columns with the names &quot;residue1&quot; and &quot;residue2&quot; respectively.</span>
</span><span id="DirectInformationData-505"><a href="#DirectInformationData-505"><span class="linenos">505</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-506"><a href="#DirectInformationData-506"><span class="linenos">506</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-507"><a href="#DirectInformationData-507"><span class="linenos">507</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-508"><a href="#DirectInformationData-508"><span class="linenos">508</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="DirectInformationData-509"><a href="#DirectInformationData-509"><span class="linenos">509</span></a><span class="sd">            Structured ndarray of DI pairs where residue 1 and residue 2 are at least 4 residues apart.</span>
</span><span id="DirectInformationData-510"><a href="#DirectInformationData-510"><span class="linenos">510</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-511"><a href="#DirectInformationData-511"><span class="linenos">511</span></a>        <span class="k">return</span> <span class="n">DI_data</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">DI_data</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">DI_data</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">]</span>
</span><span id="DirectInformationData-512"><a href="#DirectInformationData-512"><span class="linenos">512</span></a>    
</span><span id="DirectInformationData-513"><a href="#DirectInformationData-513"><span class="linenos">513</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData-514"><a href="#DirectInformationData-514"><span class="linenos">514</span></a>    <span class="k">def</span> <span class="nf">find_DI_with_residues</span><span class="p">(</span><span class="n">critical_residues_1</span> <span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">critical_residues_2</span> <span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">max_rank</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">*</span><span class="n">mapped_resi_arrs</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">list</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
</span><span id="DirectInformationData-515"><a href="#DirectInformationData-515"><span class="linenos">515</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-516"><a href="#DirectInformationData-516"><span class="linenos">516</span></a><span class="sd">        Function that takes an n number of ranked, mapped DI pairs and checks to see if they&#39;re in a list of potential residue indices.</span>
</span><span id="DirectInformationData-517"><a href="#DirectInformationData-517"><span class="linenos">517</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-518"><a href="#DirectInformationData-518"><span class="linenos">518</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-519"><a href="#DirectInformationData-519"><span class="linenos">519</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-520"><a href="#DirectInformationData-520"><span class="linenos">520</span></a><span class="sd">        critical_residues_1 : collections.abc.Iterable of int</span>
</span><span id="DirectInformationData-521"><a href="#DirectInformationData-521"><span class="linenos">521</span></a><span class="sd">            Specific residue indices that a DI pair will be compared to. If the first residue of the DI pair is not one of these indices, it will not be appended to results.</span>
</span><span id="DirectInformationData-522"><a href="#DirectInformationData-522"><span class="linenos">522</span></a><span class="sd">        crtical_residues_2 : collections.abc.Iterable of int</span>
</span><span id="DirectInformationData-523"><a href="#DirectInformationData-523"><span class="linenos">523</span></a><span class="sd">            Specific residue indices that a DI pair will be compared to. If the second residue of the DI pair is not one of these indices, it will not be appended to results.</span>
</span><span id="DirectInformationData-524"><a href="#DirectInformationData-524"><span class="linenos">524</span></a><span class="sd">        threshold : int, optional</span>
</span><span id="DirectInformationData-525"><a href="#DirectInformationData-525"><span class="linenos">525</span></a><span class="sd">            Maximum &quot;rank&quot; of the DI pair considered.</span>
</span><span id="DirectInformationData-526"><a href="#DirectInformationData-526"><span class="linenos">526</span></a><span class="sd">        *mapped_resi_arrs : tuple of numpy.ndarray</span>
</span><span id="DirectInformationData-527"><a href="#DirectInformationData-527"><span class="linenos">527</span></a><span class="sd">            Tuple of ranked, mapped pairs that are compared to critical residue indices and appended to results if in those indices and within threshold.</span>
</span><span id="DirectInformationData-528"><a href="#DirectInformationData-528"><span class="linenos">528</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-529"><a href="#DirectInformationData-529"><span class="linenos">529</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-530"><a href="#DirectInformationData-530"><span class="linenos">530</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-531"><a href="#DirectInformationData-531"><span class="linenos">531</span></a><span class="sd">        results : list of tuple of list of int, int</span>
</span><span id="DirectInformationData-532"><a href="#DirectInformationData-532"><span class="linenos">532</span></a><span class="sd">            Results which consist of tuples where the first element is a list of residue1, residue2, and DI score, whereas the second element is the rank.</span>
</span><span id="DirectInformationData-533"><a href="#DirectInformationData-533"><span class="linenos">533</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-534"><a href="#DirectInformationData-534"><span class="linenos">534</span></a>        <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="DirectInformationData-535"><a href="#DirectInformationData-535"><span class="linenos">535</span></a>        <span class="k">for</span> <span class="n">mapped_resi_arr</span> <span class="ow">in</span> <span class="n">mapped_resi_arrs</span><span class="p">:</span>
</span><span id="DirectInformationData-536"><a href="#DirectInformationData-536"><span class="linenos">536</span></a>            <span class="c1"># count_rank represents the rank of the DI pair being evaluated, iterating over every new row considered.</span>
</span><span id="DirectInformationData-537"><a href="#DirectInformationData-537"><span class="linenos">537</span></a>            <span class="n">count_rank</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="DirectInformationData-538"><a href="#DirectInformationData-538"><span class="linenos">538</span></a>            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">mapped_resi_arr</span><span class="p">:</span>
</span><span id="DirectInformationData-539"><a href="#DirectInformationData-539"><span class="linenos">539</span></a>                <span class="n">row_as_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
</span><span id="DirectInformationData-540"><a href="#DirectInformationData-540"><span class="linenos">540</span></a>                <span class="n">count_rank</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span id="DirectInformationData-541"><a href="#DirectInformationData-541"><span class="linenos">541</span></a>                <span class="k">if</span> <span class="n">max_rank</span><span class="p">:</span>
</span><span id="DirectInformationData-542"><a href="#DirectInformationData-542"><span class="linenos">542</span></a>                    <span class="k">if</span> <span class="n">row_as_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">critical_residues_1</span> <span class="ow">and</span> <span class="n">row_as_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">critical_residues_2</span> <span class="ow">and</span> <span class="n">count_rank</span> <span class="o">&lt;=</span> <span class="n">max_rank</span><span class="p">:</span>
</span><span id="DirectInformationData-543"><a href="#DirectInformationData-543"><span class="linenos">543</span></a>                        <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">row_as_list</span><span class="p">,</span> <span class="n">count_rank</span><span class="p">))</span>
</span><span id="DirectInformationData-544"><a href="#DirectInformationData-544"><span class="linenos">544</span></a>                <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData-545"><a href="#DirectInformationData-545"><span class="linenos">545</span></a>                    <span class="k">if</span> <span class="n">row_as_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">critical_residues_1</span> <span class="ow">and</span> <span class="n">row_as_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">critical_residues_2</span><span class="p">:</span>
</span><span id="DirectInformationData-546"><a href="#DirectInformationData-546"><span class="linenos">546</span></a>                        <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">row_as_list</span><span class="p">,</span> <span class="n">count_rank</span><span class="p">))</span>
</span><span id="DirectInformationData-547"><a href="#DirectInformationData-547"><span class="linenos">547</span></a>        <span class="k">return</span> <span class="n">results</span>
</span><span id="DirectInformationData-548"><a href="#DirectInformationData-548"><span class="linenos">548</span></a>
</span><span id="DirectInformationData-549"><a href="#DirectInformationData-549"><span class="linenos">549</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData-550"><a href="#DirectInformationData-550"><span class="linenos">550</span></a>    <span class="k">def</span> <span class="nf">get_dist_commands</span><span class="p">(</span><span class="n">model1</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span> <span class="n">model2</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auth_res_ids</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
</span><span id="DirectInformationData-551"><a href="#DirectInformationData-551"><span class="linenos">551</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-552"><a href="#DirectInformationData-552"><span class="linenos">552</span></a><span class="sd">        Get UCSF Chimera commands for displaying distance commands for usage in displaying distances between residue pairs. Options are present for alpha-carbon to alpha-carbon distance or for specified atom to specified atom distance.</span>
</span><span id="DirectInformationData-553"><a href="#DirectInformationData-553"><span class="linenos">553</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-554"><a href="#DirectInformationData-554"><span class="linenos">554</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-555"><a href="#DirectInformationData-555"><span class="linenos">555</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-556"><a href="#DirectInformationData-556"><span class="linenos">556</span></a><span class="sd">        model1 : str, int</span>
</span><span id="DirectInformationData-557"><a href="#DirectInformationData-557"><span class="linenos">557</span></a><span class="sd">            Number of model corresponding to the structure containing the first column of residues.</span>
</span><span id="DirectInformationData-558"><a href="#DirectInformationData-558"><span class="linenos">558</span></a><span class="sd">        model2 : str, int</span>
</span><span id="DirectInformationData-559"><a href="#DirectInformationData-559"><span class="linenos">559</span></a><span class="sd">            Number of model corresponding to the structure containing the second column of residues.</span>
</span><span id="DirectInformationData-560"><a href="#DirectInformationData-560"><span class="linenos">560</span></a><span class="sd">        chain1 : str</span>
</span><span id="DirectInformationData-561"><a href="#DirectInformationData-561"><span class="linenos">561</span></a><span class="sd">            The chain present in the structure in model1 containing the first column of residues.</span>
</span><span id="DirectInformationData-562"><a href="#DirectInformationData-562"><span class="linenos">562</span></a><span class="sd">        chain2 : str</span>
</span><span id="DirectInformationData-563"><a href="#DirectInformationData-563"><span class="linenos">563</span></a><span class="sd">            The chain present in the structure in model2 containing the second column of residues.</span>
</span><span id="DirectInformationData-564"><a href="#DirectInformationData-564"><span class="linenos">564</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="DirectInformationData-565"><a href="#DirectInformationData-565"><span class="linenos">565</span></a><span class="sd">            Structured ndarray that contains (at least) the first and second columns with the names &quot;residue1&quot; and &quot;residue2&quot; respectively. `ca_only` should be set to false and an additional &quot;atom_name1&quot; and &quot;atom_name2&quot; column should be added if atoms are specified per pair.</span>
</span><span id="DirectInformationData-566"><a href="#DirectInformationData-566"><span class="linenos">566</span></a><span class="sd">        ca_only : bool</span>
</span><span id="DirectInformationData-567"><a href="#DirectInformationData-567"><span class="linenos">567</span></a><span class="sd">            True if distance commands are for displaying distances between the two alpha-carbons of the residue pair. If false, specific_atom_names is used in lieu of &quot;CA&quot; as an atom identifier.</span>
</span><span id="DirectInformationData-568"><a href="#DirectInformationData-568"><span class="linenos">568</span></a><span class="sd">        auth_res_ids : bool</span>
</span><span id="DirectInformationData-569"><a href="#DirectInformationData-569"><span class="linenos">569</span></a><span class="sd">            If true, use the &quot;auth_residue1&quot; and &quot;auth_residue2&quot; columns instead of &quot;residue1&quot; and &quot;residue2&quot; columns. These residue ids correspond to the auth protein residue ids.</span>
</span><span id="DirectInformationData-570"><a href="#DirectInformationData-570"><span class="linenos">570</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-571"><a href="#DirectInformationData-571"><span class="linenos">571</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-572"><a href="#DirectInformationData-572"><span class="linenos">572</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-573"><a href="#DirectInformationData-573"><span class="linenos">573</span></a><span class="sd">        distance_commands : list of str</span>
</span><span id="DirectInformationData-574"><a href="#DirectInformationData-574"><span class="linenos">574</span></a><span class="sd">            List of distance commands generated between two residues with model and chain information needed, either between two alpha-carbons or the specified atoms.</span>
</span><span id="DirectInformationData-575"><a href="#DirectInformationData-575"><span class="linenos">575</span></a><span class="sd">        </span>
</span><span id="DirectInformationData-576"><a href="#DirectInformationData-576"><span class="linenos">576</span></a><span class="sd">        Note</span>
</span><span id="DirectInformationData-577"><a href="#DirectInformationData-577"><span class="linenos">577</span></a><span class="sd">        ----</span>
</span><span id="DirectInformationData-578"><a href="#DirectInformationData-578"><span class="linenos">578</span></a><span class="sd">        model1 and model2 can be equivalent if both columns involve residues referenced by the same model. The same would apply for chains if the residues are present on the same chain. </span>
</span><span id="DirectInformationData-579"><a href="#DirectInformationData-579"><span class="linenos">579</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-580"><a href="#DirectInformationData-580"><span class="linenos">580</span></a>        <span class="n">distance_commands</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="DirectInformationData-581"><a href="#DirectInformationData-581"><span class="linenos">581</span></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">pairs</span><span class="p">)[</span><span class="mi">0</span><span class="p">]):</span>
</span><span id="DirectInformationData-582"><a href="#DirectInformationData-582"><span class="linenos">582</span></a>            <span class="k">if</span> <span class="n">auth_res_ids</span><span class="p">:</span>
</span><span id="DirectInformationData-583"><a href="#DirectInformationData-583"><span class="linenos">583</span></a>                <span class="n">residue1</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;auth_residue1&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData-584"><a href="#DirectInformationData-584"><span class="linenos">584</span></a>                <span class="n">residue2</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;auth_residue2&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData-585"><a href="#DirectInformationData-585"><span class="linenos">585</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData-586"><a href="#DirectInformationData-586"><span class="linenos">586</span></a>                <span class="n">residue1</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData-587"><a href="#DirectInformationData-587"><span class="linenos">587</span></a>                <span class="n">residue2</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;residue2&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData-588"><a href="#DirectInformationData-588"><span class="linenos">588</span></a>            <span class="k">if</span> <span class="n">ca_only</span><span class="p">:</span>
</span><span id="DirectInformationData-589"><a href="#DirectInformationData-589"><span class="linenos">589</span></a>                <span class="n">distance_commands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;distance #</span><span class="si">{</span><span class="n">model1</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">residue1</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chain1</span><span class="si">}</span><span class="s2">@CA #</span><span class="si">{</span><span class="n">model2</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">residue2</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chain2</span><span class="si">}</span><span class="s2">@CA;&quot;</span><span class="p">)</span>
</span><span id="DirectInformationData-590"><a href="#DirectInformationData-590"><span class="linenos">590</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData-591"><a href="#DirectInformationData-591"><span class="linenos">591</span></a>                <span class="n">atom1</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;atom_name1&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData-592"><a href="#DirectInformationData-592"><span class="linenos">592</span></a>                <span class="n">atom2</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;atom_name2&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData-593"><a href="#DirectInformationData-593"><span class="linenos">593</span></a>                <span class="n">distance_commands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;distance #</span><span class="si">{</span><span class="n">model1</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">residue1</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chain1</span><span class="si">}</span><span class="s2">@</span><span class="si">{</span><span class="n">atom1</span><span class="si">}</span><span class="s2"> #</span><span class="si">{</span><span class="n">model2</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">residue2</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chain2</span><span class="si">}</span><span class="s2">@</span><span class="si">{</span><span class="n">atom2</span><span class="si">}</span><span class="s2">;&quot;</span><span class="p">)</span>
</span><span id="DirectInformationData-594"><a href="#DirectInformationData-594"><span class="linenos">594</span></a>        <span class="k">return</span> <span class="n">distance_commands</span>
</span><span id="DirectInformationData-595"><a href="#DirectInformationData-595"><span class="linenos">595</span></a>
</span><span id="DirectInformationData-596"><a href="#DirectInformationData-596"><span class="linenos">596</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData-597"><a href="#DirectInformationData-597"><span class="linenos">597</span></a>    <span class="k">def</span> <span class="nf">write_DI_data</span><span class="p">(</span><span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">delimiter</span><span class="p">:</span> <span class="nb">str</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">fmt</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%.3f</span><span class="s1">&#39;</span><span class="p">))</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="DirectInformationData-598"><a href="#DirectInformationData-598"><span class="linenos">598</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData-599"><a href="#DirectInformationData-599"><span class="linenos">599</span></a><span class="sd">        Writes pairs ndarray to file with specified delimiter between the pairs&#39; row elements, i.e. residue 1, residue 2, and DI score.</span>
</span><span id="DirectInformationData-600"><a href="#DirectInformationData-600"><span class="linenos">600</span></a>
</span><span id="DirectInformationData-601"><a href="#DirectInformationData-601"><span class="linenos">601</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData-602"><a href="#DirectInformationData-602"><span class="linenos">602</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData-603"><a href="#DirectInformationData-603"><span class="linenos">603</span></a><span class="sd">        filepath : str</span>
</span><span id="DirectInformationData-604"><a href="#DirectInformationData-604"><span class="linenos">604</span></a><span class="sd">            Path of the file to write DirectInformation data to.</span>
</span><span id="DirectInformationData-605"><a href="#DirectInformationData-605"><span class="linenos">605</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="DirectInformationData-606"><a href="#DirectInformationData-606"><span class="linenos">606</span></a><span class="sd">            Ndarray of at-least pairs information (residue 1, residue 2) and optionally Direct Information to write to a file via numpy.savetxt().</span>
</span><span id="DirectInformationData-607"><a href="#DirectInformationData-607"><span class="linenos">607</span></a><span class="sd">        delimiter : str</span>
</span><span id="DirectInformationData-608"><a href="#DirectInformationData-608"><span class="linenos">608</span></a><span class="sd">            Delimiter to separate columns of the pairs ndarray when writing to a file.</span>
</span><span id="DirectInformationData-609"><a href="#DirectInformationData-609"><span class="linenos">609</span></a><span class="sd">        fmt : tuple of str, default=(&#39;%d&#39;, &#39;%d&#39;, &#39;%.3f)</span>
</span><span id="DirectInformationData-610"><a href="#DirectInformationData-610"><span class="linenos">610</span></a><span class="sd">            format passed as an argument to numpy.savetxt() to define type of column and output format. Set to (&#39;%d&#39;, &#39;%d&#39;) if only two columns are present in the ndarray.</span>
</span><span id="DirectInformationData-611"><a href="#DirectInformationData-611"><span class="linenos">611</span></a>
</span><span id="DirectInformationData-612"><a href="#DirectInformationData-612"><span class="linenos">612</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData-613"><a href="#DirectInformationData-613"><span class="linenos">613</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData-614"><a href="#DirectInformationData-614"><span class="linenos">614</span></a><span class="sd">        None</span>
</span><span id="DirectInformationData-615"><a href="#DirectInformationData-615"><span class="linenos">615</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData-616"><a href="#DirectInformationData-616"><span class="linenos">616</span></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">pairs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
</span><span id="DirectInformationData-617"><a href="#DirectInformationData-617"><span class="linenos">617</span></a>            <span class="n">fmt</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">)</span>
</span><span id="DirectInformationData-618"><a href="#DirectInformationData-618"><span class="linenos">618</span></a>        <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">pairs</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="n">delimiter</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="n">fmt</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Representation and interface for Direct Information data including residue indices for a pair and its corresponding DI value represented as a 3-column ndarray.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>structured_ndarray</strong> (numpy.ndarray):
Ndarray with the shape (n,3) with dtype={'names': ('residue1', 'residue2', 'DI'), 'formats': (int, int, float, float)}</li>
</ul>

<h6 id="attributes">Attributes</h6>

<ul>
<li><strong>DI_data</strong> (numpy.ndarray):
The structured_ndarray in the parameters section where column 1 corresponds to a pair's first residue, column 2 corresponds to the pair's second residue, and column 3 corresponds to the Direct Information of the pair.</li>
</ul>
</div>


                            <div id="DirectInformationData.__init__" class="classattr">
                                        <input id="DirectInformationData.__init__-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="name">DirectInformationData</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">structured_ndarray</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span></span>)</span>

                <label class="view-source-button" for="DirectInformationData.__init__-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.__init__"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.__init__-352"><a href="#DirectInformationData.__init__-352"><span class="linenos">352</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structured_ndarray</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="DirectInformationData.__init__-353"><a href="#DirectInformationData.__init__-353"><span class="linenos">353</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">DI_data</span> <span class="o">=</span> <span class="n">structured_ndarray</span>
</span></pre></div>


    

                            </div>
                            <div id="DirectInformationData.DI_data" class="classattr">
                                <div class="attr variable">
            <span class="name">DI_data</span>

        
    </div>
    <a class="headerlink" href="#DirectInformationData.DI_data"></a>
    
    

                            </div>
                            <div id="DirectInformationData.load_from_dca_output" class="classattr">
                                        <input id="DirectInformationData.load_from_dca_output-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">load_from_dca_output</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">dca_filepath</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="n"><a href="#DirectInformationData">DirectInformationData</a></span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.load_from_dca_output-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.load_from_dca_output"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.load_from_dca_output-355"><a href="#DirectInformationData.load_from_dca_output-355"><span class="linenos">355</span></a>    <span class="nd">@staticmethod</span>    
</span><span id="DirectInformationData.load_from_dca_output-356"><a href="#DirectInformationData.load_from_dca_output-356"><span class="linenos">356</span></a>    <span class="k">def</span> <span class="nf">load_from_dca_output</span><span class="p">(</span><span class="n">dca_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DirectInformationData&#39;</span><span class="p">:</span>
</span><span id="DirectInformationData.load_from_dca_output-357"><a href="#DirectInformationData.load_from_dca_output-357"><span class="linenos">357</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.load_from_dca_output-358"><a href="#DirectInformationData.load_from_dca_output-358"><span class="linenos">358</span></a><span class="sd">        Function to generate a DirectInformationData object from the direct output of the MATLab dca function.</span>
</span><span id="DirectInformationData.load_from_dca_output-359"><a href="#DirectInformationData.load_from_dca_output-359"><span class="linenos">359</span></a>
</span><span id="DirectInformationData.load_from_dca_output-360"><a href="#DirectInformationData.load_from_dca_output-360"><span class="linenos">360</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.load_from_dca_output-361"><a href="#DirectInformationData.load_from_dca_output-361"><span class="linenos">361</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.load_from_dca_output-362"><a href="#DirectInformationData.load_from_dca_output-362"><span class="linenos">362</span></a><span class="sd">        dca_filepath : str</span>
</span><span id="DirectInformationData.load_from_dca_output-363"><a href="#DirectInformationData.load_from_dca_output-363"><span class="linenos">363</span></a><span class="sd">            Filepath of the DCA output to be read and compiled into a structured ndarray. DCA output is a 4 column text file with the following columns: (residue 1, residue 2, Mutual Information, Direct Information).</span>
</span><span id="DirectInformationData.load_from_dca_output-364"><a href="#DirectInformationData.load_from_dca_output-364"><span class="linenos">364</span></a>
</span><span id="DirectInformationData.load_from_dca_output-365"><a href="#DirectInformationData.load_from_dca_output-365"><span class="linenos">365</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.load_from_dca_output-366"><a href="#DirectInformationData.load_from_dca_output-366"><span class="linenos">366</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.load_from_dca_output-367"><a href="#DirectInformationData.load_from_dca_output-367"><span class="linenos">367</span></a><span class="sd">        DirectInformationData</span>
</span><span id="DirectInformationData.load_from_dca_output-368"><a href="#DirectInformationData.load_from_dca_output-368"><span class="linenos">368</span></a><span class="sd">            DirectInformationData object with named structured array containing residue indices and the DI value of the pair.</span>
</span><span id="DirectInformationData.load_from_dca_output-369"><a href="#DirectInformationData.load_from_dca_output-369"><span class="linenos">369</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.load_from_dca_output-370"><a href="#DirectInformationData.load_from_dca_output-370"><span class="linenos">370</span></a>        <span class="n">file_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">dca_filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="s1">&#39;MI&#39;</span><span class="p">,</span> <span class="s1">&#39;DI&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">)})</span>
</span><span id="DirectInformationData.load_from_dca_output-371"><a href="#DirectInformationData.load_from_dca_output-371"><span class="linenos">371</span></a>        <span class="k">return</span> <span class="n">DirectInformationData</span><span class="p">(</span><span class="n">file_data</span><span class="p">[[</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="s1">&#39;DI&#39;</span><span class="p">]])</span>
</span></pre></div>


            <div class="docstring"><p>Function to generate a DirectInformationData object from the direct output of the MATLab dca function.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>dca_filepath</strong> (str):
Filepath of the DCA output to be read and compiled into a structured ndarray. DCA output is a 4 column text file with the following columns: (residue 1, residue 2, Mutual Information, Direct Information).</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>DirectInformationData</strong>: DirectInformationData object with named structured array containing residue indices and the DI value of the pair.</li>
</ul>
</div>


                            </div>
                            <div id="DirectInformationData.load_from_DI_file" class="classattr">
                                        <input id="DirectInformationData.load_from_DI_file-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">load_from_DI_file</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">DI_filepath</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="n"><a href="#DirectInformationData">DirectInformationData</a></span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.load_from_DI_file-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.load_from_DI_file"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.load_from_DI_file-373"><a href="#DirectInformationData.load_from_DI_file-373"><span class="linenos">373</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData.load_from_DI_file-374"><a href="#DirectInformationData.load_from_DI_file-374"><span class="linenos">374</span></a>    <span class="k">def</span> <span class="nf">load_from_DI_file</span><span class="p">(</span><span class="n">DI_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;DirectInformationData&#39;</span><span class="p">:</span>
</span><span id="DirectInformationData.load_from_DI_file-375"><a href="#DirectInformationData.load_from_DI_file-375"><span class="linenos">375</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.load_from_DI_file-376"><a href="#DirectInformationData.load_from_DI_file-376"><span class="linenos">376</span></a><span class="sd">        Function to generate a DirectInformationData object from the modified DI-only version of the DCA output generated via the MATLab dca function.</span>
</span><span id="DirectInformationData.load_from_DI_file-377"><a href="#DirectInformationData.load_from_DI_file-377"><span class="linenos">377</span></a>
</span><span id="DirectInformationData.load_from_DI_file-378"><a href="#DirectInformationData.load_from_DI_file-378"><span class="linenos">378</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.load_from_DI_file-379"><a href="#DirectInformationData.load_from_DI_file-379"><span class="linenos">379</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.load_from_DI_file-380"><a href="#DirectInformationData.load_from_DI_file-380"><span class="linenos">380</span></a><span class="sd">        DI_filepath : str</span>
</span><span id="DirectInformationData.load_from_DI_file-381"><a href="#DirectInformationData.load_from_DI_file-381"><span class="linenos">381</span></a><span class="sd">            Filepath of the DI file to be read and compile into a structured ndarray. DI file is a 3 column text file with the following columns: (residue 1, residue 2, Direct Information).</span>
</span><span id="DirectInformationData.load_from_DI_file-382"><a href="#DirectInformationData.load_from_DI_file-382"><span class="linenos">382</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.load_from_DI_file-383"><a href="#DirectInformationData.load_from_DI_file-383"><span class="linenos">383</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.load_from_DI_file-384"><a href="#DirectInformationData.load_from_DI_file-384"><span class="linenos">384</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.load_from_DI_file-385"><a href="#DirectInformationData.load_from_DI_file-385"><span class="linenos">385</span></a><span class="sd">        DirectInformationData</span>
</span><span id="DirectInformationData.load_from_DI_file-386"><a href="#DirectInformationData.load_from_DI_file-386"><span class="linenos">386</span></a><span class="sd">            DirectInformationData object with named structured array containing residue indices and the DI value of the pair.</span>
</span><span id="DirectInformationData.load_from_DI_file-387"><a href="#DirectInformationData.load_from_DI_file-387"><span class="linenos">387</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.load_from_DI_file-388"><a href="#DirectInformationData.load_from_DI_file-388"><span class="linenos">388</span></a>        <span class="k">return</span> <span class="n">DirectInformationData</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">DI_filepath</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="s1">&#39;DI&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">)}))</span>
</span></pre></div>


            <div class="docstring"><p>Function to generate a DirectInformationData object from the modified DI-only version of the DCA output generated via the MATLab dca function.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>DI_filepath</strong> (str):
Filepath of the DI file to be read and compile into a structured ndarray. DI file is a 3 column text file with the following columns: (residue 1, residue 2, Direct Information).</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>DirectInformationData</strong>: DirectInformationData object with named structured array containing residue indices and the DI value of the pair.</li>
</ul>
</div>


                            </div>
                            <div id="DirectInformationData.load_as_ndarray" class="classattr">
                                        <input id="DirectInformationData.load_as_ndarray-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">load_as_ndarray</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">ndarray</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]],</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Iterable</span><span class="p">]]</span></span><span class="return-annotation">) -> <span class="n"><a href="#DirectInformationData">DirectInformationData</a></span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.load_as_ndarray-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.load_as_ndarray"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.load_as_ndarray-390"><a href="#DirectInformationData.load_as_ndarray-390"><span class="linenos">390</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData.load_as_ndarray-391"><a href="#DirectInformationData.load_as_ndarray-391"><span class="linenos">391</span></a>    <span class="k">def</span> <span class="nf">load_as_ndarray</span><span class="p">(</span><span class="n">ndarray</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Iterable</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="s1">&#39;DirectInformationData&#39;</span><span class="p">:</span>
</span><span id="DirectInformationData.load_as_ndarray-392"><a href="#DirectInformationData.load_as_ndarray-392"><span class="linenos">392</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.load_as_ndarray-393"><a href="#DirectInformationData.load_as_ndarray-393"><span class="linenos">393</span></a><span class="sd">        Function to generate a Direct Information object from a ndarray.</span>
</span><span id="DirectInformationData.load_as_ndarray-394"><a href="#DirectInformationData.load_as_ndarray-394"><span class="linenos">394</span></a>
</span><span id="DirectInformationData.load_as_ndarray-395"><a href="#DirectInformationData.load_as_ndarray-395"><span class="linenos">395</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.load_as_ndarray-396"><a href="#DirectInformationData.load_as_ndarray-396"><span class="linenos">396</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.load_as_ndarray-397"><a href="#DirectInformationData.load_as_ndarray-397"><span class="linenos">397</span></a><span class="sd">        ndarray : numpy.ndarray or Iterable of Iterable (excluding dict)</span>
</span><span id="DirectInformationData.load_as_ndarray-398"><a href="#DirectInformationData.load_as_ndarray-398"><span class="linenos">398</span></a><span class="sd">            An ndarray of shape (n,3) where its columns are (residue 1, residue 2, and Direct Information). Can also be parsed from an iterable of iterable provided that the aforementioned format is followed.</span>
</span><span id="DirectInformationData.load_as_ndarray-399"><a href="#DirectInformationData.load_as_ndarray-399"><span class="linenos">399</span></a>
</span><span id="DirectInformationData.load_as_ndarray-400"><a href="#DirectInformationData.load_as_ndarray-400"><span class="linenos">400</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.load_as_ndarray-401"><a href="#DirectInformationData.load_as_ndarray-401"><span class="linenos">401</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.load_as_ndarray-402"><a href="#DirectInformationData.load_as_ndarray-402"><span class="linenos">402</span></a><span class="sd">        DirectInformationData</span>
</span><span id="DirectInformationData.load_as_ndarray-403"><a href="#DirectInformationData.load_as_ndarray-403"><span class="linenos">403</span></a><span class="sd">            DirectInformationData object with named structured array containing residue indices and the DI value of the pair.</span>
</span><span id="DirectInformationData.load_as_ndarray-404"><a href="#DirectInformationData.load_as_ndarray-404"><span class="linenos">404</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.load_as_ndarray-405"><a href="#DirectInformationData.load_as_ndarray-405"><span class="linenos">405</span></a>        
</span><span id="DirectInformationData.load_as_ndarray-406"><a href="#DirectInformationData.load_as_ndarray-406"><span class="linenos">406</span></a>        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">and</span> <span class="n">ndarray</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">3</span><span class="p">:</span>
</span><span id="DirectInformationData.load_as_ndarray-407"><a href="#DirectInformationData.load_as_ndarray-407"><span class="linenos">407</span></a>            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Dimensions of numpy array supplied are different from what is expected. Please supply residue1, residue2, and DI column in int, int, float format and with shape of (n, 3).&quot;</span><span class="p">)</span>
</span><span id="DirectInformationData.load_as_ndarray-408"><a href="#DirectInformationData.load_as_ndarray-408"><span class="linenos">408</span></a>        <span class="c1"># Structured ndarrays require list of tuples for conversion.</span>
</span><span id="DirectInformationData.load_as_ndarray-409"><a href="#DirectInformationData.load_as_ndarray-409"><span class="linenos">409</span></a>        <span class="n">DI_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ndarray</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">,</span> <span class="s1">&#39;DI&#39;</span><span class="p">),</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">)})</span>
</span><span id="DirectInformationData.load_as_ndarray-410"><a href="#DirectInformationData.load_as_ndarray-410"><span class="linenos">410</span></a>        
</span><span id="DirectInformationData.load_as_ndarray-411"><a href="#DirectInformationData.load_as_ndarray-411"><span class="linenos">411</span></a>        <span class="k">return</span> <span class="n">DirectInformationData</span><span class="p">(</span><span class="n">DI_data</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Function to generate a Direct Information object from a ndarray.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>ndarray</strong> (numpy.ndarray or Iterable of Iterable (excluding dict)):
An ndarray of shape (n,3) where its columns are (residue 1, residue 2, and Direct Information). Can also be parsed from an iterable of iterable provided that the aforementioned format is followed.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>DirectInformationData</strong>: DirectInformationData object with named structured array containing residue indices and the DI value of the pair.</li>
</ul>
</div>


                            </div>
                            <div id="DirectInformationData.get_ranked_mapped_pairs" class="classattr">
                                        <input id="DirectInformationData.get_ranked_mapped_pairs-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_ranked_mapped_pairs</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">RA1</span><span class="p">:</span> <span class="n"><a href="#ResidueAlignment">ResidueAlignment</a></span>,</span><span class="param">	<span class="n">RA2</span><span class="p">:</span> <span class="n"><a href="#ResidueAlignment">ResidueAlignment</a></span>,</span><span class="param">	<span class="n">pairs_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>,</span><span class="param">	<span class="n">mirror</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>,</span><span class="param">	<span class="n">number</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.get_ranked_mapped_pairs-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.get_ranked_mapped_pairs"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.get_ranked_mapped_pairs-413"><a href="#DirectInformationData.get_ranked_mapped_pairs-413"><span class="linenos">413</span></a>    <span class="k">def</span> <span class="nf">get_ranked_mapped_pairs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">RA1</span><span class="p">:</span> <span class="n">ResidueAlignment</span><span class="p">,</span> <span class="n">RA2</span><span class="p">:</span> <span class="n">ResidueAlignment</span><span class="p">,</span> <span class="n">pairs_only</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">mirror</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">number</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-414"><a href="#DirectInformationData.get_ranked_mapped_pairs-414"><span class="linenos">414</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-415"><a href="#DirectInformationData.get_ranked_mapped_pairs-415"><span class="linenos">415</span></a><span class="sd">        Uses DirectInformationData and Pairs interface methods to obtain ranked, mapped residues. See rank_pairs() function and map_DIs() function for details on rank and mapping. Residue Alignments can be the same for intra-domain / intra-protein mapping.</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-416"><a href="#DirectInformationData.get_ranked_mapped_pairs-416"><span class="linenos">416</span></a>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-417"><a href="#DirectInformationData.get_ranked_mapped_pairs-417"><span class="linenos">417</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-418"><a href="#DirectInformationData.get_ranked_mapped_pairs-418"><span class="linenos">418</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-419"><a href="#DirectInformationData.get_ranked_mapped_pairs-419"><span class="linenos">419</span></a><span class="sd">        RA1 : ResidueAlignment</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-420"><a href="#DirectInformationData.get_ranked_mapped_pairs-420"><span class="linenos">420</span></a><span class="sd">            The ResidueAlignment used for mapping the first column of residues to the appropriate target sequence.</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-421"><a href="#DirectInformationData.get_ranked_mapped_pairs-421"><span class="linenos">421</span></a><span class="sd">        RA2 : ResidueAlignment</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-422"><a href="#DirectInformationData.get_ranked_mapped_pairs-422"><span class="linenos">422</span></a><span class="sd">            The ResidueAlignment used for mapping the second column of residues to the appropriate target sequence.</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-423"><a href="#DirectInformationData.get_ranked_mapped_pairs-423"><span class="linenos">423</span></a><span class="sd">        pairs_only : bool</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-424"><a href="#DirectInformationData.get_ranked_mapped_pairs-424"><span class="linenos">424</span></a><span class="sd">            True if the final ndarray should contain only columns 1 and 2, corresponding to the residues that constitute the pair. This would drop the DI column.</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-425"><a href="#DirectInformationData.get_ranked_mapped_pairs-425"><span class="linenos">425</span></a><span class="sd">        mirror : bool</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-426"><a href="#DirectInformationData.get_ranked_mapped_pairs-426"><span class="linenos">426</span></a><span class="sd">            See Pairs.mirror_pairs() or get_pairs() for details. NOTE: This option is overriden entirely if pairs_only is False. If true, this will produce an ndarray that has the original residue indices and repeated residue indices but with residue 1 and residue 2 switched. This is useful for plotting across the upper diagonal of a contact map.</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-427"><a href="#DirectInformationData.get_ranked_mapped_pairs-427"><span class="linenos">427</span></a><span class="sd">        number : int, None</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-428"><a href="#DirectInformationData.get_ranked_mapped_pairs-428"><span class="linenos">428</span></a><span class="sd">            Number of ranked, mapped pairs to return.</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-429"><a href="#DirectInformationData.get_ranked_mapped_pairs-429"><span class="linenos">429</span></a>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-430"><a href="#DirectInformationData.get_ranked_mapped_pairs-430"><span class="linenos">430</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-431"><a href="#DirectInformationData.get_ranked_mapped_pairs-431"><span class="linenos">431</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-432"><a href="#DirectInformationData.get_ranked_mapped_pairs-432"><span class="linenos">432</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-433"><a href="#DirectInformationData.get_ranked_mapped_pairs-433"><span class="linenos">433</span></a><span class="sd">            Structured ndarray with columns residue 1, residue 2 and optionally DI. Only has specified number of pairs if `number` is specified and mirrored pairs if `mirror` is True and pairs_only is False.</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-434"><a href="#DirectInformationData.get_ranked_mapped_pairs-434"><span class="linenos">434</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-435"><a href="#DirectInformationData.get_ranked_mapped_pairs-435"><span class="linenos">435</span></a><span class="sd">        Notes</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-436"><a href="#DirectInformationData.get_ranked_mapped_pairs-436"><span class="linenos">436</span></a><span class="sd">        -----</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-437"><a href="#DirectInformationData.get_ranked_mapped_pairs-437"><span class="linenos">437</span></a><span class="sd">        ResidueAlignments contain dictionaries like domain_to_protein to map residues produced via Direct Coupling Analysis (DCA) on an MSA generated in context to an HMM. The residues are mapped to a protein structure via alignment of the HMM hit / domain to the protein sequence.</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-438"><a href="#DirectInformationData.get_ranked_mapped_pairs-438"><span class="linenos">438</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-439"><a href="#DirectInformationData.get_ranked_mapped_pairs-439"><span class="linenos">439</span></a>        <span class="n">ranked_pairs</span> <span class="o">=</span> <span class="n">DirectInformationData</span><span class="o">.</span><span class="n">rank_pairs</span><span class="p">(</span><span class="n">DirectInformationData</span><span class="o">.</span><span class="n">nonlocal_pairs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DI_data</span><span class="p">))</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-440"><a href="#DirectInformationData.get_ranked_mapped_pairs-440"><span class="linenos">440</span></a>        <span class="n">ranked_mapped_pairs</span> <span class="o">=</span> <span class="n">DirectInformationData</span><span class="o">.</span><span class="n">map_DIs</span><span class="p">(</span><span class="n">ranked_pairs</span><span class="p">,</span> <span class="n">RA1</span><span class="p">,</span> <span class="n">RA2</span><span class="p">)</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-441"><a href="#DirectInformationData.get_ranked_mapped_pairs-441"><span class="linenos">441</span></a>        <span class="k">if</span> <span class="n">pairs_only</span><span class="p">:</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-442"><a href="#DirectInformationData.get_ranked_mapped_pairs-442"><span class="linenos">442</span></a>            <span class="k">return</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">get_pairs</span><span class="p">(</span><span class="n">ranked_mapped_pairs</span><span class="p">[[</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span> <span class="s1">&#39;residue2&#39;</span><span class="p">]],</span> <span class="n">mirror</span><span class="o">=</span><span class="n">mirror</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="n">number</span><span class="p">)</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-443"><a href="#DirectInformationData.get_ranked_mapped_pairs-443"><span class="linenos">443</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData.get_ranked_mapped_pairs-444"><a href="#DirectInformationData.get_ranked_mapped_pairs-444"><span class="linenos">444</span></a>            <span class="k">return</span> <span class="n">Pairs</span><span class="o">.</span><span class="n">get_pairs</span><span class="p">(</span><span class="n">ranked_mapped_pairs</span><span class="p">,</span> <span class="n">mirror</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="n">number</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Uses DirectInformationData and Pairs interface methods to obtain ranked, mapped residues. See rank_pairs() function and map_DIs() function for details on rank and mapping. Residue Alignments can be the same for intra-domain / intra-protein mapping.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>RA1</strong> (ResidueAlignment):
The ResidueAlignment used for mapping the first column of residues to the appropriate target sequence.</li>
<li><strong>RA2</strong> (ResidueAlignment):
The ResidueAlignment used for mapping the second column of residues to the appropriate target sequence.</li>
<li><strong>pairs_only</strong> (bool):
True if the final ndarray should contain only columns 1 and 2, corresponding to the residues that constitute the pair. This would drop the DI column.</li>
<li><strong>mirror</strong> (bool):
See <a href="#Pairs.mirror_pairs">Pairs.mirror_pairs()</a> or get_pairs() for details. NOTE: This option is overriden entirely if pairs_only is False. If true, this will produce an ndarray that has the original residue indices and repeated residue indices but with residue 1 and residue 2 switched. This is useful for plotting across the upper diagonal of a contact map.</li>
<li><strong>number</strong> (int, None):
Number of ranked, mapped pairs to return.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>numpy.ndarray</strong>: Structured ndarray with columns residue 1, residue 2 and optionally DI. Only has specified number of pairs if <code>number</code> is specified and mirrored pairs if <code>mirror</code> is True and pairs_only is False.</li>
</ul>

<h6 id="notes">Notes</h6>

<p>ResidueAlignments contain dictionaries like domain_to_protein to map residues produced via Direct Coupling Analysis (DCA) on an MSA generated in context to an HMM. The residues are mapped to a protein structure via alignment of the HMM hit / domain to the protein sequence.</p>
</div>


                            </div>
                            <div id="DirectInformationData.map_DIs" class="classattr">
                                        <input id="DirectInformationData.map_DIs-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">map_DIs</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">DI_data</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>,</span><span class="param">	<span class="n">RA1</span><span class="p">:</span> <span class="n"><a href="#ResidueAlignment">ResidueAlignment</a></span>,</span><span class="param">	<span class="n">RA2</span><span class="p">:</span> <span class="n"><a href="#ResidueAlignment">ResidueAlignment</a></span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.map_DIs-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.map_DIs"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.map_DIs-446"><a href="#DirectInformationData.map_DIs-446"><span class="linenos">446</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData.map_DIs-447"><a href="#DirectInformationData.map_DIs-447"><span class="linenos">447</span></a>    <span class="k">def</span> <span class="nf">map_DIs</span><span class="p">(</span><span class="n">DI_data</span> <span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">RA1</span><span class="p">:</span> <span class="n">ResidueAlignment</span><span class="p">,</span> <span class="n">RA2</span><span class="p">:</span> <span class="n">ResidueAlignment</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="DirectInformationData.map_DIs-448"><a href="#DirectInformationData.map_DIs-448"><span class="linenos">448</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.map_DIs-449"><a href="#DirectInformationData.map_DIs-449"><span class="linenos">449</span></a><span class="sd">        Uses domain-to-protein mappings present in the Residue Alignments provided to generate mapped representations of the residues from the DI_data structured ndarray provided.</span>
</span><span id="DirectInformationData.map_DIs-450"><a href="#DirectInformationData.map_DIs-450"><span class="linenos">450</span></a>
</span><span id="DirectInformationData.map_DIs-451"><a href="#DirectInformationData.map_DIs-451"><span class="linenos">451</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.map_DIs-452"><a href="#DirectInformationData.map_DIs-452"><span class="linenos">452</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.map_DIs-453"><a href="#DirectInformationData.map_DIs-453"><span class="linenos">453</span></a><span class="sd">        DI_data : numpy.ndarray</span>
</span><span id="DirectInformationData.map_DIs-454"><a href="#DirectInformationData.map_DIs-454"><span class="linenos">454</span></a><span class="sd">            Structured ndarray that contains columns &quot;residue1&quot; and &quot;residue2&quot;.</span>
</span><span id="DirectInformationData.map_DIs-455"><a href="#DirectInformationData.map_DIs-455"><span class="linenos">455</span></a><span class="sd">        RA1 : ResidueAlignment</span>
</span><span id="DirectInformationData.map_DIs-456"><a href="#DirectInformationData.map_DIs-456"><span class="linenos">456</span></a><span class="sd">            The ResidueAlignment used for mapping the first column of residues to the appropriate target sequence.</span>
</span><span id="DirectInformationData.map_DIs-457"><a href="#DirectInformationData.map_DIs-457"><span class="linenos">457</span></a><span class="sd">        RA2 : ResidueAlignment</span>
</span><span id="DirectInformationData.map_DIs-458"><a href="#DirectInformationData.map_DIs-458"><span class="linenos">458</span></a><span class="sd">            The ResidueAlignment used for mapping the second column of residues to the appropriate target sequence.</span>
</span><span id="DirectInformationData.map_DIs-459"><a href="#DirectInformationData.map_DIs-459"><span class="linenos">459</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.map_DIs-460"><a href="#DirectInformationData.map_DIs-460"><span class="linenos">460</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.map_DIs-461"><a href="#DirectInformationData.map_DIs-461"><span class="linenos">461</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.map_DIs-462"><a href="#DirectInformationData.map_DIs-462"><span class="linenos">462</span></a><span class="sd">        mappable_DI_data : numpy.ndarray</span>
</span><span id="DirectInformationData.map_DIs-463"><a href="#DirectInformationData.map_DIs-463"><span class="linenos">463</span></a><span class="sd">            DI_data that has been mapped to the target sequence specified in the generation of the corresponding ResidueAlignments.</span>
</span><span id="DirectInformationData.map_DIs-464"><a href="#DirectInformationData.map_DIs-464"><span class="linenos">464</span></a>
</span><span id="DirectInformationData.map_DIs-465"><a href="#DirectInformationData.map_DIs-465"><span class="linenos">465</span></a><span class="sd">        Note:</span>
</span><span id="DirectInformationData.map_DIs-466"><a href="#DirectInformationData.map_DIs-466"><span class="linenos">466</span></a><span class="sd">            Residues that do not map to the target sequence of the ResidueAlignment are dropped.</span>
</span><span id="DirectInformationData.map_DIs-467"><a href="#DirectInformationData.map_DIs-467"><span class="linenos">467</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.map_DIs-468"><a href="#DirectInformationData.map_DIs-468"><span class="linenos">468</span></a>        <span class="c1"># Alternative approach is to just use get function instead of [x] and default to np.nan and drop nans row-wise.</span>
</span><span id="DirectInformationData.map_DIs-469"><a href="#DirectInformationData.map_DIs-469"><span class="linenos">469</span></a>        <span class="n">mapping_key_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">DI_data</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">],</span> <span class="nb">list</span><span class="p">(</span><span class="n">RA1</span><span class="o">.</span><span class="n">domain_to_protein</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">DI_data</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">],</span> <span class="nb">list</span><span class="p">(</span><span class="n">RA2</span><span class="o">.</span><span class="n">domain_to_protein</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
</span><span id="DirectInformationData.map_DIs-470"><a href="#DirectInformationData.map_DIs-470"><span class="linenos">470</span></a>        <span class="n">mappable_DI_data</span> <span class="o">=</span> <span class="n">DI_data</span><span class="p">[</span><span class="n">mapping_key_mask</span><span class="p">]</span>
</span><span id="DirectInformationData.map_DIs-471"><a href="#DirectInformationData.map_DIs-471"><span class="linenos">471</span></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mappable_DI_data</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="DirectInformationData.map_DIs-472"><a href="#DirectInformationData.map_DIs-472"><span class="linenos">472</span></a>            <span class="k">return</span> <span class="n">mappable_DI_data</span>
</span><span id="DirectInformationData.map_DIs-473"><a href="#DirectInformationData.map_DIs-473"><span class="linenos">473</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData.map_DIs-474"><a href="#DirectInformationData.map_DIs-474"><span class="linenos">474</span></a>            <span class="n">mappable_DI_data</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">RA1</span><span class="o">.</span><span class="n">domain_to_protein</span><span class="p">[</span><span class="n">x</span><span class="p">])(</span><span class="n">mappable_DI_data</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">])</span>
</span><span id="DirectInformationData.map_DIs-475"><a href="#DirectInformationData.map_DIs-475"><span class="linenos">475</span></a>            <span class="n">mappable_DI_data</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vectorize</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">RA2</span><span class="o">.</span><span class="n">domain_to_protein</span><span class="p">[</span><span class="n">x</span><span class="p">])(</span><span class="n">mappable_DI_data</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">])</span>
</span><span id="DirectInformationData.map_DIs-476"><a href="#DirectInformationData.map_DIs-476"><span class="linenos">476</span></a>            <span class="k">return</span> <span class="n">mappable_DI_data</span>
</span></pre></div>


            <div class="docstring"><p>Uses domain-to-protein mappings present in the Residue Alignments provided to generate mapped representations of the residues from the DI_data structured ndarray provided.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>DI_data</strong> (numpy.ndarray):
Structured ndarray that contains columns "residue1" and "residue2".</li>
<li><strong>RA1</strong> (ResidueAlignment):
The ResidueAlignment used for mapping the first column of residues to the appropriate target sequence.</li>
<li><strong>RA2</strong> (ResidueAlignment):
The ResidueAlignment used for mapping the second column of residues to the appropriate target sequence.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>mappable_DI_data</strong> (numpy.ndarray):
DI_data that has been mapped to the target sequence specified in the generation of the corresponding ResidueAlignments.</li>
<li><strong>Note:</strong>: Residues that do not map to the target sequence of the ResidueAlignment are dropped.</li>
</ul>
</div>


                            </div>
                            <div id="DirectInformationData.rank_pairs" class="classattr">
                                        <input id="DirectInformationData.rank_pairs-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">rank_pairs</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">DI_data</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.rank_pairs-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.rank_pairs"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.rank_pairs-478"><a href="#DirectInformationData.rank_pairs-478"><span class="linenos">478</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData.rank_pairs-479"><a href="#DirectInformationData.rank_pairs-479"><span class="linenos">479</span></a>    <span class="k">def</span> <span class="nf">rank_pairs</span><span class="p">(</span><span class="n">DI_data</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="DirectInformationData.rank_pairs-480"><a href="#DirectInformationData.rank_pairs-480"><span class="linenos">480</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.rank_pairs-481"><a href="#DirectInformationData.rank_pairs-481"><span class="linenos">481</span></a><span class="sd">        Sorts a structured ndarray of pairs information to order the ndarray by Direct Information (DI) score.</span>
</span><span id="DirectInformationData.rank_pairs-482"><a href="#DirectInformationData.rank_pairs-482"><span class="linenos">482</span></a>
</span><span id="DirectInformationData.rank_pairs-483"><a href="#DirectInformationData.rank_pairs-483"><span class="linenos">483</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.rank_pairs-484"><a href="#DirectInformationData.rank_pairs-484"><span class="linenos">484</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.rank_pairs-485"><a href="#DirectInformationData.rank_pairs-485"><span class="linenos">485</span></a><span class="sd">        DI_data : numpy.ndarray</span>
</span><span id="DirectInformationData.rank_pairs-486"><a href="#DirectInformationData.rank_pairs-486"><span class="linenos">486</span></a><span class="sd">            Structured ndarray that contains columns with names &quot;residue1&quot;, &quot;residue2&quot;, and &quot;DI&quot; (Direct Information)</span>
</span><span id="DirectInformationData.rank_pairs-487"><a href="#DirectInformationData.rank_pairs-487"><span class="linenos">487</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.rank_pairs-488"><a href="#DirectInformationData.rank_pairs-488"><span class="linenos">488</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.rank_pairs-489"><a href="#DirectInformationData.rank_pairs-489"><span class="linenos">489</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.rank_pairs-490"><a href="#DirectInformationData.rank_pairs-490"><span class="linenos">490</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="DirectInformationData.rank_pairs-491"><a href="#DirectInformationData.rank_pairs-491"><span class="linenos">491</span></a><span class="sd">            Structured ndarray sorted upon column that is named DI in descending order.</span>
</span><span id="DirectInformationData.rank_pairs-492"><a href="#DirectInformationData.rank_pairs-492"><span class="linenos">492</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.rank_pairs-493"><a href="#DirectInformationData.rank_pairs-493"><span class="linenos">493</span></a>        <span class="c1"># [::-1] reverses the order from ascending DI Score to descending DI score.</span>
</span><span id="DirectInformationData.rank_pairs-494"><a href="#DirectInformationData.rank_pairs-494"><span class="linenos">494</span></a>        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">DI_data</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="s1">&#39;DI&#39;</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</span></pre></div>


            <div class="docstring"><p>Sorts a structured ndarray of pairs information to order the ndarray by Direct Information (DI) score.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>DI_data</strong> (numpy.ndarray):
Structured ndarray that contains columns with names "residue1", "residue2", and "DI" (Direct Information)</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>numpy.ndarray</strong>: Structured ndarray sorted upon column that is named DI in descending order.</li>
</ul>
</div>


                            </div>
                            <div id="DirectInformationData.nonlocal_pairs" class="classattr">
                                        <input id="DirectInformationData.nonlocal_pairs-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">nonlocal_pairs</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">DI_data</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.nonlocal_pairs-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.nonlocal_pairs"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.nonlocal_pairs-496"><a href="#DirectInformationData.nonlocal_pairs-496"><span class="linenos">496</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData.nonlocal_pairs-497"><a href="#DirectInformationData.nonlocal_pairs-497"><span class="linenos">497</span></a>    <span class="k">def</span> <span class="nf">nonlocal_pairs</span><span class="p">(</span><span class="n">DI_data</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="DirectInformationData.nonlocal_pairs-498"><a href="#DirectInformationData.nonlocal_pairs-498"><span class="linenos">498</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.nonlocal_pairs-499"><a href="#DirectInformationData.nonlocal_pairs-499"><span class="linenos">499</span></a><span class="sd">        Subsets a structured ndarray of pairs information to find nonlocal pairs, where residue interactions are likely not involved in secondary structure formation i.e. helices and sheet interactions. Nonlocal pairs must be at least 4 residues apart.</span>
</span><span id="DirectInformationData.nonlocal_pairs-500"><a href="#DirectInformationData.nonlocal_pairs-500"><span class="linenos">500</span></a>
</span><span id="DirectInformationData.nonlocal_pairs-501"><a href="#DirectInformationData.nonlocal_pairs-501"><span class="linenos">501</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.nonlocal_pairs-502"><a href="#DirectInformationData.nonlocal_pairs-502"><span class="linenos">502</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.nonlocal_pairs-503"><a href="#DirectInformationData.nonlocal_pairs-503"><span class="linenos">503</span></a><span class="sd">        DI_data : numpy.ndarray</span>
</span><span id="DirectInformationData.nonlocal_pairs-504"><a href="#DirectInformationData.nonlocal_pairs-504"><span class="linenos">504</span></a><span class="sd">            Structured ndarray that contains (at least) the first and second columns with the names &quot;residue1&quot; and &quot;residue2&quot; respectively.</span>
</span><span id="DirectInformationData.nonlocal_pairs-505"><a href="#DirectInformationData.nonlocal_pairs-505"><span class="linenos">505</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.nonlocal_pairs-506"><a href="#DirectInformationData.nonlocal_pairs-506"><span class="linenos">506</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.nonlocal_pairs-507"><a href="#DirectInformationData.nonlocal_pairs-507"><span class="linenos">507</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.nonlocal_pairs-508"><a href="#DirectInformationData.nonlocal_pairs-508"><span class="linenos">508</span></a><span class="sd">        numpy.ndarray</span>
</span><span id="DirectInformationData.nonlocal_pairs-509"><a href="#DirectInformationData.nonlocal_pairs-509"><span class="linenos">509</span></a><span class="sd">            Structured ndarray of DI pairs where residue 1 and residue 2 are at least 4 residues apart.</span>
</span><span id="DirectInformationData.nonlocal_pairs-510"><a href="#DirectInformationData.nonlocal_pairs-510"><span class="linenos">510</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.nonlocal_pairs-511"><a href="#DirectInformationData.nonlocal_pairs-511"><span class="linenos">511</span></a>        <span class="k">return</span> <span class="n">DI_data</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">DI_data</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">DI_data</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">]</span>
</span></pre></div>


            <div class="docstring"><p>Subsets a structured ndarray of pairs information to find nonlocal pairs, where residue interactions are likely not involved in secondary structure formation i.e. helices and sheet interactions. Nonlocal pairs must be at least 4 residues apart.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>DI_data</strong> (numpy.ndarray):
Structured ndarray that contains (at least) the first and second columns with the names "residue1" and "residue2" respectively.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>numpy.ndarray</strong>: Structured ndarray of DI pairs where residue 1 and residue 2 are at least 4 residues apart.</li>
</ul>
</div>


                            </div>
                            <div id="DirectInformationData.find_DI_with_residues" class="classattr">
                                        <input id="DirectInformationData.find_DI_with_residues-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">find_DI_with_residues</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">critical_residues_1</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span>,</span><span class="param">	<span class="n">critical_residues_2</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span>,</span><span class="param">	<span class="n">max_rank</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>,</span><span class="param">	<span class="o">*</span><span class="n">mapped_resi_arrs</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]]</span></span><span class="return-annotation">) -> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">list</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.find_DI_with_residues-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.find_DI_with_residues"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.find_DI_with_residues-513"><a href="#DirectInformationData.find_DI_with_residues-513"><span class="linenos">513</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData.find_DI_with_residues-514"><a href="#DirectInformationData.find_DI_with_residues-514"><span class="linenos">514</span></a>    <span class="k">def</span> <span class="nf">find_DI_with_residues</span><span class="p">(</span><span class="n">critical_residues_1</span> <span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">critical_residues_2</span> <span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">max_rank</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">*</span><span class="n">mapped_resi_arrs</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">list</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
</span><span id="DirectInformationData.find_DI_with_residues-515"><a href="#DirectInformationData.find_DI_with_residues-515"><span class="linenos">515</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.find_DI_with_residues-516"><a href="#DirectInformationData.find_DI_with_residues-516"><span class="linenos">516</span></a><span class="sd">        Function that takes an n number of ranked, mapped DI pairs and checks to see if they&#39;re in a list of potential residue indices.</span>
</span><span id="DirectInformationData.find_DI_with_residues-517"><a href="#DirectInformationData.find_DI_with_residues-517"><span class="linenos">517</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.find_DI_with_residues-518"><a href="#DirectInformationData.find_DI_with_residues-518"><span class="linenos">518</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.find_DI_with_residues-519"><a href="#DirectInformationData.find_DI_with_residues-519"><span class="linenos">519</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.find_DI_with_residues-520"><a href="#DirectInformationData.find_DI_with_residues-520"><span class="linenos">520</span></a><span class="sd">        critical_residues_1 : collections.abc.Iterable of int</span>
</span><span id="DirectInformationData.find_DI_with_residues-521"><a href="#DirectInformationData.find_DI_with_residues-521"><span class="linenos">521</span></a><span class="sd">            Specific residue indices that a DI pair will be compared to. If the first residue of the DI pair is not one of these indices, it will not be appended to results.</span>
</span><span id="DirectInformationData.find_DI_with_residues-522"><a href="#DirectInformationData.find_DI_with_residues-522"><span class="linenos">522</span></a><span class="sd">        crtical_residues_2 : collections.abc.Iterable of int</span>
</span><span id="DirectInformationData.find_DI_with_residues-523"><a href="#DirectInformationData.find_DI_with_residues-523"><span class="linenos">523</span></a><span class="sd">            Specific residue indices that a DI pair will be compared to. If the second residue of the DI pair is not one of these indices, it will not be appended to results.</span>
</span><span id="DirectInformationData.find_DI_with_residues-524"><a href="#DirectInformationData.find_DI_with_residues-524"><span class="linenos">524</span></a><span class="sd">        threshold : int, optional</span>
</span><span id="DirectInformationData.find_DI_with_residues-525"><a href="#DirectInformationData.find_DI_with_residues-525"><span class="linenos">525</span></a><span class="sd">            Maximum &quot;rank&quot; of the DI pair considered.</span>
</span><span id="DirectInformationData.find_DI_with_residues-526"><a href="#DirectInformationData.find_DI_with_residues-526"><span class="linenos">526</span></a><span class="sd">        *mapped_resi_arrs : tuple of numpy.ndarray</span>
</span><span id="DirectInformationData.find_DI_with_residues-527"><a href="#DirectInformationData.find_DI_with_residues-527"><span class="linenos">527</span></a><span class="sd">            Tuple of ranked, mapped pairs that are compared to critical residue indices and appended to results if in those indices and within threshold.</span>
</span><span id="DirectInformationData.find_DI_with_residues-528"><a href="#DirectInformationData.find_DI_with_residues-528"><span class="linenos">528</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.find_DI_with_residues-529"><a href="#DirectInformationData.find_DI_with_residues-529"><span class="linenos">529</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.find_DI_with_residues-530"><a href="#DirectInformationData.find_DI_with_residues-530"><span class="linenos">530</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.find_DI_with_residues-531"><a href="#DirectInformationData.find_DI_with_residues-531"><span class="linenos">531</span></a><span class="sd">        results : list of tuple of list of int, int</span>
</span><span id="DirectInformationData.find_DI_with_residues-532"><a href="#DirectInformationData.find_DI_with_residues-532"><span class="linenos">532</span></a><span class="sd">            Results which consist of tuples where the first element is a list of residue1, residue2, and DI score, whereas the second element is the rank.</span>
</span><span id="DirectInformationData.find_DI_with_residues-533"><a href="#DirectInformationData.find_DI_with_residues-533"><span class="linenos">533</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.find_DI_with_residues-534"><a href="#DirectInformationData.find_DI_with_residues-534"><span class="linenos">534</span></a>        <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="DirectInformationData.find_DI_with_residues-535"><a href="#DirectInformationData.find_DI_with_residues-535"><span class="linenos">535</span></a>        <span class="k">for</span> <span class="n">mapped_resi_arr</span> <span class="ow">in</span> <span class="n">mapped_resi_arrs</span><span class="p">:</span>
</span><span id="DirectInformationData.find_DI_with_residues-536"><a href="#DirectInformationData.find_DI_with_residues-536"><span class="linenos">536</span></a>            <span class="c1"># count_rank represents the rank of the DI pair being evaluated, iterating over every new row considered.</span>
</span><span id="DirectInformationData.find_DI_with_residues-537"><a href="#DirectInformationData.find_DI_with_residues-537"><span class="linenos">537</span></a>            <span class="n">count_rank</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="DirectInformationData.find_DI_with_residues-538"><a href="#DirectInformationData.find_DI_with_residues-538"><span class="linenos">538</span></a>            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">mapped_resi_arr</span><span class="p">:</span>
</span><span id="DirectInformationData.find_DI_with_residues-539"><a href="#DirectInformationData.find_DI_with_residues-539"><span class="linenos">539</span></a>                <span class="n">row_as_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
</span><span id="DirectInformationData.find_DI_with_residues-540"><a href="#DirectInformationData.find_DI_with_residues-540"><span class="linenos">540</span></a>                <span class="n">count_rank</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span id="DirectInformationData.find_DI_with_residues-541"><a href="#DirectInformationData.find_DI_with_residues-541"><span class="linenos">541</span></a>                <span class="k">if</span> <span class="n">max_rank</span><span class="p">:</span>
</span><span id="DirectInformationData.find_DI_with_residues-542"><a href="#DirectInformationData.find_DI_with_residues-542"><span class="linenos">542</span></a>                    <span class="k">if</span> <span class="n">row_as_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">critical_residues_1</span> <span class="ow">and</span> <span class="n">row_as_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">critical_residues_2</span> <span class="ow">and</span> <span class="n">count_rank</span> <span class="o">&lt;=</span> <span class="n">max_rank</span><span class="p">:</span>
</span><span id="DirectInformationData.find_DI_with_residues-543"><a href="#DirectInformationData.find_DI_with_residues-543"><span class="linenos">543</span></a>                        <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">row_as_list</span><span class="p">,</span> <span class="n">count_rank</span><span class="p">))</span>
</span><span id="DirectInformationData.find_DI_with_residues-544"><a href="#DirectInformationData.find_DI_with_residues-544"><span class="linenos">544</span></a>                <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData.find_DI_with_residues-545"><a href="#DirectInformationData.find_DI_with_residues-545"><span class="linenos">545</span></a>                    <span class="k">if</span> <span class="n">row_as_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">critical_residues_1</span> <span class="ow">and</span> <span class="n">row_as_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">critical_residues_2</span><span class="p">:</span>
</span><span id="DirectInformationData.find_DI_with_residues-546"><a href="#DirectInformationData.find_DI_with_residues-546"><span class="linenos">546</span></a>                        <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">row_as_list</span><span class="p">,</span> <span class="n">count_rank</span><span class="p">))</span>
</span><span id="DirectInformationData.find_DI_with_residues-547"><a href="#DirectInformationData.find_DI_with_residues-547"><span class="linenos">547</span></a>        <span class="k">return</span> <span class="n">results</span>
</span></pre></div>


            <div class="docstring"><p>Function that takes an n number of ranked, mapped DI pairs and checks to see if they're in a list of potential residue indices.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>critical_residues_1</strong> (collections.abc.Iterable of int):
Specific residue indices that a DI pair will be compared to. If the first residue of the DI pair is not one of these indices, it will not be appended to results.</li>
<li><strong>crtical_residues_2</strong> (collections.abc.Iterable of int):
Specific residue indices that a DI pair will be compared to. If the second residue of the DI pair is not one of these indices, it will not be appended to results.</li>
<li><strong>threshold</strong> (int, optional):
Maximum "rank" of the DI pair considered.</li>
<li><strong>*mapped_resi_arrs</strong> (tuple of numpy.ndarray):
Tuple of ranked, mapped pairs that are compared to critical residue indices and appended to results if in those indices and within threshold.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>results</strong> (list of tuple of list of int, int):
Results which consist of tuples where the first element is a list of residue1, residue2, and DI score, whereas the second element is the rank.</li>
</ul>
</div>


                            </div>
                            <div id="DirectInformationData.get_dist_commands" class="classattr">
                                        <input id="DirectInformationData.get_dist_commands-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">get_dist_commands</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">model1</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span>,</span><span class="param">	<span class="n">model2</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span>,</span><span class="param">	<span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">pairs</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>,</span><span class="param">	<span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>,</span><span class="param">	<span class="n">auth_res_ids</span><span class="o">=</span><span class="kc">False</span></span><span class="return-annotation">) -> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.get_dist_commands-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.get_dist_commands"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.get_dist_commands-549"><a href="#DirectInformationData.get_dist_commands-549"><span class="linenos">549</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData.get_dist_commands-550"><a href="#DirectInformationData.get_dist_commands-550"><span class="linenos">550</span></a>    <span class="k">def</span> <span class="nf">get_dist_commands</span><span class="p">(</span><span class="n">model1</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span> <span class="n">model2</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auth_res_ids</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
</span><span id="DirectInformationData.get_dist_commands-551"><a href="#DirectInformationData.get_dist_commands-551"><span class="linenos">551</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.get_dist_commands-552"><a href="#DirectInformationData.get_dist_commands-552"><span class="linenos">552</span></a><span class="sd">        Get UCSF Chimera commands for displaying distance commands for usage in displaying distances between residue pairs. Options are present for alpha-carbon to alpha-carbon distance or for specified atom to specified atom distance.</span>
</span><span id="DirectInformationData.get_dist_commands-553"><a href="#DirectInformationData.get_dist_commands-553"><span class="linenos">553</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.get_dist_commands-554"><a href="#DirectInformationData.get_dist_commands-554"><span class="linenos">554</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.get_dist_commands-555"><a href="#DirectInformationData.get_dist_commands-555"><span class="linenos">555</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.get_dist_commands-556"><a href="#DirectInformationData.get_dist_commands-556"><span class="linenos">556</span></a><span class="sd">        model1 : str, int</span>
</span><span id="DirectInformationData.get_dist_commands-557"><a href="#DirectInformationData.get_dist_commands-557"><span class="linenos">557</span></a><span class="sd">            Number of model corresponding to the structure containing the first column of residues.</span>
</span><span id="DirectInformationData.get_dist_commands-558"><a href="#DirectInformationData.get_dist_commands-558"><span class="linenos">558</span></a><span class="sd">        model2 : str, int</span>
</span><span id="DirectInformationData.get_dist_commands-559"><a href="#DirectInformationData.get_dist_commands-559"><span class="linenos">559</span></a><span class="sd">            Number of model corresponding to the structure containing the second column of residues.</span>
</span><span id="DirectInformationData.get_dist_commands-560"><a href="#DirectInformationData.get_dist_commands-560"><span class="linenos">560</span></a><span class="sd">        chain1 : str</span>
</span><span id="DirectInformationData.get_dist_commands-561"><a href="#DirectInformationData.get_dist_commands-561"><span class="linenos">561</span></a><span class="sd">            The chain present in the structure in model1 containing the first column of residues.</span>
</span><span id="DirectInformationData.get_dist_commands-562"><a href="#DirectInformationData.get_dist_commands-562"><span class="linenos">562</span></a><span class="sd">        chain2 : str</span>
</span><span id="DirectInformationData.get_dist_commands-563"><a href="#DirectInformationData.get_dist_commands-563"><span class="linenos">563</span></a><span class="sd">            The chain present in the structure in model2 containing the second column of residues.</span>
</span><span id="DirectInformationData.get_dist_commands-564"><a href="#DirectInformationData.get_dist_commands-564"><span class="linenos">564</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="DirectInformationData.get_dist_commands-565"><a href="#DirectInformationData.get_dist_commands-565"><span class="linenos">565</span></a><span class="sd">            Structured ndarray that contains (at least) the first and second columns with the names &quot;residue1&quot; and &quot;residue2&quot; respectively. `ca_only` should be set to false and an additional &quot;atom_name1&quot; and &quot;atom_name2&quot; column should be added if atoms are specified per pair.</span>
</span><span id="DirectInformationData.get_dist_commands-566"><a href="#DirectInformationData.get_dist_commands-566"><span class="linenos">566</span></a><span class="sd">        ca_only : bool</span>
</span><span id="DirectInformationData.get_dist_commands-567"><a href="#DirectInformationData.get_dist_commands-567"><span class="linenos">567</span></a><span class="sd">            True if distance commands are for displaying distances between the two alpha-carbons of the residue pair. If false, specific_atom_names is used in lieu of &quot;CA&quot; as an atom identifier.</span>
</span><span id="DirectInformationData.get_dist_commands-568"><a href="#DirectInformationData.get_dist_commands-568"><span class="linenos">568</span></a><span class="sd">        auth_res_ids : bool</span>
</span><span id="DirectInformationData.get_dist_commands-569"><a href="#DirectInformationData.get_dist_commands-569"><span class="linenos">569</span></a><span class="sd">            If true, use the &quot;auth_residue1&quot; and &quot;auth_residue2&quot; columns instead of &quot;residue1&quot; and &quot;residue2&quot; columns. These residue ids correspond to the auth protein residue ids.</span>
</span><span id="DirectInformationData.get_dist_commands-570"><a href="#DirectInformationData.get_dist_commands-570"><span class="linenos">570</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.get_dist_commands-571"><a href="#DirectInformationData.get_dist_commands-571"><span class="linenos">571</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.get_dist_commands-572"><a href="#DirectInformationData.get_dist_commands-572"><span class="linenos">572</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.get_dist_commands-573"><a href="#DirectInformationData.get_dist_commands-573"><span class="linenos">573</span></a><span class="sd">        distance_commands : list of str</span>
</span><span id="DirectInformationData.get_dist_commands-574"><a href="#DirectInformationData.get_dist_commands-574"><span class="linenos">574</span></a><span class="sd">            List of distance commands generated between two residues with model and chain information needed, either between two alpha-carbons or the specified atoms.</span>
</span><span id="DirectInformationData.get_dist_commands-575"><a href="#DirectInformationData.get_dist_commands-575"><span class="linenos">575</span></a><span class="sd">        </span>
</span><span id="DirectInformationData.get_dist_commands-576"><a href="#DirectInformationData.get_dist_commands-576"><span class="linenos">576</span></a><span class="sd">        Note</span>
</span><span id="DirectInformationData.get_dist_commands-577"><a href="#DirectInformationData.get_dist_commands-577"><span class="linenos">577</span></a><span class="sd">        ----</span>
</span><span id="DirectInformationData.get_dist_commands-578"><a href="#DirectInformationData.get_dist_commands-578"><span class="linenos">578</span></a><span class="sd">        model1 and model2 can be equivalent if both columns involve residues referenced by the same model. The same would apply for chains if the residues are present on the same chain. </span>
</span><span id="DirectInformationData.get_dist_commands-579"><a href="#DirectInformationData.get_dist_commands-579"><span class="linenos">579</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.get_dist_commands-580"><a href="#DirectInformationData.get_dist_commands-580"><span class="linenos">580</span></a>        <span class="n">distance_commands</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="DirectInformationData.get_dist_commands-581"><a href="#DirectInformationData.get_dist_commands-581"><span class="linenos">581</span></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">pairs</span><span class="p">)[</span><span class="mi">0</span><span class="p">]):</span>
</span><span id="DirectInformationData.get_dist_commands-582"><a href="#DirectInformationData.get_dist_commands-582"><span class="linenos">582</span></a>            <span class="k">if</span> <span class="n">auth_res_ids</span><span class="p">:</span>
</span><span id="DirectInformationData.get_dist_commands-583"><a href="#DirectInformationData.get_dist_commands-583"><span class="linenos">583</span></a>                <span class="n">residue1</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;auth_residue1&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData.get_dist_commands-584"><a href="#DirectInformationData.get_dist_commands-584"><span class="linenos">584</span></a>                <span class="n">residue2</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;auth_residue2&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData.get_dist_commands-585"><a href="#DirectInformationData.get_dist_commands-585"><span class="linenos">585</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData.get_dist_commands-586"><a href="#DirectInformationData.get_dist_commands-586"><span class="linenos">586</span></a>                <span class="n">residue1</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData.get_dist_commands-587"><a href="#DirectInformationData.get_dist_commands-587"><span class="linenos">587</span></a>                <span class="n">residue2</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;residue2&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData.get_dist_commands-588"><a href="#DirectInformationData.get_dist_commands-588"><span class="linenos">588</span></a>            <span class="k">if</span> <span class="n">ca_only</span><span class="p">:</span>
</span><span id="DirectInformationData.get_dist_commands-589"><a href="#DirectInformationData.get_dist_commands-589"><span class="linenos">589</span></a>                <span class="n">distance_commands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;distance #</span><span class="si">{</span><span class="n">model1</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">residue1</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chain1</span><span class="si">}</span><span class="s2">@CA #</span><span class="si">{</span><span class="n">model2</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">residue2</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chain2</span><span class="si">}</span><span class="s2">@CA;&quot;</span><span class="p">)</span>
</span><span id="DirectInformationData.get_dist_commands-590"><a href="#DirectInformationData.get_dist_commands-590"><span class="linenos">590</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="DirectInformationData.get_dist_commands-591"><a href="#DirectInformationData.get_dist_commands-591"><span class="linenos">591</span></a>                <span class="n">atom1</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;atom_name1&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData.get_dist_commands-592"><a href="#DirectInformationData.get_dist_commands-592"><span class="linenos">592</span></a>                <span class="n">atom2</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;atom_name2&#39;</span><span class="p">]</span>
</span><span id="DirectInformationData.get_dist_commands-593"><a href="#DirectInformationData.get_dist_commands-593"><span class="linenos">593</span></a>                <span class="n">distance_commands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;distance #</span><span class="si">{</span><span class="n">model1</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">residue1</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chain1</span><span class="si">}</span><span class="s2">@</span><span class="si">{</span><span class="n">atom1</span><span class="si">}</span><span class="s2"> #</span><span class="si">{</span><span class="n">model2</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">residue2</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chain2</span><span class="si">}</span><span class="s2">@</span><span class="si">{</span><span class="n">atom2</span><span class="si">}</span><span class="s2">;&quot;</span><span class="p">)</span>
</span><span id="DirectInformationData.get_dist_commands-594"><a href="#DirectInformationData.get_dist_commands-594"><span class="linenos">594</span></a>        <span class="k">return</span> <span class="n">distance_commands</span>
</span></pre></div>


            <div class="docstring"><p>Get UCSF Chimera commands for displaying distance commands for usage in displaying distances between residue pairs. Options are present for alpha-carbon to alpha-carbon distance or for specified atom to specified atom distance.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>model1</strong> (str, int):
Number of model corresponding to the structure containing the first column of residues.</li>
<li><strong>model2</strong> (str, int):
Number of model corresponding to the structure containing the second column of residues.</li>
<li><strong>chain1</strong> (str):
The chain present in the structure in model1 containing the first column of residues.</li>
<li><strong>chain2</strong> (str):
The chain present in the structure in model2 containing the second column of residues.</li>
<li><strong>pairs</strong> (numpy.ndarray):
Structured ndarray that contains (at least) the first and second columns with the names "residue1" and "residue2" respectively. <code>ca_only</code> should be set to false and an additional "atom_name1" and "atom_name2" column should be added if atoms are specified per pair.</li>
<li><strong>ca_only</strong> (bool):
True if distance commands are for displaying distances between the two alpha-carbons of the residue pair. If false, specific_atom_names is used in lieu of "CA" as an atom identifier.</li>
<li><strong>auth_res_ids</strong> (bool):
If true, use the "auth_residue1" and "auth_residue2" columns instead of "residue1" and "residue2" columns. These residue ids correspond to the auth protein residue ids.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>distance_commands</strong> (list of str):
List of distance commands generated between two residues with model and chain information needed, either between two alpha-carbons or the specified atoms.</li>
</ul>

<h6 id="note">Note</h6>

<p>model1 and model2 can be equivalent if both columns involve residues referenced by the same model. The same would apply for chains if the residues are present on the same chain.</p>
</div>


                            </div>
                            <div id="DirectInformationData.write_DI_data" class="classattr">
                                        <input id="DirectInformationData.write_DI_data-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">write_DI_data</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">pairs</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>,</span><span class="param">	<span class="n">delimiter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span>,</span><span class="param">	<span class="n">fmt</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%.3f</span><span class="s1">&#39;</span><span class="p">)</span></span><span class="return-annotation">) -> <span class="kc">None</span>:</span></span>

                <label class="view-source-button" for="DirectInformationData.write_DI_data-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#DirectInformationData.write_DI_data"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="DirectInformationData.write_DI_data-596"><a href="#DirectInformationData.write_DI_data-596"><span class="linenos">596</span></a>    <span class="nd">@staticmethod</span>
</span><span id="DirectInformationData.write_DI_data-597"><a href="#DirectInformationData.write_DI_data-597"><span class="linenos">597</span></a>    <span class="k">def</span> <span class="nf">write_DI_data</span><span class="p">(</span><span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">delimiter</span><span class="p">:</span> <span class="nb">str</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">fmt</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%.3f</span><span class="s1">&#39;</span><span class="p">))</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="DirectInformationData.write_DI_data-598"><a href="#DirectInformationData.write_DI_data-598"><span class="linenos">598</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="DirectInformationData.write_DI_data-599"><a href="#DirectInformationData.write_DI_data-599"><span class="linenos">599</span></a><span class="sd">        Writes pairs ndarray to file with specified delimiter between the pairs&#39; row elements, i.e. residue 1, residue 2, and DI score.</span>
</span><span id="DirectInformationData.write_DI_data-600"><a href="#DirectInformationData.write_DI_data-600"><span class="linenos">600</span></a>
</span><span id="DirectInformationData.write_DI_data-601"><a href="#DirectInformationData.write_DI_data-601"><span class="linenos">601</span></a><span class="sd">        Parameters</span>
</span><span id="DirectInformationData.write_DI_data-602"><a href="#DirectInformationData.write_DI_data-602"><span class="linenos">602</span></a><span class="sd">        ----------</span>
</span><span id="DirectInformationData.write_DI_data-603"><a href="#DirectInformationData.write_DI_data-603"><span class="linenos">603</span></a><span class="sd">        filepath : str</span>
</span><span id="DirectInformationData.write_DI_data-604"><a href="#DirectInformationData.write_DI_data-604"><span class="linenos">604</span></a><span class="sd">            Path of the file to write DirectInformation data to.</span>
</span><span id="DirectInformationData.write_DI_data-605"><a href="#DirectInformationData.write_DI_data-605"><span class="linenos">605</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="DirectInformationData.write_DI_data-606"><a href="#DirectInformationData.write_DI_data-606"><span class="linenos">606</span></a><span class="sd">            Ndarray of at-least pairs information (residue 1, residue 2) and optionally Direct Information to write to a file via numpy.savetxt().</span>
</span><span id="DirectInformationData.write_DI_data-607"><a href="#DirectInformationData.write_DI_data-607"><span class="linenos">607</span></a><span class="sd">        delimiter : str</span>
</span><span id="DirectInformationData.write_DI_data-608"><a href="#DirectInformationData.write_DI_data-608"><span class="linenos">608</span></a><span class="sd">            Delimiter to separate columns of the pairs ndarray when writing to a file.</span>
</span><span id="DirectInformationData.write_DI_data-609"><a href="#DirectInformationData.write_DI_data-609"><span class="linenos">609</span></a><span class="sd">        fmt : tuple of str, default=(&#39;%d&#39;, &#39;%d&#39;, &#39;%.3f)</span>
</span><span id="DirectInformationData.write_DI_data-610"><a href="#DirectInformationData.write_DI_data-610"><span class="linenos">610</span></a><span class="sd">            format passed as an argument to numpy.savetxt() to define type of column and output format. Set to (&#39;%d&#39;, &#39;%d&#39;) if only two columns are present in the ndarray.</span>
</span><span id="DirectInformationData.write_DI_data-611"><a href="#DirectInformationData.write_DI_data-611"><span class="linenos">611</span></a>
</span><span id="DirectInformationData.write_DI_data-612"><a href="#DirectInformationData.write_DI_data-612"><span class="linenos">612</span></a><span class="sd">        Returns</span>
</span><span id="DirectInformationData.write_DI_data-613"><a href="#DirectInformationData.write_DI_data-613"><span class="linenos">613</span></a><span class="sd">        -------</span>
</span><span id="DirectInformationData.write_DI_data-614"><a href="#DirectInformationData.write_DI_data-614"><span class="linenos">614</span></a><span class="sd">        None</span>
</span><span id="DirectInformationData.write_DI_data-615"><a href="#DirectInformationData.write_DI_data-615"><span class="linenos">615</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="DirectInformationData.write_DI_data-616"><a href="#DirectInformationData.write_DI_data-616"><span class="linenos">616</span></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">pairs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
</span><span id="DirectInformationData.write_DI_data-617"><a href="#DirectInformationData.write_DI_data-617"><span class="linenos">617</span></a>            <span class="n">fmt</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">%d</span><span class="s1">&#39;</span><span class="p">)</span>
</span><span id="DirectInformationData.write_DI_data-618"><a href="#DirectInformationData.write_DI_data-618"><span class="linenos">618</span></a>        <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">pairs</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="n">delimiter</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="n">fmt</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Writes pairs ndarray to file with specified delimiter between the pairs' row elements, i.e. residue 1, residue 2, and DI score.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>filepath</strong> (str):
Path of the file to write DirectInformation data to.</li>
<li><strong>pairs</strong> (numpy.ndarray):
Ndarray of at-least pairs information (residue 1, residue 2) and optionally Direct Information to write to a file via numpy.savetxt().</li>
<li><strong>delimiter</strong> (str):
Delimiter to separate columns of the pairs ndarray when writing to a file.</li>
<li><strong>fmt</strong> (tuple of str, default=('%d', '%d', '%.3f)):
format passed as an argument to numpy.savetxt() to define type of column and output format. Set to ('%d', '%d') if only two columns are present in the ndarray.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>None</strong></li>
</ul>
</div>


                            </div>
                </section>
                <section id="StructureInformation">
                            <input id="StructureInformation-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr class">
            
    <span class="def">class</span>
    <span class="name">StructureInformation</span>:

                <label class="view-source-button" for="StructureInformation-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#StructureInformation"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="StructureInformation-620"><a href="#StructureInformation-620"><span class="linenos">620</span></a><span class="k">class</span> <span class="nc">StructureInformation</span><span class="p">:</span>
</span><span id="StructureInformation-621"><a href="#StructureInformation-621"><span class="linenos">621</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="StructureInformation-622"><a href="#StructureInformation-622"><span class="linenos">622</span></a><span class="sd">    Information regarding a protein structure, obtained from a protein structure file.</span>
</span><span id="StructureInformation-623"><a href="#StructureInformation-623"><span class="linenos">623</span></a>
</span><span id="StructureInformation-624"><a href="#StructureInformation-624"><span class="linenos">624</span></a><span class="sd">    Uses fetch_pdb() to pull protein structure information from RCSB. Uses read_x_file() to supply a filepath to pull protein structure information from a file.</span>
</span><span id="StructureInformation-625"><a href="#StructureInformation-625"><span class="linenos">625</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="StructureInformation-626"><a href="#StructureInformation-626"><span class="linenos">626</span></a>    <span class="nd">@overload</span>
</span><span id="StructureInformation-627"><a href="#StructureInformation-627"><span class="linenos">627</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation-628"><a href="#StructureInformation-628"><span class="linenos">628</span></a>    <span class="k">def</span> <span class="nf">fetch_pdb</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">struc_format</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;mmcif&quot;</span><span class="p">]</span><span class="o">=</span><span class="s2">&quot;mmcif&quot;</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;MMCIFInformation&#39;</span><span class="p">:</span>
</span><span id="StructureInformation-629"><a href="#StructureInformation-629"><span class="linenos">629</span></a>        <span class="o">...</span>
</span><span id="StructureInformation-630"><a href="#StructureInformation-630"><span class="linenos">630</span></a>
</span><span id="StructureInformation-631"><a href="#StructureInformation-631"><span class="linenos">631</span></a>    <span class="nd">@overload</span>
</span><span id="StructureInformation-632"><a href="#StructureInformation-632"><span class="linenos">632</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation-633"><a href="#StructureInformation-633"><span class="linenos">633</span></a>    <span class="k">def</span> <span class="nf">fetch_pdb</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">struc_format</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;pdb&quot;</span><span class="p">],</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;PDBInformation&#39;</span><span class="p">:</span>
</span><span id="StructureInformation-634"><a href="#StructureInformation-634"><span class="linenos">634</span></a>        <span class="o">...</span>
</span><span id="StructureInformation-635"><a href="#StructureInformation-635"><span class="linenos">635</span></a>
</span><span id="StructureInformation-636"><a href="#StructureInformation-636"><span class="linenos">636</span></a>
</span><span id="StructureInformation-637"><a href="#StructureInformation-637"><span class="linenos">637</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation-638"><a href="#StructureInformation-638"><span class="linenos">638</span></a>    <span class="k">def</span> <span class="nf">fetch_pdb</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">struc_format</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;mmcif&quot;</span><span class="p">,</span> <span class="s2">&quot;pdb&quot;</span><span class="p">]</span><span class="o">=</span><span class="s2">&quot;mmcif&quot;</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="s1">&#39;MMCIFInformation&#39;</span><span class="p">,</span> <span class="s1">&#39;PDBInformation&#39;</span><span class="p">]:</span>
</span><span id="StructureInformation-639"><a href="#StructureInformation-639"><span class="linenos">639</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="StructureInformation-640"><a href="#StructureInformation-640"><span class="linenos">640</span></a><span class="sd">        Fetches PDB as mmCIF file from RCSB and compiles the information into a StructureInformation instance.</span>
</span><span id="StructureInformation-641"><a href="#StructureInformation-641"><span class="linenos">641</span></a>
</span><span id="StructureInformation-642"><a href="#StructureInformation-642"><span class="linenos">642</span></a><span class="sd">        Parameters</span>
</span><span id="StructureInformation-643"><a href="#StructureInformation-643"><span class="linenos">643</span></a><span class="sd">        ----------</span>
</span><span id="StructureInformation-644"><a href="#StructureInformation-644"><span class="linenos">644</span></a><span class="sd">        pdb_id : str</span>
</span><span id="StructureInformation-645"><a href="#StructureInformation-645"><span class="linenos">645</span></a><span class="sd">            PDB ID to be fetched from the RCSB database.</span>
</span><span id="StructureInformation-646"><a href="#StructureInformation-646"><span class="linenos">646</span></a><span class="sd">        struc_format : str</span>
</span><span id="StructureInformation-647"><a href="#StructureInformation-647"><span class="linenos">647</span></a><span class="sd">            The format of the file to pull from the RCSB database.</span>
</span><span id="StructureInformation-648"><a href="#StructureInformation-648"><span class="linenos">648</span></a><span class="sd">        model_num : int</span>
</span><span id="StructureInformation-649"><a href="#StructureInformation-649"><span class="linenos">649</span></a><span class="sd">            The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</span>
</span><span id="StructureInformation-650"><a href="#StructureInformation-650"><span class="linenos">650</span></a><span class="sd">            </span>
</span><span id="StructureInformation-651"><a href="#StructureInformation-651"><span class="linenos">651</span></a><span class="sd">        Returns</span>
</span><span id="StructureInformation-652"><a href="#StructureInformation-652"><span class="linenos">652</span></a><span class="sd">        -------</span>
</span><span id="StructureInformation-653"><a href="#StructureInformation-653"><span class="linenos">653</span></a><span class="sd">        StructureInformation</span>
</span><span id="StructureInformation-654"><a href="#StructureInformation-654"><span class="linenos">654</span></a><span class="sd">            StructureInformation generated from pdbx.get_structure() function using the pdbx file fetched from RCSB.</span>
</span><span id="StructureInformation-655"><a href="#StructureInformation-655"><span class="linenos">655</span></a><span class="sd">        </span>
</span><span id="StructureInformation-656"><a href="#StructureInformation-656"><span class="linenos">656</span></a><span class="sd">        Raises</span>
</span><span id="StructureInformation-657"><a href="#StructureInformation-657"><span class="linenos">657</span></a><span class="sd">        ------</span>
</span><span id="StructureInformation-658"><a href="#StructureInformation-658"><span class="linenos">658</span></a><span class="sd">        TypeError</span>
</span><span id="StructureInformation-659"><a href="#StructureInformation-659"><span class="linenos">659</span></a><span class="sd">            Fetched data was not found and returned None instead.</span>
</span><span id="StructureInformation-660"><a href="#StructureInformation-660"><span class="linenos">660</span></a><span class="sd">        ValueError</span>
</span><span id="StructureInformation-661"><a href="#StructureInformation-661"><span class="linenos">661</span></a><span class="sd">            Structure format may be invalid (not PDBx/mmCIF or PDB).</span>
</span><span id="StructureInformation-662"><a href="#StructureInformation-662"><span class="linenos">662</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="StructureInformation-663"><a href="#StructureInformation-663"><span class="linenos">663</span></a>        <span class="n">fetched_data</span> <span class="o">=</span> <span class="n">rcsb</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">,</span> <span class="n">struc_format</span><span class="p">)</span>
</span><span id="StructureInformation-664"><a href="#StructureInformation-664"><span class="linenos">664</span></a>        <span class="k">if</span> <span class="n">fetched_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="StructureInformation-665"><a href="#StructureInformation-665"><span class="linenos">665</span></a>            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;RCSB fetch failed. Try fetch again.&quot;</span><span class="p">)</span>
</span><span id="StructureInformation-666"><a href="#StructureInformation-666"><span class="linenos">666</span></a>        <span class="k">elif</span> <span class="n">struc_format</span> <span class="o">==</span> <span class="s2">&quot;mmcif&quot;</span><span class="p">:</span>
</span><span id="StructureInformation-667"><a href="#StructureInformation-667"><span class="linenos">667</span></a>            <span class="n">pdbx_file</span> <span class="o">=</span> <span class="n">pdbx</span><span class="o">.</span><span class="n">CIFFile</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fetched_data</span><span class="p">)</span>
</span><span id="StructureInformation-668"><a href="#StructureInformation-668"><span class="linenos">668</span></a>            <span class="k">return</span> <span class="n">MMCIFInformation</span><span class="p">(</span><span class="n">pdbx</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">pdbx_file</span><span class="o">=</span><span class="n">pdbx_file</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model_num</span><span class="p">,</span> <span class="n">use_author_fields</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="n">pdbx_file</span><span class="p">,</span> <span class="n">model_num</span><span class="p">)</span>
</span><span id="StructureInformation-669"><a href="#StructureInformation-669"><span class="linenos">669</span></a>        <span class="k">elif</span> <span class="n">struc_format</span> <span class="o">==</span> <span class="s2">&quot;pdb&quot;</span><span class="p">:</span>
</span><span id="StructureInformation-670"><a href="#StructureInformation-670"><span class="linenos">670</span></a>            <span class="n">pdb_file</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">PDBFile</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fetched_data</span><span class="p">)</span>
</span><span id="StructureInformation-671"><a href="#StructureInformation-671"><span class="linenos">671</span></a>            <span class="k">return</span> <span class="n">PDBInformation</span><span class="p">(</span><span class="n">pdb</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">pdb_file</span><span class="o">=</span><span class="n">pdb_file</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model_num</span><span class="p">),</span> <span class="n">pdb_file</span><span class="o">=</span><span class="n">pdb_file</span><span class="p">,</span> <span class="n">model_num</span><span class="o">=</span><span class="n">model_num</span><span class="p">)</span>
</span><span id="StructureInformation-672"><a href="#StructureInformation-672"><span class="linenos">672</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="StructureInformation-673"><a href="#StructureInformation-673"><span class="linenos">673</span></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;struc_format </span><span class="si">{</span><span class="n">struc_format</span><span class="si">}</span><span class="s2"> is not valid or currently supported by DCA Toolkit&quot;</span><span class="p">)</span>
</span><span id="StructureInformation-674"><a href="#StructureInformation-674"><span class="linenos">674</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation-675"><a href="#StructureInformation-675"><span class="linenos">675</span></a>    <span class="k">def</span> <span class="nf">read_mmCIF_file</span><span class="p">(</span><span class="n">pdbx_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;MMCIFInformation&#39;</span><span class="p">:</span>
</span><span id="StructureInformation-676"><a href="#StructureInformation-676"><span class="linenos">676</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="StructureInformation-677"><a href="#StructureInformation-677"><span class="linenos">677</span></a><span class="sd">        Reads PDB mmCIF file from filepath and compiles the information into a CIFInformation instance.</span>
</span><span id="StructureInformation-678"><a href="#StructureInformation-678"><span class="linenos">678</span></a>
</span><span id="StructureInformation-679"><a href="#StructureInformation-679"><span class="linenos">679</span></a><span class="sd">        Parameters</span>
</span><span id="StructureInformation-680"><a href="#StructureInformation-680"><span class="linenos">680</span></a><span class="sd">        ----------</span>
</span><span id="StructureInformation-681"><a href="#StructureInformation-681"><span class="linenos">681</span></a><span class="sd">        pdbx_filepath : str</span>
</span><span id="StructureInformation-682"><a href="#StructureInformation-682"><span class="linenos">682</span></a><span class="sd">            Filepath of the PDB mmCIF file to be read.</span>
</span><span id="StructureInformation-683"><a href="#StructureInformation-683"><span class="linenos">683</span></a><span class="sd">        model_num : int</span>
</span><span id="StructureInformation-684"><a href="#StructureInformation-684"><span class="linenos">684</span></a><span class="sd">            The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</span>
</span><span id="StructureInformation-685"><a href="#StructureInformation-685"><span class="linenos">685</span></a>
</span><span id="StructureInformation-686"><a href="#StructureInformation-686"><span class="linenos">686</span></a><span class="sd">        Returns</span>
</span><span id="StructureInformation-687"><a href="#StructureInformation-687"><span class="linenos">687</span></a><span class="sd">        -------</span>
</span><span id="StructureInformation-688"><a href="#StructureInformation-688"><span class="linenos">688</span></a><span class="sd">        CIFInformation</span>
</span><span id="StructureInformation-689"><a href="#StructureInformation-689"><span class="linenos">689</span></a><span class="sd">            CIFInformation generated from pdbx.get_structure() function using the PDBx file read from the pdbx_filepath.</span>
</span><span id="StructureInformation-690"><a href="#StructureInformation-690"><span class="linenos">690</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="StructureInformation-691"><a href="#StructureInformation-691"><span class="linenos">691</span></a>        <span class="n">pdbx_file</span> <span class="o">=</span> <span class="n">pdbx</span><span class="o">.</span><span class="n">CIFFile</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">pdbx_filepath</span><span class="p">)</span>
</span><span id="StructureInformation-692"><a href="#StructureInformation-692"><span class="linenos">692</span></a>        <span class="k">return</span> <span class="n">MMCIFInformation</span><span class="p">(</span><span class="n">pdbx</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">pdbx_file</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model_num</span><span class="p">,</span> <span class="n">use_author_fields</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="n">pdbx_file</span><span class="p">,</span> <span class="n">model_num</span><span class="p">)</span>
</span><span id="StructureInformation-693"><a href="#StructureInformation-693"><span class="linenos">693</span></a>    
</span><span id="StructureInformation-694"><a href="#StructureInformation-694"><span class="linenos">694</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation-695"><a href="#StructureInformation-695"><span class="linenos">695</span></a>    <span class="k">def</span> <span class="nf">read_pdb_file</span><span class="p">(</span><span class="n">pdb_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;PDBInformation&#39;</span><span class="p">:</span>
</span><span id="StructureInformation-696"><a href="#StructureInformation-696"><span class="linenos">696</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="StructureInformation-697"><a href="#StructureInformation-697"><span class="linenos">697</span></a><span class="sd">        Reads PDB file from filepath and compiles the information into a PDBInformation instance.</span>
</span><span id="StructureInformation-698"><a href="#StructureInformation-698"><span class="linenos">698</span></a>
</span><span id="StructureInformation-699"><a href="#StructureInformation-699"><span class="linenos">699</span></a><span class="sd">        Parameters</span>
</span><span id="StructureInformation-700"><a href="#StructureInformation-700"><span class="linenos">700</span></a><span class="sd">        ----------</span>
</span><span id="StructureInformation-701"><a href="#StructureInformation-701"><span class="linenos">701</span></a><span class="sd">        pdb_filepath : str</span>
</span><span id="StructureInformation-702"><a href="#StructureInformation-702"><span class="linenos">702</span></a><span class="sd">            Filepath of the PDB mmCIF file to be read.</span>
</span><span id="StructureInformation-703"><a href="#StructureInformation-703"><span class="linenos">703</span></a><span class="sd">        model_num : int</span>
</span><span id="StructureInformation-704"><a href="#StructureInformation-704"><span class="linenos">704</span></a><span class="sd">            The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</span>
</span><span id="StructureInformation-705"><a href="#StructureInformation-705"><span class="linenos">705</span></a>
</span><span id="StructureInformation-706"><a href="#StructureInformation-706"><span class="linenos">706</span></a><span class="sd">        Returns</span>
</span><span id="StructureInformation-707"><a href="#StructureInformation-707"><span class="linenos">707</span></a><span class="sd">        -------</span>
</span><span id="StructureInformation-708"><a href="#StructureInformation-708"><span class="linenos">708</span></a><span class="sd">        PDBInformation</span>
</span><span id="StructureInformation-709"><a href="#StructureInformation-709"><span class="linenos">709</span></a><span class="sd">            PDBInformation generated from pdb.get_structure() function using the PDB file read from the pdb_filepath.</span>
</span><span id="StructureInformation-710"><a href="#StructureInformation-710"><span class="linenos">710</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="StructureInformation-711"><a href="#StructureInformation-711"><span class="linenos">711</span></a>        <span class="n">pdb_file</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">PDBFile</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">pdb_filepath</span><span class="p">)</span>
</span><span id="StructureInformation-712"><a href="#StructureInformation-712"><span class="linenos">712</span></a>        <span class="k">return</span> <span class="n">PDBInformation</span><span class="p">(</span><span class="n">pdb</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model_num</span><span class="p">),</span> <span class="n">pdb_file</span><span class="p">,</span> <span class="n">model_num</span><span class="p">)</span>
</span><span id="StructureInformation-713"><a href="#StructureInformation-713"><span class="linenos">713</span></a>    
</span><span id="StructureInformation-714"><a href="#StructureInformation-714"><span class="linenos">714</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation-715"><a href="#StructureInformation-715"><span class="linenos">715</span></a>    <span class="k">def</span> <span class="nf">write_contacts_set</span><span class="p">(</span><span class="n">filepath</span> <span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">contacts_set</span> <span class="p">:</span> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="StructureInformation-716"><a href="#StructureInformation-716"><span class="linenos">716</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="StructureInformation-717"><a href="#StructureInformation-717"><span class="linenos">717</span></a><span class="sd">        Write the contacts generated from get_contacts or general set of tuples of pairs.</span>
</span><span id="StructureInformation-718"><a href="#StructureInformation-718"><span class="linenos">718</span></a>
</span><span id="StructureInformation-719"><a href="#StructureInformation-719"><span class="linenos">719</span></a><span class="sd">        Parameters</span>
</span><span id="StructureInformation-720"><a href="#StructureInformation-720"><span class="linenos">720</span></a><span class="sd">        ----------</span>
</span><span id="StructureInformation-721"><a href="#StructureInformation-721"><span class="linenos">721</span></a><span class="sd">        filepath : str</span>
</span><span id="StructureInformation-722"><a href="#StructureInformation-722"><span class="linenos">722</span></a><span class="sd">            Path of file to output contacts_set to.</span>
</span><span id="StructureInformation-723"><a href="#StructureInformation-723"><span class="linenos">723</span></a><span class="sd">        contacts_set : set of tuple of int, int</span>
</span><span id="StructureInformation-724"><a href="#StructureInformation-724"><span class="linenos">724</span></a><span class="sd">            Set of tuples of pairs that represent contacts.</span>
</span><span id="StructureInformation-725"><a href="#StructureInformation-725"><span class="linenos">725</span></a><span class="sd">        </span>
</span><span id="StructureInformation-726"><a href="#StructureInformation-726"><span class="linenos">726</span></a><span class="sd">        Returns</span>
</span><span id="StructureInformation-727"><a href="#StructureInformation-727"><span class="linenos">727</span></a><span class="sd">        -------</span>
</span><span id="StructureInformation-728"><a href="#StructureInformation-728"><span class="linenos">728</span></a><span class="sd">        None</span>
</span><span id="StructureInformation-729"><a href="#StructureInformation-729"><span class="linenos">729</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="StructureInformation-730"><a href="#StructureInformation-730"><span class="linenos">730</span></a>        <span class="n">contacts_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">contacts_set</span><span class="p">))</span>
</span><span id="StructureInformation-731"><a href="#StructureInformation-731"><span class="linenos">731</span></a>        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="StructureInformation-732"><a href="#StructureInformation-732"><span class="linenos">732</span></a>            <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">contacts_list</span><span class="p">:</span>
</span><span id="StructureInformation-733"><a href="#StructureInformation-733"><span class="linenos">733</span></a>                <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Information regarding a protein structure, obtained from a protein structure file.</p>

<p>Uses fetch_pdb() to pull protein structure information from RCSB. Uses read_x_file() to supply a filepath to pull protein structure information from a file.</p>
</div>


                            <div id="StructureInformation.fetch_pdb" class="classattr">
                                        <input id="StructureInformation.fetch_pdb-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">fetch_pdb</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">pdb_id</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">struc_format</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s1">&#39;mmcif&#39;</span><span class="p">,</span> <span class="s1">&#39;pdb&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;mmcif&#39;</span>,</span><span class="param">	<span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span></span><span class="return-annotation">) -> <span class="n">Union</span><span class="p">[</span><span class="n"><a href="#MMCIFInformation">MMCIFInformation</a></span><span class="p">,</span> <span class="n"><a href="#PDBInformation">PDBInformation</a></span><span class="p">]</span>:</span></span>

                <label class="view-source-button" for="StructureInformation.fetch_pdb-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#StructureInformation.fetch_pdb"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="StructureInformation.fetch_pdb-637"><a href="#StructureInformation.fetch_pdb-637"><span class="linenos">637</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation.fetch_pdb-638"><a href="#StructureInformation.fetch_pdb-638"><span class="linenos">638</span></a>    <span class="k">def</span> <span class="nf">fetch_pdb</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">struc_format</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;mmcif&quot;</span><span class="p">,</span> <span class="s2">&quot;pdb&quot;</span><span class="p">]</span><span class="o">=</span><span class="s2">&quot;mmcif&quot;</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="s1">&#39;MMCIFInformation&#39;</span><span class="p">,</span> <span class="s1">&#39;PDBInformation&#39;</span><span class="p">]:</span>
</span><span id="StructureInformation.fetch_pdb-639"><a href="#StructureInformation.fetch_pdb-639"><span class="linenos">639</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="StructureInformation.fetch_pdb-640"><a href="#StructureInformation.fetch_pdb-640"><span class="linenos">640</span></a><span class="sd">        Fetches PDB as mmCIF file from RCSB and compiles the information into a StructureInformation instance.</span>
</span><span id="StructureInformation.fetch_pdb-641"><a href="#StructureInformation.fetch_pdb-641"><span class="linenos">641</span></a>
</span><span id="StructureInformation.fetch_pdb-642"><a href="#StructureInformation.fetch_pdb-642"><span class="linenos">642</span></a><span class="sd">        Parameters</span>
</span><span id="StructureInformation.fetch_pdb-643"><a href="#StructureInformation.fetch_pdb-643"><span class="linenos">643</span></a><span class="sd">        ----------</span>
</span><span id="StructureInformation.fetch_pdb-644"><a href="#StructureInformation.fetch_pdb-644"><span class="linenos">644</span></a><span class="sd">        pdb_id : str</span>
</span><span id="StructureInformation.fetch_pdb-645"><a href="#StructureInformation.fetch_pdb-645"><span class="linenos">645</span></a><span class="sd">            PDB ID to be fetched from the RCSB database.</span>
</span><span id="StructureInformation.fetch_pdb-646"><a href="#StructureInformation.fetch_pdb-646"><span class="linenos">646</span></a><span class="sd">        struc_format : str</span>
</span><span id="StructureInformation.fetch_pdb-647"><a href="#StructureInformation.fetch_pdb-647"><span class="linenos">647</span></a><span class="sd">            The format of the file to pull from the RCSB database.</span>
</span><span id="StructureInformation.fetch_pdb-648"><a href="#StructureInformation.fetch_pdb-648"><span class="linenos">648</span></a><span class="sd">        model_num : int</span>
</span><span id="StructureInformation.fetch_pdb-649"><a href="#StructureInformation.fetch_pdb-649"><span class="linenos">649</span></a><span class="sd">            The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</span>
</span><span id="StructureInformation.fetch_pdb-650"><a href="#StructureInformation.fetch_pdb-650"><span class="linenos">650</span></a><span class="sd">            </span>
</span><span id="StructureInformation.fetch_pdb-651"><a href="#StructureInformation.fetch_pdb-651"><span class="linenos">651</span></a><span class="sd">        Returns</span>
</span><span id="StructureInformation.fetch_pdb-652"><a href="#StructureInformation.fetch_pdb-652"><span class="linenos">652</span></a><span class="sd">        -------</span>
</span><span id="StructureInformation.fetch_pdb-653"><a href="#StructureInformation.fetch_pdb-653"><span class="linenos">653</span></a><span class="sd">        StructureInformation</span>
</span><span id="StructureInformation.fetch_pdb-654"><a href="#StructureInformation.fetch_pdb-654"><span class="linenos">654</span></a><span class="sd">            StructureInformation generated from pdbx.get_structure() function using the pdbx file fetched from RCSB.</span>
</span><span id="StructureInformation.fetch_pdb-655"><a href="#StructureInformation.fetch_pdb-655"><span class="linenos">655</span></a><span class="sd">        </span>
</span><span id="StructureInformation.fetch_pdb-656"><a href="#StructureInformation.fetch_pdb-656"><span class="linenos">656</span></a><span class="sd">        Raises</span>
</span><span id="StructureInformation.fetch_pdb-657"><a href="#StructureInformation.fetch_pdb-657"><span class="linenos">657</span></a><span class="sd">        ------</span>
</span><span id="StructureInformation.fetch_pdb-658"><a href="#StructureInformation.fetch_pdb-658"><span class="linenos">658</span></a><span class="sd">        TypeError</span>
</span><span id="StructureInformation.fetch_pdb-659"><a href="#StructureInformation.fetch_pdb-659"><span class="linenos">659</span></a><span class="sd">            Fetched data was not found and returned None instead.</span>
</span><span id="StructureInformation.fetch_pdb-660"><a href="#StructureInformation.fetch_pdb-660"><span class="linenos">660</span></a><span class="sd">        ValueError</span>
</span><span id="StructureInformation.fetch_pdb-661"><a href="#StructureInformation.fetch_pdb-661"><span class="linenos">661</span></a><span class="sd">            Structure format may be invalid (not PDBx/mmCIF or PDB).</span>
</span><span id="StructureInformation.fetch_pdb-662"><a href="#StructureInformation.fetch_pdb-662"><span class="linenos">662</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="StructureInformation.fetch_pdb-663"><a href="#StructureInformation.fetch_pdb-663"><span class="linenos">663</span></a>        <span class="n">fetched_data</span> <span class="o">=</span> <span class="n">rcsb</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">pdb_id</span><span class="p">,</span> <span class="n">struc_format</span><span class="p">)</span>
</span><span id="StructureInformation.fetch_pdb-664"><a href="#StructureInformation.fetch_pdb-664"><span class="linenos">664</span></a>        <span class="k">if</span> <span class="n">fetched_data</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="StructureInformation.fetch_pdb-665"><a href="#StructureInformation.fetch_pdb-665"><span class="linenos">665</span></a>            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;RCSB fetch failed. Try fetch again.&quot;</span><span class="p">)</span>
</span><span id="StructureInformation.fetch_pdb-666"><a href="#StructureInformation.fetch_pdb-666"><span class="linenos">666</span></a>        <span class="k">elif</span> <span class="n">struc_format</span> <span class="o">==</span> <span class="s2">&quot;mmcif&quot;</span><span class="p">:</span>
</span><span id="StructureInformation.fetch_pdb-667"><a href="#StructureInformation.fetch_pdb-667"><span class="linenos">667</span></a>            <span class="n">pdbx_file</span> <span class="o">=</span> <span class="n">pdbx</span><span class="o">.</span><span class="n">CIFFile</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fetched_data</span><span class="p">)</span>
</span><span id="StructureInformation.fetch_pdb-668"><a href="#StructureInformation.fetch_pdb-668"><span class="linenos">668</span></a>            <span class="k">return</span> <span class="n">MMCIFInformation</span><span class="p">(</span><span class="n">pdbx</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">pdbx_file</span><span class="o">=</span><span class="n">pdbx_file</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model_num</span><span class="p">,</span> <span class="n">use_author_fields</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="n">pdbx_file</span><span class="p">,</span> <span class="n">model_num</span><span class="p">)</span>
</span><span id="StructureInformation.fetch_pdb-669"><a href="#StructureInformation.fetch_pdb-669"><span class="linenos">669</span></a>        <span class="k">elif</span> <span class="n">struc_format</span> <span class="o">==</span> <span class="s2">&quot;pdb&quot;</span><span class="p">:</span>
</span><span id="StructureInformation.fetch_pdb-670"><a href="#StructureInformation.fetch_pdb-670"><span class="linenos">670</span></a>            <span class="n">pdb_file</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">PDBFile</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fetched_data</span><span class="p">)</span>
</span><span id="StructureInformation.fetch_pdb-671"><a href="#StructureInformation.fetch_pdb-671"><span class="linenos">671</span></a>            <span class="k">return</span> <span class="n">PDBInformation</span><span class="p">(</span><span class="n">pdb</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">pdb_file</span><span class="o">=</span><span class="n">pdb_file</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model_num</span><span class="p">),</span> <span class="n">pdb_file</span><span class="o">=</span><span class="n">pdb_file</span><span class="p">,</span> <span class="n">model_num</span><span class="o">=</span><span class="n">model_num</span><span class="p">)</span>
</span><span id="StructureInformation.fetch_pdb-672"><a href="#StructureInformation.fetch_pdb-672"><span class="linenos">672</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="StructureInformation.fetch_pdb-673"><a href="#StructureInformation.fetch_pdb-673"><span class="linenos">673</span></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;struc_format </span><span class="si">{</span><span class="n">struc_format</span><span class="si">}</span><span class="s2"> is not valid or currently supported by DCA Toolkit&quot;</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Fetches PDB as mmCIF file from RCSB and compiles the information into a StructureInformation instance.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>pdb_id</strong> (str):
PDB ID to be fetched from the RCSB database.</li>
<li><strong>struc_format</strong> (str):
The format of the file to pull from the RCSB database.</li>
<li><strong>model_num</strong> (int):
The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>StructureInformation</strong>: StructureInformation generated from pdbx.get_structure() function using the pdbx file fetched from RCSB.</li>
</ul>

<h6 id="raises">Raises</h6>

<ul>
<li><strong>TypeError</strong>: Fetched data was not found and returned None instead.</li>
<li><strong>ValueError</strong>: Structure format may be invalid (not PDBx/mmCIF or PDB).</li>
</ul>
</div>


                            </div>
                            <div id="StructureInformation.read_mmCIF_file" class="classattr">
                                        <input id="StructureInformation.read_mmCIF_file-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">read_mmCIF_file</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">pdbx_filepath</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span></span><span class="return-annotation">) -> <span class="n"><a href="#MMCIFInformation">MMCIFInformation</a></span>:</span></span>

                <label class="view-source-button" for="StructureInformation.read_mmCIF_file-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#StructureInformation.read_mmCIF_file"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="StructureInformation.read_mmCIF_file-674"><a href="#StructureInformation.read_mmCIF_file-674"><span class="linenos">674</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation.read_mmCIF_file-675"><a href="#StructureInformation.read_mmCIF_file-675"><span class="linenos">675</span></a>    <span class="k">def</span> <span class="nf">read_mmCIF_file</span><span class="p">(</span><span class="n">pdbx_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;MMCIFInformation&#39;</span><span class="p">:</span>
</span><span id="StructureInformation.read_mmCIF_file-676"><a href="#StructureInformation.read_mmCIF_file-676"><span class="linenos">676</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="StructureInformation.read_mmCIF_file-677"><a href="#StructureInformation.read_mmCIF_file-677"><span class="linenos">677</span></a><span class="sd">        Reads PDB mmCIF file from filepath and compiles the information into a CIFInformation instance.</span>
</span><span id="StructureInformation.read_mmCIF_file-678"><a href="#StructureInformation.read_mmCIF_file-678"><span class="linenos">678</span></a>
</span><span id="StructureInformation.read_mmCIF_file-679"><a href="#StructureInformation.read_mmCIF_file-679"><span class="linenos">679</span></a><span class="sd">        Parameters</span>
</span><span id="StructureInformation.read_mmCIF_file-680"><a href="#StructureInformation.read_mmCIF_file-680"><span class="linenos">680</span></a><span class="sd">        ----------</span>
</span><span id="StructureInformation.read_mmCIF_file-681"><a href="#StructureInformation.read_mmCIF_file-681"><span class="linenos">681</span></a><span class="sd">        pdbx_filepath : str</span>
</span><span id="StructureInformation.read_mmCIF_file-682"><a href="#StructureInformation.read_mmCIF_file-682"><span class="linenos">682</span></a><span class="sd">            Filepath of the PDB mmCIF file to be read.</span>
</span><span id="StructureInformation.read_mmCIF_file-683"><a href="#StructureInformation.read_mmCIF_file-683"><span class="linenos">683</span></a><span class="sd">        model_num : int</span>
</span><span id="StructureInformation.read_mmCIF_file-684"><a href="#StructureInformation.read_mmCIF_file-684"><span class="linenos">684</span></a><span class="sd">            The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</span>
</span><span id="StructureInformation.read_mmCIF_file-685"><a href="#StructureInformation.read_mmCIF_file-685"><span class="linenos">685</span></a>
</span><span id="StructureInformation.read_mmCIF_file-686"><a href="#StructureInformation.read_mmCIF_file-686"><span class="linenos">686</span></a><span class="sd">        Returns</span>
</span><span id="StructureInformation.read_mmCIF_file-687"><a href="#StructureInformation.read_mmCIF_file-687"><span class="linenos">687</span></a><span class="sd">        -------</span>
</span><span id="StructureInformation.read_mmCIF_file-688"><a href="#StructureInformation.read_mmCIF_file-688"><span class="linenos">688</span></a><span class="sd">        CIFInformation</span>
</span><span id="StructureInformation.read_mmCIF_file-689"><a href="#StructureInformation.read_mmCIF_file-689"><span class="linenos">689</span></a><span class="sd">            CIFInformation generated from pdbx.get_structure() function using the PDBx file read from the pdbx_filepath.</span>
</span><span id="StructureInformation.read_mmCIF_file-690"><a href="#StructureInformation.read_mmCIF_file-690"><span class="linenos">690</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="StructureInformation.read_mmCIF_file-691"><a href="#StructureInformation.read_mmCIF_file-691"><span class="linenos">691</span></a>        <span class="n">pdbx_file</span> <span class="o">=</span> <span class="n">pdbx</span><span class="o">.</span><span class="n">CIFFile</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">pdbx_filepath</span><span class="p">)</span>
</span><span id="StructureInformation.read_mmCIF_file-692"><a href="#StructureInformation.read_mmCIF_file-692"><span class="linenos">692</span></a>        <span class="k">return</span> <span class="n">MMCIFInformation</span><span class="p">(</span><span class="n">pdbx</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">pdbx_file</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model_num</span><span class="p">,</span> <span class="n">use_author_fields</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="n">pdbx_file</span><span class="p">,</span> <span class="n">model_num</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Reads PDB mmCIF file from filepath and compiles the information into a CIFInformation instance.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>pdbx_filepath</strong> (str):
Filepath of the PDB mmCIF file to be read.</li>
<li><strong>model_num</strong> (int):
The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>CIFInformation</strong>: CIFInformation generated from pdbx.get_structure() function using the PDBx file read from the pdbx_filepath.</li>
</ul>
</div>


                            </div>
                            <div id="StructureInformation.read_pdb_file" class="classattr">
                                        <input id="StructureInformation.read_pdb_file-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">read_pdb_file</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">pdb_filepath</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span></span><span class="return-annotation">) -> <span class="n"><a href="#PDBInformation">PDBInformation</a></span>:</span></span>

                <label class="view-source-button" for="StructureInformation.read_pdb_file-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#StructureInformation.read_pdb_file"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="StructureInformation.read_pdb_file-694"><a href="#StructureInformation.read_pdb_file-694"><span class="linenos">694</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation.read_pdb_file-695"><a href="#StructureInformation.read_pdb_file-695"><span class="linenos">695</span></a>    <span class="k">def</span> <span class="nf">read_pdb_file</span><span class="p">(</span><span class="n">pdb_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;PDBInformation&#39;</span><span class="p">:</span>
</span><span id="StructureInformation.read_pdb_file-696"><a href="#StructureInformation.read_pdb_file-696"><span class="linenos">696</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="StructureInformation.read_pdb_file-697"><a href="#StructureInformation.read_pdb_file-697"><span class="linenos">697</span></a><span class="sd">        Reads PDB file from filepath and compiles the information into a PDBInformation instance.</span>
</span><span id="StructureInformation.read_pdb_file-698"><a href="#StructureInformation.read_pdb_file-698"><span class="linenos">698</span></a>
</span><span id="StructureInformation.read_pdb_file-699"><a href="#StructureInformation.read_pdb_file-699"><span class="linenos">699</span></a><span class="sd">        Parameters</span>
</span><span id="StructureInformation.read_pdb_file-700"><a href="#StructureInformation.read_pdb_file-700"><span class="linenos">700</span></a><span class="sd">        ----------</span>
</span><span id="StructureInformation.read_pdb_file-701"><a href="#StructureInformation.read_pdb_file-701"><span class="linenos">701</span></a><span class="sd">        pdb_filepath : str</span>
</span><span id="StructureInformation.read_pdb_file-702"><a href="#StructureInformation.read_pdb_file-702"><span class="linenos">702</span></a><span class="sd">            Filepath of the PDB mmCIF file to be read.</span>
</span><span id="StructureInformation.read_pdb_file-703"><a href="#StructureInformation.read_pdb_file-703"><span class="linenos">703</span></a><span class="sd">        model_num : int</span>
</span><span id="StructureInformation.read_pdb_file-704"><a href="#StructureInformation.read_pdb_file-704"><span class="linenos">704</span></a><span class="sd">            The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</span>
</span><span id="StructureInformation.read_pdb_file-705"><a href="#StructureInformation.read_pdb_file-705"><span class="linenos">705</span></a>
</span><span id="StructureInformation.read_pdb_file-706"><a href="#StructureInformation.read_pdb_file-706"><span class="linenos">706</span></a><span class="sd">        Returns</span>
</span><span id="StructureInformation.read_pdb_file-707"><a href="#StructureInformation.read_pdb_file-707"><span class="linenos">707</span></a><span class="sd">        -------</span>
</span><span id="StructureInformation.read_pdb_file-708"><a href="#StructureInformation.read_pdb_file-708"><span class="linenos">708</span></a><span class="sd">        PDBInformation</span>
</span><span id="StructureInformation.read_pdb_file-709"><a href="#StructureInformation.read_pdb_file-709"><span class="linenos">709</span></a><span class="sd">            PDBInformation generated from pdb.get_structure() function using the PDB file read from the pdb_filepath.</span>
</span><span id="StructureInformation.read_pdb_file-710"><a href="#StructureInformation.read_pdb_file-710"><span class="linenos">710</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="StructureInformation.read_pdb_file-711"><a href="#StructureInformation.read_pdb_file-711"><span class="linenos">711</span></a>        <span class="n">pdb_file</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">PDBFile</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">pdb_filepath</span><span class="p">)</span>
</span><span id="StructureInformation.read_pdb_file-712"><a href="#StructureInformation.read_pdb_file-712"><span class="linenos">712</span></a>        <span class="k">return</span> <span class="n">PDBInformation</span><span class="p">(</span><span class="n">pdb</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model_num</span><span class="p">),</span> <span class="n">pdb_file</span><span class="p">,</span> <span class="n">model_num</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Reads PDB file from filepath and compiles the information into a PDBInformation instance.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>pdb_filepath</strong> (str):
Filepath of the PDB mmCIF file to be read.</li>
<li><strong>model_num</strong> (int):
The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>PDBInformation</strong>: PDBInformation generated from pdb.get_structure() function using the PDB file read from the pdb_filepath.</li>
</ul>
</div>


                            </div>
                            <div id="StructureInformation.write_contacts_set" class="classattr">
                                        <input id="StructureInformation.write_contacts_set-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">write_contacts_set</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span>, </span><span class="param"><span class="n">contacts_set</span><span class="p">:</span> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span></span><span class="return-annotation">) -> <span class="kc">None</span>:</span></span>

                <label class="view-source-button" for="StructureInformation.write_contacts_set-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#StructureInformation.write_contacts_set"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="StructureInformation.write_contacts_set-714"><a href="#StructureInformation.write_contacts_set-714"><span class="linenos">714</span></a>    <span class="nd">@staticmethod</span>
</span><span id="StructureInformation.write_contacts_set-715"><a href="#StructureInformation.write_contacts_set-715"><span class="linenos">715</span></a>    <span class="k">def</span> <span class="nf">write_contacts_set</span><span class="p">(</span><span class="n">filepath</span> <span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">contacts_set</span> <span class="p">:</span> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="StructureInformation.write_contacts_set-716"><a href="#StructureInformation.write_contacts_set-716"><span class="linenos">716</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="StructureInformation.write_contacts_set-717"><a href="#StructureInformation.write_contacts_set-717"><span class="linenos">717</span></a><span class="sd">        Write the contacts generated from get_contacts or general set of tuples of pairs.</span>
</span><span id="StructureInformation.write_contacts_set-718"><a href="#StructureInformation.write_contacts_set-718"><span class="linenos">718</span></a>
</span><span id="StructureInformation.write_contacts_set-719"><a href="#StructureInformation.write_contacts_set-719"><span class="linenos">719</span></a><span class="sd">        Parameters</span>
</span><span id="StructureInformation.write_contacts_set-720"><a href="#StructureInformation.write_contacts_set-720"><span class="linenos">720</span></a><span class="sd">        ----------</span>
</span><span id="StructureInformation.write_contacts_set-721"><a href="#StructureInformation.write_contacts_set-721"><span class="linenos">721</span></a><span class="sd">        filepath : str</span>
</span><span id="StructureInformation.write_contacts_set-722"><a href="#StructureInformation.write_contacts_set-722"><span class="linenos">722</span></a><span class="sd">            Path of file to output contacts_set to.</span>
</span><span id="StructureInformation.write_contacts_set-723"><a href="#StructureInformation.write_contacts_set-723"><span class="linenos">723</span></a><span class="sd">        contacts_set : set of tuple of int, int</span>
</span><span id="StructureInformation.write_contacts_set-724"><a href="#StructureInformation.write_contacts_set-724"><span class="linenos">724</span></a><span class="sd">            Set of tuples of pairs that represent contacts.</span>
</span><span id="StructureInformation.write_contacts_set-725"><a href="#StructureInformation.write_contacts_set-725"><span class="linenos">725</span></a><span class="sd">        </span>
</span><span id="StructureInformation.write_contacts_set-726"><a href="#StructureInformation.write_contacts_set-726"><span class="linenos">726</span></a><span class="sd">        Returns</span>
</span><span id="StructureInformation.write_contacts_set-727"><a href="#StructureInformation.write_contacts_set-727"><span class="linenos">727</span></a><span class="sd">        -------</span>
</span><span id="StructureInformation.write_contacts_set-728"><a href="#StructureInformation.write_contacts_set-728"><span class="linenos">728</span></a><span class="sd">        None</span>
</span><span id="StructureInformation.write_contacts_set-729"><a href="#StructureInformation.write_contacts_set-729"><span class="linenos">729</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="StructureInformation.write_contacts_set-730"><a href="#StructureInformation.write_contacts_set-730"><span class="linenos">730</span></a>        <span class="n">contacts_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">contacts_set</span><span class="p">))</span>
</span><span id="StructureInformation.write_contacts_set-731"><a href="#StructureInformation.write_contacts_set-731"><span class="linenos">731</span></a>        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="StructureInformation.write_contacts_set-732"><a href="#StructureInformation.write_contacts_set-732"><span class="linenos">732</span></a>            <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">contacts_list</span><span class="p">:</span>
</span><span id="StructureInformation.write_contacts_set-733"><a href="#StructureInformation.write_contacts_set-733"><span class="linenos">733</span></a>                <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Write the contacts generated from get_contacts or general set of tuples of pairs.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>filepath</strong> (str):
Path of file to output contacts_set to.</li>
<li><strong>contacts_set</strong> (set of tuple of int, int):
Set of tuples of pairs that represent contacts.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>None</strong></li>
</ul>
</div>


                            </div>
                </section>
                <section id="ResidueAlignment">
                            <input id="ResidueAlignment-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr class">
            
    <span class="def">class</span>
    <span class="name">ResidueAlignment</span>:

                <label class="view-source-button" for="ResidueAlignment-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#ResidueAlignment"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="ResidueAlignment-181"><a href="#ResidueAlignment-181"><span class="linenos">181</span></a><span class="k">class</span> <span class="nc">ResidueAlignment</span><span class="p">:</span>
</span><span id="ResidueAlignment-182"><a href="#ResidueAlignment-182"><span class="linenos">182</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-183"><a href="#ResidueAlignment-183"><span class="linenos">183</span></a><span class="sd">    A representation of a residue alignment, often from a query HMM to a protein structure target sequence.</span>
</span><span id="ResidueAlignment-184"><a href="#ResidueAlignment-184"><span class="linenos">184</span></a>
</span><span id="ResidueAlignment-185"><a href="#ResidueAlignment-185"><span class="linenos">185</span></a><span class="sd">    Parameters</span>
</span><span id="ResidueAlignment-186"><a href="#ResidueAlignment-186"><span class="linenos">186</span></a><span class="sd">    ----------</span>
</span><span id="ResidueAlignment-187"><a href="#ResidueAlignment-187"><span class="linenos">187</span></a><span class="sd">    domain_name : str</span>
</span><span id="ResidueAlignment-188"><a href="#ResidueAlignment-188"><span class="linenos">188</span></a><span class="sd">        The name of the query HMM.</span>
</span><span id="ResidueAlignment-189"><a href="#ResidueAlignment-189"><span class="linenos">189</span></a><span class="sd">    protein_name : str</span>
</span><span id="ResidueAlignment-190"><a href="#ResidueAlignment-190"><span class="linenos">190</span></a><span class="sd">        The name of the target protein sequence.</span>
</span><span id="ResidueAlignment-191"><a href="#ResidueAlignment-191"><span class="linenos">191</span></a><span class="sd">    domain_start : int</span>
</span><span id="ResidueAlignment-192"><a href="#ResidueAlignment-192"><span class="linenos">192</span></a><span class="sd">        The starting index of the domain alignment in the query HMM.</span>
</span><span id="ResidueAlignment-193"><a href="#ResidueAlignment-193"><span class="linenos">193</span></a><span class="sd">    protein_start : int</span>
</span><span id="ResidueAlignment-194"><a href="#ResidueAlignment-194"><span class="linenos">194</span></a><span class="sd">        The starting index of the domain alignment in the protein target sequence.</span>
</span><span id="ResidueAlignment-195"><a href="#ResidueAlignment-195"><span class="linenos">195</span></a><span class="sd">    domain_text : str</span>
</span><span id="ResidueAlignment-196"><a href="#ResidueAlignment-196"><span class="linenos">196</span></a><span class="sd">        The sequence of the domain in the query HMM corresponding to this alignment.</span>
</span><span id="ResidueAlignment-197"><a href="#ResidueAlignment-197"><span class="linenos">197</span></a><span class="sd">    protein_text : str</span>
</span><span id="ResidueAlignment-198"><a href="#ResidueAlignment-198"><span class="linenos">198</span></a><span class="sd">        The sequence of the protein target sequence corresponding to this alignment.</span>
</span><span id="ResidueAlignment-199"><a href="#ResidueAlignment-199"><span class="linenos">199</span></a>
</span><span id="ResidueAlignment-200"><a href="#ResidueAlignment-200"><span class="linenos">200</span></a><span class="sd">    Attributes</span>
</span><span id="ResidueAlignment-201"><a href="#ResidueAlignment-201"><span class="linenos">201</span></a><span class="sd">    ----------</span>
</span><span id="ResidueAlignment-202"><a href="#ResidueAlignment-202"><span class="linenos">202</span></a><span class="sd">    reference_mapping : pandas.DataFrame</span>
</span><span id="ResidueAlignment-203"><a href="#ResidueAlignment-203"><span class="linenos">203</span></a><span class="sd">        The representation of the mapping where a row constitutes a residue pair and its indices in the format: &#39;domain_index&#39;, &#39;domain_residue&#39;, &#39;protein_residue&#39;, &#39;protein_index&#39;.</span>
</span><span id="ResidueAlignment-204"><a href="#ResidueAlignment-204"><span class="linenos">204</span></a><span class="sd">    domain_to_protein : dict[int, int]</span>
</span><span id="ResidueAlignment-205"><a href="#ResidueAlignment-205"><span class="linenos">205</span></a><span class="sd">        A dictionary allowing for mapping from indices corresponding to the query HMM and Multiple Sequence Alignment to the protein target sequence.</span>
</span><span id="ResidueAlignment-206"><a href="#ResidueAlignment-206"><span class="linenos">206</span></a><span class="sd">    protein_to_domain : dict[int, int]</span>
</span><span id="ResidueAlignment-207"><a href="#ResidueAlignment-207"><span class="linenos">207</span></a><span class="sd">        A dictionary allowing for mapping from indices corresponding to the protein target sequence to the query HMM and Multiple Sequence Alignment.</span>
</span><span id="ResidueAlignment-208"><a href="#ResidueAlignment-208"><span class="linenos">208</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-209"><a href="#ResidueAlignment-209"><span class="linenos">209</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">domain_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">protein_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">domain_start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="ResidueAlignment-210"><a href="#ResidueAlignment-210"><span class="linenos">210</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">domain_name</span> <span class="o">=</span> <span class="n">domain_name</span>
</span><span id="ResidueAlignment-211"><a href="#ResidueAlignment-211"><span class="linenos">211</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">protein_name</span> <span class="o">=</span> <span class="n">protein_name</span>
</span><span id="ResidueAlignment-212"><a href="#ResidueAlignment-212"><span class="linenos">212</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">set_reference_mapping</span><span class="p">(</span><span class="n">domain_start</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">)</span>
</span><span id="ResidueAlignment-213"><a href="#ResidueAlignment-213"><span class="linenos">213</span></a>    
</span><span id="ResidueAlignment-214"><a href="#ResidueAlignment-214"><span class="linenos">214</span></a>    <span class="k">def</span> <span class="nf">set_reference_mapping</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">domain_start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="ResidueAlignment-215"><a href="#ResidueAlignment-215"><span class="linenos">215</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-216"><a href="#ResidueAlignment-216"><span class="linenos">216</span></a><span class="sd">        Set values for reference_mapping and mapping dictionaries, domain_to_protein and protein_to_domain.</span>
</span><span id="ResidueAlignment-217"><a href="#ResidueAlignment-217"><span class="linenos">217</span></a>
</span><span id="ResidueAlignment-218"><a href="#ResidueAlignment-218"><span class="linenos">218</span></a><span class="sd">        Note</span>
</span><span id="ResidueAlignment-219"><a href="#ResidueAlignment-219"><span class="linenos">219</span></a><span class="sd">        ----</span>
</span><span id="ResidueAlignment-220"><a href="#ResidueAlignment-220"><span class="linenos">220</span></a><span class="sd">        For details on `domain_start`, `protein_start`, `domain_text`, `protein_text`, please refer to the `ResidueAlignment` docstring.</span>
</span><span id="ResidueAlignment-221"><a href="#ResidueAlignment-221"><span class="linenos">221</span></a>
</span><span id="ResidueAlignment-222"><a href="#ResidueAlignment-222"><span class="linenos">222</span></a><span class="sd">        Returns</span>
</span><span id="ResidueAlignment-223"><a href="#ResidueAlignment-223"><span class="linenos">223</span></a><span class="sd">        -------</span>
</span><span id="ResidueAlignment-224"><a href="#ResidueAlignment-224"><span class="linenos">224</span></a><span class="sd">        None</span>
</span><span id="ResidueAlignment-225"><a href="#ResidueAlignment-225"><span class="linenos">225</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-226"><a href="#ResidueAlignment-226"><span class="linenos">226</span></a>        <span class="c1"># Convert text to list variant for iteration</span>
</span><span id="ResidueAlignment-227"><a href="#ResidueAlignment-227"><span class="linenos">227</span></a>        <span class="n">domain_sequence</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">domain_text</span><span class="p">)</span>
</span><span id="ResidueAlignment-228"><a href="#ResidueAlignment-228"><span class="linenos">228</span></a>        <span class="n">protein_sequence</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">protein_text</span><span class="p">)</span>
</span><span id="ResidueAlignment-229"><a href="#ResidueAlignment-229"><span class="linenos">229</span></a>        
</span><span id="ResidueAlignment-230"><a href="#ResidueAlignment-230"><span class="linenos">230</span></a>        <span class="n">mapping_entries</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment-231"><a href="#ResidueAlignment-231"><span class="linenos">231</span></a>
</span><span id="ResidueAlignment-232"><a href="#ResidueAlignment-232"><span class="linenos">232</span></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">domain_sequence</span><span class="p">)):</span>
</span><span id="ResidueAlignment-233"><a href="#ResidueAlignment-233"><span class="linenos">233</span></a>            <span class="n">mapping_entry</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment-234"><a href="#ResidueAlignment-234"><span class="linenos">234</span></a>            <span class="c1"># Check to see if domain residue is valid, if so, we can assign the proper index.</span>
</span><span id="ResidueAlignment-235"><a href="#ResidueAlignment-235"><span class="linenos">235</span></a>            <span class="k">if</span> <span class="n">domain_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;.&#39;</span><span class="p">:</span>
</span><span id="ResidueAlignment-236"><a href="#ResidueAlignment-236"><span class="linenos">236</span></a>                <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">domain_start</span><span class="p">)</span>
</span><span id="ResidueAlignment-237"><a href="#ResidueAlignment-237"><span class="linenos">237</span></a>                <span class="n">domain_start</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span id="ResidueAlignment-238"><a href="#ResidueAlignment-238"><span class="linenos">238</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="ResidueAlignment-239"><a href="#ResidueAlignment-239"><span class="linenos">239</span></a>                <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">NA</span><span class="p">)</span>
</span><span id="ResidueAlignment-240"><a href="#ResidueAlignment-240"><span class="linenos">240</span></a>            <span class="c1"># Assign the values of the residues mapped together.</span>
</span><span id="ResidueAlignment-241"><a href="#ResidueAlignment-241"><span class="linenos">241</span></a>            <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">domain_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
</span><span id="ResidueAlignment-242"><a href="#ResidueAlignment-242"><span class="linenos">242</span></a>            <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
</span><span id="ResidueAlignment-243"><a href="#ResidueAlignment-243"><span class="linenos">243</span></a>            <span class="c1"># Check to see if protein residue is valid, if so, we can assign the proper index.</span>
</span><span id="ResidueAlignment-244"><a href="#ResidueAlignment-244"><span class="linenos">244</span></a>            <span class="k">if</span> <span class="n">protein_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
</span><span id="ResidueAlignment-245"><a href="#ResidueAlignment-245"><span class="linenos">245</span></a>                <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_start</span><span class="p">)</span>
</span><span id="ResidueAlignment-246"><a href="#ResidueAlignment-246"><span class="linenos">246</span></a>                <span class="n">protein_start</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span id="ResidueAlignment-247"><a href="#ResidueAlignment-247"><span class="linenos">247</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="ResidueAlignment-248"><a href="#ResidueAlignment-248"><span class="linenos">248</span></a>                <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">NA</span><span class="p">)</span>
</span><span id="ResidueAlignment-249"><a href="#ResidueAlignment-249"><span class="linenos">249</span></a>            <span class="c1"># Store the resulting mapping in the reference map.</span>
</span><span id="ResidueAlignment-250"><a href="#ResidueAlignment-250"><span class="linenos">250</span></a>            <span class="n">mapping_entries</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mapping_entry</span><span class="p">)</span>
</span><span id="ResidueAlignment-251"><a href="#ResidueAlignment-251"><span class="linenos">251</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">reference_mapping</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">mapping_entries</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;domain_index&#39;</span><span class="p">,</span> <span class="s1">&#39;domain_residue&#39;</span><span class="p">,</span> <span class="s1">&#39;protein_residue&#39;</span><span class="p">,</span> <span class="s1">&#39;protein_index&#39;</span><span class="p">])</span>
</span><span id="ResidueAlignment-252"><a href="#ResidueAlignment-252"><span class="linenos">252</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">reference_mapping</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_mapping</span><span class="o">.</span><span class="n">astype</span><span class="p">({</span><span class="s1">&#39;domain_index&#39;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Int32Dtype</span><span class="p">(),</span> <span class="s1">&#39;protein_index&#39;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Int32Dtype</span><span class="p">(),</span> <span class="s1">&#39;domain_residue&#39;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">StringDtype</span><span class="p">(),</span> <span class="s1">&#39;protein_residue&#39;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">StringDtype</span><span class="p">()})</span>
</span><span id="ResidueAlignment-253"><a href="#ResidueAlignment-253"><span class="linenos">253</span></a>        <span class="n">reference_mapping_notna</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_mapping</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
</span><span id="ResidueAlignment-254"><a href="#ResidueAlignment-254"><span class="linenos">254</span></a>        
</span><span id="ResidueAlignment-255"><a href="#ResidueAlignment-255"><span class="linenos">255</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">domain_to_protein</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">reference_mapping_notna</span><span class="o">.</span><span class="n">domain_index</span><span class="p">,</span> <span class="n">reference_mapping_notna</span><span class="o">.</span><span class="n">protein_index</span><span class="p">))</span>
</span><span id="ResidueAlignment-256"><a href="#ResidueAlignment-256"><span class="linenos">256</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">protein_to_domain</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">reference_mapping_notna</span><span class="o">.</span><span class="n">protein_index</span><span class="p">,</span> <span class="n">reference_mapping_notna</span><span class="o">.</span><span class="n">domain_index</span><span class="p">))</span>
</span><span id="ResidueAlignment-257"><a href="#ResidueAlignment-257"><span class="linenos">257</span></a>
</span><span id="ResidueAlignment-258"><a href="#ResidueAlignment-258"><span class="linenos">258</span></a>    <span class="nd">@staticmethod</span>
</span><span id="ResidueAlignment-259"><a href="#ResidueAlignment-259"><span class="linenos">259</span></a>    <span class="k">def</span> <span class="nf">load_from_align_file</span><span class="p">(</span><span class="n">align_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;ResidueAlignment&#39;</span><span class="p">:</span>
</span><span id="ResidueAlignment-260"><a href="#ResidueAlignment-260"><span class="linenos">260</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-261"><a href="#ResidueAlignment-261"><span class="linenos">261</span></a><span class="sd">        Generate ResidueAlignment from a standard align file generated from HMM scan.</span>
</span><span id="ResidueAlignment-262"><a href="#ResidueAlignment-262"><span class="linenos">262</span></a>
</span><span id="ResidueAlignment-263"><a href="#ResidueAlignment-263"><span class="linenos">263</span></a><span class="sd">        Parameters</span>
</span><span id="ResidueAlignment-264"><a href="#ResidueAlignment-264"><span class="linenos">264</span></a><span class="sd">        ----------</span>
</span><span id="ResidueAlignment-265"><a href="#ResidueAlignment-265"><span class="linenos">265</span></a><span class="sd">        align_filepath : str</span>
</span><span id="ResidueAlignment-266"><a href="#ResidueAlignment-266"><span class="linenos">266</span></a><span class="sd">            Filepath of the align file generated from a scan file produced via hmmscan.</span>
</span><span id="ResidueAlignment-267"><a href="#ResidueAlignment-267"><span class="linenos">267</span></a>
</span><span id="ResidueAlignment-268"><a href="#ResidueAlignment-268"><span class="linenos">268</span></a><span class="sd">        Returns</span>
</span><span id="ResidueAlignment-269"><a href="#ResidueAlignment-269"><span class="linenos">269</span></a><span class="sd">        -------</span>
</span><span id="ResidueAlignment-270"><a href="#ResidueAlignment-270"><span class="linenos">270</span></a><span class="sd">        ResidueAlignment</span>
</span><span id="ResidueAlignment-271"><a href="#ResidueAlignment-271"><span class="linenos">271</span></a><span class="sd">            ResidueAlignment with domain and protein starting indices and corresponding sequence texts.</span>
</span><span id="ResidueAlignment-272"><a href="#ResidueAlignment-272"><span class="linenos">272</span></a>
</span><span id="ResidueAlignment-273"><a href="#ResidueAlignment-273"><span class="linenos">273</span></a><span class="sd">        File Format</span>
</span><span id="ResidueAlignment-274"><a href="#ResidueAlignment-274"><span class="linenos">274</span></a><span class="sd">        -----------</span>
</span><span id="ResidueAlignment-275"><a href="#ResidueAlignment-275"><span class="linenos">275</span></a><span class="sd">        Domain_name</span>
</span><span id="ResidueAlignment-276"><a href="#ResidueAlignment-276"><span class="linenos">276</span></a><span class="sd">        1</span>
</span><span id="ResidueAlignment-277"><a href="#ResidueAlignment-277"><span class="linenos">277</span></a><span class="sd">        XXXXXXXXXXXXXXXXXXXX</span>
</span><span id="ResidueAlignment-278"><a href="#ResidueAlignment-278"><span class="linenos">278</span></a><span class="sd">        20</span>
</span><span id="ResidueAlignment-279"><a href="#ResidueAlignment-279"><span class="linenos">279</span></a>
</span><span id="ResidueAlignment-280"><a href="#ResidueAlignment-280"><span class="linenos">280</span></a><span class="sd">        Protein_name</span>
</span><span id="ResidueAlignment-281"><a href="#ResidueAlignment-281"><span class="linenos">281</span></a><span class="sd">        70</span>
</span><span id="ResidueAlignment-282"><a href="#ResidueAlignment-282"><span class="linenos">282</span></a><span class="sd">        XXXXXXXXXXXXXXXXXXXX</span>
</span><span id="ResidueAlignment-283"><a href="#ResidueAlignment-283"><span class="linenos">283</span></a><span class="sd">        89</span>
</span><span id="ResidueAlignment-284"><a href="#ResidueAlignment-284"><span class="linenos">284</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-285"><a href="#ResidueAlignment-285"><span class="linenos">285</span></a>        <span class="c1"># Read the alignment file and parse the important information from each alignment entry.</span>
</span><span id="ResidueAlignment-286"><a href="#ResidueAlignment-286"><span class="linenos">286</span></a>        <span class="n">alignment_entries</span> <span class="o">=</span> <span class="n">ResidueAlignment</span><span class="o">.</span><span class="n">read_align_file</span><span class="p">(</span><span class="n">align_filepath</span><span class="p">)</span>
</span><span id="ResidueAlignment-287"><a href="#ResidueAlignment-287"><span class="linenos">287</span></a>        <span class="n">hmm_entry</span><span class="p">,</span> <span class="n">protein_entry</span> <span class="o">=</span> <span class="n">alignment_entries</span>
</span><span id="ResidueAlignment-288"><a href="#ResidueAlignment-288"><span class="linenos">288</span></a>        <span class="n">domain_name</span><span class="p">,</span> <span class="n">domain_start</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">hmm_entry</span>
</span><span id="ResidueAlignment-289"><a href="#ResidueAlignment-289"><span class="linenos">289</span></a>        <span class="n">protein_name</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">protein_entry</span>
</span><span id="ResidueAlignment-290"><a href="#ResidueAlignment-290"><span class="linenos">290</span></a>    
</span><span id="ResidueAlignment-291"><a href="#ResidueAlignment-291"><span class="linenos">291</span></a>        <span class="c1"># Convert to ints for iteration</span>
</span><span id="ResidueAlignment-292"><a href="#ResidueAlignment-292"><span class="linenos">292</span></a>        <span class="n">domain_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">domain_start</span><span class="p">)</span>
</span><span id="ResidueAlignment-293"><a href="#ResidueAlignment-293"><span class="linenos">293</span></a>        <span class="n">protein_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">protein_start</span><span class="p">)</span>
</span><span id="ResidueAlignment-294"><a href="#ResidueAlignment-294"><span class="linenos">294</span></a>
</span><span id="ResidueAlignment-295"><a href="#ResidueAlignment-295"><span class="linenos">295</span></a>        <span class="k">return</span> <span class="n">ResidueAlignment</span><span class="p">(</span><span class="n">domain_name</span><span class="p">,</span> <span class="n">protein_name</span><span class="p">,</span> <span class="n">domain_start</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">)</span>
</span><span id="ResidueAlignment-296"><a href="#ResidueAlignment-296"><span class="linenos">296</span></a>
</span><span id="ResidueAlignment-297"><a href="#ResidueAlignment-297"><span class="linenos">297</span></a>    <span class="nd">@staticmethod</span>
</span><span id="ResidueAlignment-298"><a href="#ResidueAlignment-298"><span class="linenos">298</span></a>    <span class="k">def</span> <span class="nf">read_align_file</span><span class="p">(</span><span class="n">align_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
</span><span id="ResidueAlignment-299"><a href="#ResidueAlignment-299"><span class="linenos">299</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-300"><a href="#ResidueAlignment-300"><span class="linenos">300</span></a><span class="sd">        Reads standard align file, where a scan file is selected for a particular domain and processed into an align file format. Details are present in produce_align_from_scan().</span>
</span><span id="ResidueAlignment-301"><a href="#ResidueAlignment-301"><span class="linenos">301</span></a><span class="sd">        </span>
</span><span id="ResidueAlignment-302"><a href="#ResidueAlignment-302"><span class="linenos">302</span></a><span class="sd">        Parameters</span>
</span><span id="ResidueAlignment-303"><a href="#ResidueAlignment-303"><span class="linenos">303</span></a><span class="sd">        ----------</span>
</span><span id="ResidueAlignment-304"><a href="#ResidueAlignment-304"><span class="linenos">304</span></a><span class="sd">        align_filepath : str</span>
</span><span id="ResidueAlignment-305"><a href="#ResidueAlignment-305"><span class="linenos">305</span></a><span class="sd">            Filepath and filename of alignment file that contains information on the domain / protein of interest and its mapping to a protein&#39;s structural sequence</span>
</span><span id="ResidueAlignment-306"><a href="#ResidueAlignment-306"><span class="linenos">306</span></a><span class="sd">        </span>
</span><span id="ResidueAlignment-307"><a href="#ResidueAlignment-307"><span class="linenos">307</span></a><span class="sd">        Returns</span>
</span><span id="ResidueAlignment-308"><a href="#ResidueAlignment-308"><span class="linenos">308</span></a><span class="sd">        -------</span>
</span><span id="ResidueAlignment-309"><a href="#ResidueAlignment-309"><span class="linenos">309</span></a><span class="sd">        alignment_entries : list of list of strings</span>
</span><span id="ResidueAlignment-310"><a href="#ResidueAlignment-310"><span class="linenos">310</span></a><span class="sd">            list of associated lines (one which corresponds to the HMM produced sequence and its indices and one that corresponds to the protein&#39;s seqeuence and its indices), which are also contained in a list.</span>
</span><span id="ResidueAlignment-311"><a href="#ResidueAlignment-311"><span class="linenos">311</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-312"><a href="#ResidueAlignment-312"><span class="linenos">312</span></a>        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">align_filepath</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="ResidueAlignment-313"><a href="#ResidueAlignment-313"><span class="linenos">313</span></a>            <span class="n">alignment_entries</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment-314"><a href="#ResidueAlignment-314"><span class="linenos">314</span></a>            <span class="n">current_entry</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment-315"><a href="#ResidueAlignment-315"><span class="linenos">315</span></a>            <span class="n">line_count</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="ResidueAlignment-316"><a href="#ResidueAlignment-316"><span class="linenos">316</span></a>            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="ResidueAlignment-317"><a href="#ResidueAlignment-317"><span class="linenos">317</span></a>                <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span id="ResidueAlignment-318"><a href="#ResidueAlignment-318"><span class="linenos">318</span></a>                <span class="k">if</span> <span class="n">line</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
</span><span id="ResidueAlignment-319"><a href="#ResidueAlignment-319"><span class="linenos">319</span></a>                    <span class="n">line_count</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span id="ResidueAlignment-320"><a href="#ResidueAlignment-320"><span class="linenos">320</span></a>                    <span class="n">current_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span id="ResidueAlignment-321"><a href="#ResidueAlignment-321"><span class="linenos">321</span></a>                <span class="k">if</span> <span class="n">line_count</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>
</span><span id="ResidueAlignment-322"><a href="#ResidueAlignment-322"><span class="linenos">322</span></a>                    <span class="n">line_count</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="ResidueAlignment-323"><a href="#ResidueAlignment-323"><span class="linenos">323</span></a>                    <span class="n">alignment_entries</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_entry</span><span class="p">)</span>
</span><span id="ResidueAlignment-324"><a href="#ResidueAlignment-324"><span class="linenos">324</span></a>                    <span class="n">current_entry</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment-325"><a href="#ResidueAlignment-325"><span class="linenos">325</span></a>        <span class="k">return</span> <span class="n">alignment_entries</span>
</span><span id="ResidueAlignment-326"><a href="#ResidueAlignment-326"><span class="linenos">326</span></a>    
</span><span id="ResidueAlignment-327"><a href="#ResidueAlignment-327"><span class="linenos">327</span></a>    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
</span><span id="ResidueAlignment-328"><a href="#ResidueAlignment-328"><span class="linenos">328</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-329"><a href="#ResidueAlignment-329"><span class="linenos">329</span></a><span class="sd">        Returns string representation of the ResidueAlignment pandas DataFrame in tab-separated value (tsv) format.</span>
</span><span id="ResidueAlignment-330"><a href="#ResidueAlignment-330"><span class="linenos">330</span></a><span class="sd">        </span>
</span><span id="ResidueAlignment-331"><a href="#ResidueAlignment-331"><span class="linenos">331</span></a><span class="sd">        Returns</span>
</span><span id="ResidueAlignment-332"><a href="#ResidueAlignment-332"><span class="linenos">332</span></a><span class="sd">        -------</span>
</span><span id="ResidueAlignment-333"><a href="#ResidueAlignment-333"><span class="linenos">333</span></a><span class="sd">        str</span>
</span><span id="ResidueAlignment-334"><a href="#ResidueAlignment-334"><span class="linenos">334</span></a><span class="sd">            reference_mapping pandas DataFrame exported to TSV format via the to_csv(sep=&quot;\t&quot;) function from pandas.</span>
</span><span id="ResidueAlignment-335"><a href="#ResidueAlignment-335"><span class="linenos">335</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="ResidueAlignment-336"><a href="#ResidueAlignment-336"><span class="linenos">336</span></a>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_mapping</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>A representation of a residue alignment, often from a query HMM to a protein structure target sequence.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>domain_name</strong> (str):
The name of the query HMM.</li>
<li><strong>protein_name</strong> (str):
The name of the target protein sequence.</li>
<li><strong>domain_start</strong> (int):
The starting index of the domain alignment in the query HMM.</li>
<li><strong>protein_start</strong> (int):
The starting index of the domain alignment in the protein target sequence.</li>
<li><strong>domain_text</strong> (str):
The sequence of the domain in the query HMM corresponding to this alignment.</li>
<li><strong>protein_text</strong> (str):
The sequence of the protein target sequence corresponding to this alignment.</li>
</ul>

<h6 id="attributes">Attributes</h6>

<ul>
<li><strong>reference_mapping</strong> (pandas.DataFrame):
The representation of the mapping where a row constitutes a residue pair and its indices in the format: 'domain_index', 'domain_residue', 'protein_residue', 'protein_index'.</li>
<li><strong>domain_to_protein</strong> (dict[int, int]):
A dictionary allowing for mapping from indices corresponding to the query HMM and Multiple Sequence Alignment to the protein target sequence.</li>
<li><strong>protein_to_domain</strong> (dict[int, int]):
A dictionary allowing for mapping from indices corresponding to the protein target sequence to the query HMM and Multiple Sequence Alignment.</li>
</ul>
</div>


                            <div id="ResidueAlignment.__init__" class="classattr">
                                        <input id="ResidueAlignment.__init__-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="name">ResidueAlignment</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">domain_name</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">protein_name</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">domain_start</span><span class="p">:</span> <span class="nb">int</span>,</span><span class="param">	<span class="n">protein_start</span><span class="p">:</span> <span class="nb">int</span>,</span><span class="param">	<span class="n">domain_text</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">protein_text</span><span class="p">:</span> <span class="nb">str</span></span>)</span>

                <label class="view-source-button" for="ResidueAlignment.__init__-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#ResidueAlignment.__init__"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="ResidueAlignment.__init__-209"><a href="#ResidueAlignment.__init__-209"><span class="linenos">209</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">domain_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">protein_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">domain_start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="ResidueAlignment.__init__-210"><a href="#ResidueAlignment.__init__-210"><span class="linenos">210</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">domain_name</span> <span class="o">=</span> <span class="n">domain_name</span>
</span><span id="ResidueAlignment.__init__-211"><a href="#ResidueAlignment.__init__-211"><span class="linenos">211</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">protein_name</span> <span class="o">=</span> <span class="n">protein_name</span>
</span><span id="ResidueAlignment.__init__-212"><a href="#ResidueAlignment.__init__-212"><span class="linenos">212</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">set_reference_mapping</span><span class="p">(</span><span class="n">domain_start</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">)</span>
</span></pre></div>


    

                            </div>
                            <div id="ResidueAlignment.domain_name" class="classattr">
                                <div class="attr variable">
            <span class="name">domain_name</span>

        
    </div>
    <a class="headerlink" href="#ResidueAlignment.domain_name"></a>
    
    

                            </div>
                            <div id="ResidueAlignment.protein_name" class="classattr">
                                <div class="attr variable">
            <span class="name">protein_name</span>

        
    </div>
    <a class="headerlink" href="#ResidueAlignment.protein_name"></a>
    
    

                            </div>
                            <div id="ResidueAlignment.set_reference_mapping" class="classattr">
                                        <input id="ResidueAlignment.set_reference_mapping-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">set_reference_mapping</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">domain_start</span><span class="p">:</span> <span class="nb">int</span>,</span><span class="param">	<span class="n">protein_start</span><span class="p">:</span> <span class="nb">int</span>,</span><span class="param">	<span class="n">domain_text</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">protein_text</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="kc">None</span>:</span></span>

                <label class="view-source-button" for="ResidueAlignment.set_reference_mapping-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#ResidueAlignment.set_reference_mapping"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="ResidueAlignment.set_reference_mapping-214"><a href="#ResidueAlignment.set_reference_mapping-214"><span class="linenos">214</span></a>    <span class="k">def</span> <span class="nf">set_reference_mapping</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">domain_start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="ResidueAlignment.set_reference_mapping-215"><a href="#ResidueAlignment.set_reference_mapping-215"><span class="linenos">215</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="ResidueAlignment.set_reference_mapping-216"><a href="#ResidueAlignment.set_reference_mapping-216"><span class="linenos">216</span></a><span class="sd">        Set values for reference_mapping and mapping dictionaries, domain_to_protein and protein_to_domain.</span>
</span><span id="ResidueAlignment.set_reference_mapping-217"><a href="#ResidueAlignment.set_reference_mapping-217"><span class="linenos">217</span></a>
</span><span id="ResidueAlignment.set_reference_mapping-218"><a href="#ResidueAlignment.set_reference_mapping-218"><span class="linenos">218</span></a><span class="sd">        Note</span>
</span><span id="ResidueAlignment.set_reference_mapping-219"><a href="#ResidueAlignment.set_reference_mapping-219"><span class="linenos">219</span></a><span class="sd">        ----</span>
</span><span id="ResidueAlignment.set_reference_mapping-220"><a href="#ResidueAlignment.set_reference_mapping-220"><span class="linenos">220</span></a><span class="sd">        For details on `domain_start`, `protein_start`, `domain_text`, `protein_text`, please refer to the `ResidueAlignment` docstring.</span>
</span><span id="ResidueAlignment.set_reference_mapping-221"><a href="#ResidueAlignment.set_reference_mapping-221"><span class="linenos">221</span></a>
</span><span id="ResidueAlignment.set_reference_mapping-222"><a href="#ResidueAlignment.set_reference_mapping-222"><span class="linenos">222</span></a><span class="sd">        Returns</span>
</span><span id="ResidueAlignment.set_reference_mapping-223"><a href="#ResidueAlignment.set_reference_mapping-223"><span class="linenos">223</span></a><span class="sd">        -------</span>
</span><span id="ResidueAlignment.set_reference_mapping-224"><a href="#ResidueAlignment.set_reference_mapping-224"><span class="linenos">224</span></a><span class="sd">        None</span>
</span><span id="ResidueAlignment.set_reference_mapping-225"><a href="#ResidueAlignment.set_reference_mapping-225"><span class="linenos">225</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="ResidueAlignment.set_reference_mapping-226"><a href="#ResidueAlignment.set_reference_mapping-226"><span class="linenos">226</span></a>        <span class="c1"># Convert text to list variant for iteration</span>
</span><span id="ResidueAlignment.set_reference_mapping-227"><a href="#ResidueAlignment.set_reference_mapping-227"><span class="linenos">227</span></a>        <span class="n">domain_sequence</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">domain_text</span><span class="p">)</span>
</span><span id="ResidueAlignment.set_reference_mapping-228"><a href="#ResidueAlignment.set_reference_mapping-228"><span class="linenos">228</span></a>        <span class="n">protein_sequence</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">protein_text</span><span class="p">)</span>
</span><span id="ResidueAlignment.set_reference_mapping-229"><a href="#ResidueAlignment.set_reference_mapping-229"><span class="linenos">229</span></a>        
</span><span id="ResidueAlignment.set_reference_mapping-230"><a href="#ResidueAlignment.set_reference_mapping-230"><span class="linenos">230</span></a>        <span class="n">mapping_entries</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment.set_reference_mapping-231"><a href="#ResidueAlignment.set_reference_mapping-231"><span class="linenos">231</span></a>
</span><span id="ResidueAlignment.set_reference_mapping-232"><a href="#ResidueAlignment.set_reference_mapping-232"><span class="linenos">232</span></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">domain_sequence</span><span class="p">)):</span>
</span><span id="ResidueAlignment.set_reference_mapping-233"><a href="#ResidueAlignment.set_reference_mapping-233"><span class="linenos">233</span></a>            <span class="n">mapping_entry</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment.set_reference_mapping-234"><a href="#ResidueAlignment.set_reference_mapping-234"><span class="linenos">234</span></a>            <span class="c1"># Check to see if domain residue is valid, if so, we can assign the proper index.</span>
</span><span id="ResidueAlignment.set_reference_mapping-235"><a href="#ResidueAlignment.set_reference_mapping-235"><span class="linenos">235</span></a>            <span class="k">if</span> <span class="n">domain_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;.&#39;</span><span class="p">:</span>
</span><span id="ResidueAlignment.set_reference_mapping-236"><a href="#ResidueAlignment.set_reference_mapping-236"><span class="linenos">236</span></a>                <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">domain_start</span><span class="p">)</span>
</span><span id="ResidueAlignment.set_reference_mapping-237"><a href="#ResidueAlignment.set_reference_mapping-237"><span class="linenos">237</span></a>                <span class="n">domain_start</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span id="ResidueAlignment.set_reference_mapping-238"><a href="#ResidueAlignment.set_reference_mapping-238"><span class="linenos">238</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="ResidueAlignment.set_reference_mapping-239"><a href="#ResidueAlignment.set_reference_mapping-239"><span class="linenos">239</span></a>                <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">NA</span><span class="p">)</span>
</span><span id="ResidueAlignment.set_reference_mapping-240"><a href="#ResidueAlignment.set_reference_mapping-240"><span class="linenos">240</span></a>            <span class="c1"># Assign the values of the residues mapped together.</span>
</span><span id="ResidueAlignment.set_reference_mapping-241"><a href="#ResidueAlignment.set_reference_mapping-241"><span class="linenos">241</span></a>            <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">domain_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
</span><span id="ResidueAlignment.set_reference_mapping-242"><a href="#ResidueAlignment.set_reference_mapping-242"><span class="linenos">242</span></a>            <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
</span><span id="ResidueAlignment.set_reference_mapping-243"><a href="#ResidueAlignment.set_reference_mapping-243"><span class="linenos">243</span></a>            <span class="c1"># Check to see if protein residue is valid, if so, we can assign the proper index.</span>
</span><span id="ResidueAlignment.set_reference_mapping-244"><a href="#ResidueAlignment.set_reference_mapping-244"><span class="linenos">244</span></a>            <span class="k">if</span> <span class="n">protein_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
</span><span id="ResidueAlignment.set_reference_mapping-245"><a href="#ResidueAlignment.set_reference_mapping-245"><span class="linenos">245</span></a>                <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_start</span><span class="p">)</span>
</span><span id="ResidueAlignment.set_reference_mapping-246"><a href="#ResidueAlignment.set_reference_mapping-246"><span class="linenos">246</span></a>                <span class="n">protein_start</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span id="ResidueAlignment.set_reference_mapping-247"><a href="#ResidueAlignment.set_reference_mapping-247"><span class="linenos">247</span></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="ResidueAlignment.set_reference_mapping-248"><a href="#ResidueAlignment.set_reference_mapping-248"><span class="linenos">248</span></a>                <span class="n">mapping_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">NA</span><span class="p">)</span>
</span><span id="ResidueAlignment.set_reference_mapping-249"><a href="#ResidueAlignment.set_reference_mapping-249"><span class="linenos">249</span></a>            <span class="c1"># Store the resulting mapping in the reference map.</span>
</span><span id="ResidueAlignment.set_reference_mapping-250"><a href="#ResidueAlignment.set_reference_mapping-250"><span class="linenos">250</span></a>            <span class="n">mapping_entries</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mapping_entry</span><span class="p">)</span>
</span><span id="ResidueAlignment.set_reference_mapping-251"><a href="#ResidueAlignment.set_reference_mapping-251"><span class="linenos">251</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">reference_mapping</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">mapping_entries</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;domain_index&#39;</span><span class="p">,</span> <span class="s1">&#39;domain_residue&#39;</span><span class="p">,</span> <span class="s1">&#39;protein_residue&#39;</span><span class="p">,</span> <span class="s1">&#39;protein_index&#39;</span><span class="p">])</span>
</span><span id="ResidueAlignment.set_reference_mapping-252"><a href="#ResidueAlignment.set_reference_mapping-252"><span class="linenos">252</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">reference_mapping</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_mapping</span><span class="o">.</span><span class="n">astype</span><span class="p">({</span><span class="s1">&#39;domain_index&#39;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Int32Dtype</span><span class="p">(),</span> <span class="s1">&#39;protein_index&#39;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Int32Dtype</span><span class="p">(),</span> <span class="s1">&#39;domain_residue&#39;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">StringDtype</span><span class="p">(),</span> <span class="s1">&#39;protein_residue&#39;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">StringDtype</span><span class="p">()})</span>
</span><span id="ResidueAlignment.set_reference_mapping-253"><a href="#ResidueAlignment.set_reference_mapping-253"><span class="linenos">253</span></a>        <span class="n">reference_mapping_notna</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_mapping</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
</span><span id="ResidueAlignment.set_reference_mapping-254"><a href="#ResidueAlignment.set_reference_mapping-254"><span class="linenos">254</span></a>        
</span><span id="ResidueAlignment.set_reference_mapping-255"><a href="#ResidueAlignment.set_reference_mapping-255"><span class="linenos">255</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">domain_to_protein</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">reference_mapping_notna</span><span class="o">.</span><span class="n">domain_index</span><span class="p">,</span> <span class="n">reference_mapping_notna</span><span class="o">.</span><span class="n">protein_index</span><span class="p">))</span>
</span><span id="ResidueAlignment.set_reference_mapping-256"><a href="#ResidueAlignment.set_reference_mapping-256"><span class="linenos">256</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">protein_to_domain</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">reference_mapping_notna</span><span class="o">.</span><span class="n">protein_index</span><span class="p">,</span> <span class="n">reference_mapping_notna</span><span class="o">.</span><span class="n">domain_index</span><span class="p">))</span>
</span></pre></div>


            <div class="docstring"><p>Set values for reference_mapping and mapping dictionaries, domain_to_protein and protein_to_domain.</p>

<h6 id="note">Note</h6>

<p>For details on <code>domain_start</code>, <code>protein_start</code>, <code>domain_text</code>, <code>protein_text</code>, please refer to the <code><a href="#ResidueAlignment">ResidueAlignment</a></code> docstring.</p>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>None</strong></li>
</ul>
</div>


                            </div>
                            <div id="ResidueAlignment.load_from_align_file" class="classattr">
                                        <input id="ResidueAlignment.load_from_align_file-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">load_from_align_file</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">align_filepath</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="n"><a href="#ResidueAlignment">ResidueAlignment</a></span>:</span></span>

                <label class="view-source-button" for="ResidueAlignment.load_from_align_file-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#ResidueAlignment.load_from_align_file"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="ResidueAlignment.load_from_align_file-258"><a href="#ResidueAlignment.load_from_align_file-258"><span class="linenos">258</span></a>    <span class="nd">@staticmethod</span>
</span><span id="ResidueAlignment.load_from_align_file-259"><a href="#ResidueAlignment.load_from_align_file-259"><span class="linenos">259</span></a>    <span class="k">def</span> <span class="nf">load_from_align_file</span><span class="p">(</span><span class="n">align_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;ResidueAlignment&#39;</span><span class="p">:</span>
</span><span id="ResidueAlignment.load_from_align_file-260"><a href="#ResidueAlignment.load_from_align_file-260"><span class="linenos">260</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="ResidueAlignment.load_from_align_file-261"><a href="#ResidueAlignment.load_from_align_file-261"><span class="linenos">261</span></a><span class="sd">        Generate ResidueAlignment from a standard align file generated from HMM scan.</span>
</span><span id="ResidueAlignment.load_from_align_file-262"><a href="#ResidueAlignment.load_from_align_file-262"><span class="linenos">262</span></a>
</span><span id="ResidueAlignment.load_from_align_file-263"><a href="#ResidueAlignment.load_from_align_file-263"><span class="linenos">263</span></a><span class="sd">        Parameters</span>
</span><span id="ResidueAlignment.load_from_align_file-264"><a href="#ResidueAlignment.load_from_align_file-264"><span class="linenos">264</span></a><span class="sd">        ----------</span>
</span><span id="ResidueAlignment.load_from_align_file-265"><a href="#ResidueAlignment.load_from_align_file-265"><span class="linenos">265</span></a><span class="sd">        align_filepath : str</span>
</span><span id="ResidueAlignment.load_from_align_file-266"><a href="#ResidueAlignment.load_from_align_file-266"><span class="linenos">266</span></a><span class="sd">            Filepath of the align file generated from a scan file produced via hmmscan.</span>
</span><span id="ResidueAlignment.load_from_align_file-267"><a href="#ResidueAlignment.load_from_align_file-267"><span class="linenos">267</span></a>
</span><span id="ResidueAlignment.load_from_align_file-268"><a href="#ResidueAlignment.load_from_align_file-268"><span class="linenos">268</span></a><span class="sd">        Returns</span>
</span><span id="ResidueAlignment.load_from_align_file-269"><a href="#ResidueAlignment.load_from_align_file-269"><span class="linenos">269</span></a><span class="sd">        -------</span>
</span><span id="ResidueAlignment.load_from_align_file-270"><a href="#ResidueAlignment.load_from_align_file-270"><span class="linenos">270</span></a><span class="sd">        ResidueAlignment</span>
</span><span id="ResidueAlignment.load_from_align_file-271"><a href="#ResidueAlignment.load_from_align_file-271"><span class="linenos">271</span></a><span class="sd">            ResidueAlignment with domain and protein starting indices and corresponding sequence texts.</span>
</span><span id="ResidueAlignment.load_from_align_file-272"><a href="#ResidueAlignment.load_from_align_file-272"><span class="linenos">272</span></a>
</span><span id="ResidueAlignment.load_from_align_file-273"><a href="#ResidueAlignment.load_from_align_file-273"><span class="linenos">273</span></a><span class="sd">        File Format</span>
</span><span id="ResidueAlignment.load_from_align_file-274"><a href="#ResidueAlignment.load_from_align_file-274"><span class="linenos">274</span></a><span class="sd">        -----------</span>
</span><span id="ResidueAlignment.load_from_align_file-275"><a href="#ResidueAlignment.load_from_align_file-275"><span class="linenos">275</span></a><span class="sd">        Domain_name</span>
</span><span id="ResidueAlignment.load_from_align_file-276"><a href="#ResidueAlignment.load_from_align_file-276"><span class="linenos">276</span></a><span class="sd">        1</span>
</span><span id="ResidueAlignment.load_from_align_file-277"><a href="#ResidueAlignment.load_from_align_file-277"><span class="linenos">277</span></a><span class="sd">        XXXXXXXXXXXXXXXXXXXX</span>
</span><span id="ResidueAlignment.load_from_align_file-278"><a href="#ResidueAlignment.load_from_align_file-278"><span class="linenos">278</span></a><span class="sd">        20</span>
</span><span id="ResidueAlignment.load_from_align_file-279"><a href="#ResidueAlignment.load_from_align_file-279"><span class="linenos">279</span></a>
</span><span id="ResidueAlignment.load_from_align_file-280"><a href="#ResidueAlignment.load_from_align_file-280"><span class="linenos">280</span></a><span class="sd">        Protein_name</span>
</span><span id="ResidueAlignment.load_from_align_file-281"><a href="#ResidueAlignment.load_from_align_file-281"><span class="linenos">281</span></a><span class="sd">        70</span>
</span><span id="ResidueAlignment.load_from_align_file-282"><a href="#ResidueAlignment.load_from_align_file-282"><span class="linenos">282</span></a><span class="sd">        XXXXXXXXXXXXXXXXXXXX</span>
</span><span id="ResidueAlignment.load_from_align_file-283"><a href="#ResidueAlignment.load_from_align_file-283"><span class="linenos">283</span></a><span class="sd">        89</span>
</span><span id="ResidueAlignment.load_from_align_file-284"><a href="#ResidueAlignment.load_from_align_file-284"><span class="linenos">284</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="ResidueAlignment.load_from_align_file-285"><a href="#ResidueAlignment.load_from_align_file-285"><span class="linenos">285</span></a>        <span class="c1"># Read the alignment file and parse the important information from each alignment entry.</span>
</span><span id="ResidueAlignment.load_from_align_file-286"><a href="#ResidueAlignment.load_from_align_file-286"><span class="linenos">286</span></a>        <span class="n">alignment_entries</span> <span class="o">=</span> <span class="n">ResidueAlignment</span><span class="o">.</span><span class="n">read_align_file</span><span class="p">(</span><span class="n">align_filepath</span><span class="p">)</span>
</span><span id="ResidueAlignment.load_from_align_file-287"><a href="#ResidueAlignment.load_from_align_file-287"><span class="linenos">287</span></a>        <span class="n">hmm_entry</span><span class="p">,</span> <span class="n">protein_entry</span> <span class="o">=</span> <span class="n">alignment_entries</span>
</span><span id="ResidueAlignment.load_from_align_file-288"><a href="#ResidueAlignment.load_from_align_file-288"><span class="linenos">288</span></a>        <span class="n">domain_name</span><span class="p">,</span> <span class="n">domain_start</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">hmm_entry</span>
</span><span id="ResidueAlignment.load_from_align_file-289"><a href="#ResidueAlignment.load_from_align_file-289"><span class="linenos">289</span></a>        <span class="n">protein_name</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">protein_entry</span>
</span><span id="ResidueAlignment.load_from_align_file-290"><a href="#ResidueAlignment.load_from_align_file-290"><span class="linenos">290</span></a>    
</span><span id="ResidueAlignment.load_from_align_file-291"><a href="#ResidueAlignment.load_from_align_file-291"><span class="linenos">291</span></a>        <span class="c1"># Convert to ints for iteration</span>
</span><span id="ResidueAlignment.load_from_align_file-292"><a href="#ResidueAlignment.load_from_align_file-292"><span class="linenos">292</span></a>        <span class="n">domain_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">domain_start</span><span class="p">)</span>
</span><span id="ResidueAlignment.load_from_align_file-293"><a href="#ResidueAlignment.load_from_align_file-293"><span class="linenos">293</span></a>        <span class="n">protein_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">protein_start</span><span class="p">)</span>
</span><span id="ResidueAlignment.load_from_align_file-294"><a href="#ResidueAlignment.load_from_align_file-294"><span class="linenos">294</span></a>
</span><span id="ResidueAlignment.load_from_align_file-295"><a href="#ResidueAlignment.load_from_align_file-295"><span class="linenos">295</span></a>        <span class="k">return</span> <span class="n">ResidueAlignment</span><span class="p">(</span><span class="n">domain_name</span><span class="p">,</span> <span class="n">protein_name</span><span class="p">,</span> <span class="n">domain_start</span><span class="p">,</span> <span class="n">protein_start</span><span class="p">,</span> <span class="n">domain_text</span><span class="p">,</span> <span class="n">protein_text</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Generate ResidueAlignment from a standard align file generated from HMM scan.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>align_filepath</strong> (str):
Filepath of the align file generated from a scan file produced via hmmscan.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>ResidueAlignment</strong>: ResidueAlignment with domain and protein starting indices and corresponding sequence texts.</li>
</ul>

<h6 id="file-format">File Format</h6>

<p>Domain_name
1
XXXXXXXXXXXXXXXXXXXX
20</p>

<p>Protein_name
70
XXXXXXXXXXXXXXXXXXXX
89</p>
</div>


                            </div>
                            <div id="ResidueAlignment.read_align_file" class="classattr">
                                        <input id="ResidueAlignment.read_align_file-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">read_align_file</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">align_filepath</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="ResidueAlignment.read_align_file-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#ResidueAlignment.read_align_file"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="ResidueAlignment.read_align_file-297"><a href="#ResidueAlignment.read_align_file-297"><span class="linenos">297</span></a>    <span class="nd">@staticmethod</span>
</span><span id="ResidueAlignment.read_align_file-298"><a href="#ResidueAlignment.read_align_file-298"><span class="linenos">298</span></a>    <span class="k">def</span> <span class="nf">read_align_file</span><span class="p">(</span><span class="n">align_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
</span><span id="ResidueAlignment.read_align_file-299"><a href="#ResidueAlignment.read_align_file-299"><span class="linenos">299</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="ResidueAlignment.read_align_file-300"><a href="#ResidueAlignment.read_align_file-300"><span class="linenos">300</span></a><span class="sd">        Reads standard align file, where a scan file is selected for a particular domain and processed into an align file format. Details are present in produce_align_from_scan().</span>
</span><span id="ResidueAlignment.read_align_file-301"><a href="#ResidueAlignment.read_align_file-301"><span class="linenos">301</span></a><span class="sd">        </span>
</span><span id="ResidueAlignment.read_align_file-302"><a href="#ResidueAlignment.read_align_file-302"><span class="linenos">302</span></a><span class="sd">        Parameters</span>
</span><span id="ResidueAlignment.read_align_file-303"><a href="#ResidueAlignment.read_align_file-303"><span class="linenos">303</span></a><span class="sd">        ----------</span>
</span><span id="ResidueAlignment.read_align_file-304"><a href="#ResidueAlignment.read_align_file-304"><span class="linenos">304</span></a><span class="sd">        align_filepath : str</span>
</span><span id="ResidueAlignment.read_align_file-305"><a href="#ResidueAlignment.read_align_file-305"><span class="linenos">305</span></a><span class="sd">            Filepath and filename of alignment file that contains information on the domain / protein of interest and its mapping to a protein&#39;s structural sequence</span>
</span><span id="ResidueAlignment.read_align_file-306"><a href="#ResidueAlignment.read_align_file-306"><span class="linenos">306</span></a><span class="sd">        </span>
</span><span id="ResidueAlignment.read_align_file-307"><a href="#ResidueAlignment.read_align_file-307"><span class="linenos">307</span></a><span class="sd">        Returns</span>
</span><span id="ResidueAlignment.read_align_file-308"><a href="#ResidueAlignment.read_align_file-308"><span class="linenos">308</span></a><span class="sd">        -------</span>
</span><span id="ResidueAlignment.read_align_file-309"><a href="#ResidueAlignment.read_align_file-309"><span class="linenos">309</span></a><span class="sd">        alignment_entries : list of list of strings</span>
</span><span id="ResidueAlignment.read_align_file-310"><a href="#ResidueAlignment.read_align_file-310"><span class="linenos">310</span></a><span class="sd">            list of associated lines (one which corresponds to the HMM produced sequence and its indices and one that corresponds to the protein&#39;s seqeuence and its indices), which are also contained in a list.</span>
</span><span id="ResidueAlignment.read_align_file-311"><a href="#ResidueAlignment.read_align_file-311"><span class="linenos">311</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="ResidueAlignment.read_align_file-312"><a href="#ResidueAlignment.read_align_file-312"><span class="linenos">312</span></a>        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">align_filepath</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="ResidueAlignment.read_align_file-313"><a href="#ResidueAlignment.read_align_file-313"><span class="linenos">313</span></a>            <span class="n">alignment_entries</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment.read_align_file-314"><a href="#ResidueAlignment.read_align_file-314"><span class="linenos">314</span></a>            <span class="n">current_entry</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment.read_align_file-315"><a href="#ResidueAlignment.read_align_file-315"><span class="linenos">315</span></a>            <span class="n">line_count</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="ResidueAlignment.read_align_file-316"><a href="#ResidueAlignment.read_align_file-316"><span class="linenos">316</span></a>            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="ResidueAlignment.read_align_file-317"><a href="#ResidueAlignment.read_align_file-317"><span class="linenos">317</span></a>                <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</span><span id="ResidueAlignment.read_align_file-318"><a href="#ResidueAlignment.read_align_file-318"><span class="linenos">318</span></a>                <span class="k">if</span> <span class="n">line</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
</span><span id="ResidueAlignment.read_align_file-319"><a href="#ResidueAlignment.read_align_file-319"><span class="linenos">319</span></a>                    <span class="n">line_count</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span id="ResidueAlignment.read_align_file-320"><a href="#ResidueAlignment.read_align_file-320"><span class="linenos">320</span></a>                    <span class="n">current_entry</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span id="ResidueAlignment.read_align_file-321"><a href="#ResidueAlignment.read_align_file-321"><span class="linenos">321</span></a>                <span class="k">if</span> <span class="n">line_count</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>
</span><span id="ResidueAlignment.read_align_file-322"><a href="#ResidueAlignment.read_align_file-322"><span class="linenos">322</span></a>                    <span class="n">line_count</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="ResidueAlignment.read_align_file-323"><a href="#ResidueAlignment.read_align_file-323"><span class="linenos">323</span></a>                    <span class="n">alignment_entries</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_entry</span><span class="p">)</span>
</span><span id="ResidueAlignment.read_align_file-324"><a href="#ResidueAlignment.read_align_file-324"><span class="linenos">324</span></a>                    <span class="n">current_entry</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="ResidueAlignment.read_align_file-325"><a href="#ResidueAlignment.read_align_file-325"><span class="linenos">325</span></a>        <span class="k">return</span> <span class="n">alignment_entries</span>
</span></pre></div>


            <div class="docstring"><p>Reads standard align file, where a scan file is selected for a particular domain and processed into an align file format. Details are present in produce_align_from_scan().</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>align_filepath</strong> (str):
Filepath and filename of alignment file that contains information on the domain / protein of interest and its mapping to a protein's structural sequence</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>alignment_entries</strong> (list of list of strings):
list of associated lines (one which corresponds to the HMM produced sequence and its indices and one that corresponds to the protein's seqeuence and its indices), which are also contained in a list.</li>
</ul>
</div>


                            </div>
                </section>
                <section id="MMCIFInformation">
                            <input id="MMCIFInformation-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr class">
            
    <span class="def">class</span>
    <span class="name">MMCIFInformation</span><wbr>(<span class="base"><a href="#StructureInformation">dcatoolkit.StructureInformation</a></span>):

                <label class="view-source-button" for="MMCIFInformation-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation-735"><a href="#MMCIFInformation-735"><span class="linenos"> 735</span></a><span class="k">class</span> <span class="nc">MMCIFInformation</span><span class="p">(</span><span class="n">StructureInformation</span><span class="p">):</span>
</span><span id="MMCIFInformation-736"><a href="#MMCIFInformation-736"><span class="linenos"> 736</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-737"><a href="#MMCIFInformation-737"><span class="linenos"> 737</span></a><span class="sd">    Information regarding a protein structure, obtained from a protein structure file.</span>
</span><span id="MMCIFInformation-738"><a href="#MMCIFInformation-738"><span class="linenos"> 738</span></a>
</span><span id="MMCIFInformation-739"><a href="#MMCIFInformation-739"><span class="linenos"> 739</span></a><span class="sd">    Parameters</span>
</span><span id="MMCIFInformation-740"><a href="#MMCIFInformation-740"><span class="linenos"> 740</span></a><span class="sd">    ----------</span>
</span><span id="MMCIFInformation-741"><a href="#MMCIFInformation-741"><span class="linenos"> 741</span></a><span class="sd">    structure : biotite.structure.AtomArray</span>
</span><span id="MMCIFInformation-742"><a href="#MMCIFInformation-742"><span class="linenos"> 742</span></a><span class="sd">        Structure obtained from an RCSB entry with a provided pdbx/mmcif file with a specified model number.</span>
</span><span id="MMCIFInformation-743"><a href="#MMCIFInformation-743"><span class="linenos"> 743</span></a><span class="sd">    pdbx_file : biotite.io.pdbx.CIFFile</span>
</span><span id="MMCIFInformation-744"><a href="#MMCIFInformation-744"><span class="linenos"> 744</span></a><span class="sd">        mmCIF file that contains generic information and atomic information of the protein structure categorized into mmCIF blocks.</span>
</span><span id="MMCIFInformation-745"><a href="#MMCIFInformation-745"><span class="linenos"> 745</span></a><span class="sd">    model_num : int</span>
</span><span id="MMCIFInformation-746"><a href="#MMCIFInformation-746"><span class="linenos"> 746</span></a><span class="sd">        The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</span>
</span><span id="MMCIFInformation-747"><a href="#MMCIFInformation-747"><span class="linenos"> 747</span></a>
</span><span id="MMCIFInformation-748"><a href="#MMCIFInformation-748"><span class="linenos"> 748</span></a><span class="sd">    Attributes</span>
</span><span id="MMCIFInformation-749"><a href="#MMCIFInformation-749"><span class="linenos"> 749</span></a><span class="sd">    ----------</span>
</span><span id="MMCIFInformation-750"><a href="#MMCIFInformation-750"><span class="linenos"> 750</span></a><span class="sd">    self.full_sequences : dict of str, biotite.sequence.ProteinSequence</span>
</span><span id="MMCIFInformation-751"><a href="#MMCIFInformation-751"><span class="linenos"> 751</span></a><span class="sd">        The full protein sequences from the pdbx file used to generate the structure stored in a dictionary where auth_chain_id is the key and the ProteinSequence object is the value.</span>
</span><span id="MMCIFInformation-752"><a href="#MMCIFInformation-752"><span class="linenos"> 752</span></a><span class="sd">    self.non_missing_sequences : dict of str, biotite.sequence.ProteinSequence</span>
</span><span id="MMCIFInformation-753"><a href="#MMCIFInformation-753"><span class="linenos"> 753</span></a><span class="sd">        The protein sequences, without missing residues, compiled in the structure of the StructureInformation instance stored in a dictionary where chain_id is the key and the sequence string is the value.</span>
</span><span id="MMCIFInformation-754"><a href="#MMCIFInformation-754"><span class="linenos"> 754</span></a><span class="sd">    self.atom_data : numpy.ndarray, optional</span>
</span><span id="MMCIFInformation-755"><a href="#MMCIFInformation-755"><span class="linenos"> 755</span></a><span class="sd">        Entries in the format &#39;ATOM&#39;, residue index, chain ID, auth residue index, auth chain ID, model number</span>
</span><span id="MMCIFInformation-756"><a href="#MMCIFInformation-756"><span class="linenos"> 756</span></a><span class="sd">    self.het_atom_data : numpy.ndarray, optional</span>
</span><span id="MMCIFInformation-757"><a href="#MMCIFInformation-757"><span class="linenos"> 757</span></a><span class="sd">        Array of entries in the format &#39;HETATM&#39;, residue index, chain ID, auth residue index, auth chain ID, model number</span>
</span><span id="MMCIFInformation-758"><a href="#MMCIFInformation-758"><span class="linenos"> 758</span></a><span class="sd">    self.unique_chains : numpy.ndarray, optional</span>
</span><span id="MMCIFInformation-759"><a href="#MMCIFInformation-759"><span class="linenos"> 759</span></a><span class="sd">        Array of unique asym_id entries which corresponds to unique chain IDs.</span>
</span><span id="MMCIFInformation-760"><a href="#MMCIFInformation-760"><span class="linenos"> 760</span></a><span class="sd">    self.chain_auth_dict : dict of str, str, optional</span>
</span><span id="MMCIFInformation-761"><a href="#MMCIFInformation-761"><span class="linenos"> 761</span></a><span class="sd">        Uses chain id as a key and provides auth chain id as a value.</span>
</span><span id="MMCIFInformation-762"><a href="#MMCIFInformation-762"><span class="linenos"> 762</span></a><span class="sd">    self.auth_chain_dict : dict of str, str, optional</span>
</span><span id="MMCIFInformation-763"><a href="#MMCIFInformation-763"><span class="linenos"> 763</span></a><span class="sd">        Uses auth chain id as a key and provides original/label chain id as a value. </span>
</span><span id="MMCIFInformation-764"><a href="#MMCIFInformation-764"><span class="linenos"> 764</span></a><span class="sd">    self.res_auth_dict : dict of str, tuple of int, int or optional </span>
</span><span id="MMCIFInformation-765"><a href="#MMCIFInformation-765"><span class="linenos"> 765</span></a><span class="sd">        Uses chain id as a key and an array of residue index and auth residue index as a value.</span>
</span><span id="MMCIFInformation-766"><a href="#MMCIFInformation-766"><span class="linenos"> 766</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-767"><a href="#MMCIFInformation-767"><span class="linenos"> 767</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structure</span><span class="p">,</span> <span class="n">pdbx_file</span><span class="p">:</span> <span class="n">pdbx</span><span class="o">.</span><span class="n">CIFFile</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
</span><span id="MMCIFInformation-768"><a href="#MMCIFInformation-768"><span class="linenos"> 768</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">structure</span>
</span><span id="MMCIFInformation-769"><a href="#MMCIFInformation-769"><span class="linenos"> 769</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">pdbx_file</span> <span class="o">=</span> <span class="n">pdbx_file</span>
</span><span id="MMCIFInformation-770"><a href="#MMCIFInformation-770"><span class="linenos"> 770</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">model_num</span> <span class="o">=</span> <span class="n">model_num</span>
</span><span id="MMCIFInformation-771"><a href="#MMCIFInformation-771"><span class="linenos"> 771</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">full_sequences</span> <span class="o">=</span> <span class="n">pdbx</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">pdbx_file</span><span class="p">)</span>
</span><span id="MMCIFInformation-772"><a href="#MMCIFInformation-772"><span class="linenos"> 772</span></a>        <span class="n">non_hetero_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="MMCIFInformation-773"><a href="#MMCIFInformation-773"><span class="linenos"> 773</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">chain</span><span class="p">):</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">chain</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">struc</span><span class="o">.</span><span class="n">get_chains</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">),</span> <span class="n">struc</span><span class="o">.</span><span class="n">to_sequence</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))}</span>
</span><span id="MMCIFInformation-774"><a href="#MMCIFInformation-774"><span class="linenos"> 774</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">generate_auth_info</span><span class="p">()</span>
</span><span id="MMCIFInformation-775"><a href="#MMCIFInformation-775"><span class="linenos"> 775</span></a>
</span><span id="MMCIFInformation-776"><a href="#MMCIFInformation-776"><span class="linenos"> 776</span></a>    <span class="k">def</span> <span class="nf">generate_auth_info</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="MMCIFInformation-777"><a href="#MMCIFInformation-777"><span class="linenos"> 777</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-778"><a href="#MMCIFInformation-778"><span class="linenos"> 778</span></a><span class="sd">        Ran as part of constructor function. Generates information needed to access auth information including auth_seq_id and auth_asym_id, which correspond to alternative chain ids and alternative residue indices.</span>
</span><span id="MMCIFInformation-779"><a href="#MMCIFInformation-779"><span class="linenos"> 779</span></a><span class="sd">        </span>
</span><span id="MMCIFInformation-780"><a href="#MMCIFInformation-780"><span class="linenos"> 780</span></a><span class="sd">        Note</span>
</span><span id="MMCIFInformation-781"><a href="#MMCIFInformation-781"><span class="linenos"> 781</span></a><span class="sd">        ----</span>
</span><span id="MMCIFInformation-782"><a href="#MMCIFInformation-782"><span class="linenos"> 782</span></a><span class="sd">        See attributes for details.</span>
</span><span id="MMCIFInformation-783"><a href="#MMCIFInformation-783"><span class="linenos"> 783</span></a><span class="sd">        </span>
</span><span id="MMCIFInformation-784"><a href="#MMCIFInformation-784"><span class="linenos"> 784</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation-785"><a href="#MMCIFInformation-785"><span class="linenos"> 785</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation-786"><a href="#MMCIFInformation-786"><span class="linenos"> 786</span></a><span class="sd">        None</span>
</span><span id="MMCIFInformation-787"><a href="#MMCIFInformation-787"><span class="linenos"> 787</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-788"><a href="#MMCIFInformation-788"><span class="linenos"> 788</span></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pdbx_file</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="MMCIFInformation-789"><a href="#MMCIFInformation-789"><span class="linenos"> 789</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">first_block</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pdbx_file</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MMCIFInformation-790"><a href="#MMCIFInformation-790"><span class="linenos"> 790</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pdbx_file</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">first_block</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;atom_site&#39;</span><span class="p">)</span>
</span><span id="MMCIFInformation-791"><a href="#MMCIFInformation-791"><span class="linenos"> 791</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">chain_auth_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="MMCIFInformation-792"><a href="#MMCIFInformation-792"><span class="linenos"> 792</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="MMCIFInformation-793"><a href="#MMCIFInformation-793"><span class="linenos"> 793</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="MMCIFInformation-794"><a href="#MMCIFInformation-794"><span class="linenos"> 794</span></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span><span class="p">:</span>
</span><span id="MMCIFInformation-795"><a href="#MMCIFInformation-795"><span class="linenos"> 795</span></a>                <span class="n">group_pdbs</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation-796"><a href="#MMCIFInformation-796"><span class="linenos"> 796</span></a>                <span class="n">seq_ids</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation-797"><a href="#MMCIFInformation-797"><span class="linenos"> 797</span></a>                <span class="n">asym_ids</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation-798"><a href="#MMCIFInformation-798"><span class="linenos"> 798</span></a>                <span class="n">auth_seq_ids</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation-799"><a href="#MMCIFInformation-799"><span class="linenos"> 799</span></a>                <span class="n">auth_asym_ids</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation-800"><a href="#MMCIFInformation-800"><span class="linenos"> 800</span></a>                <span class="n">model_nums</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation-801"><a href="#MMCIFInformation-801"><span class="linenos"> 801</span></a>                <span class="k">for</span> <span class="n">col_name</span><span class="p">,</span> <span class="n">col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
</span><span id="MMCIFInformation-802"><a href="#MMCIFInformation-802"><span class="linenos"> 802</span></a>                    <span class="k">if</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;group_PDB&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation-803"><a href="#MMCIFInformation-803"><span class="linenos"> 803</span></a>                        <span class="n">group_pdbs</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation-804"><a href="#MMCIFInformation-804"><span class="linenos"> 804</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;label_seq_id&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation-805"><a href="#MMCIFInformation-805"><span class="linenos"> 805</span></a>                        <span class="n">seq_ids</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation-806"><a href="#MMCIFInformation-806"><span class="linenos"> 806</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;label_asym_id&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation-807"><a href="#MMCIFInformation-807"><span class="linenos"> 807</span></a>                        <span class="n">asym_ids</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation-808"><a href="#MMCIFInformation-808"><span class="linenos"> 808</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;auth_seq_id&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation-809"><a href="#MMCIFInformation-809"><span class="linenos"> 809</span></a>                        <span class="n">auth_seq_ids</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation-810"><a href="#MMCIFInformation-810"><span class="linenos"> 810</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;auth_asym_id&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation-811"><a href="#MMCIFInformation-811"><span class="linenos"> 811</span></a>                        <span class="n">auth_asym_ids</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation-812"><a href="#MMCIFInformation-812"><span class="linenos"> 812</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;pdbx_PDB_model_num&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation-813"><a href="#MMCIFInformation-813"><span class="linenos"> 813</span></a>                        <span class="n">model_nums</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation-814"><a href="#MMCIFInformation-814"><span class="linenos"> 814</span></a>
</span><span id="MMCIFInformation-815"><a href="#MMCIFInformation-815"><span class="linenos"> 815</span></a>                <span class="n">atom_site_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">((</span><span class="n">group_pdbs</span><span class="p">,</span> <span class="n">seq_ids</span><span class="p">,</span> <span class="n">asym_ids</span><span class="p">,</span> <span class="n">auth_seq_ids</span><span class="p">,</span> <span class="n">auth_asym_ids</span><span class="p">,</span> <span class="n">model_nums</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</span><span id="MMCIFInformation-816"><a href="#MMCIFInformation-816"><span class="linenos"> 816</span></a>                <span class="n">atom_site_data</span> <span class="o">=</span> <span class="n">atom_site_data</span><span class="p">[</span><span class="n">atom_site_data</span><span class="p">[:,</span><span class="mi">5</span><span class="p">]</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_num</span><span class="p">)]</span>
</span><span id="MMCIFInformation-817"><a href="#MMCIFInformation-817"><span class="linenos"> 817</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">atom_data</span> <span class="o">=</span> <span class="n">atom_site_data</span><span class="p">[</span><span class="n">atom_site_data</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ATOM&quot;</span><span class="p">]</span>
</span><span id="MMCIFInformation-818"><a href="#MMCIFInformation-818"><span class="linenos"> 818</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">het_atom_data</span> <span class="o">=</span> <span class="n">atom_site_data</span><span class="p">[</span><span class="n">atom_site_data</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;HETATM&quot;</span><span class="p">]</span>
</span><span id="MMCIFInformation-819"><a href="#MMCIFInformation-819"><span class="linenos"> 819</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">atom_data</span><span class="p">[:,</span><span class="mi">2</span><span class="p">])</span>
</span><span id="MMCIFInformation-820"><a href="#MMCIFInformation-820"><span class="linenos"> 820</span></a>                <span class="k">for</span> <span class="n">unique_chain</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span><span class="p">:</span>
</span><span id="MMCIFInformation-821"><a href="#MMCIFInformation-821"><span class="linenos"> 821</span></a>                    <span class="n">unique_entry</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">atom_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">atom_data</span><span class="p">[:,</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="n">unique_chain</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MMCIFInformation-822"><a href="#MMCIFInformation-822"><span class="linenos"> 822</span></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">chain_auth_dict</span><span class="p">[</span><span class="n">unique_entry</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> <span class="o">=</span> <span class="n">unique_entry</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
</span><span id="MMCIFInformation-823"><a href="#MMCIFInformation-823"><span class="linenos"> 823</span></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">unique_entry</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span> <span class="o">=</span> <span class="n">unique_entry</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
</span><span id="MMCIFInformation-824"><a href="#MMCIFInformation-824"><span class="linenos"> 824</span></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">unique_entry</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> <span class="o">=</span> <span class="n">unique_entry</span><span class="p">[[</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">]]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)</span>
</span><span id="MMCIFInformation-825"><a href="#MMCIFInformation-825"><span class="linenos"> 825</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation-826"><a href="#MMCIFInformation-826"><span class="linenos"> 826</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="MMCIFInformation-827"><a href="#MMCIFInformation-827"><span class="linenos"> 827</span></a>    
</span><span id="MMCIFInformation-828"><a href="#MMCIFInformation-828"><span class="linenos"> 828</span></a>    <span class="k">def</span> <span class="nf">get_full_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
</span><span id="MMCIFInformation-829"><a href="#MMCIFInformation-829"><span class="linenos"> 829</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-830"><a href="#MMCIFInformation-830"><span class="linenos"> 830</span></a><span class="sd">        Get full sequence, including missing residues, from the specified chain off of RCSB.</span>
</span><span id="MMCIFInformation-831"><a href="#MMCIFInformation-831"><span class="linenos"> 831</span></a>
</span><span id="MMCIFInformation-832"><a href="#MMCIFInformation-832"><span class="linenos"> 832</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation-833"><a href="#MMCIFInformation-833"><span class="linenos"> 833</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation-834"><a href="#MMCIFInformation-834"><span class="linenos"> 834</span></a><span class="sd">        chain_id : str</span>
</span><span id="MMCIFInformation-835"><a href="#MMCIFInformation-835"><span class="linenos"> 835</span></a><span class="sd">            Chain id supplied. The full sequence, including missing residues, of this chain will be returned.</span>
</span><span id="MMCIFInformation-836"><a href="#MMCIFInformation-836"><span class="linenos"> 836</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation-837"><a href="#MMCIFInformation-837"><span class="linenos"> 837</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation-838"><a href="#MMCIFInformation-838"><span class="linenos"> 838</span></a>
</span><span id="MMCIFInformation-839"><a href="#MMCIFInformation-839"><span class="linenos"> 839</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation-840"><a href="#MMCIFInformation-840"><span class="linenos"> 840</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation-841"><a href="#MMCIFInformation-841"><span class="linenos"> 841</span></a><span class="sd">        str</span>
</span><span id="MMCIFInformation-842"><a href="#MMCIFInformation-842"><span class="linenos"> 842</span></a><span class="sd">            The full sequence, including missing residues, of the chain specified.</span>
</span><span id="MMCIFInformation-843"><a href="#MMCIFInformation-843"><span class="linenos"> 843</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-844"><a href="#MMCIFInformation-844"><span class="linenos"> 844</span></a>        <span class="k">if</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span>
</span><span id="MMCIFInformation-845"><a href="#MMCIFInformation-845"><span class="linenos"> 845</span></a>            <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">full_sequences</span><span class="p">[</span><span class="n">chain_id</span><span class="p">])</span>
</span><span id="MMCIFInformation-846"><a href="#MMCIFInformation-846"><span class="linenos"> 846</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation-847"><a href="#MMCIFInformation-847"><span class="linenos"> 847</span></a>            <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">full_sequences</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chain_auth_dict</span><span class="p">[</span><span class="n">chain_id</span><span class="p">]])</span>
</span><span id="MMCIFInformation-848"><a href="#MMCIFInformation-848"><span class="linenos"> 848</span></a>        
</span><span id="MMCIFInformation-849"><a href="#MMCIFInformation-849"><span class="linenos"> 849</span></a>    <span class="k">def</span> <span class="nf">get_non_missing_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
</span><span id="MMCIFInformation-850"><a href="#MMCIFInformation-850"><span class="linenos"> 850</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-851"><a href="#MMCIFInformation-851"><span class="linenos"> 851</span></a><span class="sd">        Get sequence, including only non-missing residues, from the specified chain off of RCSB.</span>
</span><span id="MMCIFInformation-852"><a href="#MMCIFInformation-852"><span class="linenos"> 852</span></a>
</span><span id="MMCIFInformation-853"><a href="#MMCIFInformation-853"><span class="linenos"> 853</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation-854"><a href="#MMCIFInformation-854"><span class="linenos"> 854</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation-855"><a href="#MMCIFInformation-855"><span class="linenos"> 855</span></a><span class="sd">        chain_id : str</span>
</span><span id="MMCIFInformation-856"><a href="#MMCIFInformation-856"><span class="linenos"> 856</span></a><span class="sd">            Chain id supplied. The full sequence, including only non-missing residues, of this chain will be returned.</span>
</span><span id="MMCIFInformation-857"><a href="#MMCIFInformation-857"><span class="linenos"> 857</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation-858"><a href="#MMCIFInformation-858"><span class="linenos"> 858</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation-859"><a href="#MMCIFInformation-859"><span class="linenos"> 859</span></a>
</span><span id="MMCIFInformation-860"><a href="#MMCIFInformation-860"><span class="linenos"> 860</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation-861"><a href="#MMCIFInformation-861"><span class="linenos"> 861</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation-862"><a href="#MMCIFInformation-862"><span class="linenos"> 862</span></a><span class="sd">        str</span>
</span><span id="MMCIFInformation-863"><a href="#MMCIFInformation-863"><span class="linenos"> 863</span></a><span class="sd">            The full sequence, including only non-missing residues, of the chain specified.</span>
</span><span id="MMCIFInformation-864"><a href="#MMCIFInformation-864"><span class="linenos"> 864</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-865"><a href="#MMCIFInformation-865"><span class="linenos"> 865</span></a>        <span class="k">if</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span>
</span><span id="MMCIFInformation-866"><a href="#MMCIFInformation-866"><span class="linenos"> 866</span></a>            <span class="n">original_chain_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain_id</span><span class="p">]</span>
</span><span id="MMCIFInformation-867"><a href="#MMCIFInformation-867"><span class="linenos"> 867</span></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span><span class="p">[</span><span class="n">original_chain_id</span><span class="p">]</span>
</span><span id="MMCIFInformation-868"><a href="#MMCIFInformation-868"><span class="linenos"> 868</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation-869"><a href="#MMCIFInformation-869"><span class="linenos"> 869</span></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span><span class="p">[</span><span class="n">chain_id</span><span class="p">]</span>
</span><span id="MMCIFInformation-870"><a href="#MMCIFInformation-870"><span class="linenos"> 870</span></a>        
</span><span id="MMCIFInformation-871"><a href="#MMCIFInformation-871"><span class="linenos"> 871</span></a>    <span class="k">def</span> <span class="nf">get_chain_specific_structure</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
</span><span id="MMCIFInformation-872"><a href="#MMCIFInformation-872"><span class="linenos"> 872</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-873"><a href="#MMCIFInformation-873"><span class="linenos"> 873</span></a><span class="sd">        Subsets structure attribute to select for chain specific portions of the structure.</span>
</span><span id="MMCIFInformation-874"><a href="#MMCIFInformation-874"><span class="linenos"> 874</span></a>
</span><span id="MMCIFInformation-875"><a href="#MMCIFInformation-875"><span class="linenos"> 875</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation-876"><a href="#MMCIFInformation-876"><span class="linenos"> 876</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation-877"><a href="#MMCIFInformation-877"><span class="linenos"> 877</span></a><span class="sd">        ca_only : bool</span>
</span><span id="MMCIFInformation-878"><a href="#MMCIFInformation-878"><span class="linenos"> 878</span></a><span class="sd">            If true, the structure will also be subsetted for atom entries where the atom_name annotation is &quot;CA&quot; (referring to alpha-carbons)</span>
</span><span id="MMCIFInformation-879"><a href="#MMCIFInformation-879"><span class="linenos"> 879</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation-880"><a href="#MMCIFInformation-880"><span class="linenos"> 880</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation-881"><a href="#MMCIFInformation-881"><span class="linenos"> 881</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation-882"><a href="#MMCIFInformation-882"><span class="linenos"> 882</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="MMCIFInformation-883"><a href="#MMCIFInformation-883"><span class="linenos"> 883</span></a><span class="sd">        remove_hetero : bool, default=True</span>
</span><span id="MMCIFInformation-884"><a href="#MMCIFInformation-884"><span class="linenos"> 884</span></a><span class="sd">            If true, the structure will also be subsetted for atom entries where the hetero annotation is False, thus removing heteroatoms.</span>
</span><span id="MMCIFInformation-885"><a href="#MMCIFInformation-885"><span class="linenos"> 885</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation-886"><a href="#MMCIFInformation-886"><span class="linenos"> 886</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation-887"><a href="#MMCIFInformation-887"><span class="linenos"> 887</span></a>
</span><span id="MMCIFInformation-888"><a href="#MMCIFInformation-888"><span class="linenos"> 888</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation-889"><a href="#MMCIFInformation-889"><span class="linenos"> 889</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation-890"><a href="#MMCIFInformation-890"><span class="linenos"> 890</span></a><span class="sd">        tuple of biotite.structure.AtomArray, biotite.structure.AtomArray</span>
</span><span id="MMCIFInformation-891"><a href="#MMCIFInformation-891"><span class="linenos"> 891</span></a><span class="sd">            Two AtomArrays that refer to atoms in the first chain and second chain, respectively without accounting for the presence of heteroatoms if `remove_hetero` is True.</span>
</span><span id="MMCIFInformation-892"><a href="#MMCIFInformation-892"><span class="linenos"> 892</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-893"><a href="#MMCIFInformation-893"><span class="linenos"> 893</span></a>        <span class="k">if</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span>
</span><span id="MMCIFInformation-894"><a href="#MMCIFInformation-894"><span class="linenos"> 894</span></a>            <span class="n">chain1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain1</span><span class="p">]</span>
</span><span id="MMCIFInformation-895"><a href="#MMCIFInformation-895"><span class="linenos"> 895</span></a>            <span class="n">chain2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain2</span><span class="p">]</span>
</span><span id="MMCIFInformation-896"><a href="#MMCIFInformation-896"><span class="linenos"> 896</span></a>
</span><span id="MMCIFInformation-897"><a href="#MMCIFInformation-897"><span class="linenos"> 897</span></a>        <span class="n">selected_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span>
</span><span id="MMCIFInformation-898"><a href="#MMCIFInformation-898"><span class="linenos"> 898</span></a>        <span class="k">if</span> <span class="n">remove_hetero</span><span class="p">:</span>
</span><span id="MMCIFInformation-899"><a href="#MMCIFInformation-899"><span class="linenos"> 899</span></a>            <span class="c1"># Remove hetero atoms via hetero column of structure ndarray</span>
</span><span id="MMCIFInformation-900"><a href="#MMCIFInformation-900"><span class="linenos"> 900</span></a>            <span class="n">selected_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="MMCIFInformation-901"><a href="#MMCIFInformation-901"><span class="linenos"> 901</span></a>        <span class="k">if</span> <span class="n">ca_only</span><span class="p">:</span>
</span><span id="MMCIFInformation-902"><a href="#MMCIFInformation-902"><span class="linenos"> 902</span></a>            <span class="c1"># Consider selection of alpha-carbon atoms only</span>
</span><span id="MMCIFInformation-903"><a href="#MMCIFInformation-903"><span class="linenos"> 903</span></a>            <span class="n">selected_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">atom_name</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span><span class="p">]</span>
</span><span id="MMCIFInformation-904"><a href="#MMCIFInformation-904"><span class="linenos"> 904</span></a>        <span class="n">chain1_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain1</span><span class="p">]</span>
</span><span id="MMCIFInformation-905"><a href="#MMCIFInformation-905"><span class="linenos"> 905</span></a>        <span class="n">chain2_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain2</span><span class="p">]</span>
</span><span id="MMCIFInformation-906"><a href="#MMCIFInformation-906"><span class="linenos"> 906</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">)</span>
</span><span id="MMCIFInformation-907"><a href="#MMCIFInformation-907"><span class="linenos"> 907</span></a>
</span><span id="MMCIFInformation-908"><a href="#MMCIFInformation-908"><span class="linenos"> 908</span></a>    <span class="k">def</span> <span class="nf">generate_dist_matrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="MMCIFInformation-909"><a href="#MMCIFInformation-909"><span class="linenos"> 909</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-910"><a href="#MMCIFInformation-910"><span class="linenos"> 910</span></a><span class="sd">        Generates distance matrix between two chains in the structure attribute.</span>
</span><span id="MMCIFInformation-911"><a href="#MMCIFInformation-911"><span class="linenos"> 911</span></a>
</span><span id="MMCIFInformation-912"><a href="#MMCIFInformation-912"><span class="linenos"> 912</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation-913"><a href="#MMCIFInformation-913"><span class="linenos"> 913</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation-914"><a href="#MMCIFInformation-914"><span class="linenos"> 914</span></a><span class="sd">        ca_only : bool</span>
</span><span id="MMCIFInformation-915"><a href="#MMCIFInformation-915"><span class="linenos"> 915</span></a><span class="sd">            If True, only atoms that have the name &quot;CA&quot; are selected in the chains the distance matrix is calculated between.</span>
</span><span id="MMCIFInformation-916"><a href="#MMCIFInformation-916"><span class="linenos"> 916</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation-917"><a href="#MMCIFInformation-917"><span class="linenos"> 917</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation-918"><a href="#MMCIFInformation-918"><span class="linenos"> 918</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation-919"><a href="#MMCIFInformation-919"><span class="linenos"> 919</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation-920"><a href="#MMCIFInformation-920"><span class="linenos"> 920</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation-921"><a href="#MMCIFInformation-921"><span class="linenos"> 921</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation-922"><a href="#MMCIFInformation-922"><span class="linenos"> 922</span></a>
</span><span id="MMCIFInformation-923"><a href="#MMCIFInformation-923"><span class="linenos"> 923</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation-924"><a href="#MMCIFInformation-924"><span class="linenos"> 924</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation-925"><a href="#MMCIFInformation-925"><span class="linenos"> 925</span></a><span class="sd">        tuple of biotite.structure.AtomArray, biotite.structure.AtomArray, numpy.ndarray</span>
</span><span id="MMCIFInformation-926"><a href="#MMCIFInformation-926"><span class="linenos"> 926</span></a><span class="sd">            Tuple containing the chain 1 structure, the chain 2 structure, and the distance matrix of chain 1 and chain 2&#39;s pairwise distances.</span>
</span><span id="MMCIFInformation-927"><a href="#MMCIFInformation-927"><span class="linenos"> 927</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-928"><a href="#MMCIFInformation-928"><span class="linenos"> 928</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chain_specific_structure</span><span class="p">(</span><span class="n">ca_only</span><span class="p">,</span> <span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation-929"><a href="#MMCIFInformation-929"><span class="linenos"> 929</span></a>        <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">cdist</span><span class="p">(</span><span class="n">chain1_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>
</span><span id="MMCIFInformation-930"><a href="#MMCIFInformation-930"><span class="linenos"> 930</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">,</span> <span class="n">dist_matrix</span><span class="p">)</span>
</span><span id="MMCIFInformation-931"><a href="#MMCIFInformation-931"><span class="linenos"> 931</span></a>    
</span><span id="MMCIFInformation-932"><a href="#MMCIFInformation-932"><span class="linenos"> 932</span></a>    <span class="k">def</span> <span class="nf">get_shift_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
</span><span id="MMCIFInformation-933"><a href="#MMCIFInformation-933"><span class="linenos"> 933</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-934"><a href="#MMCIFInformation-934"><span class="linenos"> 934</span></a><span class="sd">        Get shift values needed for production of auth residue ids.</span>
</span><span id="MMCIFInformation-935"><a href="#MMCIFInformation-935"><span class="linenos"> 935</span></a><span class="sd">        </span>
</span><span id="MMCIFInformation-936"><a href="#MMCIFInformation-936"><span class="linenos"> 936</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation-937"><a href="#MMCIFInformation-937"><span class="linenos"> 937</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation-938"><a href="#MMCIFInformation-938"><span class="linenos"> 938</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation-939"><a href="#MMCIFInformation-939"><span class="linenos"> 939</span></a><span class="sd">            Name of the chain id present referring to the second column of residues.</span>
</span><span id="MMCIFInformation-940"><a href="#MMCIFInformation-940"><span class="linenos"> 940</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation-941"><a href="#MMCIFInformation-941"><span class="linenos"> 941</span></a><span class="sd">            Name of the chain id present referring to the second column of residues.</span>
</span><span id="MMCIFInformation-942"><a href="#MMCIFInformation-942"><span class="linenos"> 942</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation-943"><a href="#MMCIFInformation-943"><span class="linenos"> 943</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation-944"><a href="#MMCIFInformation-944"><span class="linenos"> 944</span></a>
</span><span id="MMCIFInformation-945"><a href="#MMCIFInformation-945"><span class="linenos"> 945</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation-946"><a href="#MMCIFInformation-946"><span class="linenos"> 946</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation-947"><a href="#MMCIFInformation-947"><span class="linenos"> 947</span></a><span class="sd">        (shift1, shift2) : tuple of int, int</span>
</span><span id="MMCIFInformation-948"><a href="#MMCIFInformation-948"><span class="linenos"> 948</span></a><span class="sd">            Tuple containing both shift values, the difference between the auth_res_id and res_id.</span>
</span><span id="MMCIFInformation-949"><a href="#MMCIFInformation-949"><span class="linenos"> 949</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-950"><a href="#MMCIFInformation-950"><span class="linenos"> 950</span></a>        <span class="k">if</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span>
</span><span id="MMCIFInformation-951"><a href="#MMCIFInformation-951"><span class="linenos"> 951</span></a>            <span class="n">chain1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain1</span><span class="p">]</span>
</span><span id="MMCIFInformation-952"><a href="#MMCIFInformation-952"><span class="linenos"> 952</span></a>            <span class="n">chain2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain2</span><span class="p">]</span>
</span><span id="MMCIFInformation-953"><a href="#MMCIFInformation-953"><span class="linenos"> 953</span></a>
</span><span id="MMCIFInformation-954"><a href="#MMCIFInformation-954"><span class="linenos"> 954</span></a>        <span class="n">shift1</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="MMCIFInformation-955"><a href="#MMCIFInformation-955"><span class="linenos"> 955</span></a>        <span class="n">shift2</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="MMCIFInformation-956"><a href="#MMCIFInformation-956"><span class="linenos"> 956</span></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span><span class="p">:</span>
</span><span id="MMCIFInformation-957"><a href="#MMCIFInformation-957"><span class="linenos"> 957</span></a>            <span class="n">shift1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">chain1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">chain1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MMCIFInformation-958"><a href="#MMCIFInformation-958"><span class="linenos"> 958</span></a>            <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">chain2</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">chain2</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MMCIFInformation-959"><a href="#MMCIFInformation-959"><span class="linenos"> 959</span></a>            <span class="k">return</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span>
</span><span id="MMCIFInformation-960"><a href="#MMCIFInformation-960"><span class="linenos"> 960</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation-961"><a href="#MMCIFInformation-961"><span class="linenos"> 961</span></a>            <span class="k">return</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span>
</span><span id="MMCIFInformation-962"><a href="#MMCIFInformation-962"><span class="linenos"> 962</span></a>
</span><span id="MMCIFInformation-963"><a href="#MMCIFInformation-963"><span class="linenos"> 963</span></a>    <span class="k">def</span> <span class="nf">get_min_dist_atom_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="MMCIFInformation-964"><a href="#MMCIFInformation-964"><span class="linenos"> 964</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-965"><a href="#MMCIFInformation-965"><span class="linenos"> 965</span></a><span class="sd">        Generate a ndarray of residue ids and their corresponding atom names such that the distance is the minimum between the initial residues provided.</span>
</span><span id="MMCIFInformation-966"><a href="#MMCIFInformation-966"><span class="linenos"> 966</span></a>
</span><span id="MMCIFInformation-967"><a href="#MMCIFInformation-967"><span class="linenos"> 967</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation-968"><a href="#MMCIFInformation-968"><span class="linenos"> 968</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation-969"><a href="#MMCIFInformation-969"><span class="linenos"> 969</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="MMCIFInformation-970"><a href="#MMCIFInformation-970"><span class="linenos"> 970</span></a><span class="sd">            Pairs structured ndarray with &quot;residue1&quot; and &quot;residue2&quot; columns.</span>
</span><span id="MMCIFInformation-971"><a href="#MMCIFInformation-971"><span class="linenos"> 971</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation-972"><a href="#MMCIFInformation-972"><span class="linenos"> 972</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation-973"><a href="#MMCIFInformation-973"><span class="linenos"> 973</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation-974"><a href="#MMCIFInformation-974"><span class="linenos"> 974</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="MMCIFInformation-975"><a href="#MMCIFInformation-975"><span class="linenos"> 975</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation-976"><a href="#MMCIFInformation-976"><span class="linenos"> 976</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation-977"><a href="#MMCIFInformation-977"><span class="linenos"> 977</span></a><span class="sd">            </span>
</span><span id="MMCIFInformation-978"><a href="#MMCIFInformation-978"><span class="linenos"> 978</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation-979"><a href="#MMCIFInformation-979"><span class="linenos"> 979</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation-980"><a href="#MMCIFInformation-980"><span class="linenos"> 980</span></a><span class="sd">        min_dist_pairs_atoms_arr : numpy.ndarray</span>
</span><span id="MMCIFInformation-981"><a href="#MMCIFInformation-981"><span class="linenos"> 981</span></a><span class="sd">            Structured ndarray that has residue indices, auth residue indices (corresponding to the protein numbering), and atomic names in the format {&#39;names&#39;: [&#39;residue1&#39;,&#39;residue2&#39;,&#39;auth_residue1&#39;,&#39;auth_residue2&#39;,&#39;atom_name1&#39;,&#39;atom_name2&#39;], &#39;formats&#39;: [int,int,int,int,&#39;&lt;U10&#39;,&#39;&lt;U10&#39;]}</span>
</span><span id="MMCIFInformation-982"><a href="#MMCIFInformation-982"><span class="linenos"> 982</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-983"><a href="#MMCIFInformation-983"><span class="linenos"> 983</span></a>        <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_shift_values</span><span class="p">(</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation-984"><a href="#MMCIFInformation-984"><span class="linenos"> 984</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chain_specific_structure</span><span class="p">(</span><span class="n">ca_only</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">chain1</span><span class="o">=</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="o">=</span><span class="n">chain2</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation-985"><a href="#MMCIFInformation-985"><span class="linenos"> 985</span></a>        <span class="n">min_dist_pairs_atoms</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation-986"><a href="#MMCIFInformation-986"><span class="linenos"> 986</span></a>        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">:</span>
</span><span id="MMCIFInformation-987"><a href="#MMCIFInformation-987"><span class="linenos"> 987</span></a>            <span class="c1"># Obtain structure information for chains 1 and 2</span>
</span><span id="MMCIFInformation-988"><a href="#MMCIFInformation-988"><span class="linenos"> 988</span></a>            <span class="n">chain1_res1_structure</span> <span class="o">=</span> <span class="n">chain1_structure</span><span class="p">[</span><span class="n">chain1_structure</span><span class="o">.</span><span class="n">res_id</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]]</span>
</span><span id="MMCIFInformation-989"><a href="#MMCIFInformation-989"><span class="linenos"> 989</span></a>            <span class="n">chain2_res2_structure</span> <span class="o">=</span> <span class="n">chain2_structure</span><span class="p">[</span><span class="n">chain2_structure</span><span class="o">.</span><span class="n">res_id</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]]</span>
</span><span id="MMCIFInformation-990"><a href="#MMCIFInformation-990"><span class="linenos"> 990</span></a>            
</span><span id="MMCIFInformation-991"><a href="#MMCIFInformation-991"><span class="linenos"> 991</span></a>            <span class="c1"># Calculate a distance matrix and find the indices of the minimal value in the matrix</span>
</span><span id="MMCIFInformation-992"><a href="#MMCIFInformation-992"><span class="linenos"> 992</span></a>            <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">cdist</span><span class="p">(</span><span class="n">chain1_res1_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="n">chain2_res2_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>
</span><span id="MMCIFInformation-993"><a href="#MMCIFInformation-993"><span class="linenos"> 993</span></a>            
</span><span id="MMCIFInformation-994"><a href="#MMCIFInformation-994"><span class="linenos"> 994</span></a>            <span class="n">ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unravel_index</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">dist_matrix</span><span class="p">),</span> <span class="n">dist_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</span><span id="MMCIFInformation-995"><a href="#MMCIFInformation-995"><span class="linenos"> 995</span></a>            <span class="c1"># Use the indices to access the atom in the atom array and get the correct atom name.</span>
</span><span id="MMCIFInformation-996"><a href="#MMCIFInformation-996"><span class="linenos"> 996</span></a>            <span class="c1"># Generate the auth ids of the residues in the pairs ndarray</span>
</span><span id="MMCIFInformation-997"><a href="#MMCIFInformation-997"><span class="linenos"> 997</span></a>            <span class="n">auth_res_id1</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">shift1</span>
</span><span id="MMCIFInformation-998"><a href="#MMCIFInformation-998"><span class="linenos"> 998</span></a>            <span class="n">auth_res_id2</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">shift2</span>
</span><span id="MMCIFInformation-999"><a href="#MMCIFInformation-999"><span class="linenos"> 999</span></a>            <span class="n">min_dist_pairs_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">],</span> <span class="n">auth_res_id1</span><span class="p">,</span> <span class="n">auth_res_id2</span><span class="p">,</span> <span class="n">chain1_res1_structure</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">atom_name</span><span class="p">,</span> <span class="n">chain2_res2_structure</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">atom_name</span><span class="p">))</span>
</span><span id="MMCIFInformation-1000"><a href="#MMCIFInformation-1000"><span class="linenos">1000</span></a>        <span class="n">min_dist_pairs_atoms_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">min_dist_pairs_atoms</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span><span class="s1">&#39;residue2&#39;</span><span class="p">,</span><span class="s1">&#39;auth_residue1&#39;</span><span class="p">,</span><span class="s1">&#39;auth_residue2&#39;</span><span class="p">,</span><span class="s1">&#39;atom_name1&#39;</span><span class="p">,</span><span class="s1">&#39;atom_name2&#39;</span><span class="p">],</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="s1">&#39;&lt;U10&#39;</span><span class="p">,</span><span class="s1">&#39;&lt;U10&#39;</span><span class="p">]})</span>
</span><span id="MMCIFInformation-1001"><a href="#MMCIFInformation-1001"><span class="linenos">1001</span></a>        <span class="k">return</span> <span class="n">min_dist_pairs_atoms_arr</span>
</span><span id="MMCIFInformation-1002"><a href="#MMCIFInformation-1002"><span class="linenos">1002</span></a>
</span><span id="MMCIFInformation-1003"><a href="#MMCIFInformation-1003"><span class="linenos">1003</span></a>    <span class="k">def</span> <span class="nf">get_contacts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_contacts</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
</span><span id="MMCIFInformation-1004"><a href="#MMCIFInformation-1004"><span class="linenos">1004</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-1005"><a href="#MMCIFInformation-1005"><span class="linenos">1005</span></a><span class="sd">        Get contacts from the structure attribute where the distance between two residues is less than the threshold.</span>
</span><span id="MMCIFInformation-1006"><a href="#MMCIFInformation-1006"><span class="linenos">1006</span></a>
</span><span id="MMCIFInformation-1007"><a href="#MMCIFInformation-1007"><span class="linenos">1007</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation-1008"><a href="#MMCIFInformation-1008"><span class="linenos">1008</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation-1009"><a href="#MMCIFInformation-1009"><span class="linenos">1009</span></a><span class="sd">        ca_only : bool</span>
</span><span id="MMCIFInformation-1010"><a href="#MMCIFInformation-1010"><span class="linenos">1010</span></a><span class="sd">            If true, only consider alpha-carbon to alpha-carbon distances. </span>
</span><span id="MMCIFInformation-1011"><a href="#MMCIFInformation-1011"><span class="linenos">1011</span></a><span class="sd">        threshold : float</span>
</span><span id="MMCIFInformation-1012"><a href="#MMCIFInformation-1012"><span class="linenos">1012</span></a><span class="sd">            Maximum distance to consider between two atoms.</span>
</span><span id="MMCIFInformation-1013"><a href="#MMCIFInformation-1013"><span class="linenos">1013</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation-1014"><a href="#MMCIFInformation-1014"><span class="linenos">1014</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation-1015"><a href="#MMCIFInformation-1015"><span class="linenos">1015</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation-1016"><a href="#MMCIFInformation-1016"><span class="linenos">1016</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="MMCIFInformation-1017"><a href="#MMCIFInformation-1017"><span class="linenos">1017</span></a><span class="sd">        auth_contacts : bool</span>
</span><span id="MMCIFInformation-1018"><a href="#MMCIFInformation-1018"><span class="linenos">1018</span></a><span class="sd">            True if you want alt_ids for residues indices, False if cif residue indexing is needed.</span>
</span><span id="MMCIFInformation-1019"><a href="#MMCIFInformation-1019"><span class="linenos">1019</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation-1020"><a href="#MMCIFInformation-1020"><span class="linenos">1020</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation-1021"><a href="#MMCIFInformation-1021"><span class="linenos">1021</span></a>
</span><span id="MMCIFInformation-1022"><a href="#MMCIFInformation-1022"><span class="linenos">1022</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation-1023"><a href="#MMCIFInformation-1023"><span class="linenos">1023</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation-1024"><a href="#MMCIFInformation-1024"><span class="linenos">1024</span></a><span class="sd">        contacts_set : set of tuple of ints</span>
</span><span id="MMCIFInformation-1025"><a href="#MMCIFInformation-1025"><span class="linenos">1025</span></a><span class="sd">            Set of contacts, tuples with &quot;residue1&quot; and &quot;residue2&quot; from the structure that are within the distance threshold.</span>
</span><span id="MMCIFInformation-1026"><a href="#MMCIFInformation-1026"><span class="linenos">1026</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation-1027"><a href="#MMCIFInformation-1027"><span class="linenos">1027</span></a>        
</span><span id="MMCIFInformation-1028"><a href="#MMCIFInformation-1028"><span class="linenos">1028</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">,</span> <span class="n">dist_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_dist_matrix</span><span class="p">(</span><span class="n">ca_only</span><span class="p">,</span> <span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation-1029"><a href="#MMCIFInformation-1029"><span class="linenos">1029</span></a>        <span class="n">thresh_ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">dist_matrix</span> <span class="o">&lt;=</span> <span class="n">threshold</span><span class="p">)</span>
</span><span id="MMCIFInformation-1030"><a href="#MMCIFInformation-1030"><span class="linenos">1030</span></a>        <span class="n">contacts_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
</span><span id="MMCIFInformation-1031"><a href="#MMCIFInformation-1031"><span class="linenos">1031</span></a>        <span class="k">for</span> <span class="n">indices</span> <span class="ow">in</span> <span class="n">thresh_ind</span><span class="p">:</span>
</span><span id="MMCIFInformation-1032"><a href="#MMCIFInformation-1032"><span class="linenos">1032</span></a>            <span class="n">chain1_atom</span> <span class="o">=</span> <span class="n">chain1_structure</span><span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
</span><span id="MMCIFInformation-1033"><a href="#MMCIFInformation-1033"><span class="linenos">1033</span></a>            <span class="n">chain2_atom</span> <span class="o">=</span> <span class="n">chain2_structure</span><span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
</span><span id="MMCIFInformation-1034"><a href="#MMCIFInformation-1034"><span class="linenos">1034</span></a>            <span class="n">res1</span> <span class="o">=</span> <span class="n">chain1_atom</span><span class="o">.</span><span class="n">res_id</span>
</span><span id="MMCIFInformation-1035"><a href="#MMCIFInformation-1035"><span class="linenos">1035</span></a>            <span class="n">res2</span> <span class="o">=</span> <span class="n">chain2_atom</span><span class="o">.</span><span class="n">res_id</span>
</span><span id="MMCIFInformation-1036"><a href="#MMCIFInformation-1036"><span class="linenos">1036</span></a>            <span class="k">if</span> <span class="ow">not</span><span class="p">(</span><span class="n">chain1</span><span class="o">==</span><span class="n">chain2</span> <span class="ow">and</span> <span class="n">res1</span> <span class="o">&gt;=</span> <span class="n">res2</span><span class="p">):</span>
</span><span id="MMCIFInformation-1037"><a href="#MMCIFInformation-1037"><span class="linenos">1037</span></a>                <span class="k">if</span> <span class="n">auth_contacts</span><span class="p">:</span>
</span><span id="MMCIFInformation-1038"><a href="#MMCIFInformation-1038"><span class="linenos">1038</span></a>                    <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_shift_values</span><span class="p">(</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation-1039"><a href="#MMCIFInformation-1039"><span class="linenos">1039</span></a>                    <span class="n">contacts_set</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">res1</span> <span class="o">+</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">res2</span> <span class="o">+</span> <span class="n">shift2</span><span class="p">))</span>
</span><span id="MMCIFInformation-1040"><a href="#MMCIFInformation-1040"><span class="linenos">1040</span></a>                <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation-1041"><a href="#MMCIFInformation-1041"><span class="linenos">1041</span></a>                    <span class="n">contacts_set</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">res1</span><span class="p">,</span> <span class="n">res2</span><span class="p">))</span>
</span><span id="MMCIFInformation-1042"><a href="#MMCIFInformation-1042"><span class="linenos">1042</span></a>        <span class="k">return</span> <span class="n">contacts_set</span>
</span></pre></div>


            <div class="docstring"><p>Information regarding a protein structure, obtained from a protein structure file.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>structure</strong> (biotite.structure.AtomArray):
Structure obtained from an RCSB entry with a provided pdbx/mmcif file with a specified model number.</li>
<li><strong>pdbx_file</strong> (biotite.io.pdbx.CIFFile):
mmCIF file that contains generic information and atomic information of the protein structure categorized into mmCIF blocks.</li>
<li><strong>model_num</strong> (int):
The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</li>
</ul>

<h6 id="attributes">Attributes</h6>

<ul>
<li><strong>self.full_sequences</strong> (dict of str, biotite.sequence.ProteinSequence):
The full protein sequences from the pdbx file used to generate the structure stored in a dictionary where auth_chain_id is the key and the ProteinSequence object is the value.</li>
<li><strong>self.non_missing_sequences</strong> (dict of str, biotite.sequence.ProteinSequence):
The protein sequences, without missing residues, compiled in the structure of the StructureInformation instance stored in a dictionary where chain_id is the key and the sequence string is the value.</li>
<li><strong>self.atom_data</strong> (numpy.ndarray, optional):
Entries in the format 'ATOM', residue index, chain ID, auth residue index, auth chain ID, model number</li>
<li><strong>self.het_atom_data</strong> (numpy.ndarray, optional):
Array of entries in the format 'HETATM', residue index, chain ID, auth residue index, auth chain ID, model number</li>
<li><strong>self.unique_chains</strong> (numpy.ndarray, optional):
Array of unique asym_id entries which corresponds to unique chain IDs.</li>
<li><strong>self.chain_auth_dict</strong> (dict of str, str, optional):
Uses chain id as a key and provides auth chain id as a value.</li>
<li><strong>self.auth_chain_dict</strong> (dict of str, str, optional):
Uses auth chain id as a key and provides original/label chain id as a value.</li>
<li><strong>self.res_auth_dict</strong> (dict of str, tuple of int, int or optional):
Uses chain id as a key and an array of residue index and auth residue index as a value.</li>
</ul>
</div>


                            <div id="MMCIFInformation.__init__" class="classattr">
                                        <input id="MMCIFInformation.__init__-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="name">MMCIFInformation</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">structure</span>,</span><span class="param">	<span class="n">pdbx_file</span><span class="p">:</span> <span class="n">biotite</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">pdbx</span><span class="o">.</span><span class="n">CIFFile</span>,</span><span class="param">	<span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span></span>)</span>

                <label class="view-source-button" for="MMCIFInformation.__init__-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation.__init__"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation.__init__-767"><a href="#MMCIFInformation.__init__-767"><span class="linenos">767</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structure</span><span class="p">,</span> <span class="n">pdbx_file</span><span class="p">:</span> <span class="n">pdbx</span><span class="o">.</span><span class="n">CIFFile</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
</span><span id="MMCIFInformation.__init__-768"><a href="#MMCIFInformation.__init__-768"><span class="linenos">768</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">structure</span>
</span><span id="MMCIFInformation.__init__-769"><a href="#MMCIFInformation.__init__-769"><span class="linenos">769</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">pdbx_file</span> <span class="o">=</span> <span class="n">pdbx_file</span>
</span><span id="MMCIFInformation.__init__-770"><a href="#MMCIFInformation.__init__-770"><span class="linenos">770</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">model_num</span> <span class="o">=</span> <span class="n">model_num</span>
</span><span id="MMCIFInformation.__init__-771"><a href="#MMCIFInformation.__init__-771"><span class="linenos">771</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">full_sequences</span> <span class="o">=</span> <span class="n">pdbx</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">pdbx_file</span><span class="p">)</span>
</span><span id="MMCIFInformation.__init__-772"><a href="#MMCIFInformation.__init__-772"><span class="linenos">772</span></a>        <span class="n">non_hetero_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="MMCIFInformation.__init__-773"><a href="#MMCIFInformation.__init__-773"><span class="linenos">773</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">chain</span><span class="p">):</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">chain</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">struc</span><span class="o">.</span><span class="n">get_chains</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">),</span> <span class="n">struc</span><span class="o">.</span><span class="n">to_sequence</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))}</span>
</span><span id="MMCIFInformation.__init__-774"><a href="#MMCIFInformation.__init__-774"><span class="linenos">774</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">generate_auth_info</span><span class="p">()</span>
</span></pre></div>


    

                            </div>
                            <div id="MMCIFInformation.structure" class="classattr">
                                <div class="attr variable">
            <span class="name">structure</span>

        
    </div>
    <a class="headerlink" href="#MMCIFInformation.structure"></a>
    
    

                            </div>
                            <div id="MMCIFInformation.pdbx_file" class="classattr">
                                <div class="attr variable">
            <span class="name">pdbx_file</span>

        
    </div>
    <a class="headerlink" href="#MMCIFInformation.pdbx_file"></a>
    
    

                            </div>
                            <div id="MMCIFInformation.model_num" class="classattr">
                                <div class="attr variable">
            <span class="name">model_num</span>

        
    </div>
    <a class="headerlink" href="#MMCIFInformation.model_num"></a>
    
    

                            </div>
                            <div id="MMCIFInformation.full_sequences" class="classattr">
                                <div class="attr variable">
            <span class="name">full_sequences</span>

        
    </div>
    <a class="headerlink" href="#MMCIFInformation.full_sequences"></a>
    
    

                            </div>
                            <div id="MMCIFInformation.non_missing_sequences" class="classattr">
                                <div class="attr variable">
            <span class="name">non_missing_sequences</span>

        
    </div>
    <a class="headerlink" href="#MMCIFInformation.non_missing_sequences"></a>
    
    

                            </div>
                            <div id="MMCIFInformation.generate_auth_info" class="classattr">
                                        <input id="MMCIFInformation.generate_auth_info-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">generate_auth_info</span><span class="signature pdoc-code condensed">(<span class="param"><span class="bp">self</span></span><span class="return-annotation">) -> <span class="kc">None</span>:</span></span>

                <label class="view-source-button" for="MMCIFInformation.generate_auth_info-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation.generate_auth_info"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation.generate_auth_info-776"><a href="#MMCIFInformation.generate_auth_info-776"><span class="linenos">776</span></a>    <span class="k">def</span> <span class="nf">generate_auth_info</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-777"><a href="#MMCIFInformation.generate_auth_info-777"><span class="linenos">777</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.generate_auth_info-778"><a href="#MMCIFInformation.generate_auth_info-778"><span class="linenos">778</span></a><span class="sd">        Ran as part of constructor function. Generates information needed to access auth information including auth_seq_id and auth_asym_id, which correspond to alternative chain ids and alternative residue indices.</span>
</span><span id="MMCIFInformation.generate_auth_info-779"><a href="#MMCIFInformation.generate_auth_info-779"><span class="linenos">779</span></a><span class="sd">        </span>
</span><span id="MMCIFInformation.generate_auth_info-780"><a href="#MMCIFInformation.generate_auth_info-780"><span class="linenos">780</span></a><span class="sd">        Note</span>
</span><span id="MMCIFInformation.generate_auth_info-781"><a href="#MMCIFInformation.generate_auth_info-781"><span class="linenos">781</span></a><span class="sd">        ----</span>
</span><span id="MMCIFInformation.generate_auth_info-782"><a href="#MMCIFInformation.generate_auth_info-782"><span class="linenos">782</span></a><span class="sd">        See attributes for details.</span>
</span><span id="MMCIFInformation.generate_auth_info-783"><a href="#MMCIFInformation.generate_auth_info-783"><span class="linenos">783</span></a><span class="sd">        </span>
</span><span id="MMCIFInformation.generate_auth_info-784"><a href="#MMCIFInformation.generate_auth_info-784"><span class="linenos">784</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation.generate_auth_info-785"><a href="#MMCIFInformation.generate_auth_info-785"><span class="linenos">785</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation.generate_auth_info-786"><a href="#MMCIFInformation.generate_auth_info-786"><span class="linenos">786</span></a><span class="sd">        None</span>
</span><span id="MMCIFInformation.generate_auth_info-787"><a href="#MMCIFInformation.generate_auth_info-787"><span class="linenos">787</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.generate_auth_info-788"><a href="#MMCIFInformation.generate_auth_info-788"><span class="linenos">788</span></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pdbx_file</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-789"><a href="#MMCIFInformation.generate_auth_info-789"><span class="linenos">789</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">first_block</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pdbx_file</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MMCIFInformation.generate_auth_info-790"><a href="#MMCIFInformation.generate_auth_info-790"><span class="linenos">790</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pdbx_file</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">first_block</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;atom_site&#39;</span><span class="p">)</span>
</span><span id="MMCIFInformation.generate_auth_info-791"><a href="#MMCIFInformation.generate_auth_info-791"><span class="linenos">791</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">chain_auth_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="MMCIFInformation.generate_auth_info-792"><a href="#MMCIFInformation.generate_auth_info-792"><span class="linenos">792</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="MMCIFInformation.generate_auth_info-793"><a href="#MMCIFInformation.generate_auth_info-793"><span class="linenos">793</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="MMCIFInformation.generate_auth_info-794"><a href="#MMCIFInformation.generate_auth_info-794"><span class="linenos">794</span></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-795"><a href="#MMCIFInformation.generate_auth_info-795"><span class="linenos">795</span></a>                <span class="n">group_pdbs</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation.generate_auth_info-796"><a href="#MMCIFInformation.generate_auth_info-796"><span class="linenos">796</span></a>                <span class="n">seq_ids</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation.generate_auth_info-797"><a href="#MMCIFInformation.generate_auth_info-797"><span class="linenos">797</span></a>                <span class="n">asym_ids</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation.generate_auth_info-798"><a href="#MMCIFInformation.generate_auth_info-798"><span class="linenos">798</span></a>                <span class="n">auth_seq_ids</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation.generate_auth_info-799"><a href="#MMCIFInformation.generate_auth_info-799"><span class="linenos">799</span></a>                <span class="n">auth_asym_ids</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation.generate_auth_info-800"><a href="#MMCIFInformation.generate_auth_info-800"><span class="linenos">800</span></a>                <span class="n">model_nums</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation.generate_auth_info-801"><a href="#MMCIFInformation.generate_auth_info-801"><span class="linenos">801</span></a>                <span class="k">for</span> <span class="n">col_name</span><span class="p">,</span> <span class="n">col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
</span><span id="MMCIFInformation.generate_auth_info-802"><a href="#MMCIFInformation.generate_auth_info-802"><span class="linenos">802</span></a>                    <span class="k">if</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;group_PDB&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-803"><a href="#MMCIFInformation.generate_auth_info-803"><span class="linenos">803</span></a>                        <span class="n">group_pdbs</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation.generate_auth_info-804"><a href="#MMCIFInformation.generate_auth_info-804"><span class="linenos">804</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;label_seq_id&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-805"><a href="#MMCIFInformation.generate_auth_info-805"><span class="linenos">805</span></a>                        <span class="n">seq_ids</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation.generate_auth_info-806"><a href="#MMCIFInformation.generate_auth_info-806"><span class="linenos">806</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;label_asym_id&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-807"><a href="#MMCIFInformation.generate_auth_info-807"><span class="linenos">807</span></a>                        <span class="n">asym_ids</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation.generate_auth_info-808"><a href="#MMCIFInformation.generate_auth_info-808"><span class="linenos">808</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;auth_seq_id&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-809"><a href="#MMCIFInformation.generate_auth_info-809"><span class="linenos">809</span></a>                        <span class="n">auth_seq_ids</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation.generate_auth_info-810"><a href="#MMCIFInformation.generate_auth_info-810"><span class="linenos">810</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;auth_asym_id&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-811"><a href="#MMCIFInformation.generate_auth_info-811"><span class="linenos">811</span></a>                        <span class="n">auth_asym_ids</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation.generate_auth_info-812"><a href="#MMCIFInformation.generate_auth_info-812"><span class="linenos">812</span></a>                    <span class="k">elif</span> <span class="n">col_name</span> <span class="o">==</span> <span class="s1">&#39;pdbx_PDB_model_num&#39;</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-813"><a href="#MMCIFInformation.generate_auth_info-813"><span class="linenos">813</span></a>                        <span class="n">model_nums</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">as_array</span><span class="p">()</span>
</span><span id="MMCIFInformation.generate_auth_info-814"><a href="#MMCIFInformation.generate_auth_info-814"><span class="linenos">814</span></a>
</span><span id="MMCIFInformation.generate_auth_info-815"><a href="#MMCIFInformation.generate_auth_info-815"><span class="linenos">815</span></a>                <span class="n">atom_site_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">((</span><span class="n">group_pdbs</span><span class="p">,</span> <span class="n">seq_ids</span><span class="p">,</span> <span class="n">asym_ids</span><span class="p">,</span> <span class="n">auth_seq_ids</span><span class="p">,</span> <span class="n">auth_asym_ids</span><span class="p">,</span> <span class="n">model_nums</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</span><span id="MMCIFInformation.generate_auth_info-816"><a href="#MMCIFInformation.generate_auth_info-816"><span class="linenos">816</span></a>                <span class="n">atom_site_data</span> <span class="o">=</span> <span class="n">atom_site_data</span><span class="p">[</span><span class="n">atom_site_data</span><span class="p">[:,</span><span class="mi">5</span><span class="p">]</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_num</span><span class="p">)]</span>
</span><span id="MMCIFInformation.generate_auth_info-817"><a href="#MMCIFInformation.generate_auth_info-817"><span class="linenos">817</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">atom_data</span> <span class="o">=</span> <span class="n">atom_site_data</span><span class="p">[</span><span class="n">atom_site_data</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ATOM&quot;</span><span class="p">]</span>
</span><span id="MMCIFInformation.generate_auth_info-818"><a href="#MMCIFInformation.generate_auth_info-818"><span class="linenos">818</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">het_atom_data</span> <span class="o">=</span> <span class="n">atom_site_data</span><span class="p">[</span><span class="n">atom_site_data</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;HETATM&quot;</span><span class="p">]</span>
</span><span id="MMCIFInformation.generate_auth_info-819"><a href="#MMCIFInformation.generate_auth_info-819"><span class="linenos">819</span></a>                <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">atom_data</span><span class="p">[:,</span><span class="mi">2</span><span class="p">])</span>
</span><span id="MMCIFInformation.generate_auth_info-820"><a href="#MMCIFInformation.generate_auth_info-820"><span class="linenos">820</span></a>                <span class="k">for</span> <span class="n">unique_chain</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-821"><a href="#MMCIFInformation.generate_auth_info-821"><span class="linenos">821</span></a>                    <span class="n">unique_entry</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">atom_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">atom_data</span><span class="p">[:,</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="n">unique_chain</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MMCIFInformation.generate_auth_info-822"><a href="#MMCIFInformation.generate_auth_info-822"><span class="linenos">822</span></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">chain_auth_dict</span><span class="p">[</span><span class="n">unique_entry</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> <span class="o">=</span> <span class="n">unique_entry</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
</span><span id="MMCIFInformation.generate_auth_info-823"><a href="#MMCIFInformation.generate_auth_info-823"><span class="linenos">823</span></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">unique_entry</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span> <span class="o">=</span> <span class="n">unique_entry</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
</span><span id="MMCIFInformation.generate_auth_info-824"><a href="#MMCIFInformation.generate_auth_info-824"><span class="linenos">824</span></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">unique_entry</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> <span class="o">=</span> <span class="n">unique_entry</span><span class="p">[[</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">]]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)</span>
</span><span id="MMCIFInformation.generate_auth_info-825"><a href="#MMCIFInformation.generate_auth_info-825"><span class="linenos">825</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation.generate_auth_info-826"><a href="#MMCIFInformation.generate_auth_info-826"><span class="linenos">826</span></a>            <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span> <span class="o">=</span> <span class="kc">None</span>
</span></pre></div>


            <div class="docstring"><p>Ran as part of constructor function. Generates information needed to access auth information including auth_seq_id and auth_asym_id, which correspond to alternative chain ids and alternative residue indices.</p>

<h6 id="note">Note</h6>

<p>See attributes for details.</p>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>None</strong></li>
</ul>
</div>


                            </div>
                            <div id="MMCIFInformation.get_full_sequence" class="classattr">
                                        <input id="MMCIFInformation.get_full_sequence-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_full_sequence</span><span class="signature pdoc-code condensed">(<span class="param"><span class="bp">self</span>, </span><span class="param"><span class="n">chain_id</span><span class="p">:</span> <span class="nb">str</span>, </span><span class="param"><span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></span><span class="return-annotation">) -> <span class="nb">str</span>:</span></span>

                <label class="view-source-button" for="MMCIFInformation.get_full_sequence-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation.get_full_sequence"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation.get_full_sequence-828"><a href="#MMCIFInformation.get_full_sequence-828"><span class="linenos">828</span></a>    <span class="k">def</span> <span class="nf">get_full_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_full_sequence-829"><a href="#MMCIFInformation.get_full_sequence-829"><span class="linenos">829</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_full_sequence-830"><a href="#MMCIFInformation.get_full_sequence-830"><span class="linenos">830</span></a><span class="sd">        Get full sequence, including missing residues, from the specified chain off of RCSB.</span>
</span><span id="MMCIFInformation.get_full_sequence-831"><a href="#MMCIFInformation.get_full_sequence-831"><span class="linenos">831</span></a>
</span><span id="MMCIFInformation.get_full_sequence-832"><a href="#MMCIFInformation.get_full_sequence-832"><span class="linenos">832</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation.get_full_sequence-833"><a href="#MMCIFInformation.get_full_sequence-833"><span class="linenos">833</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation.get_full_sequence-834"><a href="#MMCIFInformation.get_full_sequence-834"><span class="linenos">834</span></a><span class="sd">        chain_id : str</span>
</span><span id="MMCIFInformation.get_full_sequence-835"><a href="#MMCIFInformation.get_full_sequence-835"><span class="linenos">835</span></a><span class="sd">            Chain id supplied. The full sequence, including missing residues, of this chain will be returned.</span>
</span><span id="MMCIFInformation.get_full_sequence-836"><a href="#MMCIFInformation.get_full_sequence-836"><span class="linenos">836</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation.get_full_sequence-837"><a href="#MMCIFInformation.get_full_sequence-837"><span class="linenos">837</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation.get_full_sequence-838"><a href="#MMCIFInformation.get_full_sequence-838"><span class="linenos">838</span></a>
</span><span id="MMCIFInformation.get_full_sequence-839"><a href="#MMCIFInformation.get_full_sequence-839"><span class="linenos">839</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation.get_full_sequence-840"><a href="#MMCIFInformation.get_full_sequence-840"><span class="linenos">840</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation.get_full_sequence-841"><a href="#MMCIFInformation.get_full_sequence-841"><span class="linenos">841</span></a><span class="sd">        str</span>
</span><span id="MMCIFInformation.get_full_sequence-842"><a href="#MMCIFInformation.get_full_sequence-842"><span class="linenos">842</span></a><span class="sd">            The full sequence, including missing residues, of the chain specified.</span>
</span><span id="MMCIFInformation.get_full_sequence-843"><a href="#MMCIFInformation.get_full_sequence-843"><span class="linenos">843</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_full_sequence-844"><a href="#MMCIFInformation.get_full_sequence-844"><span class="linenos">844</span></a>        <span class="k">if</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_full_sequence-845"><a href="#MMCIFInformation.get_full_sequence-845"><span class="linenos">845</span></a>            <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">full_sequences</span><span class="p">[</span><span class="n">chain_id</span><span class="p">])</span>
</span><span id="MMCIFInformation.get_full_sequence-846"><a href="#MMCIFInformation.get_full_sequence-846"><span class="linenos">846</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_full_sequence-847"><a href="#MMCIFInformation.get_full_sequence-847"><span class="linenos">847</span></a>            <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">full_sequences</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chain_auth_dict</span><span class="p">[</span><span class="n">chain_id</span><span class="p">]])</span>
</span></pre></div>


            <div class="docstring"><p>Get full sequence, including missing residues, from the specified chain off of RCSB.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>chain_id</strong> (str):
Chain id supplied. The full sequence, including missing residues, of this chain will be returned.</li>
<li><strong>auth_chain_id_supplied</strong> (bool):
If True, the chain_id supplied is the auth chain id found on the RCSB website.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>str</strong>: The full sequence, including missing residues, of the chain specified.</li>
</ul>
</div>


                            </div>
                            <div id="MMCIFInformation.get_non_missing_sequence" class="classattr">
                                        <input id="MMCIFInformation.get_non_missing_sequence-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_non_missing_sequence</span><span class="signature pdoc-code condensed">(<span class="param"><span class="bp">self</span>, </span><span class="param"><span class="n">chain_id</span><span class="p">:</span> <span class="nb">str</span>, </span><span class="param"><span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></span><span class="return-annotation">) -> <span class="nb">str</span>:</span></span>

                <label class="view-source-button" for="MMCIFInformation.get_non_missing_sequence-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation.get_non_missing_sequence"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation.get_non_missing_sequence-849"><a href="#MMCIFInformation.get_non_missing_sequence-849"><span class="linenos">849</span></a>    <span class="k">def</span> <span class="nf">get_non_missing_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-850"><a href="#MMCIFInformation.get_non_missing_sequence-850"><span class="linenos">850</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-851"><a href="#MMCIFInformation.get_non_missing_sequence-851"><span class="linenos">851</span></a><span class="sd">        Get sequence, including only non-missing residues, from the specified chain off of RCSB.</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-852"><a href="#MMCIFInformation.get_non_missing_sequence-852"><span class="linenos">852</span></a>
</span><span id="MMCIFInformation.get_non_missing_sequence-853"><a href="#MMCIFInformation.get_non_missing_sequence-853"><span class="linenos">853</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-854"><a href="#MMCIFInformation.get_non_missing_sequence-854"><span class="linenos">854</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-855"><a href="#MMCIFInformation.get_non_missing_sequence-855"><span class="linenos">855</span></a><span class="sd">        chain_id : str</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-856"><a href="#MMCIFInformation.get_non_missing_sequence-856"><span class="linenos">856</span></a><span class="sd">            Chain id supplied. The full sequence, including only non-missing residues, of this chain will be returned.</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-857"><a href="#MMCIFInformation.get_non_missing_sequence-857"><span class="linenos">857</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-858"><a href="#MMCIFInformation.get_non_missing_sequence-858"><span class="linenos">858</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-859"><a href="#MMCIFInformation.get_non_missing_sequence-859"><span class="linenos">859</span></a>
</span><span id="MMCIFInformation.get_non_missing_sequence-860"><a href="#MMCIFInformation.get_non_missing_sequence-860"><span class="linenos">860</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-861"><a href="#MMCIFInformation.get_non_missing_sequence-861"><span class="linenos">861</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-862"><a href="#MMCIFInformation.get_non_missing_sequence-862"><span class="linenos">862</span></a><span class="sd">        str</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-863"><a href="#MMCIFInformation.get_non_missing_sequence-863"><span class="linenos">863</span></a><span class="sd">            The full sequence, including only non-missing residues, of the chain specified.</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-864"><a href="#MMCIFInformation.get_non_missing_sequence-864"><span class="linenos">864</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-865"><a href="#MMCIFInformation.get_non_missing_sequence-865"><span class="linenos">865</span></a>        <span class="k">if</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-866"><a href="#MMCIFInformation.get_non_missing_sequence-866"><span class="linenos">866</span></a>            <span class="n">original_chain_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain_id</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-867"><a href="#MMCIFInformation.get_non_missing_sequence-867"><span class="linenos">867</span></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span><span class="p">[</span><span class="n">original_chain_id</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-868"><a href="#MMCIFInformation.get_non_missing_sequence-868"><span class="linenos">868</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_non_missing_sequence-869"><a href="#MMCIFInformation.get_non_missing_sequence-869"><span class="linenos">869</span></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span><span class="p">[</span><span class="n">chain_id</span><span class="p">]</span>
</span></pre></div>


            <div class="docstring"><p>Get sequence, including only non-missing residues, from the specified chain off of RCSB.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>chain_id</strong> (str):
Chain id supplied. The full sequence, including only non-missing residues, of this chain will be returned.</li>
<li><strong>auth_chain_id_supplied</strong> (bool):
If True, the chain_id supplied is the auth chain id found on the RCSB website.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>str</strong>: The full sequence, including only non-missing residues, of the chain specified.</li>
</ul>
</div>


                            </div>
                            <div id="MMCIFInformation.get_chain_specific_structure" class="classattr">
                                        <input id="MMCIFInformation.get_chain_specific_structure-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_chain_specific_structure</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span>,</span><span class="param">	<span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span>,</span><span class="param">	<span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></span><span class="return-annotation">) -> <span class="nb">tuple</span>:</span></span>

                <label class="view-source-button" for="MMCIFInformation.get_chain_specific_structure-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation.get_chain_specific_structure"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation.get_chain_specific_structure-871"><a href="#MMCIFInformation.get_chain_specific_structure-871"><span class="linenos">871</span></a>    <span class="k">def</span> <span class="nf">get_chain_specific_structure</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-872"><a href="#MMCIFInformation.get_chain_specific_structure-872"><span class="linenos">872</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-873"><a href="#MMCIFInformation.get_chain_specific_structure-873"><span class="linenos">873</span></a><span class="sd">        Subsets structure attribute to select for chain specific portions of the structure.</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-874"><a href="#MMCIFInformation.get_chain_specific_structure-874"><span class="linenos">874</span></a>
</span><span id="MMCIFInformation.get_chain_specific_structure-875"><a href="#MMCIFInformation.get_chain_specific_structure-875"><span class="linenos">875</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-876"><a href="#MMCIFInformation.get_chain_specific_structure-876"><span class="linenos">876</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-877"><a href="#MMCIFInformation.get_chain_specific_structure-877"><span class="linenos">877</span></a><span class="sd">        ca_only : bool</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-878"><a href="#MMCIFInformation.get_chain_specific_structure-878"><span class="linenos">878</span></a><span class="sd">            If true, the structure will also be subsetted for atom entries where the atom_name annotation is &quot;CA&quot; (referring to alpha-carbons)</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-879"><a href="#MMCIFInformation.get_chain_specific_structure-879"><span class="linenos">879</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-880"><a href="#MMCIFInformation.get_chain_specific_structure-880"><span class="linenos">880</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-881"><a href="#MMCIFInformation.get_chain_specific_structure-881"><span class="linenos">881</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-882"><a href="#MMCIFInformation.get_chain_specific_structure-882"><span class="linenos">882</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-883"><a href="#MMCIFInformation.get_chain_specific_structure-883"><span class="linenos">883</span></a><span class="sd">        remove_hetero : bool, default=True</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-884"><a href="#MMCIFInformation.get_chain_specific_structure-884"><span class="linenos">884</span></a><span class="sd">            If true, the structure will also be subsetted for atom entries where the hetero annotation is False, thus removing heteroatoms.</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-885"><a href="#MMCIFInformation.get_chain_specific_structure-885"><span class="linenos">885</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-886"><a href="#MMCIFInformation.get_chain_specific_structure-886"><span class="linenos">886</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-887"><a href="#MMCIFInformation.get_chain_specific_structure-887"><span class="linenos">887</span></a>
</span><span id="MMCIFInformation.get_chain_specific_structure-888"><a href="#MMCIFInformation.get_chain_specific_structure-888"><span class="linenos">888</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-889"><a href="#MMCIFInformation.get_chain_specific_structure-889"><span class="linenos">889</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-890"><a href="#MMCIFInformation.get_chain_specific_structure-890"><span class="linenos">890</span></a><span class="sd">        tuple of biotite.structure.AtomArray, biotite.structure.AtomArray</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-891"><a href="#MMCIFInformation.get_chain_specific_structure-891"><span class="linenos">891</span></a><span class="sd">            Two AtomArrays that refer to atoms in the first chain and second chain, respectively without accounting for the presence of heteroatoms if `remove_hetero` is True.</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-892"><a href="#MMCIFInformation.get_chain_specific_structure-892"><span class="linenos">892</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-893"><a href="#MMCIFInformation.get_chain_specific_structure-893"><span class="linenos">893</span></a>        <span class="k">if</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-894"><a href="#MMCIFInformation.get_chain_specific_structure-894"><span class="linenos">894</span></a>            <span class="n">chain1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain1</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-895"><a href="#MMCIFInformation.get_chain_specific_structure-895"><span class="linenos">895</span></a>            <span class="n">chain2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain2</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-896"><a href="#MMCIFInformation.get_chain_specific_structure-896"><span class="linenos">896</span></a>
</span><span id="MMCIFInformation.get_chain_specific_structure-897"><a href="#MMCIFInformation.get_chain_specific_structure-897"><span class="linenos">897</span></a>        <span class="n">selected_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-898"><a href="#MMCIFInformation.get_chain_specific_structure-898"><span class="linenos">898</span></a>        <span class="k">if</span> <span class="n">remove_hetero</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-899"><a href="#MMCIFInformation.get_chain_specific_structure-899"><span class="linenos">899</span></a>            <span class="c1"># Remove hetero atoms via hetero column of structure ndarray</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-900"><a href="#MMCIFInformation.get_chain_specific_structure-900"><span class="linenos">900</span></a>            <span class="n">selected_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-901"><a href="#MMCIFInformation.get_chain_specific_structure-901"><span class="linenos">901</span></a>        <span class="k">if</span> <span class="n">ca_only</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-902"><a href="#MMCIFInformation.get_chain_specific_structure-902"><span class="linenos">902</span></a>            <span class="c1"># Consider selection of alpha-carbon atoms only</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-903"><a href="#MMCIFInformation.get_chain_specific_structure-903"><span class="linenos">903</span></a>            <span class="n">selected_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">atom_name</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-904"><a href="#MMCIFInformation.get_chain_specific_structure-904"><span class="linenos">904</span></a>        <span class="n">chain1_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain1</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-905"><a href="#MMCIFInformation.get_chain_specific_structure-905"><span class="linenos">905</span></a>        <span class="n">chain2_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain2</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_chain_specific_structure-906"><a href="#MMCIFInformation.get_chain_specific_structure-906"><span class="linenos">906</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Subsets structure attribute to select for chain specific portions of the structure.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>ca_only</strong> (bool):
If true, the structure will also be subsetted for atom entries where the atom_name annotation is "CA" (referring to alpha-carbons)</li>
<li><strong>chain1</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
<li><strong>chain2</strong> (str):
Chain id corresponding to the second column of residues in the structure.</li>
<li><strong>remove_hetero</strong> (bool, default=True):
If true, the structure will also be subsetted for atom entries where the hetero annotation is False, thus removing heteroatoms.</li>
<li><strong>auth_chain_id_supplied</strong> (bool):
If True, the chain_id supplied is the auth chain id found on the RCSB website.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>tuple of biotite.structure.AtomArray, biotite.structure.AtomArray</strong>: Two AtomArrays that refer to atoms in the first chain and second chain, respectively without accounting for the presence of heteroatoms if <code>remove_hetero</code> is True.</li>
</ul>
</div>


                            </div>
                            <div id="MMCIFInformation.generate_dist_matrix" class="classattr">
                                        <input id="MMCIFInformation.generate_dist_matrix-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">generate_dist_matrix</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span>,</span><span class="param">	<span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></span><span class="return-annotation">):</span></span>

                <label class="view-source-button" for="MMCIFInformation.generate_dist_matrix-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation.generate_dist_matrix"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation.generate_dist_matrix-908"><a href="#MMCIFInformation.generate_dist_matrix-908"><span class="linenos">908</span></a>    <span class="k">def</span> <span class="nf">generate_dist_matrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="MMCIFInformation.generate_dist_matrix-909"><a href="#MMCIFInformation.generate_dist_matrix-909"><span class="linenos">909</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.generate_dist_matrix-910"><a href="#MMCIFInformation.generate_dist_matrix-910"><span class="linenos">910</span></a><span class="sd">        Generates distance matrix between two chains in the structure attribute.</span>
</span><span id="MMCIFInformation.generate_dist_matrix-911"><a href="#MMCIFInformation.generate_dist_matrix-911"><span class="linenos">911</span></a>
</span><span id="MMCIFInformation.generate_dist_matrix-912"><a href="#MMCIFInformation.generate_dist_matrix-912"><span class="linenos">912</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation.generate_dist_matrix-913"><a href="#MMCIFInformation.generate_dist_matrix-913"><span class="linenos">913</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation.generate_dist_matrix-914"><a href="#MMCIFInformation.generate_dist_matrix-914"><span class="linenos">914</span></a><span class="sd">        ca_only : bool</span>
</span><span id="MMCIFInformation.generate_dist_matrix-915"><a href="#MMCIFInformation.generate_dist_matrix-915"><span class="linenos">915</span></a><span class="sd">            If True, only atoms that have the name &quot;CA&quot; are selected in the chains the distance matrix is calculated between.</span>
</span><span id="MMCIFInformation.generate_dist_matrix-916"><a href="#MMCIFInformation.generate_dist_matrix-916"><span class="linenos">916</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation.generate_dist_matrix-917"><a href="#MMCIFInformation.generate_dist_matrix-917"><span class="linenos">917</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation.generate_dist_matrix-918"><a href="#MMCIFInformation.generate_dist_matrix-918"><span class="linenos">918</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation.generate_dist_matrix-919"><a href="#MMCIFInformation.generate_dist_matrix-919"><span class="linenos">919</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation.generate_dist_matrix-920"><a href="#MMCIFInformation.generate_dist_matrix-920"><span class="linenos">920</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation.generate_dist_matrix-921"><a href="#MMCIFInformation.generate_dist_matrix-921"><span class="linenos">921</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation.generate_dist_matrix-922"><a href="#MMCIFInformation.generate_dist_matrix-922"><span class="linenos">922</span></a>
</span><span id="MMCIFInformation.generate_dist_matrix-923"><a href="#MMCIFInformation.generate_dist_matrix-923"><span class="linenos">923</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation.generate_dist_matrix-924"><a href="#MMCIFInformation.generate_dist_matrix-924"><span class="linenos">924</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation.generate_dist_matrix-925"><a href="#MMCIFInformation.generate_dist_matrix-925"><span class="linenos">925</span></a><span class="sd">        tuple of biotite.structure.AtomArray, biotite.structure.AtomArray, numpy.ndarray</span>
</span><span id="MMCIFInformation.generate_dist_matrix-926"><a href="#MMCIFInformation.generate_dist_matrix-926"><span class="linenos">926</span></a><span class="sd">            Tuple containing the chain 1 structure, the chain 2 structure, and the distance matrix of chain 1 and chain 2&#39;s pairwise distances.</span>
</span><span id="MMCIFInformation.generate_dist_matrix-927"><a href="#MMCIFInformation.generate_dist_matrix-927"><span class="linenos">927</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.generate_dist_matrix-928"><a href="#MMCIFInformation.generate_dist_matrix-928"><span class="linenos">928</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chain_specific_structure</span><span class="p">(</span><span class="n">ca_only</span><span class="p">,</span> <span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation.generate_dist_matrix-929"><a href="#MMCIFInformation.generate_dist_matrix-929"><span class="linenos">929</span></a>        <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">cdist</span><span class="p">(</span><span class="n">chain1_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>
</span><span id="MMCIFInformation.generate_dist_matrix-930"><a href="#MMCIFInformation.generate_dist_matrix-930"><span class="linenos">930</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">,</span> <span class="n">dist_matrix</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Generates distance matrix between two chains in the structure attribute.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>ca_only</strong> (bool):
If True, only atoms that have the name "CA" are selected in the chains the distance matrix is calculated between.</li>
<li><strong>chain1</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
<li><strong>chain2</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
<li><strong>auth_chain_id_supplied</strong> (bool):
If True, the chain_id supplied is the auth chain id found on the RCSB website.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>tuple of biotite.structure.AtomArray, biotite.structure.AtomArray, numpy.ndarray</strong>: Tuple containing the chain 1 structure, the chain 2 structure, and the distance matrix of chain 1 and chain 2's pairwise distances.</li>
</ul>
</div>


                            </div>
                            <div id="MMCIFInformation.get_shift_values" class="classattr">
                                        <input id="MMCIFInformation.get_shift_values-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_shift_values</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></span><span class="return-annotation">) -> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span>:</span></span>

                <label class="view-source-button" for="MMCIFInformation.get_shift_values-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation.get_shift_values"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation.get_shift_values-932"><a href="#MMCIFInformation.get_shift_values-932"><span class="linenos">932</span></a>    <span class="k">def</span> <span class="nf">get_shift_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
</span><span id="MMCIFInformation.get_shift_values-933"><a href="#MMCIFInformation.get_shift_values-933"><span class="linenos">933</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_shift_values-934"><a href="#MMCIFInformation.get_shift_values-934"><span class="linenos">934</span></a><span class="sd">        Get shift values needed for production of auth residue ids.</span>
</span><span id="MMCIFInformation.get_shift_values-935"><a href="#MMCIFInformation.get_shift_values-935"><span class="linenos">935</span></a><span class="sd">        </span>
</span><span id="MMCIFInformation.get_shift_values-936"><a href="#MMCIFInformation.get_shift_values-936"><span class="linenos">936</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation.get_shift_values-937"><a href="#MMCIFInformation.get_shift_values-937"><span class="linenos">937</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation.get_shift_values-938"><a href="#MMCIFInformation.get_shift_values-938"><span class="linenos">938</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation.get_shift_values-939"><a href="#MMCIFInformation.get_shift_values-939"><span class="linenos">939</span></a><span class="sd">            Name of the chain id present referring to the second column of residues.</span>
</span><span id="MMCIFInformation.get_shift_values-940"><a href="#MMCIFInformation.get_shift_values-940"><span class="linenos">940</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation.get_shift_values-941"><a href="#MMCIFInformation.get_shift_values-941"><span class="linenos">941</span></a><span class="sd">            Name of the chain id present referring to the second column of residues.</span>
</span><span id="MMCIFInformation.get_shift_values-942"><a href="#MMCIFInformation.get_shift_values-942"><span class="linenos">942</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation.get_shift_values-943"><a href="#MMCIFInformation.get_shift_values-943"><span class="linenos">943</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation.get_shift_values-944"><a href="#MMCIFInformation.get_shift_values-944"><span class="linenos">944</span></a>
</span><span id="MMCIFInformation.get_shift_values-945"><a href="#MMCIFInformation.get_shift_values-945"><span class="linenos">945</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation.get_shift_values-946"><a href="#MMCIFInformation.get_shift_values-946"><span class="linenos">946</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation.get_shift_values-947"><a href="#MMCIFInformation.get_shift_values-947"><span class="linenos">947</span></a><span class="sd">        (shift1, shift2) : tuple of int, int</span>
</span><span id="MMCIFInformation.get_shift_values-948"><a href="#MMCIFInformation.get_shift_values-948"><span class="linenos">948</span></a><span class="sd">            Tuple containing both shift values, the difference between the auth_res_id and res_id.</span>
</span><span id="MMCIFInformation.get_shift_values-949"><a href="#MMCIFInformation.get_shift_values-949"><span class="linenos">949</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_shift_values-950"><a href="#MMCIFInformation.get_shift_values-950"><span class="linenos">950</span></a>        <span class="k">if</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_shift_values-951"><a href="#MMCIFInformation.get_shift_values-951"><span class="linenos">951</span></a>            <span class="n">chain1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain1</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_shift_values-952"><a href="#MMCIFInformation.get_shift_values-952"><span class="linenos">952</span></a>            <span class="n">chain2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">auth_chain_dict</span><span class="p">[</span><span class="n">chain2</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_shift_values-953"><a href="#MMCIFInformation.get_shift_values-953"><span class="linenos">953</span></a>
</span><span id="MMCIFInformation.get_shift_values-954"><a href="#MMCIFInformation.get_shift_values-954"><span class="linenos">954</span></a>        <span class="n">shift1</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="MMCIFInformation.get_shift_values-955"><a href="#MMCIFInformation.get_shift_values-955"><span class="linenos">955</span></a>        <span class="n">shift2</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="MMCIFInformation.get_shift_values-956"><a href="#MMCIFInformation.get_shift_values-956"><span class="linenos">956</span></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">atom_site_category</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_shift_values-957"><a href="#MMCIFInformation.get_shift_values-957"><span class="linenos">957</span></a>            <span class="n">shift1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">chain1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">chain1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_shift_values-958"><a href="#MMCIFInformation.get_shift_values-958"><span class="linenos">958</span></a>            <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">chain2</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">res_auth_dict</span><span class="p">[</span><span class="n">chain2</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MMCIFInformation.get_shift_values-959"><a href="#MMCIFInformation.get_shift_values-959"><span class="linenos">959</span></a>            <span class="k">return</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span>
</span><span id="MMCIFInformation.get_shift_values-960"><a href="#MMCIFInformation.get_shift_values-960"><span class="linenos">960</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_shift_values-961"><a href="#MMCIFInformation.get_shift_values-961"><span class="linenos">961</span></a>            <span class="k">return</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span>
</span></pre></div>


            <div class="docstring"><p>Get shift values needed for production of auth residue ids.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>chain1</strong> (str):
Name of the chain id present referring to the second column of residues.</li>
<li><strong>chain2</strong> (str):
Name of the chain id present referring to the second column of residues.</li>
<li><strong>auth_chain_id_supplied</strong> (bool):
If True, the chain_id supplied is the auth chain id found on the RCSB website.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>(shift1, shift2)</strong> (tuple of int, int):
Tuple containing both shift values, the difference between the auth_res_id and res_id.</li>
</ul>
</div>


                            </div>
                            <div id="MMCIFInformation.get_min_dist_atom_info" class="classattr">
                                        <input id="MMCIFInformation.get_min_dist_atom_info-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_min_dist_atom_info</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">pairs</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>,</span><span class="param">	<span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="MMCIFInformation.get_min_dist_atom_info-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation.get_min_dist_atom_info"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation.get_min_dist_atom_info-963"><a href="#MMCIFInformation.get_min_dist_atom_info-963"><span class="linenos"> 963</span></a>    <span class="k">def</span> <span class="nf">get_min_dist_atom_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-964"><a href="#MMCIFInformation.get_min_dist_atom_info-964"><span class="linenos"> 964</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-965"><a href="#MMCIFInformation.get_min_dist_atom_info-965"><span class="linenos"> 965</span></a><span class="sd">        Generate a ndarray of residue ids and their corresponding atom names such that the distance is the minimum between the initial residues provided.</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-966"><a href="#MMCIFInformation.get_min_dist_atom_info-966"><span class="linenos"> 966</span></a>
</span><span id="MMCIFInformation.get_min_dist_atom_info-967"><a href="#MMCIFInformation.get_min_dist_atom_info-967"><span class="linenos"> 967</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-968"><a href="#MMCIFInformation.get_min_dist_atom_info-968"><span class="linenos"> 968</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-969"><a href="#MMCIFInformation.get_min_dist_atom_info-969"><span class="linenos"> 969</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-970"><a href="#MMCIFInformation.get_min_dist_atom_info-970"><span class="linenos"> 970</span></a><span class="sd">            Pairs structured ndarray with &quot;residue1&quot; and &quot;residue2&quot; columns.</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-971"><a href="#MMCIFInformation.get_min_dist_atom_info-971"><span class="linenos"> 971</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-972"><a href="#MMCIFInformation.get_min_dist_atom_info-972"><span class="linenos"> 972</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-973"><a href="#MMCIFInformation.get_min_dist_atom_info-973"><span class="linenos"> 973</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-974"><a href="#MMCIFInformation.get_min_dist_atom_info-974"><span class="linenos"> 974</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-975"><a href="#MMCIFInformation.get_min_dist_atom_info-975"><span class="linenos"> 975</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-976"><a href="#MMCIFInformation.get_min_dist_atom_info-976"><span class="linenos"> 976</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-977"><a href="#MMCIFInformation.get_min_dist_atom_info-977"><span class="linenos"> 977</span></a><span class="sd">            </span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-978"><a href="#MMCIFInformation.get_min_dist_atom_info-978"><span class="linenos"> 978</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-979"><a href="#MMCIFInformation.get_min_dist_atom_info-979"><span class="linenos"> 979</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-980"><a href="#MMCIFInformation.get_min_dist_atom_info-980"><span class="linenos"> 980</span></a><span class="sd">        min_dist_pairs_atoms_arr : numpy.ndarray</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-981"><a href="#MMCIFInformation.get_min_dist_atom_info-981"><span class="linenos"> 981</span></a><span class="sd">            Structured ndarray that has residue indices, auth residue indices (corresponding to the protein numbering), and atomic names in the format {&#39;names&#39;: [&#39;residue1&#39;,&#39;residue2&#39;,&#39;auth_residue1&#39;,&#39;auth_residue2&#39;,&#39;atom_name1&#39;,&#39;atom_name2&#39;], &#39;formats&#39;: [int,int,int,int,&#39;&lt;U10&#39;,&#39;&lt;U10&#39;]}</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-982"><a href="#MMCIFInformation.get_min_dist_atom_info-982"><span class="linenos"> 982</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-983"><a href="#MMCIFInformation.get_min_dist_atom_info-983"><span class="linenos"> 983</span></a>        <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_shift_values</span><span class="p">(</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-984"><a href="#MMCIFInformation.get_min_dist_atom_info-984"><span class="linenos"> 984</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chain_specific_structure</span><span class="p">(</span><span class="n">ca_only</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">chain1</span><span class="o">=</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="o">=</span><span class="n">chain2</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-985"><a href="#MMCIFInformation.get_min_dist_atom_info-985"><span class="linenos"> 985</span></a>        <span class="n">min_dist_pairs_atoms</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-986"><a href="#MMCIFInformation.get_min_dist_atom_info-986"><span class="linenos"> 986</span></a>        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-987"><a href="#MMCIFInformation.get_min_dist_atom_info-987"><span class="linenos"> 987</span></a>            <span class="c1"># Obtain structure information for chains 1 and 2</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-988"><a href="#MMCIFInformation.get_min_dist_atom_info-988"><span class="linenos"> 988</span></a>            <span class="n">chain1_res1_structure</span> <span class="o">=</span> <span class="n">chain1_structure</span><span class="p">[</span><span class="n">chain1_structure</span><span class="o">.</span><span class="n">res_id</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]]</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-989"><a href="#MMCIFInformation.get_min_dist_atom_info-989"><span class="linenos"> 989</span></a>            <span class="n">chain2_res2_structure</span> <span class="o">=</span> <span class="n">chain2_structure</span><span class="p">[</span><span class="n">chain2_structure</span><span class="o">.</span><span class="n">res_id</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]]</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-990"><a href="#MMCIFInformation.get_min_dist_atom_info-990"><span class="linenos"> 990</span></a>            
</span><span id="MMCIFInformation.get_min_dist_atom_info-991"><a href="#MMCIFInformation.get_min_dist_atom_info-991"><span class="linenos"> 991</span></a>            <span class="c1"># Calculate a distance matrix and find the indices of the minimal value in the matrix</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-992"><a href="#MMCIFInformation.get_min_dist_atom_info-992"><span class="linenos"> 992</span></a>            <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">cdist</span><span class="p">(</span><span class="n">chain1_res1_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="n">chain2_res2_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-993"><a href="#MMCIFInformation.get_min_dist_atom_info-993"><span class="linenos"> 993</span></a>            
</span><span id="MMCIFInformation.get_min_dist_atom_info-994"><a href="#MMCIFInformation.get_min_dist_atom_info-994"><span class="linenos"> 994</span></a>            <span class="n">ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unravel_index</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">dist_matrix</span><span class="p">),</span> <span class="n">dist_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-995"><a href="#MMCIFInformation.get_min_dist_atom_info-995"><span class="linenos"> 995</span></a>            <span class="c1"># Use the indices to access the atom in the atom array and get the correct atom name.</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-996"><a href="#MMCIFInformation.get_min_dist_atom_info-996"><span class="linenos"> 996</span></a>            <span class="c1"># Generate the auth ids of the residues in the pairs ndarray</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-997"><a href="#MMCIFInformation.get_min_dist_atom_info-997"><span class="linenos"> 997</span></a>            <span class="n">auth_res_id1</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">shift1</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-998"><a href="#MMCIFInformation.get_min_dist_atom_info-998"><span class="linenos"> 998</span></a>            <span class="n">auth_res_id2</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">shift2</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-999"><a href="#MMCIFInformation.get_min_dist_atom_info-999"><span class="linenos"> 999</span></a>            <span class="n">min_dist_pairs_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">],</span> <span class="n">auth_res_id1</span><span class="p">,</span> <span class="n">auth_res_id2</span><span class="p">,</span> <span class="n">chain1_res1_structure</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">atom_name</span><span class="p">,</span> <span class="n">chain2_res2_structure</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">atom_name</span><span class="p">))</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-1000"><a href="#MMCIFInformation.get_min_dist_atom_info-1000"><span class="linenos">1000</span></a>        <span class="n">min_dist_pairs_atoms_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">min_dist_pairs_atoms</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span><span class="s1">&#39;residue2&#39;</span><span class="p">,</span><span class="s1">&#39;auth_residue1&#39;</span><span class="p">,</span><span class="s1">&#39;auth_residue2&#39;</span><span class="p">,</span><span class="s1">&#39;atom_name1&#39;</span><span class="p">,</span><span class="s1">&#39;atom_name2&#39;</span><span class="p">],</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="s1">&#39;&lt;U10&#39;</span><span class="p">,</span><span class="s1">&#39;&lt;U10&#39;</span><span class="p">]})</span>
</span><span id="MMCIFInformation.get_min_dist_atom_info-1001"><a href="#MMCIFInformation.get_min_dist_atom_info-1001"><span class="linenos">1001</span></a>        <span class="k">return</span> <span class="n">min_dist_pairs_atoms_arr</span>
</span></pre></div>


            <div class="docstring"><p>Generate a ndarray of residue ids and their corresponding atom names such that the distance is the minimum between the initial residues provided.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>pairs</strong> (numpy.ndarray):
Pairs structured ndarray with "residue1" and "residue2" columns.</li>
<li><strong>chain1</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
<li><strong>chain2</strong> (str):
Chain id corresponding to the second column of residues in the structure.</li>
<li><strong>auth_chain_id_supplied</strong> (bool):
If True, the chain_id supplied is the auth chain id found on the RCSB website.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>min_dist_pairs_atoms_arr</strong> (numpy.ndarray):
Structured ndarray that has residue indices, auth residue indices (corresponding to the protein numbering), and atomic names in the format {'names': ['residue1','residue2','auth_residue1','auth_residue2','atom_name1','atom_name2'], 'formats': [int,int,int,int,'<U10','<U10']}</li>
</ul>
</div>


                            </div>
                            <div id="MMCIFInformation.get_contacts" class="classattr">
                                        <input id="MMCIFInformation.get_contacts-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_contacts</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span>,</span><span class="param">	<span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span>,</span><span class="param">	<span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">auth_contacts</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>,</span><span class="param">	<span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></span><span class="return-annotation">) -> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="MMCIFInformation.get_contacts-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MMCIFInformation.get_contacts"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MMCIFInformation.get_contacts-1003"><a href="#MMCIFInformation.get_contacts-1003"><span class="linenos">1003</span></a>    <span class="k">def</span> <span class="nf">get_contacts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_contacts</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
</span><span id="MMCIFInformation.get_contacts-1004"><a href="#MMCIFInformation.get_contacts-1004"><span class="linenos">1004</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_contacts-1005"><a href="#MMCIFInformation.get_contacts-1005"><span class="linenos">1005</span></a><span class="sd">        Get contacts from the structure attribute where the distance between two residues is less than the threshold.</span>
</span><span id="MMCIFInformation.get_contacts-1006"><a href="#MMCIFInformation.get_contacts-1006"><span class="linenos">1006</span></a>
</span><span id="MMCIFInformation.get_contacts-1007"><a href="#MMCIFInformation.get_contacts-1007"><span class="linenos">1007</span></a><span class="sd">        Parameters</span>
</span><span id="MMCIFInformation.get_contacts-1008"><a href="#MMCIFInformation.get_contacts-1008"><span class="linenos">1008</span></a><span class="sd">        ----------</span>
</span><span id="MMCIFInformation.get_contacts-1009"><a href="#MMCIFInformation.get_contacts-1009"><span class="linenos">1009</span></a><span class="sd">        ca_only : bool</span>
</span><span id="MMCIFInformation.get_contacts-1010"><a href="#MMCIFInformation.get_contacts-1010"><span class="linenos">1010</span></a><span class="sd">            If true, only consider alpha-carbon to alpha-carbon distances. </span>
</span><span id="MMCIFInformation.get_contacts-1011"><a href="#MMCIFInformation.get_contacts-1011"><span class="linenos">1011</span></a><span class="sd">        threshold : float</span>
</span><span id="MMCIFInformation.get_contacts-1012"><a href="#MMCIFInformation.get_contacts-1012"><span class="linenos">1012</span></a><span class="sd">            Maximum distance to consider between two atoms.</span>
</span><span id="MMCIFInformation.get_contacts-1013"><a href="#MMCIFInformation.get_contacts-1013"><span class="linenos">1013</span></a><span class="sd">        chain1 : str</span>
</span><span id="MMCIFInformation.get_contacts-1014"><a href="#MMCIFInformation.get_contacts-1014"><span class="linenos">1014</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="MMCIFInformation.get_contacts-1015"><a href="#MMCIFInformation.get_contacts-1015"><span class="linenos">1015</span></a><span class="sd">        chain2 : str</span>
</span><span id="MMCIFInformation.get_contacts-1016"><a href="#MMCIFInformation.get_contacts-1016"><span class="linenos">1016</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="MMCIFInformation.get_contacts-1017"><a href="#MMCIFInformation.get_contacts-1017"><span class="linenos">1017</span></a><span class="sd">        auth_contacts : bool</span>
</span><span id="MMCIFInformation.get_contacts-1018"><a href="#MMCIFInformation.get_contacts-1018"><span class="linenos">1018</span></a><span class="sd">            True if you want alt_ids for residues indices, False if cif residue indexing is needed.</span>
</span><span id="MMCIFInformation.get_contacts-1019"><a href="#MMCIFInformation.get_contacts-1019"><span class="linenos">1019</span></a><span class="sd">        auth_chain_id_supplied : bool</span>
</span><span id="MMCIFInformation.get_contacts-1020"><a href="#MMCIFInformation.get_contacts-1020"><span class="linenos">1020</span></a><span class="sd">            If True, the chain_id supplied is the auth chain id found on the RCSB website.</span>
</span><span id="MMCIFInformation.get_contacts-1021"><a href="#MMCIFInformation.get_contacts-1021"><span class="linenos">1021</span></a>
</span><span id="MMCIFInformation.get_contacts-1022"><a href="#MMCIFInformation.get_contacts-1022"><span class="linenos">1022</span></a><span class="sd">        Returns</span>
</span><span id="MMCIFInformation.get_contacts-1023"><a href="#MMCIFInformation.get_contacts-1023"><span class="linenos">1023</span></a><span class="sd">        -------</span>
</span><span id="MMCIFInformation.get_contacts-1024"><a href="#MMCIFInformation.get_contacts-1024"><span class="linenos">1024</span></a><span class="sd">        contacts_set : set of tuple of ints</span>
</span><span id="MMCIFInformation.get_contacts-1025"><a href="#MMCIFInformation.get_contacts-1025"><span class="linenos">1025</span></a><span class="sd">            Set of contacts, tuples with &quot;residue1&quot; and &quot;residue2&quot; from the structure that are within the distance threshold.</span>
</span><span id="MMCIFInformation.get_contacts-1026"><a href="#MMCIFInformation.get_contacts-1026"><span class="linenos">1026</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MMCIFInformation.get_contacts-1027"><a href="#MMCIFInformation.get_contacts-1027"><span class="linenos">1027</span></a>        
</span><span id="MMCIFInformation.get_contacts-1028"><a href="#MMCIFInformation.get_contacts-1028"><span class="linenos">1028</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">,</span> <span class="n">dist_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_dist_matrix</span><span class="p">(</span><span class="n">ca_only</span><span class="p">,</span> <span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation.get_contacts-1029"><a href="#MMCIFInformation.get_contacts-1029"><span class="linenos">1029</span></a>        <span class="n">thresh_ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">dist_matrix</span> <span class="o">&lt;=</span> <span class="n">threshold</span><span class="p">)</span>
</span><span id="MMCIFInformation.get_contacts-1030"><a href="#MMCIFInformation.get_contacts-1030"><span class="linenos">1030</span></a>        <span class="n">contacts_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
</span><span id="MMCIFInformation.get_contacts-1031"><a href="#MMCIFInformation.get_contacts-1031"><span class="linenos">1031</span></a>        <span class="k">for</span> <span class="n">indices</span> <span class="ow">in</span> <span class="n">thresh_ind</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_contacts-1032"><a href="#MMCIFInformation.get_contacts-1032"><span class="linenos">1032</span></a>            <span class="n">chain1_atom</span> <span class="o">=</span> <span class="n">chain1_structure</span><span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
</span><span id="MMCIFInformation.get_contacts-1033"><a href="#MMCIFInformation.get_contacts-1033"><span class="linenos">1033</span></a>            <span class="n">chain2_atom</span> <span class="o">=</span> <span class="n">chain2_structure</span><span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
</span><span id="MMCIFInformation.get_contacts-1034"><a href="#MMCIFInformation.get_contacts-1034"><span class="linenos">1034</span></a>            <span class="n">res1</span> <span class="o">=</span> <span class="n">chain1_atom</span><span class="o">.</span><span class="n">res_id</span>
</span><span id="MMCIFInformation.get_contacts-1035"><a href="#MMCIFInformation.get_contacts-1035"><span class="linenos">1035</span></a>            <span class="n">res2</span> <span class="o">=</span> <span class="n">chain2_atom</span><span class="o">.</span><span class="n">res_id</span>
</span><span id="MMCIFInformation.get_contacts-1036"><a href="#MMCIFInformation.get_contacts-1036"><span class="linenos">1036</span></a>            <span class="k">if</span> <span class="ow">not</span><span class="p">(</span><span class="n">chain1</span><span class="o">==</span><span class="n">chain2</span> <span class="ow">and</span> <span class="n">res1</span> <span class="o">&gt;=</span> <span class="n">res2</span><span class="p">):</span>
</span><span id="MMCIFInformation.get_contacts-1037"><a href="#MMCIFInformation.get_contacts-1037"><span class="linenos">1037</span></a>                <span class="k">if</span> <span class="n">auth_contacts</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_contacts-1038"><a href="#MMCIFInformation.get_contacts-1038"><span class="linenos">1038</span></a>                    <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_shift_values</span><span class="p">(</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">auth_chain_id_supplied</span><span class="o">=</span><span class="n">auth_chain_id_supplied</span><span class="p">)</span>
</span><span id="MMCIFInformation.get_contacts-1039"><a href="#MMCIFInformation.get_contacts-1039"><span class="linenos">1039</span></a>                    <span class="n">contacts_set</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">res1</span> <span class="o">+</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">res2</span> <span class="o">+</span> <span class="n">shift2</span><span class="p">))</span>
</span><span id="MMCIFInformation.get_contacts-1040"><a href="#MMCIFInformation.get_contacts-1040"><span class="linenos">1040</span></a>                <span class="k">else</span><span class="p">:</span>
</span><span id="MMCIFInformation.get_contacts-1041"><a href="#MMCIFInformation.get_contacts-1041"><span class="linenos">1041</span></a>                    <span class="n">contacts_set</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">res1</span><span class="p">,</span> <span class="n">res2</span><span class="p">))</span>
</span><span id="MMCIFInformation.get_contacts-1042"><a href="#MMCIFInformation.get_contacts-1042"><span class="linenos">1042</span></a>        <span class="k">return</span> <span class="n">contacts_set</span>
</span></pre></div>


            <div class="docstring"><p>Get contacts from the structure attribute where the distance between two residues is less than the threshold.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>ca_only</strong> (bool):
If true, only consider alpha-carbon to alpha-carbon distances.</li>
<li><strong>threshold</strong> (float):
Maximum distance to consider between two atoms.</li>
<li><strong>chain1</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
<li><strong>chain2</strong> (str):
Chain id corresponding to the second column of residues in the structure.</li>
<li><strong>auth_contacts</strong> (bool):
True if you want alt_ids for residues indices, False if cif residue indexing is needed.</li>
<li><strong>auth_chain_id_supplied</strong> (bool):
If True, the chain_id supplied is the auth chain id found on the RCSB website.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>contacts_set</strong> (set of tuple of ints):
Set of contacts, tuples with "residue1" and "residue2" from the structure that are within the distance threshold.</li>
</ul>
</div>


                            </div>
                            <div class="inherited">
                                <h5>Inherited Members</h5>
                                <dl>
                                    <div><dt><a href="#StructureInformation">StructureInformation</a></dt>
                                <dd id="MMCIFInformation.fetch_pdb" class="function"><a href="#StructureInformation.fetch_pdb">fetch_pdb</a></dd>
                <dd id="MMCIFInformation.read_mmCIF_file" class="function"><a href="#StructureInformation.read_mmCIF_file">read_mmCIF_file</a></dd>
                <dd id="MMCIFInformation.read_pdb_file" class="function"><a href="#StructureInformation.read_pdb_file">read_pdb_file</a></dd>
                <dd id="MMCIFInformation.write_contacts_set" class="function"><a href="#StructureInformation.write_contacts_set">write_contacts_set</a></dd>

            </div>
                                </dl>
                            </div>
                </section>
                <section id="PDBInformation">
                            <input id="PDBInformation-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr class">
            
    <span class="def">class</span>
    <span class="name">PDBInformation</span><wbr>(<span class="base"><a href="#StructureInformation">dcatoolkit.StructureInformation</a></span>):

                <label class="view-source-button" for="PDBInformation-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#PDBInformation"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="PDBInformation-1044"><a href="#PDBInformation-1044"><span class="linenos">1044</span></a><span class="k">class</span> <span class="nc">PDBInformation</span><span class="p">(</span><span class="n">StructureInformation</span><span class="p">):</span>
</span><span id="PDBInformation-1045"><a href="#PDBInformation-1045"><span class="linenos">1045</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation-1046"><a href="#PDBInformation-1046"><span class="linenos">1046</span></a><span class="sd">    Information regarding a protein structure, obtained from a protein structure file.</span>
</span><span id="PDBInformation-1047"><a href="#PDBInformation-1047"><span class="linenos">1047</span></a>
</span><span id="PDBInformation-1048"><a href="#PDBInformation-1048"><span class="linenos">1048</span></a><span class="sd">    Parameters</span>
</span><span id="PDBInformation-1049"><a href="#PDBInformation-1049"><span class="linenos">1049</span></a><span class="sd">    ----------</span>
</span><span id="PDBInformation-1050"><a href="#PDBInformation-1050"><span class="linenos">1050</span></a><span class="sd">    structure : biotite.structure.AtomArray</span>
</span><span id="PDBInformation-1051"><a href="#PDBInformation-1051"><span class="linenos">1051</span></a><span class="sd">        Structure obtained from an RCSB entry with a provided pdbx/mmcif file with a specified model number.</span>
</span><span id="PDBInformation-1052"><a href="#PDBInformation-1052"><span class="linenos">1052</span></a><span class="sd">    pdb_file : biotite.io.pdb.PDBFile</span>
</span><span id="PDBInformation-1053"><a href="#PDBInformation-1053"><span class="linenos">1053</span></a><span class="sd">        mmCIF file that contains generic information and atomic information of the protein structure categorized into mmCIF blocks.</span>
</span><span id="PDBInformation-1054"><a href="#PDBInformation-1054"><span class="linenos">1054</span></a><span class="sd">    model_num : int</span>
</span><span id="PDBInformation-1055"><a href="#PDBInformation-1055"><span class="linenos">1055</span></a><span class="sd">        The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</span>
</span><span id="PDBInformation-1056"><a href="#PDBInformation-1056"><span class="linenos">1056</span></a>
</span><span id="PDBInformation-1057"><a href="#PDBInformation-1057"><span class="linenos">1057</span></a><span class="sd">    Attributes</span>
</span><span id="PDBInformation-1058"><a href="#PDBInformation-1058"><span class="linenos">1058</span></a><span class="sd">    ----------</span>
</span><span id="PDBInformation-1059"><a href="#PDBInformation-1059"><span class="linenos">1059</span></a><span class="sd">    self.non_missing_sequences : dict of str, biotite.sequence.ProteinSequence</span>
</span><span id="PDBInformation-1060"><a href="#PDBInformation-1060"><span class="linenos">1060</span></a><span class="sd">        The protein sequences, without missing residues, compiled in the structure of the StructureInformation instance stored in a dictionary where chain_id is the key and the sequence string is the value.</span>
</span><span id="PDBInformation-1061"><a href="#PDBInformation-1061"><span class="linenos">1061</span></a><span class="sd">    </span>
</span><span id="PDBInformation-1062"><a href="#PDBInformation-1062"><span class="linenos">1062</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="PDBInformation-1063"><a href="#PDBInformation-1063"><span class="linenos">1063</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structure</span><span class="p">,</span> <span class="n">pdb_file</span><span class="p">:</span> <span class="n">pdb</span><span class="o">.</span><span class="n">PDBFile</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
</span><span id="PDBInformation-1064"><a href="#PDBInformation-1064"><span class="linenos">1064</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">structure</span>
</span><span id="PDBInformation-1065"><a href="#PDBInformation-1065"><span class="linenos">1065</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">pdb_file</span> <span class="o">=</span> <span class="n">pdb_file</span>
</span><span id="PDBInformation-1066"><a href="#PDBInformation-1066"><span class="linenos">1066</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">model_num</span> <span class="o">=</span> <span class="n">model_num</span>
</span><span id="PDBInformation-1067"><a href="#PDBInformation-1067"><span class="linenos">1067</span></a>        <span class="n">non_hetero_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="PDBInformation-1068"><a href="#PDBInformation-1068"><span class="linenos">1068</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">chain</span><span class="p">):</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">chain</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">struc</span><span class="o">.</span><span class="n">get_chains</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">),</span> <span class="n">struc</span><span class="o">.</span><span class="n">to_sequence</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))}</span>
</span><span id="PDBInformation-1069"><a href="#PDBInformation-1069"><span class="linenos">1069</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span> <span class="o">=</span> <span class="n">struc</span><span class="o">.</span><span class="n">get_chains</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">)</span>
</span><span id="PDBInformation-1070"><a href="#PDBInformation-1070"><span class="linenos">1070</span></a>
</span><span id="PDBInformation-1071"><a href="#PDBInformation-1071"><span class="linenos">1071</span></a>    <span class="k">def</span> <span class="nf">get_non_missing_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
</span><span id="PDBInformation-1072"><a href="#PDBInformation-1072"><span class="linenos">1072</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation-1073"><a href="#PDBInformation-1073"><span class="linenos">1073</span></a><span class="sd">        Get sequence, including only non-missing residues, from the specified chain.</span>
</span><span id="PDBInformation-1074"><a href="#PDBInformation-1074"><span class="linenos">1074</span></a>
</span><span id="PDBInformation-1075"><a href="#PDBInformation-1075"><span class="linenos">1075</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation-1076"><a href="#PDBInformation-1076"><span class="linenos">1076</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation-1077"><a href="#PDBInformation-1077"><span class="linenos">1077</span></a><span class="sd">        chain_id : str</span>
</span><span id="PDBInformation-1078"><a href="#PDBInformation-1078"><span class="linenos">1078</span></a><span class="sd">            Chain id supplied. The full sequence, including only non-missing residues, of this chain will be returned.</span>
</span><span id="PDBInformation-1079"><a href="#PDBInformation-1079"><span class="linenos">1079</span></a>
</span><span id="PDBInformation-1080"><a href="#PDBInformation-1080"><span class="linenos">1080</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation-1081"><a href="#PDBInformation-1081"><span class="linenos">1081</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation-1082"><a href="#PDBInformation-1082"><span class="linenos">1082</span></a><span class="sd">        str</span>
</span><span id="PDBInformation-1083"><a href="#PDBInformation-1083"><span class="linenos">1083</span></a><span class="sd">            The full sequence, including only non-missing residues, of the chain specified.</span>
</span><span id="PDBInformation-1084"><a href="#PDBInformation-1084"><span class="linenos">1084</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation-1085"><a href="#PDBInformation-1085"><span class="linenos">1085</span></a>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span><span class="p">[</span><span class="n">chain_id</span><span class="p">]</span>
</span><span id="PDBInformation-1086"><a href="#PDBInformation-1086"><span class="linenos">1086</span></a>    
</span><span id="PDBInformation-1087"><a href="#PDBInformation-1087"><span class="linenos">1087</span></a>    <span class="k">def</span> <span class="nf">get_chain_specific_structure</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
</span><span id="PDBInformation-1088"><a href="#PDBInformation-1088"><span class="linenos">1088</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation-1089"><a href="#PDBInformation-1089"><span class="linenos">1089</span></a><span class="sd">        Subsets structure attribute to select for chain specific portions of the structure.</span>
</span><span id="PDBInformation-1090"><a href="#PDBInformation-1090"><span class="linenos">1090</span></a>
</span><span id="PDBInformation-1091"><a href="#PDBInformation-1091"><span class="linenos">1091</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation-1092"><a href="#PDBInformation-1092"><span class="linenos">1092</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation-1093"><a href="#PDBInformation-1093"><span class="linenos">1093</span></a><span class="sd">        ca_only : bool</span>
</span><span id="PDBInformation-1094"><a href="#PDBInformation-1094"><span class="linenos">1094</span></a><span class="sd">            If true, the structure will also be subsetted for atom entries where the atom_name annotation is &quot;CA&quot; (referring to alpha-carbons)</span>
</span><span id="PDBInformation-1095"><a href="#PDBInformation-1095"><span class="linenos">1095</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation-1096"><a href="#PDBInformation-1096"><span class="linenos">1096</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation-1097"><a href="#PDBInformation-1097"><span class="linenos">1097</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation-1098"><a href="#PDBInformation-1098"><span class="linenos">1098</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="PDBInformation-1099"><a href="#PDBInformation-1099"><span class="linenos">1099</span></a><span class="sd">        remove_hetero : bool, default=True</span>
</span><span id="PDBInformation-1100"><a href="#PDBInformation-1100"><span class="linenos">1100</span></a><span class="sd">            If true, the structure will also be subsetted for atom entries where the hetero annotation is False, thus removing heteroatoms.</span>
</span><span id="PDBInformation-1101"><a href="#PDBInformation-1101"><span class="linenos">1101</span></a>
</span><span id="PDBInformation-1102"><a href="#PDBInformation-1102"><span class="linenos">1102</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation-1103"><a href="#PDBInformation-1103"><span class="linenos">1103</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation-1104"><a href="#PDBInformation-1104"><span class="linenos">1104</span></a><span class="sd">        tuple of biotite.structure.AtomArray, biotite.structure.AtomArray</span>
</span><span id="PDBInformation-1105"><a href="#PDBInformation-1105"><span class="linenos">1105</span></a><span class="sd">            Two AtomArrays that refer to atoms in the first chain and second chain, respectively without accounting for the presence of heteroatoms if `remove_hetero` is True.</span>
</span><span id="PDBInformation-1106"><a href="#PDBInformation-1106"><span class="linenos">1106</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation-1107"><a href="#PDBInformation-1107"><span class="linenos">1107</span></a>
</span><span id="PDBInformation-1108"><a href="#PDBInformation-1108"><span class="linenos">1108</span></a>        <span class="n">selected_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span>
</span><span id="PDBInformation-1109"><a href="#PDBInformation-1109"><span class="linenos">1109</span></a>        <span class="k">if</span> <span class="n">remove_hetero</span><span class="p">:</span>
</span><span id="PDBInformation-1110"><a href="#PDBInformation-1110"><span class="linenos">1110</span></a>            <span class="c1"># Remove hetero atoms via hetero column of structure ndarray</span>
</span><span id="PDBInformation-1111"><a href="#PDBInformation-1111"><span class="linenos">1111</span></a>            <span class="n">selected_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="PDBInformation-1112"><a href="#PDBInformation-1112"><span class="linenos">1112</span></a>        <span class="k">if</span> <span class="n">ca_only</span><span class="p">:</span>
</span><span id="PDBInformation-1113"><a href="#PDBInformation-1113"><span class="linenos">1113</span></a>            <span class="c1"># Consider selection of alpha-carbon atoms only</span>
</span><span id="PDBInformation-1114"><a href="#PDBInformation-1114"><span class="linenos">1114</span></a>            <span class="n">selected_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">atom_name</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span><span class="p">]</span>
</span><span id="PDBInformation-1115"><a href="#PDBInformation-1115"><span class="linenos">1115</span></a>        <span class="n">chain1_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain1</span><span class="p">]</span>
</span><span id="PDBInformation-1116"><a href="#PDBInformation-1116"><span class="linenos">1116</span></a>        <span class="n">chain2_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain2</span><span class="p">]</span>
</span><span id="PDBInformation-1117"><a href="#PDBInformation-1117"><span class="linenos">1117</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">)</span>
</span><span id="PDBInformation-1118"><a href="#PDBInformation-1118"><span class="linenos">1118</span></a>    
</span><span id="PDBInformation-1119"><a href="#PDBInformation-1119"><span class="linenos">1119</span></a>    <span class="k">def</span> <span class="nf">generate_dist_matrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
</span><span id="PDBInformation-1120"><a href="#PDBInformation-1120"><span class="linenos">1120</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation-1121"><a href="#PDBInformation-1121"><span class="linenos">1121</span></a><span class="sd">        Generates distance matrix between two chains in the structure attribute.</span>
</span><span id="PDBInformation-1122"><a href="#PDBInformation-1122"><span class="linenos">1122</span></a>
</span><span id="PDBInformation-1123"><a href="#PDBInformation-1123"><span class="linenos">1123</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation-1124"><a href="#PDBInformation-1124"><span class="linenos">1124</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation-1125"><a href="#PDBInformation-1125"><span class="linenos">1125</span></a><span class="sd">        ca_only : bool</span>
</span><span id="PDBInformation-1126"><a href="#PDBInformation-1126"><span class="linenos">1126</span></a><span class="sd">            If True, only atoms that have the name &quot;CA&quot; are selected in the chains the distance matrix is calculated between.</span>
</span><span id="PDBInformation-1127"><a href="#PDBInformation-1127"><span class="linenos">1127</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation-1128"><a href="#PDBInformation-1128"><span class="linenos">1128</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation-1129"><a href="#PDBInformation-1129"><span class="linenos">1129</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation-1130"><a href="#PDBInformation-1130"><span class="linenos">1130</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation-1131"><a href="#PDBInformation-1131"><span class="linenos">1131</span></a>
</span><span id="PDBInformation-1132"><a href="#PDBInformation-1132"><span class="linenos">1132</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation-1133"><a href="#PDBInformation-1133"><span class="linenos">1133</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation-1134"><a href="#PDBInformation-1134"><span class="linenos">1134</span></a><span class="sd">        tuple of biotite.structure.AtomArray, biotite.structure.AtomArray, numpy.ndarray</span>
</span><span id="PDBInformation-1135"><a href="#PDBInformation-1135"><span class="linenos">1135</span></a><span class="sd">            Tuple containing the chain 1 structure, the chain 2 structure, and the distance matrix of chain 1 and chain 2&#39;s pairwise distances.</span>
</span><span id="PDBInformation-1136"><a href="#PDBInformation-1136"><span class="linenos">1136</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation-1137"><a href="#PDBInformation-1137"><span class="linenos">1137</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chain_specific_structure</span><span class="p">(</span><span class="n">ca_only</span><span class="p">,</span> <span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="PDBInformation-1138"><a href="#PDBInformation-1138"><span class="linenos">1138</span></a>        <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">cdist</span><span class="p">(</span><span class="n">chain1_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>
</span><span id="PDBInformation-1139"><a href="#PDBInformation-1139"><span class="linenos">1139</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">,</span> <span class="n">dist_matrix</span><span class="p">)</span>
</span><span id="PDBInformation-1140"><a href="#PDBInformation-1140"><span class="linenos">1140</span></a>
</span><span id="PDBInformation-1141"><a href="#PDBInformation-1141"><span class="linenos">1141</span></a>    <span class="k">def</span> <span class="nf">get_shift_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
</span><span id="PDBInformation-1142"><a href="#PDBInformation-1142"><span class="linenos">1142</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation-1143"><a href="#PDBInformation-1143"><span class="linenos">1143</span></a><span class="sd">        Get shift values needed for production of auth residue ids.</span>
</span><span id="PDBInformation-1144"><a href="#PDBInformation-1144"><span class="linenos">1144</span></a><span class="sd">        </span>
</span><span id="PDBInformation-1145"><a href="#PDBInformation-1145"><span class="linenos">1145</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation-1146"><a href="#PDBInformation-1146"><span class="linenos">1146</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation-1147"><a href="#PDBInformation-1147"><span class="linenos">1147</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation-1148"><a href="#PDBInformation-1148"><span class="linenos">1148</span></a><span class="sd">            Name of the chain id present in the struct_ref_seq block of cif files referring to the second column of residues.</span>
</span><span id="PDBInformation-1149"><a href="#PDBInformation-1149"><span class="linenos">1149</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation-1150"><a href="#PDBInformation-1150"><span class="linenos">1150</span></a><span class="sd">            Name of the chain id present in the struct_ref_seq block of cif files referring to the second column of residues.</span>
</span><span id="PDBInformation-1151"><a href="#PDBInformation-1151"><span class="linenos">1151</span></a>
</span><span id="PDBInformation-1152"><a href="#PDBInformation-1152"><span class="linenos">1152</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation-1153"><a href="#PDBInformation-1153"><span class="linenos">1153</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation-1154"><a href="#PDBInformation-1154"><span class="linenos">1154</span></a><span class="sd">        (shift1, shift2) : tuple of int, int</span>
</span><span id="PDBInformation-1155"><a href="#PDBInformation-1155"><span class="linenos">1155</span></a><span class="sd">            Tuple containing both shift values, the difference between the auth_res_id and res_id.</span>
</span><span id="PDBInformation-1156"><a href="#PDBInformation-1156"><span class="linenos">1156</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation-1157"><a href="#PDBInformation-1157"><span class="linenos">1157</span></a>        <span class="n">non_hetero_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="PDBInformation-1158"><a href="#PDBInformation-1158"><span class="linenos">1158</span></a>        <span class="n">shift1</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="PDBInformation-1159"><a href="#PDBInformation-1159"><span class="linenos">1159</span></a>        <span class="n">shift2</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="PDBInformation-1160"><a href="#PDBInformation-1160"><span class="linenos">1160</span></a>        <span class="k">if</span> <span class="n">chain1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span> <span class="ow">and</span> <span class="n">chain2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span><span class="p">:</span>
</span><span id="PDBInformation-1161"><a href="#PDBInformation-1161"><span class="linenos">1161</span></a>            <span class="n">shift1</span> <span class="o">=</span> <span class="n">non_hetero_structure</span><span class="p">[</span><span class="n">non_hetero_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">res_id</span> <span class="o">-</span> <span class="mi">1</span>
</span><span id="PDBInformation-1162"><a href="#PDBInformation-1162"><span class="linenos">1162</span></a>            <span class="n">shift2</span> <span class="o">=</span> <span class="n">non_hetero_structure</span><span class="p">[</span><span class="n">non_hetero_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain2</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">res_id</span> <span class="o">-</span> <span class="mi">1</span>
</span><span id="PDBInformation-1163"><a href="#PDBInformation-1163"><span class="linenos">1163</span></a>            <span class="k">return</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span>
</span><span id="PDBInformation-1164"><a href="#PDBInformation-1164"><span class="linenos">1164</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="PDBInformation-1165"><a href="#PDBInformation-1165"><span class="linenos">1165</span></a>            <span class="k">return</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span>
</span><span id="PDBInformation-1166"><a href="#PDBInformation-1166"><span class="linenos">1166</span></a>
</span><span id="PDBInformation-1167"><a href="#PDBInformation-1167"><span class="linenos">1167</span></a>    <span class="k">def</span> <span class="nf">get_min_dist_atom_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="PDBInformation-1168"><a href="#PDBInformation-1168"><span class="linenos">1168</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation-1169"><a href="#PDBInformation-1169"><span class="linenos">1169</span></a><span class="sd">        Generate a ndarray of residue ids and their corresponding atom names such that the distance is the minimum between the initial residues provided.</span>
</span><span id="PDBInformation-1170"><a href="#PDBInformation-1170"><span class="linenos">1170</span></a>
</span><span id="PDBInformation-1171"><a href="#PDBInformation-1171"><span class="linenos">1171</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation-1172"><a href="#PDBInformation-1172"><span class="linenos">1172</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation-1173"><a href="#PDBInformation-1173"><span class="linenos">1173</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="PDBInformation-1174"><a href="#PDBInformation-1174"><span class="linenos">1174</span></a><span class="sd">            Pairs structured ndarray with &quot;residue1&quot; and &quot;residue2&quot; columns.</span>
</span><span id="PDBInformation-1175"><a href="#PDBInformation-1175"><span class="linenos">1175</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation-1176"><a href="#PDBInformation-1176"><span class="linenos">1176</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation-1177"><a href="#PDBInformation-1177"><span class="linenos">1177</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation-1178"><a href="#PDBInformation-1178"><span class="linenos">1178</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="PDBInformation-1179"><a href="#PDBInformation-1179"><span class="linenos">1179</span></a><span class="sd">            </span>
</span><span id="PDBInformation-1180"><a href="#PDBInformation-1180"><span class="linenos">1180</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation-1181"><a href="#PDBInformation-1181"><span class="linenos">1181</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation-1182"><a href="#PDBInformation-1182"><span class="linenos">1182</span></a><span class="sd">        min_dist_pairs_atoms_arr : numpy.ndarray</span>
</span><span id="PDBInformation-1183"><a href="#PDBInformation-1183"><span class="linenos">1183</span></a><span class="sd">            Structured ndarray that has residue indices, auth residue indices (corresponding to the protein numbering), and atomic names in the format {&#39;names&#39;: [&#39;residue1&#39;,&#39;residue2&#39;,&#39;auth_residue1&#39;,&#39;auth_residue2&#39;,&#39;atom_name1&#39;,&#39;atom_name2&#39;], &#39;formats&#39;: [int,int,int,int,&#39;&lt;U10&#39;,&#39;&lt;U10&#39;]}</span>
</span><span id="PDBInformation-1184"><a href="#PDBInformation-1184"><span class="linenos">1184</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation-1185"><a href="#PDBInformation-1185"><span class="linenos">1185</span></a>        <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_shift_values</span><span class="p">(</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">)</span>
</span><span id="PDBInformation-1186"><a href="#PDBInformation-1186"><span class="linenos">1186</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chain_specific_structure</span><span class="p">(</span><span class="n">ca_only</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">chain1</span><span class="o">=</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="o">=</span><span class="n">chain2</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="PDBInformation-1187"><a href="#PDBInformation-1187"><span class="linenos">1187</span></a>        <span class="n">min_dist_pairs_atoms</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="PDBInformation-1188"><a href="#PDBInformation-1188"><span class="linenos">1188</span></a>        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">:</span>
</span><span id="PDBInformation-1189"><a href="#PDBInformation-1189"><span class="linenos">1189</span></a>            <span class="c1"># Obtain structure information for chains 1 and 2</span>
</span><span id="PDBInformation-1190"><a href="#PDBInformation-1190"><span class="linenos">1190</span></a>            <span class="n">chain1_res1_structure</span> <span class="o">=</span> <span class="n">chain1_structure</span><span class="p">[</span><span class="n">chain1_structure</span><span class="o">.</span><span class="n">res_id</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]]</span>
</span><span id="PDBInformation-1191"><a href="#PDBInformation-1191"><span class="linenos">1191</span></a>            <span class="n">chain2_res2_structure</span> <span class="o">=</span> <span class="n">chain2_structure</span><span class="p">[</span><span class="n">chain2_structure</span><span class="o">.</span><span class="n">res_id</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]]</span>
</span><span id="PDBInformation-1192"><a href="#PDBInformation-1192"><span class="linenos">1192</span></a>            
</span><span id="PDBInformation-1193"><a href="#PDBInformation-1193"><span class="linenos">1193</span></a>            <span class="c1"># Calculate a distance matrix and find the indices of the minimal value in the matrix</span>
</span><span id="PDBInformation-1194"><a href="#PDBInformation-1194"><span class="linenos">1194</span></a>            <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">cdist</span><span class="p">(</span><span class="n">chain1_res1_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="n">chain2_res2_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>
</span><span id="PDBInformation-1195"><a href="#PDBInformation-1195"><span class="linenos">1195</span></a>            
</span><span id="PDBInformation-1196"><a href="#PDBInformation-1196"><span class="linenos">1196</span></a>            <span class="n">ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unravel_index</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">dist_matrix</span><span class="p">),</span> <span class="n">dist_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</span><span id="PDBInformation-1197"><a href="#PDBInformation-1197"><span class="linenos">1197</span></a>            <span class="c1"># Use the indices to access the atom in the atom array and get the correct atom name.</span>
</span><span id="PDBInformation-1198"><a href="#PDBInformation-1198"><span class="linenos">1198</span></a>            <span class="c1"># Generate the auth ids of the residues in the pairs ndarray</span>
</span><span id="PDBInformation-1199"><a href="#PDBInformation-1199"><span class="linenos">1199</span></a>            <span class="n">orig_res_id1</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">shift1</span>
</span><span id="PDBInformation-1200"><a href="#PDBInformation-1200"><span class="linenos">1200</span></a>            <span class="n">orig_res_id2</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">shift2</span>
</span><span id="PDBInformation-1201"><a href="#PDBInformation-1201"><span class="linenos">1201</span></a>            <span class="n">min_dist_pairs_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">orig_res_id1</span><span class="p">,</span> <span class="n">orig_res_id2</span><span class="p">,</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">],</span> <span class="n">chain1_res1_structure</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">atom_name</span><span class="p">,</span> <span class="n">chain2_res2_structure</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">atom_name</span><span class="p">))</span>
</span><span id="PDBInformation-1202"><a href="#PDBInformation-1202"><span class="linenos">1202</span></a>        <span class="n">min_dist_pairs_atoms_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">min_dist_pairs_atoms</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span><span class="s1">&#39;residue2&#39;</span><span class="p">,</span><span class="s1">&#39;auth_residue1&#39;</span><span class="p">,</span><span class="s1">&#39;auth_residue2&#39;</span><span class="p">,</span><span class="s1">&#39;atom_name1&#39;</span><span class="p">,</span><span class="s1">&#39;atom_name2&#39;</span><span class="p">],</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="s1">&#39;&lt;U10&#39;</span><span class="p">,</span><span class="s1">&#39;&lt;U10&#39;</span><span class="p">]})</span>
</span><span id="PDBInformation-1203"><a href="#PDBInformation-1203"><span class="linenos">1203</span></a>        <span class="k">return</span> <span class="n">min_dist_pairs_atoms_arr</span>    
</span><span id="PDBInformation-1204"><a href="#PDBInformation-1204"><span class="linenos">1204</span></a>
</span><span id="PDBInformation-1205"><a href="#PDBInformation-1205"><span class="linenos">1205</span></a>    <span class="k">def</span> <span class="nf">get_contacts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_contacts</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
</span><span id="PDBInformation-1206"><a href="#PDBInformation-1206"><span class="linenos">1206</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation-1207"><a href="#PDBInformation-1207"><span class="linenos">1207</span></a><span class="sd">        Get contacts from the structure attribute where the distance between two residues is less than the threshold.</span>
</span><span id="PDBInformation-1208"><a href="#PDBInformation-1208"><span class="linenos">1208</span></a>
</span><span id="PDBInformation-1209"><a href="#PDBInformation-1209"><span class="linenos">1209</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation-1210"><a href="#PDBInformation-1210"><span class="linenos">1210</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation-1211"><a href="#PDBInformation-1211"><span class="linenos">1211</span></a><span class="sd">        ca_only : bool</span>
</span><span id="PDBInformation-1212"><a href="#PDBInformation-1212"><span class="linenos">1212</span></a><span class="sd">            If true, only consider alpha-carbon to alpha-carbon distances. </span>
</span><span id="PDBInformation-1213"><a href="#PDBInformation-1213"><span class="linenos">1213</span></a><span class="sd">        threshold : float</span>
</span><span id="PDBInformation-1214"><a href="#PDBInformation-1214"><span class="linenos">1214</span></a><span class="sd">            Maximum distance to consider between two atoms.</span>
</span><span id="PDBInformation-1215"><a href="#PDBInformation-1215"><span class="linenos">1215</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation-1216"><a href="#PDBInformation-1216"><span class="linenos">1216</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation-1217"><a href="#PDBInformation-1217"><span class="linenos">1217</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation-1218"><a href="#PDBInformation-1218"><span class="linenos">1218</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="PDBInformation-1219"><a href="#PDBInformation-1219"><span class="linenos">1219</span></a><span class="sd">        auth_contacts : bool</span>
</span><span id="PDBInformation-1220"><a href="#PDBInformation-1220"><span class="linenos">1220</span></a><span class="sd">            True if you want alt_ids for residues indices, False if cif residue indexing is needed.</span>
</span><span id="PDBInformation-1221"><a href="#PDBInformation-1221"><span class="linenos">1221</span></a>
</span><span id="PDBInformation-1222"><a href="#PDBInformation-1222"><span class="linenos">1222</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation-1223"><a href="#PDBInformation-1223"><span class="linenos">1223</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation-1224"><a href="#PDBInformation-1224"><span class="linenos">1224</span></a><span class="sd">        contacts_set : set of tuple of ints</span>
</span><span id="PDBInformation-1225"><a href="#PDBInformation-1225"><span class="linenos">1225</span></a><span class="sd">            Set of contacts, tuples with &quot;residue1&quot; and &quot;residue2&quot; from the structure that are within the distance threshold.</span>
</span><span id="PDBInformation-1226"><a href="#PDBInformation-1226"><span class="linenos">1226</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation-1227"><a href="#PDBInformation-1227"><span class="linenos">1227</span></a>        
</span><span id="PDBInformation-1228"><a href="#PDBInformation-1228"><span class="linenos">1228</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">,</span> <span class="n">dist_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_dist_matrix</span><span class="p">(</span><span class="n">ca_only</span><span class="p">,</span> <span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">)</span>
</span><span id="PDBInformation-1229"><a href="#PDBInformation-1229"><span class="linenos">1229</span></a>        <span class="n">thresh_ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">dist_matrix</span> <span class="o">&lt;=</span> <span class="n">threshold</span><span class="p">)</span>
</span><span id="PDBInformation-1230"><a href="#PDBInformation-1230"><span class="linenos">1230</span></a>        <span class="n">contacts_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
</span><span id="PDBInformation-1231"><a href="#PDBInformation-1231"><span class="linenos">1231</span></a>        <span class="k">for</span> <span class="n">indices</span> <span class="ow">in</span> <span class="n">thresh_ind</span><span class="p">:</span>
</span><span id="PDBInformation-1232"><a href="#PDBInformation-1232"><span class="linenos">1232</span></a>            <span class="n">chain1_atom</span> <span class="o">=</span> <span class="n">chain1_structure</span><span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
</span><span id="PDBInformation-1233"><a href="#PDBInformation-1233"><span class="linenos">1233</span></a>            <span class="n">chain2_atom</span> <span class="o">=</span> <span class="n">chain2_structure</span><span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
</span><span id="PDBInformation-1234"><a href="#PDBInformation-1234"><span class="linenos">1234</span></a>            <span class="n">res1</span> <span class="o">=</span> <span class="n">chain1_atom</span><span class="o">.</span><span class="n">res_id</span>
</span><span id="PDBInformation-1235"><a href="#PDBInformation-1235"><span class="linenos">1235</span></a>            <span class="n">res2</span> <span class="o">=</span> <span class="n">chain2_atom</span><span class="o">.</span><span class="n">res_id</span>
</span><span id="PDBInformation-1236"><a href="#PDBInformation-1236"><span class="linenos">1236</span></a>            <span class="k">if</span> <span class="ow">not</span><span class="p">(</span><span class="n">chain1</span><span class="o">==</span><span class="n">chain2</span> <span class="ow">and</span> <span class="n">res1</span> <span class="o">&gt;=</span> <span class="n">res2</span><span class="p">):</span>
</span><span id="PDBInformation-1237"><a href="#PDBInformation-1237"><span class="linenos">1237</span></a>                <span class="k">if</span> <span class="ow">not</span> <span class="n">auth_contacts</span><span class="p">:</span>
</span><span id="PDBInformation-1238"><a href="#PDBInformation-1238"><span class="linenos">1238</span></a>                    <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_shift_values</span><span class="p">(</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">)</span>
</span><span id="PDBInformation-1239"><a href="#PDBInformation-1239"><span class="linenos">1239</span></a>                    <span class="n">contacts_set</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">res1</span> <span class="o">-</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">res2</span> <span class="o">-</span> <span class="n">shift2</span><span class="p">))</span>
</span><span id="PDBInformation-1240"><a href="#PDBInformation-1240"><span class="linenos">1240</span></a>                <span class="k">else</span><span class="p">:</span>
</span><span id="PDBInformation-1241"><a href="#PDBInformation-1241"><span class="linenos">1241</span></a>                    <span class="n">contacts_set</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">res1</span><span class="p">,</span> <span class="n">res2</span><span class="p">))</span>
</span><span id="PDBInformation-1242"><a href="#PDBInformation-1242"><span class="linenos">1242</span></a>        <span class="k">return</span> <span class="n">contacts_set</span>
</span></pre></div>


            <div class="docstring"><p>Information regarding a protein structure, obtained from a protein structure file.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>structure</strong> (biotite.structure.AtomArray):
Structure obtained from an RCSB entry with a provided pdbx/mmcif file with a specified model number.</li>
<li><strong>pdb_file</strong> (biotite.io.pdb.PDBFile):
mmCIF file that contains generic information and atomic information of the protein structure categorized into mmCIF blocks.</li>
<li><strong>model_num</strong> (int):
The model number to access from the PDB to ensure an AtomArray is returned containing the atom information of the protein structure.</li>
</ul>

<h6 id="attributes">Attributes</h6>

<ul>
<li><strong>self.non_missing_sequences</strong> (dict of str, biotite.sequence.ProteinSequence):
The protein sequences, without missing residues, compiled in the structure of the StructureInformation instance stored in a dictionary where chain_id is the key and the sequence string is the value.</li>
</ul>
</div>


                            <div id="PDBInformation.__init__" class="classattr">
                                        <input id="PDBInformation.__init__-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="name">PDBInformation</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">structure</span>,</span><span class="param">	<span class="n">pdb_file</span><span class="p">:</span> <span class="n">biotite</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">pdb</span><span class="o">.</span><span class="n">PDBFile</span>,</span><span class="param">	<span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span></span>)</span>

                <label class="view-source-button" for="PDBInformation.__init__-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#PDBInformation.__init__"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="PDBInformation.__init__-1063"><a href="#PDBInformation.__init__-1063"><span class="linenos">1063</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structure</span><span class="p">,</span> <span class="n">pdb_file</span><span class="p">:</span> <span class="n">pdb</span><span class="o">.</span><span class="n">PDBFile</span><span class="p">,</span> <span class="n">model_num</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
</span><span id="PDBInformation.__init__-1064"><a href="#PDBInformation.__init__-1064"><span class="linenos">1064</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">structure</span>
</span><span id="PDBInformation.__init__-1065"><a href="#PDBInformation.__init__-1065"><span class="linenos">1065</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">pdb_file</span> <span class="o">=</span> <span class="n">pdb_file</span>
</span><span id="PDBInformation.__init__-1066"><a href="#PDBInformation.__init__-1066"><span class="linenos">1066</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">model_num</span> <span class="o">=</span> <span class="n">model_num</span>
</span><span id="PDBInformation.__init__-1067"><a href="#PDBInformation.__init__-1067"><span class="linenos">1067</span></a>        <span class="n">non_hetero_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="PDBInformation.__init__-1068"><a href="#PDBInformation.__init__-1068"><span class="linenos">1068</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">chain</span><span class="p">):</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">chain</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">struc</span><span class="o">.</span><span class="n">get_chains</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">),</span> <span class="n">struc</span><span class="o">.</span><span class="n">to_sequence</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))}</span>
</span><span id="PDBInformation.__init__-1069"><a href="#PDBInformation.__init__-1069"><span class="linenos">1069</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span> <span class="o">=</span> <span class="n">struc</span><span class="o">.</span><span class="n">get_chains</span><span class="p">(</span><span class="n">non_hetero_structure</span><span class="p">)</span>
</span></pre></div>


    

                            </div>
                            <div id="PDBInformation.structure" class="classattr">
                                <div class="attr variable">
            <span class="name">structure</span>

        
    </div>
    <a class="headerlink" href="#PDBInformation.structure"></a>
    
    

                            </div>
                            <div id="PDBInformation.pdb_file" class="classattr">
                                <div class="attr variable">
            <span class="name">pdb_file</span>

        
    </div>
    <a class="headerlink" href="#PDBInformation.pdb_file"></a>
    
    

                            </div>
                            <div id="PDBInformation.model_num" class="classattr">
                                <div class="attr variable">
            <span class="name">model_num</span>

        
    </div>
    <a class="headerlink" href="#PDBInformation.model_num"></a>
    
    

                            </div>
                            <div id="PDBInformation.non_missing_sequences" class="classattr">
                                <div class="attr variable">
            <span class="name">non_missing_sequences</span>

        
    </div>
    <a class="headerlink" href="#PDBInformation.non_missing_sequences"></a>
    
    

                            </div>
                            <div id="PDBInformation.unique_chains" class="classattr">
                                <div class="attr variable">
            <span class="name">unique_chains</span>

        
    </div>
    <a class="headerlink" href="#PDBInformation.unique_chains"></a>
    
    

                            </div>
                            <div id="PDBInformation.get_non_missing_sequence" class="classattr">
                                        <input id="PDBInformation.get_non_missing_sequence-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_non_missing_sequence</span><span class="signature pdoc-code condensed">(<span class="param"><span class="bp">self</span>, </span><span class="param"><span class="n">chain_id</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="nb">str</span>:</span></span>

                <label class="view-source-button" for="PDBInformation.get_non_missing_sequence-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#PDBInformation.get_non_missing_sequence"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="PDBInformation.get_non_missing_sequence-1071"><a href="#PDBInformation.get_non_missing_sequence-1071"><span class="linenos">1071</span></a>    <span class="k">def</span> <span class="nf">get_non_missing_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
</span><span id="PDBInformation.get_non_missing_sequence-1072"><a href="#PDBInformation.get_non_missing_sequence-1072"><span class="linenos">1072</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_non_missing_sequence-1073"><a href="#PDBInformation.get_non_missing_sequence-1073"><span class="linenos">1073</span></a><span class="sd">        Get sequence, including only non-missing residues, from the specified chain.</span>
</span><span id="PDBInformation.get_non_missing_sequence-1074"><a href="#PDBInformation.get_non_missing_sequence-1074"><span class="linenos">1074</span></a>
</span><span id="PDBInformation.get_non_missing_sequence-1075"><a href="#PDBInformation.get_non_missing_sequence-1075"><span class="linenos">1075</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation.get_non_missing_sequence-1076"><a href="#PDBInformation.get_non_missing_sequence-1076"><span class="linenos">1076</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation.get_non_missing_sequence-1077"><a href="#PDBInformation.get_non_missing_sequence-1077"><span class="linenos">1077</span></a><span class="sd">        chain_id : str</span>
</span><span id="PDBInformation.get_non_missing_sequence-1078"><a href="#PDBInformation.get_non_missing_sequence-1078"><span class="linenos">1078</span></a><span class="sd">            Chain id supplied. The full sequence, including only non-missing residues, of this chain will be returned.</span>
</span><span id="PDBInformation.get_non_missing_sequence-1079"><a href="#PDBInformation.get_non_missing_sequence-1079"><span class="linenos">1079</span></a>
</span><span id="PDBInformation.get_non_missing_sequence-1080"><a href="#PDBInformation.get_non_missing_sequence-1080"><span class="linenos">1080</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation.get_non_missing_sequence-1081"><a href="#PDBInformation.get_non_missing_sequence-1081"><span class="linenos">1081</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation.get_non_missing_sequence-1082"><a href="#PDBInformation.get_non_missing_sequence-1082"><span class="linenos">1082</span></a><span class="sd">        str</span>
</span><span id="PDBInformation.get_non_missing_sequence-1083"><a href="#PDBInformation.get_non_missing_sequence-1083"><span class="linenos">1083</span></a><span class="sd">            The full sequence, including only non-missing residues, of the chain specified.</span>
</span><span id="PDBInformation.get_non_missing_sequence-1084"><a href="#PDBInformation.get_non_missing_sequence-1084"><span class="linenos">1084</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_non_missing_sequence-1085"><a href="#PDBInformation.get_non_missing_sequence-1085"><span class="linenos">1085</span></a>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_missing_sequences</span><span class="p">[</span><span class="n">chain_id</span><span class="p">]</span>
</span></pre></div>


            <div class="docstring"><p>Get sequence, including only non-missing residues, from the specified chain.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>chain_id</strong> (str):
Chain id supplied. The full sequence, including only non-missing residues, of this chain will be returned.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>str</strong>: The full sequence, including only non-missing residues, of the chain specified.</li>
</ul>
</div>


                            </div>
                            <div id="PDBInformation.get_chain_specific_structure" class="classattr">
                                        <input id="PDBInformation.get_chain_specific_structure-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_chain_specific_structure</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span>,</span><span class="param">	<span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span></span><span class="return-annotation">) -> <span class="nb">tuple</span>:</span></span>

                <label class="view-source-button" for="PDBInformation.get_chain_specific_structure-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#PDBInformation.get_chain_specific_structure"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="PDBInformation.get_chain_specific_structure-1087"><a href="#PDBInformation.get_chain_specific_structure-1087"><span class="linenos">1087</span></a>    <span class="k">def</span> <span class="nf">get_chain_specific_structure</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
</span><span id="PDBInformation.get_chain_specific_structure-1088"><a href="#PDBInformation.get_chain_specific_structure-1088"><span class="linenos">1088</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_chain_specific_structure-1089"><a href="#PDBInformation.get_chain_specific_structure-1089"><span class="linenos">1089</span></a><span class="sd">        Subsets structure attribute to select for chain specific portions of the structure.</span>
</span><span id="PDBInformation.get_chain_specific_structure-1090"><a href="#PDBInformation.get_chain_specific_structure-1090"><span class="linenos">1090</span></a>
</span><span id="PDBInformation.get_chain_specific_structure-1091"><a href="#PDBInformation.get_chain_specific_structure-1091"><span class="linenos">1091</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation.get_chain_specific_structure-1092"><a href="#PDBInformation.get_chain_specific_structure-1092"><span class="linenos">1092</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation.get_chain_specific_structure-1093"><a href="#PDBInformation.get_chain_specific_structure-1093"><span class="linenos">1093</span></a><span class="sd">        ca_only : bool</span>
</span><span id="PDBInformation.get_chain_specific_structure-1094"><a href="#PDBInformation.get_chain_specific_structure-1094"><span class="linenos">1094</span></a><span class="sd">            If true, the structure will also be subsetted for atom entries where the atom_name annotation is &quot;CA&quot; (referring to alpha-carbons)</span>
</span><span id="PDBInformation.get_chain_specific_structure-1095"><a href="#PDBInformation.get_chain_specific_structure-1095"><span class="linenos">1095</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation.get_chain_specific_structure-1096"><a href="#PDBInformation.get_chain_specific_structure-1096"><span class="linenos">1096</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation.get_chain_specific_structure-1097"><a href="#PDBInformation.get_chain_specific_structure-1097"><span class="linenos">1097</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation.get_chain_specific_structure-1098"><a href="#PDBInformation.get_chain_specific_structure-1098"><span class="linenos">1098</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="PDBInformation.get_chain_specific_structure-1099"><a href="#PDBInformation.get_chain_specific_structure-1099"><span class="linenos">1099</span></a><span class="sd">        remove_hetero : bool, default=True</span>
</span><span id="PDBInformation.get_chain_specific_structure-1100"><a href="#PDBInformation.get_chain_specific_structure-1100"><span class="linenos">1100</span></a><span class="sd">            If true, the structure will also be subsetted for atom entries where the hetero annotation is False, thus removing heteroatoms.</span>
</span><span id="PDBInformation.get_chain_specific_structure-1101"><a href="#PDBInformation.get_chain_specific_structure-1101"><span class="linenos">1101</span></a>
</span><span id="PDBInformation.get_chain_specific_structure-1102"><a href="#PDBInformation.get_chain_specific_structure-1102"><span class="linenos">1102</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation.get_chain_specific_structure-1103"><a href="#PDBInformation.get_chain_specific_structure-1103"><span class="linenos">1103</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation.get_chain_specific_structure-1104"><a href="#PDBInformation.get_chain_specific_structure-1104"><span class="linenos">1104</span></a><span class="sd">        tuple of biotite.structure.AtomArray, biotite.structure.AtomArray</span>
</span><span id="PDBInformation.get_chain_specific_structure-1105"><a href="#PDBInformation.get_chain_specific_structure-1105"><span class="linenos">1105</span></a><span class="sd">            Two AtomArrays that refer to atoms in the first chain and second chain, respectively without accounting for the presence of heteroatoms if `remove_hetero` is True.</span>
</span><span id="PDBInformation.get_chain_specific_structure-1106"><a href="#PDBInformation.get_chain_specific_structure-1106"><span class="linenos">1106</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_chain_specific_structure-1107"><a href="#PDBInformation.get_chain_specific_structure-1107"><span class="linenos">1107</span></a>
</span><span id="PDBInformation.get_chain_specific_structure-1108"><a href="#PDBInformation.get_chain_specific_structure-1108"><span class="linenos">1108</span></a>        <span class="n">selected_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span>
</span><span id="PDBInformation.get_chain_specific_structure-1109"><a href="#PDBInformation.get_chain_specific_structure-1109"><span class="linenos">1109</span></a>        <span class="k">if</span> <span class="n">remove_hetero</span><span class="p">:</span>
</span><span id="PDBInformation.get_chain_specific_structure-1110"><a href="#PDBInformation.get_chain_specific_structure-1110"><span class="linenos">1110</span></a>            <span class="c1"># Remove hetero atoms via hetero column of structure ndarray</span>
</span><span id="PDBInformation.get_chain_specific_structure-1111"><a href="#PDBInformation.get_chain_specific_structure-1111"><span class="linenos">1111</span></a>            <span class="n">selected_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="PDBInformation.get_chain_specific_structure-1112"><a href="#PDBInformation.get_chain_specific_structure-1112"><span class="linenos">1112</span></a>        <span class="k">if</span> <span class="n">ca_only</span><span class="p">:</span>
</span><span id="PDBInformation.get_chain_specific_structure-1113"><a href="#PDBInformation.get_chain_specific_structure-1113"><span class="linenos">1113</span></a>            <span class="c1"># Consider selection of alpha-carbon atoms only</span>
</span><span id="PDBInformation.get_chain_specific_structure-1114"><a href="#PDBInformation.get_chain_specific_structure-1114"><span class="linenos">1114</span></a>            <span class="n">selected_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">atom_name</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span><span class="p">]</span>
</span><span id="PDBInformation.get_chain_specific_structure-1115"><a href="#PDBInformation.get_chain_specific_structure-1115"><span class="linenos">1115</span></a>        <span class="n">chain1_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain1</span><span class="p">]</span>
</span><span id="PDBInformation.get_chain_specific_structure-1116"><a href="#PDBInformation.get_chain_specific_structure-1116"><span class="linenos">1116</span></a>        <span class="n">chain2_structure</span> <span class="o">=</span> <span class="n">selected_structure</span><span class="p">[</span><span class="n">selected_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain2</span><span class="p">]</span>
</span><span id="PDBInformation.get_chain_specific_structure-1117"><a href="#PDBInformation.get_chain_specific_structure-1117"><span class="linenos">1117</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Subsets structure attribute to select for chain specific portions of the structure.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>ca_only</strong> (bool):
If true, the structure will also be subsetted for atom entries where the atom_name annotation is "CA" (referring to alpha-carbons)</li>
<li><strong>chain1</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
<li><strong>chain2</strong> (str):
Chain id corresponding to the second column of residues in the structure.</li>
<li><strong>remove_hetero</strong> (bool, default=True):
If true, the structure will also be subsetted for atom entries where the hetero annotation is False, thus removing heteroatoms.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>tuple of biotite.structure.AtomArray, biotite.structure.AtomArray</strong>: Two AtomArrays that refer to atoms in the first chain and second chain, respectively without accounting for the presence of heteroatoms if <code>remove_hetero</code> is True.</li>
</ul>
</div>


                            </div>
                            <div id="PDBInformation.generate_dist_matrix" class="classattr">
                                        <input id="PDBInformation.generate_dist_matrix-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">generate_dist_matrix</span><span class="signature pdoc-code condensed">(<span class="param"><span class="bp">self</span>, </span><span class="param"><span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span>, </span><span class="param"><span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>, </span><span class="param"><span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">):</span></span>

                <label class="view-source-button" for="PDBInformation.generate_dist_matrix-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#PDBInformation.generate_dist_matrix"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="PDBInformation.generate_dist_matrix-1119"><a href="#PDBInformation.generate_dist_matrix-1119"><span class="linenos">1119</span></a>    <span class="k">def</span> <span class="nf">generate_dist_matrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
</span><span id="PDBInformation.generate_dist_matrix-1120"><a href="#PDBInformation.generate_dist_matrix-1120"><span class="linenos">1120</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation.generate_dist_matrix-1121"><a href="#PDBInformation.generate_dist_matrix-1121"><span class="linenos">1121</span></a><span class="sd">        Generates distance matrix between two chains in the structure attribute.</span>
</span><span id="PDBInformation.generate_dist_matrix-1122"><a href="#PDBInformation.generate_dist_matrix-1122"><span class="linenos">1122</span></a>
</span><span id="PDBInformation.generate_dist_matrix-1123"><a href="#PDBInformation.generate_dist_matrix-1123"><span class="linenos">1123</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation.generate_dist_matrix-1124"><a href="#PDBInformation.generate_dist_matrix-1124"><span class="linenos">1124</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation.generate_dist_matrix-1125"><a href="#PDBInformation.generate_dist_matrix-1125"><span class="linenos">1125</span></a><span class="sd">        ca_only : bool</span>
</span><span id="PDBInformation.generate_dist_matrix-1126"><a href="#PDBInformation.generate_dist_matrix-1126"><span class="linenos">1126</span></a><span class="sd">            If True, only atoms that have the name &quot;CA&quot; are selected in the chains the distance matrix is calculated between.</span>
</span><span id="PDBInformation.generate_dist_matrix-1127"><a href="#PDBInformation.generate_dist_matrix-1127"><span class="linenos">1127</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation.generate_dist_matrix-1128"><a href="#PDBInformation.generate_dist_matrix-1128"><span class="linenos">1128</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation.generate_dist_matrix-1129"><a href="#PDBInformation.generate_dist_matrix-1129"><span class="linenos">1129</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation.generate_dist_matrix-1130"><a href="#PDBInformation.generate_dist_matrix-1130"><span class="linenos">1130</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation.generate_dist_matrix-1131"><a href="#PDBInformation.generate_dist_matrix-1131"><span class="linenos">1131</span></a>
</span><span id="PDBInformation.generate_dist_matrix-1132"><a href="#PDBInformation.generate_dist_matrix-1132"><span class="linenos">1132</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation.generate_dist_matrix-1133"><a href="#PDBInformation.generate_dist_matrix-1133"><span class="linenos">1133</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation.generate_dist_matrix-1134"><a href="#PDBInformation.generate_dist_matrix-1134"><span class="linenos">1134</span></a><span class="sd">        tuple of biotite.structure.AtomArray, biotite.structure.AtomArray, numpy.ndarray</span>
</span><span id="PDBInformation.generate_dist_matrix-1135"><a href="#PDBInformation.generate_dist_matrix-1135"><span class="linenos">1135</span></a><span class="sd">            Tuple containing the chain 1 structure, the chain 2 structure, and the distance matrix of chain 1 and chain 2&#39;s pairwise distances.</span>
</span><span id="PDBInformation.generate_dist_matrix-1136"><a href="#PDBInformation.generate_dist_matrix-1136"><span class="linenos">1136</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation.generate_dist_matrix-1137"><a href="#PDBInformation.generate_dist_matrix-1137"><span class="linenos">1137</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chain_specific_structure</span><span class="p">(</span><span class="n">ca_only</span><span class="p">,</span> <span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="PDBInformation.generate_dist_matrix-1138"><a href="#PDBInformation.generate_dist_matrix-1138"><span class="linenos">1138</span></a>        <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">cdist</span><span class="p">(</span><span class="n">chain1_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>
</span><span id="PDBInformation.generate_dist_matrix-1139"><a href="#PDBInformation.generate_dist_matrix-1139"><span class="linenos">1139</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">,</span> <span class="n">dist_matrix</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Generates distance matrix between two chains in the structure attribute.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>ca_only</strong> (bool):
If True, only atoms that have the name "CA" are selected in the chains the distance matrix is calculated between.</li>
<li><strong>chain1</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
<li><strong>chain2</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>tuple of biotite.structure.AtomArray, biotite.structure.AtomArray, numpy.ndarray</strong>: Tuple containing the chain 1 structure, the chain 2 structure, and the distance matrix of chain 1 and chain 2's pairwise distances.</li>
</ul>
</div>


                            </div>
                            <div id="PDBInformation.get_shift_values" class="classattr">
                                        <input id="PDBInformation.get_shift_values-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_shift_values</span><span class="signature pdoc-code condensed">(<span class="param"><span class="bp">self</span>, </span><span class="param"><span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>, </span><span class="param"><span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span>:</span></span>

                <label class="view-source-button" for="PDBInformation.get_shift_values-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#PDBInformation.get_shift_values"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="PDBInformation.get_shift_values-1141"><a href="#PDBInformation.get_shift_values-1141"><span class="linenos">1141</span></a>    <span class="k">def</span> <span class="nf">get_shift_values</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
</span><span id="PDBInformation.get_shift_values-1142"><a href="#PDBInformation.get_shift_values-1142"><span class="linenos">1142</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_shift_values-1143"><a href="#PDBInformation.get_shift_values-1143"><span class="linenos">1143</span></a><span class="sd">        Get shift values needed for production of auth residue ids.</span>
</span><span id="PDBInformation.get_shift_values-1144"><a href="#PDBInformation.get_shift_values-1144"><span class="linenos">1144</span></a><span class="sd">        </span>
</span><span id="PDBInformation.get_shift_values-1145"><a href="#PDBInformation.get_shift_values-1145"><span class="linenos">1145</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation.get_shift_values-1146"><a href="#PDBInformation.get_shift_values-1146"><span class="linenos">1146</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation.get_shift_values-1147"><a href="#PDBInformation.get_shift_values-1147"><span class="linenos">1147</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation.get_shift_values-1148"><a href="#PDBInformation.get_shift_values-1148"><span class="linenos">1148</span></a><span class="sd">            Name of the chain id present in the struct_ref_seq block of cif files referring to the second column of residues.</span>
</span><span id="PDBInformation.get_shift_values-1149"><a href="#PDBInformation.get_shift_values-1149"><span class="linenos">1149</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation.get_shift_values-1150"><a href="#PDBInformation.get_shift_values-1150"><span class="linenos">1150</span></a><span class="sd">            Name of the chain id present in the struct_ref_seq block of cif files referring to the second column of residues.</span>
</span><span id="PDBInformation.get_shift_values-1151"><a href="#PDBInformation.get_shift_values-1151"><span class="linenos">1151</span></a>
</span><span id="PDBInformation.get_shift_values-1152"><a href="#PDBInformation.get_shift_values-1152"><span class="linenos">1152</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation.get_shift_values-1153"><a href="#PDBInformation.get_shift_values-1153"><span class="linenos">1153</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation.get_shift_values-1154"><a href="#PDBInformation.get_shift_values-1154"><span class="linenos">1154</span></a><span class="sd">        (shift1, shift2) : tuple of int, int</span>
</span><span id="PDBInformation.get_shift_values-1155"><a href="#PDBInformation.get_shift_values-1155"><span class="linenos">1155</span></a><span class="sd">            Tuple containing both shift values, the difference between the auth_res_id and res_id.</span>
</span><span id="PDBInformation.get_shift_values-1156"><a href="#PDBInformation.get_shift_values-1156"><span class="linenos">1156</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_shift_values-1157"><a href="#PDBInformation.get_shift_values-1157"><span class="linenos">1157</span></a>        <span class="n">non_hetero_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="o">.</span><span class="n">hetero</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
</span><span id="PDBInformation.get_shift_values-1158"><a href="#PDBInformation.get_shift_values-1158"><span class="linenos">1158</span></a>        <span class="n">shift1</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="PDBInformation.get_shift_values-1159"><a href="#PDBInformation.get_shift_values-1159"><span class="linenos">1159</span></a>        <span class="n">shift2</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="PDBInformation.get_shift_values-1160"><a href="#PDBInformation.get_shift_values-1160"><span class="linenos">1160</span></a>        <span class="k">if</span> <span class="n">chain1</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span> <span class="ow">and</span> <span class="n">chain2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_chains</span><span class="p">:</span>
</span><span id="PDBInformation.get_shift_values-1161"><a href="#PDBInformation.get_shift_values-1161"><span class="linenos">1161</span></a>            <span class="n">shift1</span> <span class="o">=</span> <span class="n">non_hetero_structure</span><span class="p">[</span><span class="n">non_hetero_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">res_id</span> <span class="o">-</span> <span class="mi">1</span>
</span><span id="PDBInformation.get_shift_values-1162"><a href="#PDBInformation.get_shift_values-1162"><span class="linenos">1162</span></a>            <span class="n">shift2</span> <span class="o">=</span> <span class="n">non_hetero_structure</span><span class="p">[</span><span class="n">non_hetero_structure</span><span class="o">.</span><span class="n">chain_id</span> <span class="o">==</span> <span class="n">chain2</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">res_id</span> <span class="o">-</span> <span class="mi">1</span>
</span><span id="PDBInformation.get_shift_values-1163"><a href="#PDBInformation.get_shift_values-1163"><span class="linenos">1163</span></a>            <span class="k">return</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span>
</span><span id="PDBInformation.get_shift_values-1164"><a href="#PDBInformation.get_shift_values-1164"><span class="linenos">1164</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="PDBInformation.get_shift_values-1165"><a href="#PDBInformation.get_shift_values-1165"><span class="linenos">1165</span></a>            <span class="k">return</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span>
</span></pre></div>


            <div class="docstring"><p>Get shift values needed for production of auth residue ids.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>chain1</strong> (str):
Name of the chain id present in the struct_ref_seq block of cif files referring to the second column of residues.</li>
<li><strong>chain2</strong> (str):
Name of the chain id present in the struct_ref_seq block of cif files referring to the second column of residues.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>(shift1, shift2)</strong> (tuple of int, int):
Tuple containing both shift values, the difference between the auth_res_id and res_id.</li>
</ul>
</div>


                            </div>
                            <div id="PDBInformation.get_min_dist_atom_info" class="classattr">
                                        <input id="PDBInformation.get_min_dist_atom_info-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_min_dist_atom_info</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">pairs</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>,</span><span class="param">	<span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="n">typing</span><span class="o">.</span><span class="n">Any</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">[</span><span class="o">+</span><span class="n">_ScalarType_co</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="PDBInformation.get_min_dist_atom_info-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#PDBInformation.get_min_dist_atom_info"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="PDBInformation.get_min_dist_atom_info-1167"><a href="#PDBInformation.get_min_dist_atom_info-1167"><span class="linenos">1167</span></a>    <span class="k">def</span> <span class="nf">get_min_dist_atom_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pairs</span><span class="p">:</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">:</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1168"><a href="#PDBInformation.get_min_dist_atom_info-1168"><span class="linenos">1168</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1169"><a href="#PDBInformation.get_min_dist_atom_info-1169"><span class="linenos">1169</span></a><span class="sd">        Generate a ndarray of residue ids and their corresponding atom names such that the distance is the minimum between the initial residues provided.</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1170"><a href="#PDBInformation.get_min_dist_atom_info-1170"><span class="linenos">1170</span></a>
</span><span id="PDBInformation.get_min_dist_atom_info-1171"><a href="#PDBInformation.get_min_dist_atom_info-1171"><span class="linenos">1171</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1172"><a href="#PDBInformation.get_min_dist_atom_info-1172"><span class="linenos">1172</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1173"><a href="#PDBInformation.get_min_dist_atom_info-1173"><span class="linenos">1173</span></a><span class="sd">        pairs : numpy.ndarray</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1174"><a href="#PDBInformation.get_min_dist_atom_info-1174"><span class="linenos">1174</span></a><span class="sd">            Pairs structured ndarray with &quot;residue1&quot; and &quot;residue2&quot; columns.</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1175"><a href="#PDBInformation.get_min_dist_atom_info-1175"><span class="linenos">1175</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1176"><a href="#PDBInformation.get_min_dist_atom_info-1176"><span class="linenos">1176</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1177"><a href="#PDBInformation.get_min_dist_atom_info-1177"><span class="linenos">1177</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1178"><a href="#PDBInformation.get_min_dist_atom_info-1178"><span class="linenos">1178</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1179"><a href="#PDBInformation.get_min_dist_atom_info-1179"><span class="linenos">1179</span></a><span class="sd">            </span>
</span><span id="PDBInformation.get_min_dist_atom_info-1180"><a href="#PDBInformation.get_min_dist_atom_info-1180"><span class="linenos">1180</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1181"><a href="#PDBInformation.get_min_dist_atom_info-1181"><span class="linenos">1181</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1182"><a href="#PDBInformation.get_min_dist_atom_info-1182"><span class="linenos">1182</span></a><span class="sd">        min_dist_pairs_atoms_arr : numpy.ndarray</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1183"><a href="#PDBInformation.get_min_dist_atom_info-1183"><span class="linenos">1183</span></a><span class="sd">            Structured ndarray that has residue indices, auth residue indices (corresponding to the protein numbering), and atomic names in the format {&#39;names&#39;: [&#39;residue1&#39;,&#39;residue2&#39;,&#39;auth_residue1&#39;,&#39;auth_residue2&#39;,&#39;atom_name1&#39;,&#39;atom_name2&#39;], &#39;formats&#39;: [int,int,int,int,&#39;&lt;U10&#39;,&#39;&lt;U10&#39;]}</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1184"><a href="#PDBInformation.get_min_dist_atom_info-1184"><span class="linenos">1184</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1185"><a href="#PDBInformation.get_min_dist_atom_info-1185"><span class="linenos">1185</span></a>        <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_shift_values</span><span class="p">(</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">)</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1186"><a href="#PDBInformation.get_min_dist_atom_info-1186"><span class="linenos">1186</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chain_specific_structure</span><span class="p">(</span><span class="n">ca_only</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">chain1</span><span class="o">=</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="o">=</span><span class="n">chain2</span><span class="p">,</span> <span class="n">remove_hetero</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1187"><a href="#PDBInformation.get_min_dist_atom_info-1187"><span class="linenos">1187</span></a>        <span class="n">min_dist_pairs_atoms</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1188"><a href="#PDBInformation.get_min_dist_atom_info-1188"><span class="linenos">1188</span></a>        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">:</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1189"><a href="#PDBInformation.get_min_dist_atom_info-1189"><span class="linenos">1189</span></a>            <span class="c1"># Obtain structure information for chains 1 and 2</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1190"><a href="#PDBInformation.get_min_dist_atom_info-1190"><span class="linenos">1190</span></a>            <span class="n">chain1_res1_structure</span> <span class="o">=</span> <span class="n">chain1_structure</span><span class="p">[</span><span class="n">chain1_structure</span><span class="o">.</span><span class="n">res_id</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]]</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1191"><a href="#PDBInformation.get_min_dist_atom_info-1191"><span class="linenos">1191</span></a>            <span class="n">chain2_res2_structure</span> <span class="o">=</span> <span class="n">chain2_structure</span><span class="p">[</span><span class="n">chain2_structure</span><span class="o">.</span><span class="n">res_id</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]]</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1192"><a href="#PDBInformation.get_min_dist_atom_info-1192"><span class="linenos">1192</span></a>            
</span><span id="PDBInformation.get_min_dist_atom_info-1193"><a href="#PDBInformation.get_min_dist_atom_info-1193"><span class="linenos">1193</span></a>            <span class="c1"># Calculate a distance matrix and find the indices of the minimal value in the matrix</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1194"><a href="#PDBInformation.get_min_dist_atom_info-1194"><span class="linenos">1194</span></a>            <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">cdist</span><span class="p">(</span><span class="n">chain1_res1_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="n">chain2_res2_structure</span><span class="o">.</span><span class="n">coord</span><span class="p">)</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1195"><a href="#PDBInformation.get_min_dist_atom_info-1195"><span class="linenos">1195</span></a>            
</span><span id="PDBInformation.get_min_dist_atom_info-1196"><a href="#PDBInformation.get_min_dist_atom_info-1196"><span class="linenos">1196</span></a>            <span class="n">ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unravel_index</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">dist_matrix</span><span class="p">),</span> <span class="n">dist_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1197"><a href="#PDBInformation.get_min_dist_atom_info-1197"><span class="linenos">1197</span></a>            <span class="c1"># Use the indices to access the atom in the atom array and get the correct atom name.</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1198"><a href="#PDBInformation.get_min_dist_atom_info-1198"><span class="linenos">1198</span></a>            <span class="c1"># Generate the auth ids of the residues in the pairs ndarray</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1199"><a href="#PDBInformation.get_min_dist_atom_info-1199"><span class="linenos">1199</span></a>            <span class="n">orig_res_id1</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">shift1</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1200"><a href="#PDBInformation.get_min_dist_atom_info-1200"><span class="linenos">1200</span></a>            <span class="n">orig_res_id2</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">shift2</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1201"><a href="#PDBInformation.get_min_dist_atom_info-1201"><span class="linenos">1201</span></a>            <span class="n">min_dist_pairs_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">orig_res_id1</span><span class="p">,</span> <span class="n">orig_res_id2</span><span class="p">,</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;residue2&#39;</span><span class="p">],</span> <span class="n">chain1_res1_structure</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">atom_name</span><span class="p">,</span> <span class="n">chain2_res2_structure</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">atom_name</span><span class="p">))</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1202"><a href="#PDBInformation.get_min_dist_atom_info-1202"><span class="linenos">1202</span></a>        <span class="n">min_dist_pairs_atoms_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">min_dist_pairs_atoms</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;names&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;residue1&#39;</span><span class="p">,</span><span class="s1">&#39;residue2&#39;</span><span class="p">,</span><span class="s1">&#39;auth_residue1&#39;</span><span class="p">,</span><span class="s1">&#39;auth_residue2&#39;</span><span class="p">,</span><span class="s1">&#39;atom_name1&#39;</span><span class="p">,</span><span class="s1">&#39;atom_name2&#39;</span><span class="p">],</span> <span class="s1">&#39;formats&#39;</span><span class="p">:</span> <span class="p">[</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="nb">int</span><span class="p">,</span><span class="s1">&#39;&lt;U10&#39;</span><span class="p">,</span><span class="s1">&#39;&lt;U10&#39;</span><span class="p">]})</span>
</span><span id="PDBInformation.get_min_dist_atom_info-1203"><a href="#PDBInformation.get_min_dist_atom_info-1203"><span class="linenos">1203</span></a>        <span class="k">return</span> <span class="n">min_dist_pairs_atoms_arr</span>    
</span></pre></div>


            <div class="docstring"><p>Generate a ndarray of residue ids and their corresponding atom names such that the distance is the minimum between the initial residues provided.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>pairs</strong> (numpy.ndarray):
Pairs structured ndarray with "residue1" and "residue2" columns.</li>
<li><strong>chain1</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
<li><strong>chain2</strong> (str):
Chain id corresponding to the second column of residues in the structure.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>min_dist_pairs_atoms_arr</strong> (numpy.ndarray):
Structured ndarray that has residue indices, auth residue indices (corresponding to the protein numbering), and atomic names in the format {'names': ['residue1','residue2','auth_residue1','auth_residue2','atom_name1','atom_name2'], 'formats': [int,int,int,int,'<U10','<U10']}</li>
</ul>
</div>


                            </div>
                            <div id="PDBInformation.get_contacts" class="classattr">
                                        <input id="PDBInformation.get_contacts-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">get_contacts</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="bp">self</span>,</span><span class="param">	<span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span>,</span><span class="param">	<span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span>,</span><span class="param">	<span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">auth_contacts</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span></span><span class="return-annotation">) -> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="PDBInformation.get_contacts-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#PDBInformation.get_contacts"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="PDBInformation.get_contacts-1205"><a href="#PDBInformation.get_contacts-1205"><span class="linenos">1205</span></a>    <span class="k">def</span> <span class="nf">get_contacts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ca_only</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">chain1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">auth_contacts</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
</span><span id="PDBInformation.get_contacts-1206"><a href="#PDBInformation.get_contacts-1206"><span class="linenos">1206</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_contacts-1207"><a href="#PDBInformation.get_contacts-1207"><span class="linenos">1207</span></a><span class="sd">        Get contacts from the structure attribute where the distance between two residues is less than the threshold.</span>
</span><span id="PDBInformation.get_contacts-1208"><a href="#PDBInformation.get_contacts-1208"><span class="linenos">1208</span></a>
</span><span id="PDBInformation.get_contacts-1209"><a href="#PDBInformation.get_contacts-1209"><span class="linenos">1209</span></a><span class="sd">        Parameters</span>
</span><span id="PDBInformation.get_contacts-1210"><a href="#PDBInformation.get_contacts-1210"><span class="linenos">1210</span></a><span class="sd">        ----------</span>
</span><span id="PDBInformation.get_contacts-1211"><a href="#PDBInformation.get_contacts-1211"><span class="linenos">1211</span></a><span class="sd">        ca_only : bool</span>
</span><span id="PDBInformation.get_contacts-1212"><a href="#PDBInformation.get_contacts-1212"><span class="linenos">1212</span></a><span class="sd">            If true, only consider alpha-carbon to alpha-carbon distances. </span>
</span><span id="PDBInformation.get_contacts-1213"><a href="#PDBInformation.get_contacts-1213"><span class="linenos">1213</span></a><span class="sd">        threshold : float</span>
</span><span id="PDBInformation.get_contacts-1214"><a href="#PDBInformation.get_contacts-1214"><span class="linenos">1214</span></a><span class="sd">            Maximum distance to consider between two atoms.</span>
</span><span id="PDBInformation.get_contacts-1215"><a href="#PDBInformation.get_contacts-1215"><span class="linenos">1215</span></a><span class="sd">        chain1 : str</span>
</span><span id="PDBInformation.get_contacts-1216"><a href="#PDBInformation.get_contacts-1216"><span class="linenos">1216</span></a><span class="sd">            Chain id corresponding to the first column of residues in the structure.</span>
</span><span id="PDBInformation.get_contacts-1217"><a href="#PDBInformation.get_contacts-1217"><span class="linenos">1217</span></a><span class="sd">        chain2 : str</span>
</span><span id="PDBInformation.get_contacts-1218"><a href="#PDBInformation.get_contacts-1218"><span class="linenos">1218</span></a><span class="sd">            Chain id corresponding to the second column of residues in the structure.</span>
</span><span id="PDBInformation.get_contacts-1219"><a href="#PDBInformation.get_contacts-1219"><span class="linenos">1219</span></a><span class="sd">        auth_contacts : bool</span>
</span><span id="PDBInformation.get_contacts-1220"><a href="#PDBInformation.get_contacts-1220"><span class="linenos">1220</span></a><span class="sd">            True if you want alt_ids for residues indices, False if cif residue indexing is needed.</span>
</span><span id="PDBInformation.get_contacts-1221"><a href="#PDBInformation.get_contacts-1221"><span class="linenos">1221</span></a>
</span><span id="PDBInformation.get_contacts-1222"><a href="#PDBInformation.get_contacts-1222"><span class="linenos">1222</span></a><span class="sd">        Returns</span>
</span><span id="PDBInformation.get_contacts-1223"><a href="#PDBInformation.get_contacts-1223"><span class="linenos">1223</span></a><span class="sd">        -------</span>
</span><span id="PDBInformation.get_contacts-1224"><a href="#PDBInformation.get_contacts-1224"><span class="linenos">1224</span></a><span class="sd">        contacts_set : set of tuple of ints</span>
</span><span id="PDBInformation.get_contacts-1225"><a href="#PDBInformation.get_contacts-1225"><span class="linenos">1225</span></a><span class="sd">            Set of contacts, tuples with &quot;residue1&quot; and &quot;residue2&quot; from the structure that are within the distance threshold.</span>
</span><span id="PDBInformation.get_contacts-1226"><a href="#PDBInformation.get_contacts-1226"><span class="linenos">1226</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="PDBInformation.get_contacts-1227"><a href="#PDBInformation.get_contacts-1227"><span class="linenos">1227</span></a>        
</span><span id="PDBInformation.get_contacts-1228"><a href="#PDBInformation.get_contacts-1228"><span class="linenos">1228</span></a>        <span class="n">chain1_structure</span><span class="p">,</span> <span class="n">chain2_structure</span><span class="p">,</span> <span class="n">dist_matrix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_dist_matrix</span><span class="p">(</span><span class="n">ca_only</span><span class="p">,</span> <span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">)</span>
</span><span id="PDBInformation.get_contacts-1229"><a href="#PDBInformation.get_contacts-1229"><span class="linenos">1229</span></a>        <span class="n">thresh_ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">dist_matrix</span> <span class="o">&lt;=</span> <span class="n">threshold</span><span class="p">)</span>
</span><span id="PDBInformation.get_contacts-1230"><a href="#PDBInformation.get_contacts-1230"><span class="linenos">1230</span></a>        <span class="n">contacts_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
</span><span id="PDBInformation.get_contacts-1231"><a href="#PDBInformation.get_contacts-1231"><span class="linenos">1231</span></a>        <span class="k">for</span> <span class="n">indices</span> <span class="ow">in</span> <span class="n">thresh_ind</span><span class="p">:</span>
</span><span id="PDBInformation.get_contacts-1232"><a href="#PDBInformation.get_contacts-1232"><span class="linenos">1232</span></a>            <span class="n">chain1_atom</span> <span class="o">=</span> <span class="n">chain1_structure</span><span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
</span><span id="PDBInformation.get_contacts-1233"><a href="#PDBInformation.get_contacts-1233"><span class="linenos">1233</span></a>            <span class="n">chain2_atom</span> <span class="o">=</span> <span class="n">chain2_structure</span><span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
</span><span id="PDBInformation.get_contacts-1234"><a href="#PDBInformation.get_contacts-1234"><span class="linenos">1234</span></a>            <span class="n">res1</span> <span class="o">=</span> <span class="n">chain1_atom</span><span class="o">.</span><span class="n">res_id</span>
</span><span id="PDBInformation.get_contacts-1235"><a href="#PDBInformation.get_contacts-1235"><span class="linenos">1235</span></a>            <span class="n">res2</span> <span class="o">=</span> <span class="n">chain2_atom</span><span class="o">.</span><span class="n">res_id</span>
</span><span id="PDBInformation.get_contacts-1236"><a href="#PDBInformation.get_contacts-1236"><span class="linenos">1236</span></a>            <span class="k">if</span> <span class="ow">not</span><span class="p">(</span><span class="n">chain1</span><span class="o">==</span><span class="n">chain2</span> <span class="ow">and</span> <span class="n">res1</span> <span class="o">&gt;=</span> <span class="n">res2</span><span class="p">):</span>
</span><span id="PDBInformation.get_contacts-1237"><a href="#PDBInformation.get_contacts-1237"><span class="linenos">1237</span></a>                <span class="k">if</span> <span class="ow">not</span> <span class="n">auth_contacts</span><span class="p">:</span>
</span><span id="PDBInformation.get_contacts-1238"><a href="#PDBInformation.get_contacts-1238"><span class="linenos">1238</span></a>                    <span class="n">shift1</span><span class="p">,</span> <span class="n">shift2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_shift_values</span><span class="p">(</span><span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span><span class="p">)</span>
</span><span id="PDBInformation.get_contacts-1239"><a href="#PDBInformation.get_contacts-1239"><span class="linenos">1239</span></a>                    <span class="n">contacts_set</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">res1</span> <span class="o">-</span> <span class="n">shift1</span><span class="p">,</span> <span class="n">res2</span> <span class="o">-</span> <span class="n">shift2</span><span class="p">))</span>
</span><span id="PDBInformation.get_contacts-1240"><a href="#PDBInformation.get_contacts-1240"><span class="linenos">1240</span></a>                <span class="k">else</span><span class="p">:</span>
</span><span id="PDBInformation.get_contacts-1241"><a href="#PDBInformation.get_contacts-1241"><span class="linenos">1241</span></a>                    <span class="n">contacts_set</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">res1</span><span class="p">,</span> <span class="n">res2</span><span class="p">))</span>
</span><span id="PDBInformation.get_contacts-1242"><a href="#PDBInformation.get_contacts-1242"><span class="linenos">1242</span></a>        <span class="k">return</span> <span class="n">contacts_set</span>
</span></pre></div>


            <div class="docstring"><p>Get contacts from the structure attribute where the distance between two residues is less than the threshold.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>ca_only</strong> (bool):
If true, only consider alpha-carbon to alpha-carbon distances.</li>
<li><strong>threshold</strong> (float):
Maximum distance to consider between two atoms.</li>
<li><strong>chain1</strong> (str):
Chain id corresponding to the first column of residues in the structure.</li>
<li><strong>chain2</strong> (str):
Chain id corresponding to the second column of residues in the structure.</li>
<li><strong>auth_contacts</strong> (bool):
True if you want alt_ids for residues indices, False if cif residue indexing is needed.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>contacts_set</strong> (set of tuple of ints):
Set of contacts, tuples with "residue1" and "residue2" from the structure that are within the distance threshold.</li>
</ul>
</div>


                            </div>
                            <div class="inherited">
                                <h5>Inherited Members</h5>
                                <dl>
                                    <div><dt><a href="#StructureInformation">StructureInformation</a></dt>
                                <dd id="PDBInformation.fetch_pdb" class="function"><a href="#StructureInformation.fetch_pdb">fetch_pdb</a></dd>
                <dd id="PDBInformation.read_mmCIF_file" class="function"><a href="#StructureInformation.read_mmCIF_file">read_mmCIF_file</a></dd>
                <dd id="PDBInformation.read_pdb_file" class="function"><a href="#StructureInformation.read_pdb_file">read_pdb_file</a></dd>
                <dd id="PDBInformation.write_contacts_set" class="function"><a href="#StructureInformation.write_contacts_set">write_contacts_set</a></dd>

            </div>
                                </dl>
                            </div>
                </section>
                <section id="MSATools">
                            <input id="MSATools-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr class">
            
    <span class="def">class</span>
    <span class="name">MSATools</span>:

                <label class="view-source-button" for="MSATools-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MSATools"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MSATools-8"><a href="#MSATools-8"><span class="linenos">  8</span></a><span class="k">class</span> <span class="nc">MSATools</span><span class="p">:</span>
</span><span id="MSATools-9"><a href="#MSATools-9"><span class="linenos">  9</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools-10"><a href="#MSATools-10"><span class="linenos"> 10</span></a><span class="sd">    Tools and interface for encapsulating MSA data and providing functionality for filtering and analysis.</span>
</span><span id="MSATools-11"><a href="#MSATools-11"><span class="linenos"> 11</span></a>
</span><span id="MSATools-12"><a href="#MSATools-12"><span class="linenos"> 12</span></a><span class="sd">    Parameters</span>
</span><span id="MSATools-13"><a href="#MSATools-13"><span class="linenos"> 13</span></a><span class="sd">    ----------</span>
</span><span id="MSATools-14"><a href="#MSATools-14"><span class="linenos"> 14</span></a><span class="sd">    MSA : list of tuple of str, str</span>
</span><span id="MSATools-15"><a href="#MSATools-15"><span class="linenos"> 15</span></a><span class="sd">        Loaded MSA that is a list of tuples where the first element is the header and the second element is its corresponding sequence.</span>
</span><span id="MSATools-16"><a href="#MSATools-16"><span class="linenos"> 16</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="MSATools-17"><a href="#MSATools-17"><span class="linenos"> 17</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">MSA</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]):</span>
</span><span id="MSATools-18"><a href="#MSATools-18"><span class="linenos"> 18</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span> <span class="o">=</span> <span class="n">MSA</span>
</span><span id="MSATools-19"><a href="#MSATools-19"><span class="linenos"> 19</span></a>    
</span><span id="MSATools-20"><a href="#MSATools-20"><span class="linenos"> 20</span></a>    <span class="nd">@staticmethod</span>
</span><span id="MSATools-21"><a href="#MSATools-21"><span class="linenos"> 21</span></a>    <span class="k">def</span> <span class="nf">load_from_file</span><span class="p">(</span><span class="n">msa_source</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="s1">&#39;MSATools&#39;</span><span class="p">:</span>
</span><span id="MSATools-22"><a href="#MSATools-22"><span class="linenos"> 22</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools-23"><a href="#MSATools-23"><span class="linenos"> 23</span></a><span class="sd">        Generates MSATools object from an MSA file in &quot;.afa&quot; format.</span>
</span><span id="MSATools-24"><a href="#MSATools-24"><span class="linenos"> 24</span></a>
</span><span id="MSATools-25"><a href="#MSATools-25"><span class="linenos"> 25</span></a><span class="sd">        Parameters</span>
</span><span id="MSATools-26"><a href="#MSATools-26"><span class="linenos"> 26</span></a><span class="sd">        ----------</span>
</span><span id="MSATools-27"><a href="#MSATools-27"><span class="linenos"> 27</span></a><span class="sd">        msa_source : str or io.IOBase</span>
</span><span id="MSATools-28"><a href="#MSATools-28"><span class="linenos"> 28</span></a><span class="sd">            Filepath or IOBase of the MSA in &quot;.afa&quot; format that is provided.</span>
</span><span id="MSATools-29"><a href="#MSATools-29"><span class="linenos"> 29</span></a><span class="sd">        </span>
</span><span id="MSATools-30"><a href="#MSATools-30"><span class="linenos"> 30</span></a><span class="sd">        Returns</span>
</span><span id="MSATools-31"><a href="#MSATools-31"><span class="linenos"> 31</span></a><span class="sd">        -------</span>
</span><span id="MSATools-32"><a href="#MSATools-32"><span class="linenos"> 32</span></a><span class="sd">        MSATools</span>
</span><span id="MSATools-33"><a href="#MSATools-33"><span class="linenos"> 33</span></a><span class="sd">            An MSATools instance with the appropriate list of (header, sequence) tuples where sequences are simplified and converted to single line format.</span>
</span><span id="MSATools-34"><a href="#MSATools-34"><span class="linenos"> 34</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools-35"><a href="#MSATools-35"><span class="linenos"> 35</span></a>        <span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
</span><span id="MSATools-36"><a href="#MSATools-36"><span class="linenos"> 36</span></a>        <span class="n">msa_entries</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MSATools-37"><a href="#MSATools-37"><span class="linenos"> 37</span></a>        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">msa_source</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
</span><span id="MSATools-38"><a href="#MSATools-38"><span class="linenos"> 38</span></a>            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">msa_source</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="MSATools-39"><a href="#MSATools-39"><span class="linenos"> 39</span></a>                <span class="n">data</span> <span class="o">=</span> <span class="n">fs</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
</span><span id="MSATools-40"><a href="#MSATools-40"><span class="linenos"> 40</span></a>        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">msa_source</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">BytesIO</span><span class="p">):</span>
</span><span id="MSATools-41"><a href="#MSATools-41"><span class="linenos"> 41</span></a>            <span class="n">data</span> <span class="o">=</span> <span class="n">msa_source</span><span class="o">.</span><span class="n">getvalue</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
</span><span id="MSATools-42"><a href="#MSATools-42"><span class="linenos"> 42</span></a>        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">msa_source</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">):</span>
</span><span id="MSATools-43"><a href="#MSATools-43"><span class="linenos"> 43</span></a>            <span class="n">data</span> <span class="o">=</span> <span class="n">msa_source</span><span class="o">.</span><span class="n">getvalue</span><span class="p">()</span>
</span><span id="MSATools-44"><a href="#MSATools-44"><span class="linenos"> 44</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MSATools-45"><a href="#MSATools-45"><span class="linenos"> 45</span></a>            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;msa_file is not bytesIO, stringIO, or a filepath.&quot;</span><span class="p">)</span>
</span><span id="MSATools-46"><a href="#MSATools-46"><span class="linenos"> 46</span></a>        <span class="n">split_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
</span><span id="MSATools-47"><a href="#MSATools-47"><span class="linenos"> 47</span></a>        <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">split_data</span><span class="p">:</span>
</span><span id="MSATools-48"><a href="#MSATools-48"><span class="linenos"> 48</span></a>            <span class="n">line_split_entry</span> <span class="o">=</span> <span class="n">entry</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="MSATools-49"><a href="#MSATools-49"><span class="linenos"> 49</span></a>            <span class="n">header</span> <span class="o">=</span> <span class="n">line_split_entry</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MSATools-50"><a href="#MSATools-50"><span class="linenos"> 50</span></a>            <span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_split_entry</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
</span><span id="MSATools-51"><a href="#MSATools-51"><span class="linenos"> 51</span></a>            <span class="n">msa_entries</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="s2">&quot;&gt;&quot;</span><span class="o">+</span><span class="n">header</span><span class="p">,</span> <span class="n">sequence</span><span class="p">))</span>
</span><span id="MSATools-52"><a href="#MSATools-52"><span class="linenos"> 52</span></a>        <span class="k">return</span> <span class="n">MSATools</span><span class="p">(</span><span class="n">msa_entries</span><span class="p">)</span>
</span><span id="MSATools-53"><a href="#MSATools-53"><span class="linenos"> 53</span></a>
</span><span id="MSATools-54"><a href="#MSATools-54"><span class="linenos"> 54</span></a>    <span class="nd">@staticmethod</span>
</span><span id="MSATools-55"><a href="#MSATools-55"><span class="linenos"> 55</span></a>    <span class="k">def</span> <span class="nf">get_sequence_max_cont_gaps</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
</span><span id="MSATools-56"><a href="#MSATools-56"><span class="linenos"> 56</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools-57"><a href="#MSATools-57"><span class="linenos"> 57</span></a><span class="sd">        Find maximum number of continuous gaps in a specific sequence.</span>
</span><span id="MSATools-58"><a href="#MSATools-58"><span class="linenos"> 58</span></a>
</span><span id="MSATools-59"><a href="#MSATools-59"><span class="linenos"> 59</span></a><span class="sd">        Parameters</span>
</span><span id="MSATools-60"><a href="#MSATools-60"><span class="linenos"> 60</span></a><span class="sd">        ----------</span>
</span><span id="MSATools-61"><a href="#MSATools-61"><span class="linenos"> 61</span></a><span class="sd">        sequence : str</span>
</span><span id="MSATools-62"><a href="#MSATools-62"><span class="linenos"> 62</span></a><span class="sd">            Sequence of characters, potentially containing multiple of &#39;-&#39;, a gap character.</span>
</span><span id="MSATools-63"><a href="#MSATools-63"><span class="linenos"> 63</span></a><span class="sd">        </span>
</span><span id="MSATools-64"><a href="#MSATools-64"><span class="linenos"> 64</span></a><span class="sd">        Returns</span>
</span><span id="MSATools-65"><a href="#MSATools-65"><span class="linenos"> 65</span></a><span class="sd">        -------</span>
</span><span id="MSATools-66"><a href="#MSATools-66"><span class="linenos"> 66</span></a><span class="sd">        int</span>
</span><span id="MSATools-67"><a href="#MSATools-67"><span class="linenos"> 67</span></a><span class="sd">            The maximum number of continuous gaps in a sequence.</span>
</span><span id="MSATools-68"><a href="#MSATools-68"><span class="linenos"> 68</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools-69"><a href="#MSATools-69"><span class="linenos"> 69</span></a>        <span class="n">dash_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;-+&quot;</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</span><span id="MSATools-70"><a href="#MSATools-70"><span class="linenos"> 70</span></a>        <span class="n">gap_counts</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">match</span><span class="p">)</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">dash_match</span><span class="p">]</span>
</span><span id="MSATools-71"><a href="#MSATools-71"><span class="linenos"> 71</span></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gap_counts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="MSATools-72"><a href="#MSATools-72"><span class="linenos"> 72</span></a>            <span class="k">return</span> <span class="nb">max</span><span class="p">(</span><span class="n">gap_counts</span><span class="p">)</span>
</span><span id="MSATools-73"><a href="#MSATools-73"><span class="linenos"> 73</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MSATools-74"><a href="#MSATools-74"><span class="linenos"> 74</span></a>            <span class="k">return</span> <span class="mi">0</span>
</span><span id="MSATools-75"><a href="#MSATools-75"><span class="linenos"> 75</span></a>    
</span><span id="MSATools-76"><a href="#MSATools-76"><span class="linenos"> 76</span></a>    <span class="k">def</span> <span class="nf">gap_frequency</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]:</span>
</span><span id="MSATools-77"><a href="#MSATools-77"><span class="linenos"> 77</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools-78"><a href="#MSATools-78"><span class="linenos"> 78</span></a><span class="sd">        Calculates the frequency of maximum continuous gaps throughout the MSA where the key corresponds to the number of continous gaps and the value corresponds to the number of sequences or the cumulative percentage of their sequences.</span>
</span><span id="MSATools-79"><a href="#MSATools-79"><span class="linenos"> 79</span></a>
</span><span id="MSATools-80"><a href="#MSATools-80"><span class="linenos"> 80</span></a><span class="sd">        Returns</span>
</span><span id="MSATools-81"><a href="#MSATools-81"><span class="linenos"> 81</span></a><span class="sd">        -------</span>
</span><span id="MSATools-82"><a href="#MSATools-82"><span class="linenos"> 82</span></a><span class="sd">        tuple of dict of int, int and dict of int, int</span>
</span><span id="MSATools-83"><a href="#MSATools-83"><span class="linenos"> 83</span></a><span class="sd">            Two element tuple where first element is a frequency count dictionary and the second element is a cumulative percentage of sequences with a specific maximum number of continous gaps.</span>
</span><span id="MSATools-84"><a href="#MSATools-84"><span class="linenos"> 84</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools-85"><a href="#MSATools-85"><span class="linenos"> 85</span></a>        <span class="n">max_gap_counts</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MSATools-86"><a href="#MSATools-86"><span class="linenos"> 86</span></a>        <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools-87"><a href="#MSATools-87"><span class="linenos"> 87</span></a>            <span class="n">max_gap_counts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">MSATools</span><span class="o">.</span><span class="n">get_sequence_max_cont_gaps</span><span class="p">(</span><span class="n">sequence</span><span class="p">))</span>
</span><span id="MSATools-88"><a href="#MSATools-88"><span class="linenos"> 88</span></a>        <span class="n">frequency_count_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">max_gap_counts</span><span class="p">))</span>
</span><span id="MSATools-89"><a href="#MSATools-89"><span class="linenos"> 89</span></a>        <span class="n">cumul_perc_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="MSATools-90"><a href="#MSATools-90"><span class="linenos"> 90</span></a>        <span class="n">cumul_count</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="MSATools-91"><a href="#MSATools-91"><span class="linenos"> 91</span></a>        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">frequency_count_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
</span><span id="MSATools-92"><a href="#MSATools-92"><span class="linenos"> 92</span></a>            <span class="n">value</span> <span class="o">=</span> <span class="n">frequency_count_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
</span><span id="MSATools-93"><a href="#MSATools-93"><span class="linenos"> 93</span></a>            <span class="n">cumul_count</span> <span class="o">+=</span> <span class="n">value</span>
</span><span id="MSATools-94"><a href="#MSATools-94"><span class="linenos"> 94</span></a>            <span class="n">cumul_perc_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">cumul_count</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">)</span>
</span><span id="MSATools-95"><a href="#MSATools-95"><span class="linenos"> 95</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">frequency_count_dict</span><span class="p">,</span> <span class="n">cumul_perc_dict</span><span class="p">)</span>
</span><span id="MSATools-96"><a href="#MSATools-96"><span class="linenos"> 96</span></a>    
</span><span id="MSATools-97"><a href="#MSATools-97"><span class="linenos"> 97</span></a>    <span class="k">def</span> <span class="nf">filter_by_continuous_gaps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">max_gaps</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
</span><span id="MSATools-98"><a href="#MSATools-98"><span class="linenos"> 98</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools-99"><a href="#MSATools-99"><span class="linenos"> 99</span></a><span class="sd">        Filter out entries in your MSA by the number of maximum continuous gaps specified unless None is provided. Also, removes .s and lowercase letters from the sequence.</span>
</span><span id="MSATools-100"><a href="#MSATools-100"><span class="linenos">100</span></a>
</span><span id="MSATools-101"><a href="#MSATools-101"><span class="linenos">101</span></a><span class="sd">        Parameters</span>
</span><span id="MSATools-102"><a href="#MSATools-102"><span class="linenos">102</span></a><span class="sd">        ----------</span>
</span><span id="MSATools-103"><a href="#MSATools-103"><span class="linenos">103</span></a><span class="sd">        max_gaps : int</span>
</span><span id="MSATools-104"><a href="#MSATools-104"><span class="linenos">104</span></a><span class="sd">            The maximum allowed number of continuous gaps in a sequence</span>
</span><span id="MSATools-105"><a href="#MSATools-105"><span class="linenos">105</span></a>
</span><span id="MSATools-106"><a href="#MSATools-106"><span class="linenos">106</span></a><span class="sd">        Returns</span>
</span><span id="MSATools-107"><a href="#MSATools-107"><span class="linenos">107</span></a><span class="sd">        -------</span>
</span><span id="MSATools-108"><a href="#MSATools-108"><span class="linenos">108</span></a><span class="sd">        list of tuple of str, str</span>
</span><span id="MSATools-109"><a href="#MSATools-109"><span class="linenos">109</span></a><span class="sd">            List of entries that are valid in that their sequences&#39; number of maximum continuous gaps is within the threshold supplied as `max_gaps`. </span>
</span><span id="MSATools-110"><a href="#MSATools-110"><span class="linenos">110</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools-111"><a href="#MSATools-111"><span class="linenos">111</span></a>        <span class="n">table</span> <span class="o">=</span> <span class="nb">str</span><span class="o">.</span><span class="n">maketrans</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">string</span><span class="o">.</span><span class="n">ascii_lowercase</span><span class="o">+</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
</span><span id="MSATools-112"><a href="#MSATools-112"><span class="linenos">112</span></a>        <span class="k">if</span> <span class="n">max_gaps</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="MSATools-113"><a href="#MSATools-113"><span class="linenos">113</span></a>            <span class="n">kept_entries</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MSATools-114"><a href="#MSATools-114"><span class="linenos">114</span></a>            <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools-115"><a href="#MSATools-115"><span class="linenos">115</span></a>                <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>
</span><span id="MSATools-116"><a href="#MSATools-116"><span class="linenos">116</span></a>                <span class="n">kept_entries</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">header</span><span class="p">,</span> <span class="n">sequence</span><span class="p">))</span>
</span><span id="MSATools-117"><a href="#MSATools-117"><span class="linenos">117</span></a>            <span class="k">return</span> <span class="n">kept_entries</span>
</span><span id="MSATools-118"><a href="#MSATools-118"><span class="linenos">118</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MSATools-119"><a href="#MSATools-119"><span class="linenos">119</span></a>            <span class="n">kept_entries</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MSATools-120"><a href="#MSATools-120"><span class="linenos">120</span></a>            <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools-121"><a href="#MSATools-121"><span class="linenos">121</span></a>                <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>
</span><span id="MSATools-122"><a href="#MSATools-122"><span class="linenos">122</span></a>                <span class="k">if</span> <span class="n">MSATools</span><span class="o">.</span><span class="n">get_sequence_max_cont_gaps</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">max_gaps</span><span class="p">:</span>
</span><span id="MSATools-123"><a href="#MSATools-123"><span class="linenos">123</span></a>                    <span class="n">kept_entries</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">header</span><span class="p">,</span> <span class="n">sequence</span><span class="p">))</span>
</span><span id="MSATools-124"><a href="#MSATools-124"><span class="linenos">124</span></a>        <span class="k">return</span> <span class="n">kept_entries</span>
</span><span id="MSATools-125"><a href="#MSATools-125"><span class="linenos">125</span></a>
</span><span id="MSATools-126"><a href="#MSATools-126"><span class="linenos">126</span></a>    <span class="k">def</span> <span class="nf">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">destination</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="MSATools-127"><a href="#MSATools-127"><span class="linenos">127</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools-128"><a href="#MSATools-128"><span class="linenos">128</span></a><span class="sd">        Writes this MSA&#39;s headers and sequences to the destination specified.</span>
</span><span id="MSATools-129"><a href="#MSATools-129"><span class="linenos">129</span></a><span class="sd">        </span>
</span><span id="MSATools-130"><a href="#MSATools-130"><span class="linenos">130</span></a><span class="sd">        Parameters</span>
</span><span id="MSATools-131"><a href="#MSATools-131"><span class="linenos">131</span></a><span class="sd">        ----------</span>
</span><span id="MSATools-132"><a href="#MSATools-132"><span class="linenos">132</span></a><span class="sd">        destination : str or io.IOBase</span>
</span><span id="MSATools-133"><a href="#MSATools-133"><span class="linenos">133</span></a><span class="sd">            Filepath or IO to write the MSA supplied to.</span>
</span><span id="MSATools-134"><a href="#MSATools-134"><span class="linenos">134</span></a>
</span><span id="MSATools-135"><a href="#MSATools-135"><span class="linenos">135</span></a><span class="sd">        Returns</span>
</span><span id="MSATools-136"><a href="#MSATools-136"><span class="linenos">136</span></a><span class="sd">        -------</span>
</span><span id="MSATools-137"><a href="#MSATools-137"><span class="linenos">137</span></a><span class="sd">        None</span>
</span><span id="MSATools-138"><a href="#MSATools-138"><span class="linenos">138</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools-139"><a href="#MSATools-139"><span class="linenos">139</span></a>        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
</span><span id="MSATools-140"><a href="#MSATools-140"><span class="linenos">140</span></a>            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="MSATools-141"><a href="#MSATools-141"><span class="linenos">141</span></a>                <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools-142"><a href="#MSATools-142"><span class="linenos">142</span></a>                    <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
</span><span id="MSATools-143"><a href="#MSATools-143"><span class="linenos">143</span></a>                    <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="MSATools-144"><a href="#MSATools-144"><span class="linenos">144</span></a>                    <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span id="MSATools-145"><a href="#MSATools-145"><span class="linenos">145</span></a>                    <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="MSATools-146"><a href="#MSATools-146"><span class="linenos">146</span></a>        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">):</span>
</span><span id="MSATools-147"><a href="#MSATools-147"><span class="linenos">147</span></a>            <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools-148"><a href="#MSATools-148"><span class="linenos">148</span></a>                <span class="n">destination</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
</span><span id="MSATools-149"><a href="#MSATools-149"><span class="linenos">149</span></a>                <span class="n">destination</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="MSATools-150"><a href="#MSATools-150"><span class="linenos">150</span></a>                <span class="n">destination</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span id="MSATools-151"><a href="#MSATools-151"><span class="linenos">151</span></a>                <span class="n">destination</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="MSATools-152"><a href="#MSATools-152"><span class="linenos">152</span></a>
</span><span id="MSATools-153"><a href="#MSATools-153"><span class="linenos">153</span></a>    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="MSATools-154"><a href="#MSATools-154"><span class="linenos">154</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools-155"><a href="#MSATools-155"><span class="linenos">155</span></a><span class="sd">        Returns the number of sequences, and equivalently, the number of headers in the MSA.</span>
</span><span id="MSATools-156"><a href="#MSATools-156"><span class="linenos">156</span></a>
</span><span id="MSATools-157"><a href="#MSATools-157"><span class="linenos">157</span></a><span class="sd">        Returns</span>
</span><span id="MSATools-158"><a href="#MSATools-158"><span class="linenos">158</span></a><span class="sd">        -------</span>
</span><span id="MSATools-159"><a href="#MSATools-159"><span class="linenos">159</span></a><span class="sd">        int</span>
</span><span id="MSATools-160"><a href="#MSATools-160"><span class="linenos">160</span></a><span class="sd">            length of the MSA list of header, sequence tuples.</span>
</span><span id="MSATools-161"><a href="#MSATools-161"><span class="linenos">161</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools-162"><a href="#MSATools-162"><span class="linenos">162</span></a>        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Tools and interface for encapsulating MSA data and providing functionality for filtering and analysis.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>MSA</strong> (list of tuple of str, str):
Loaded MSA that is a list of tuples where the first element is the header and the second element is its corresponding sequence.</li>
</ul>
</div>


                            <div id="MSATools.__init__" class="classattr">
                                        <input id="MSATools.__init__-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="name">MSATools</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">MSA</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span></span>)</span>

                <label class="view-source-button" for="MSATools.__init__-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MSATools.__init__"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MSATools.__init__-17"><a href="#MSATools.__init__-17"><span class="linenos">17</span></a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">MSA</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]):</span>
</span><span id="MSATools.__init__-18"><a href="#MSATools.__init__-18"><span class="linenos">18</span></a>        <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span> <span class="o">=</span> <span class="n">MSA</span>
</span></pre></div>


    

                            </div>
                            <div id="MSATools.MSA" class="classattr">
                                <div class="attr variable">
            <span class="name">MSA</span>

        
    </div>
    <a class="headerlink" href="#MSATools.MSA"></a>
    
    

                            </div>
                            <div id="MSATools.load_from_file" class="classattr">
                                        <input id="MSATools.load_from_file-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">load_from_file</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">msa_source</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">]</span></span><span class="return-annotation">) -> <span class="n"><a href="#MSATools">MSATools</a></span>:</span></span>

                <label class="view-source-button" for="MSATools.load_from_file-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MSATools.load_from_file"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MSATools.load_from_file-20"><a href="#MSATools.load_from_file-20"><span class="linenos">20</span></a>    <span class="nd">@staticmethod</span>
</span><span id="MSATools.load_from_file-21"><a href="#MSATools.load_from_file-21"><span class="linenos">21</span></a>    <span class="k">def</span> <span class="nf">load_from_file</span><span class="p">(</span><span class="n">msa_source</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="s1">&#39;MSATools&#39;</span><span class="p">:</span>
</span><span id="MSATools.load_from_file-22"><a href="#MSATools.load_from_file-22"><span class="linenos">22</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools.load_from_file-23"><a href="#MSATools.load_from_file-23"><span class="linenos">23</span></a><span class="sd">        Generates MSATools object from an MSA file in &quot;.afa&quot; format.</span>
</span><span id="MSATools.load_from_file-24"><a href="#MSATools.load_from_file-24"><span class="linenos">24</span></a>
</span><span id="MSATools.load_from_file-25"><a href="#MSATools.load_from_file-25"><span class="linenos">25</span></a><span class="sd">        Parameters</span>
</span><span id="MSATools.load_from_file-26"><a href="#MSATools.load_from_file-26"><span class="linenos">26</span></a><span class="sd">        ----------</span>
</span><span id="MSATools.load_from_file-27"><a href="#MSATools.load_from_file-27"><span class="linenos">27</span></a><span class="sd">        msa_source : str or io.IOBase</span>
</span><span id="MSATools.load_from_file-28"><a href="#MSATools.load_from_file-28"><span class="linenos">28</span></a><span class="sd">            Filepath or IOBase of the MSA in &quot;.afa&quot; format that is provided.</span>
</span><span id="MSATools.load_from_file-29"><a href="#MSATools.load_from_file-29"><span class="linenos">29</span></a><span class="sd">        </span>
</span><span id="MSATools.load_from_file-30"><a href="#MSATools.load_from_file-30"><span class="linenos">30</span></a><span class="sd">        Returns</span>
</span><span id="MSATools.load_from_file-31"><a href="#MSATools.load_from_file-31"><span class="linenos">31</span></a><span class="sd">        -------</span>
</span><span id="MSATools.load_from_file-32"><a href="#MSATools.load_from_file-32"><span class="linenos">32</span></a><span class="sd">        MSATools</span>
</span><span id="MSATools.load_from_file-33"><a href="#MSATools.load_from_file-33"><span class="linenos">33</span></a><span class="sd">            An MSATools instance with the appropriate list of (header, sequence) tuples where sequences are simplified and converted to single line format.</span>
</span><span id="MSATools.load_from_file-34"><a href="#MSATools.load_from_file-34"><span class="linenos">34</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools.load_from_file-35"><a href="#MSATools.load_from_file-35"><span class="linenos">35</span></a>        <span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
</span><span id="MSATools.load_from_file-36"><a href="#MSATools.load_from_file-36"><span class="linenos">36</span></a>        <span class="n">msa_entries</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MSATools.load_from_file-37"><a href="#MSATools.load_from_file-37"><span class="linenos">37</span></a>        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">msa_source</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
</span><span id="MSATools.load_from_file-38"><a href="#MSATools.load_from_file-38"><span class="linenos">38</span></a>            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">msa_source</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="MSATools.load_from_file-39"><a href="#MSATools.load_from_file-39"><span class="linenos">39</span></a>                <span class="n">data</span> <span class="o">=</span> <span class="n">fs</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
</span><span id="MSATools.load_from_file-40"><a href="#MSATools.load_from_file-40"><span class="linenos">40</span></a>        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">msa_source</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">BytesIO</span><span class="p">):</span>
</span><span id="MSATools.load_from_file-41"><a href="#MSATools.load_from_file-41"><span class="linenos">41</span></a>            <span class="n">data</span> <span class="o">=</span> <span class="n">msa_source</span><span class="o">.</span><span class="n">getvalue</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
</span><span id="MSATools.load_from_file-42"><a href="#MSATools.load_from_file-42"><span class="linenos">42</span></a>        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">msa_source</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">):</span>
</span><span id="MSATools.load_from_file-43"><a href="#MSATools.load_from_file-43"><span class="linenos">43</span></a>            <span class="n">data</span> <span class="o">=</span> <span class="n">msa_source</span><span class="o">.</span><span class="n">getvalue</span><span class="p">()</span>
</span><span id="MSATools.load_from_file-44"><a href="#MSATools.load_from_file-44"><span class="linenos">44</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MSATools.load_from_file-45"><a href="#MSATools.load_from_file-45"><span class="linenos">45</span></a>            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;msa_file is not bytesIO, stringIO, or a filepath.&quot;</span><span class="p">)</span>
</span><span id="MSATools.load_from_file-46"><a href="#MSATools.load_from_file-46"><span class="linenos">46</span></a>        <span class="n">split_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
</span><span id="MSATools.load_from_file-47"><a href="#MSATools.load_from_file-47"><span class="linenos">47</span></a>        <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">split_data</span><span class="p">:</span>
</span><span id="MSATools.load_from_file-48"><a href="#MSATools.load_from_file-48"><span class="linenos">48</span></a>            <span class="n">line_split_entry</span> <span class="o">=</span> <span class="n">entry</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="MSATools.load_from_file-49"><a href="#MSATools.load_from_file-49"><span class="linenos">49</span></a>            <span class="n">header</span> <span class="o">=</span> <span class="n">line_split_entry</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span><span id="MSATools.load_from_file-50"><a href="#MSATools.load_from_file-50"><span class="linenos">50</span></a>            <span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_split_entry</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
</span><span id="MSATools.load_from_file-51"><a href="#MSATools.load_from_file-51"><span class="linenos">51</span></a>            <span class="n">msa_entries</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="s2">&quot;&gt;&quot;</span><span class="o">+</span><span class="n">header</span><span class="p">,</span> <span class="n">sequence</span><span class="p">))</span>
</span><span id="MSATools.load_from_file-52"><a href="#MSATools.load_from_file-52"><span class="linenos">52</span></a>        <span class="k">return</span> <span class="n">MSATools</span><span class="p">(</span><span class="n">msa_entries</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Generates MSATools object from an MSA file in ".afa" format.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>msa_source</strong> (str or io.IOBase):
Filepath or IOBase of the MSA in ".afa" format that is provided.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>MSATools</strong>: An MSATools instance with the appropriate list of (header, sequence) tuples where sequences are simplified and converted to single line format.</li>
</ul>
</div>


                            </div>
                            <div id="MSATools.get_sequence_max_cont_gaps" class="classattr">
                                        <input id="MSATools.get_sequence_max_cont_gaps-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
                    <div class="decorator">@staticmethod</div>

        <span class="def">def</span>
        <span class="name">get_sequence_max_cont_gaps</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="nb">int</span>:</span></span>

                <label class="view-source-button" for="MSATools.get_sequence_max_cont_gaps-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MSATools.get_sequence_max_cont_gaps"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MSATools.get_sequence_max_cont_gaps-54"><a href="#MSATools.get_sequence_max_cont_gaps-54"><span class="linenos">54</span></a>    <span class="nd">@staticmethod</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-55"><a href="#MSATools.get_sequence_max_cont_gaps-55"><span class="linenos">55</span></a>    <span class="k">def</span> <span class="nf">get_sequence_max_cont_gaps</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-56"><a href="#MSATools.get_sequence_max_cont_gaps-56"><span class="linenos">56</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-57"><a href="#MSATools.get_sequence_max_cont_gaps-57"><span class="linenos">57</span></a><span class="sd">        Find maximum number of continuous gaps in a specific sequence.</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-58"><a href="#MSATools.get_sequence_max_cont_gaps-58"><span class="linenos">58</span></a>
</span><span id="MSATools.get_sequence_max_cont_gaps-59"><a href="#MSATools.get_sequence_max_cont_gaps-59"><span class="linenos">59</span></a><span class="sd">        Parameters</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-60"><a href="#MSATools.get_sequence_max_cont_gaps-60"><span class="linenos">60</span></a><span class="sd">        ----------</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-61"><a href="#MSATools.get_sequence_max_cont_gaps-61"><span class="linenos">61</span></a><span class="sd">        sequence : str</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-62"><a href="#MSATools.get_sequence_max_cont_gaps-62"><span class="linenos">62</span></a><span class="sd">            Sequence of characters, potentially containing multiple of &#39;-&#39;, a gap character.</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-63"><a href="#MSATools.get_sequence_max_cont_gaps-63"><span class="linenos">63</span></a><span class="sd">        </span>
</span><span id="MSATools.get_sequence_max_cont_gaps-64"><a href="#MSATools.get_sequence_max_cont_gaps-64"><span class="linenos">64</span></a><span class="sd">        Returns</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-65"><a href="#MSATools.get_sequence_max_cont_gaps-65"><span class="linenos">65</span></a><span class="sd">        -------</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-66"><a href="#MSATools.get_sequence_max_cont_gaps-66"><span class="linenos">66</span></a><span class="sd">        int</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-67"><a href="#MSATools.get_sequence_max_cont_gaps-67"><span class="linenos">67</span></a><span class="sd">            The maximum number of continuous gaps in a sequence.</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-68"><a href="#MSATools.get_sequence_max_cont_gaps-68"><span class="linenos">68</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-69"><a href="#MSATools.get_sequence_max_cont_gaps-69"><span class="linenos">69</span></a>        <span class="n">dash_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;-+&quot;</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-70"><a href="#MSATools.get_sequence_max_cont_gaps-70"><span class="linenos">70</span></a>        <span class="n">gap_counts</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">match</span><span class="p">)</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">dash_match</span><span class="p">]</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-71"><a href="#MSATools.get_sequence_max_cont_gaps-71"><span class="linenos">71</span></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gap_counts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-72"><a href="#MSATools.get_sequence_max_cont_gaps-72"><span class="linenos">72</span></a>            <span class="k">return</span> <span class="nb">max</span><span class="p">(</span><span class="n">gap_counts</span><span class="p">)</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-73"><a href="#MSATools.get_sequence_max_cont_gaps-73"><span class="linenos">73</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MSATools.get_sequence_max_cont_gaps-74"><a href="#MSATools.get_sequence_max_cont_gaps-74"><span class="linenos">74</span></a>            <span class="k">return</span> <span class="mi">0</span>
</span></pre></div>


            <div class="docstring"><p>Find maximum number of continuous gaps in a specific sequence.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>sequence</strong> (str):
Sequence of characters, potentially containing multiple of '-', a gap character.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>int</strong>: The maximum number of continuous gaps in a sequence.</li>
</ul>
</div>


                            </div>
                            <div id="MSATools.gap_frequency" class="classattr">
                                        <input id="MSATools.gap_frequency-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">gap_frequency</span><span class="signature pdoc-code condensed">(<span class="param"><span class="bp">self</span></span><span class="return-annotation">) -> <span class="nb">tuple</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="MSATools.gap_frequency-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MSATools.gap_frequency"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MSATools.gap_frequency-76"><a href="#MSATools.gap_frequency-76"><span class="linenos">76</span></a>    <span class="k">def</span> <span class="nf">gap_frequency</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]:</span>
</span><span id="MSATools.gap_frequency-77"><a href="#MSATools.gap_frequency-77"><span class="linenos">77</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools.gap_frequency-78"><a href="#MSATools.gap_frequency-78"><span class="linenos">78</span></a><span class="sd">        Calculates the frequency of maximum continuous gaps throughout the MSA where the key corresponds to the number of continous gaps and the value corresponds to the number of sequences or the cumulative percentage of their sequences.</span>
</span><span id="MSATools.gap_frequency-79"><a href="#MSATools.gap_frequency-79"><span class="linenos">79</span></a>
</span><span id="MSATools.gap_frequency-80"><a href="#MSATools.gap_frequency-80"><span class="linenos">80</span></a><span class="sd">        Returns</span>
</span><span id="MSATools.gap_frequency-81"><a href="#MSATools.gap_frequency-81"><span class="linenos">81</span></a><span class="sd">        -------</span>
</span><span id="MSATools.gap_frequency-82"><a href="#MSATools.gap_frequency-82"><span class="linenos">82</span></a><span class="sd">        tuple of dict of int, int and dict of int, int</span>
</span><span id="MSATools.gap_frequency-83"><a href="#MSATools.gap_frequency-83"><span class="linenos">83</span></a><span class="sd">            Two element tuple where first element is a frequency count dictionary and the second element is a cumulative percentage of sequences with a specific maximum number of continous gaps.</span>
</span><span id="MSATools.gap_frequency-84"><a href="#MSATools.gap_frequency-84"><span class="linenos">84</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools.gap_frequency-85"><a href="#MSATools.gap_frequency-85"><span class="linenos">85</span></a>        <span class="n">max_gap_counts</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MSATools.gap_frequency-86"><a href="#MSATools.gap_frequency-86"><span class="linenos">86</span></a>        <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools.gap_frequency-87"><a href="#MSATools.gap_frequency-87"><span class="linenos">87</span></a>            <span class="n">max_gap_counts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">MSATools</span><span class="o">.</span><span class="n">get_sequence_max_cont_gaps</span><span class="p">(</span><span class="n">sequence</span><span class="p">))</span>
</span><span id="MSATools.gap_frequency-88"><a href="#MSATools.gap_frequency-88"><span class="linenos">88</span></a>        <span class="n">frequency_count_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">max_gap_counts</span><span class="p">))</span>
</span><span id="MSATools.gap_frequency-89"><a href="#MSATools.gap_frequency-89"><span class="linenos">89</span></a>        <span class="n">cumul_perc_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="MSATools.gap_frequency-90"><a href="#MSATools.gap_frequency-90"><span class="linenos">90</span></a>        <span class="n">cumul_count</span> <span class="o">=</span> <span class="mi">0</span>
</span><span id="MSATools.gap_frequency-91"><a href="#MSATools.gap_frequency-91"><span class="linenos">91</span></a>        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">frequency_count_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
</span><span id="MSATools.gap_frequency-92"><a href="#MSATools.gap_frequency-92"><span class="linenos">92</span></a>            <span class="n">value</span> <span class="o">=</span> <span class="n">frequency_count_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
</span><span id="MSATools.gap_frequency-93"><a href="#MSATools.gap_frequency-93"><span class="linenos">93</span></a>            <span class="n">cumul_count</span> <span class="o">+=</span> <span class="n">value</span>
</span><span id="MSATools.gap_frequency-94"><a href="#MSATools.gap_frequency-94"><span class="linenos">94</span></a>            <span class="n">cumul_perc_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">cumul_count</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">)</span>
</span><span id="MSATools.gap_frequency-95"><a href="#MSATools.gap_frequency-95"><span class="linenos">95</span></a>        <span class="k">return</span> <span class="p">(</span><span class="n">frequency_count_dict</span><span class="p">,</span> <span class="n">cumul_perc_dict</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Calculates the frequency of maximum continuous gaps throughout the MSA where the key corresponds to the number of continous gaps and the value corresponds to the number of sequences or the cumulative percentage of their sequences.</p>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>tuple of dict of int, int and dict of int, int</strong>: Two element tuple where first element is a frequency count dictionary and the second element is a cumulative percentage of sequences with a specific maximum number of continous gaps.</li>
</ul>
</div>


                            </div>
                            <div id="MSATools.filter_by_continuous_gaps" class="classattr">
                                        <input id="MSATools.filter_by_continuous_gaps-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">filter_by_continuous_gaps</span><span class="signature pdoc-code condensed">(<span class="param"><span class="bp">self</span>, </span><span class="param"><span class="n">max_gaps</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span></span><span class="return-annotation">) -> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span>:</span></span>

                <label class="view-source-button" for="MSATools.filter_by_continuous_gaps-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MSATools.filter_by_continuous_gaps"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MSATools.filter_by_continuous_gaps-97"><a href="#MSATools.filter_by_continuous_gaps-97"><span class="linenos"> 97</span></a>    <span class="k">def</span> <span class="nf">filter_by_continuous_gaps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">max_gaps</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
</span><span id="MSATools.filter_by_continuous_gaps-98"><a href="#MSATools.filter_by_continuous_gaps-98"><span class="linenos"> 98</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools.filter_by_continuous_gaps-99"><a href="#MSATools.filter_by_continuous_gaps-99"><span class="linenos"> 99</span></a><span class="sd">        Filter out entries in your MSA by the number of maximum continuous gaps specified unless None is provided. Also, removes .s and lowercase letters from the sequence.</span>
</span><span id="MSATools.filter_by_continuous_gaps-100"><a href="#MSATools.filter_by_continuous_gaps-100"><span class="linenos">100</span></a>
</span><span id="MSATools.filter_by_continuous_gaps-101"><a href="#MSATools.filter_by_continuous_gaps-101"><span class="linenos">101</span></a><span class="sd">        Parameters</span>
</span><span id="MSATools.filter_by_continuous_gaps-102"><a href="#MSATools.filter_by_continuous_gaps-102"><span class="linenos">102</span></a><span class="sd">        ----------</span>
</span><span id="MSATools.filter_by_continuous_gaps-103"><a href="#MSATools.filter_by_continuous_gaps-103"><span class="linenos">103</span></a><span class="sd">        max_gaps : int</span>
</span><span id="MSATools.filter_by_continuous_gaps-104"><a href="#MSATools.filter_by_continuous_gaps-104"><span class="linenos">104</span></a><span class="sd">            The maximum allowed number of continuous gaps in a sequence</span>
</span><span id="MSATools.filter_by_continuous_gaps-105"><a href="#MSATools.filter_by_continuous_gaps-105"><span class="linenos">105</span></a>
</span><span id="MSATools.filter_by_continuous_gaps-106"><a href="#MSATools.filter_by_continuous_gaps-106"><span class="linenos">106</span></a><span class="sd">        Returns</span>
</span><span id="MSATools.filter_by_continuous_gaps-107"><a href="#MSATools.filter_by_continuous_gaps-107"><span class="linenos">107</span></a><span class="sd">        -------</span>
</span><span id="MSATools.filter_by_continuous_gaps-108"><a href="#MSATools.filter_by_continuous_gaps-108"><span class="linenos">108</span></a><span class="sd">        list of tuple of str, str</span>
</span><span id="MSATools.filter_by_continuous_gaps-109"><a href="#MSATools.filter_by_continuous_gaps-109"><span class="linenos">109</span></a><span class="sd">            List of entries that are valid in that their sequences&#39; number of maximum continuous gaps is within the threshold supplied as `max_gaps`. </span>
</span><span id="MSATools.filter_by_continuous_gaps-110"><a href="#MSATools.filter_by_continuous_gaps-110"><span class="linenos">110</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools.filter_by_continuous_gaps-111"><a href="#MSATools.filter_by_continuous_gaps-111"><span class="linenos">111</span></a>        <span class="n">table</span> <span class="o">=</span> <span class="nb">str</span><span class="o">.</span><span class="n">maketrans</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">string</span><span class="o">.</span><span class="n">ascii_lowercase</span><span class="o">+</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
</span><span id="MSATools.filter_by_continuous_gaps-112"><a href="#MSATools.filter_by_continuous_gaps-112"><span class="linenos">112</span></a>        <span class="k">if</span> <span class="n">max_gaps</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="MSATools.filter_by_continuous_gaps-113"><a href="#MSATools.filter_by_continuous_gaps-113"><span class="linenos">113</span></a>            <span class="n">kept_entries</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MSATools.filter_by_continuous_gaps-114"><a href="#MSATools.filter_by_continuous_gaps-114"><span class="linenos">114</span></a>            <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools.filter_by_continuous_gaps-115"><a href="#MSATools.filter_by_continuous_gaps-115"><span class="linenos">115</span></a>                <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>
</span><span id="MSATools.filter_by_continuous_gaps-116"><a href="#MSATools.filter_by_continuous_gaps-116"><span class="linenos">116</span></a>                <span class="n">kept_entries</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">header</span><span class="p">,</span> <span class="n">sequence</span><span class="p">))</span>
</span><span id="MSATools.filter_by_continuous_gaps-117"><a href="#MSATools.filter_by_continuous_gaps-117"><span class="linenos">117</span></a>            <span class="k">return</span> <span class="n">kept_entries</span>
</span><span id="MSATools.filter_by_continuous_gaps-118"><a href="#MSATools.filter_by_continuous_gaps-118"><span class="linenos">118</span></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="MSATools.filter_by_continuous_gaps-119"><a href="#MSATools.filter_by_continuous_gaps-119"><span class="linenos">119</span></a>            <span class="n">kept_entries</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="MSATools.filter_by_continuous_gaps-120"><a href="#MSATools.filter_by_continuous_gaps-120"><span class="linenos">120</span></a>            <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools.filter_by_continuous_gaps-121"><a href="#MSATools.filter_by_continuous_gaps-121"><span class="linenos">121</span></a>                <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>
</span><span id="MSATools.filter_by_continuous_gaps-122"><a href="#MSATools.filter_by_continuous_gaps-122"><span class="linenos">122</span></a>                <span class="k">if</span> <span class="n">MSATools</span><span class="o">.</span><span class="n">get_sequence_max_cont_gaps</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">max_gaps</span><span class="p">:</span>
</span><span id="MSATools.filter_by_continuous_gaps-123"><a href="#MSATools.filter_by_continuous_gaps-123"><span class="linenos">123</span></a>                    <span class="n">kept_entries</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">header</span><span class="p">,</span> <span class="n">sequence</span><span class="p">))</span>
</span><span id="MSATools.filter_by_continuous_gaps-124"><a href="#MSATools.filter_by_continuous_gaps-124"><span class="linenos">124</span></a>        <span class="k">return</span> <span class="n">kept_entries</span>
</span></pre></div>


            <div class="docstring"><p>Filter out entries in your MSA by the number of maximum continuous gaps specified unless None is provided. Also, removes .s and lowercase letters from the sequence.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>max_gaps</strong> (int):
The maximum allowed number of continuous gaps in a sequence</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>list of tuple of str, str</strong>: List of entries that are valid in that their sequences' number of maximum continuous gaps is within the threshold supplied as <code>max_gaps</code>.</li>
</ul>
</div>


                            </div>
                            <div id="MSATools.write" class="classattr">
                                        <input id="MSATools.write-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">write</span><span class="signature pdoc-code condensed">(<span class="param"><span class="bp">self</span>, </span><span class="param"><span class="n">destination</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">]</span></span><span class="return-annotation">) -> <span class="kc">None</span>:</span></span>

                <label class="view-source-button" for="MSATools.write-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#MSATools.write"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="MSATools.write-126"><a href="#MSATools.write-126"><span class="linenos">126</span></a>    <span class="k">def</span> <span class="nf">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">destination</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="MSATools.write-127"><a href="#MSATools.write-127"><span class="linenos">127</span></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="MSATools.write-128"><a href="#MSATools.write-128"><span class="linenos">128</span></a><span class="sd">        Writes this MSA&#39;s headers and sequences to the destination specified.</span>
</span><span id="MSATools.write-129"><a href="#MSATools.write-129"><span class="linenos">129</span></a><span class="sd">        </span>
</span><span id="MSATools.write-130"><a href="#MSATools.write-130"><span class="linenos">130</span></a><span class="sd">        Parameters</span>
</span><span id="MSATools.write-131"><a href="#MSATools.write-131"><span class="linenos">131</span></a><span class="sd">        ----------</span>
</span><span id="MSATools.write-132"><a href="#MSATools.write-132"><span class="linenos">132</span></a><span class="sd">        destination : str or io.IOBase</span>
</span><span id="MSATools.write-133"><a href="#MSATools.write-133"><span class="linenos">133</span></a><span class="sd">            Filepath or IO to write the MSA supplied to.</span>
</span><span id="MSATools.write-134"><a href="#MSATools.write-134"><span class="linenos">134</span></a>
</span><span id="MSATools.write-135"><a href="#MSATools.write-135"><span class="linenos">135</span></a><span class="sd">        Returns</span>
</span><span id="MSATools.write-136"><a href="#MSATools.write-136"><span class="linenos">136</span></a><span class="sd">        -------</span>
</span><span id="MSATools.write-137"><a href="#MSATools.write-137"><span class="linenos">137</span></a><span class="sd">        None</span>
</span><span id="MSATools.write-138"><a href="#MSATools.write-138"><span class="linenos">138</span></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="MSATools.write-139"><a href="#MSATools.write-139"><span class="linenos">139</span></a>        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
</span><span id="MSATools.write-140"><a href="#MSATools.write-140"><span class="linenos">140</span></a>            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fs</span><span class="p">:</span>
</span><span id="MSATools.write-141"><a href="#MSATools.write-141"><span class="linenos">141</span></a>                <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools.write-142"><a href="#MSATools.write-142"><span class="linenos">142</span></a>                    <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
</span><span id="MSATools.write-143"><a href="#MSATools.write-143"><span class="linenos">143</span></a>                    <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="MSATools.write-144"><a href="#MSATools.write-144"><span class="linenos">144</span></a>                    <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span id="MSATools.write-145"><a href="#MSATools.write-145"><span class="linenos">145</span></a>                    <span class="n">fs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="MSATools.write-146"><a href="#MSATools.write-146"><span class="linenos">146</span></a>        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">destination</span><span class="p">,</span> <span class="n">io</span><span class="o">.</span><span class="n">IOBase</span><span class="p">):</span>
</span><span id="MSATools.write-147"><a href="#MSATools.write-147"><span class="linenos">147</span></a>            <span class="k">for</span> <span class="n">header</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">MSA</span><span class="p">:</span>
</span><span id="MSATools.write-148"><a href="#MSATools.write-148"><span class="linenos">148</span></a>                <span class="n">destination</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
</span><span id="MSATools.write-149"><a href="#MSATools.write-149"><span class="linenos">149</span></a>                <span class="n">destination</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="MSATools.write-150"><a href="#MSATools.write-150"><span class="linenos">150</span></a>                <span class="n">destination</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
</span><span id="MSATools.write-151"><a href="#MSATools.write-151"><span class="linenos">151</span></a>                <span class="n">destination</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Writes this MSA's headers and sequences to the destination specified.</p>

<h6 id="parameters">Parameters</h6>

<ul>
<li><strong>destination</strong> (str or io.IOBase):
Filepath or IO to write the MSA supplied to.</li>
</ul>

<h6 id="returns">Returns</h6>

<ul>
<li><strong>None</strong></li>
</ul>
</div>


                            </div>
                </section>
    </main>
</body>
</html>